
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

    <title>Sera concentrations &#8212; polyclonal 4.1 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css" />
    <link rel="stylesheet" type="text/css" href="_static/plot_directive.css" />
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/sphinx_highlight.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Details on fitting" href="optimization.html" />
    <link rel="prev" title="Library mutation rate" href="mutation_rate.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars and line breaks on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
    white-space: pre;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt .copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
.jp-RenderedHTMLCommon table,
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
.jp-RenderedHTMLCommon thead,
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
.jp-RenderedHTMLCommon tr,
.jp-RenderedHTMLCommon th,
.jp-RenderedHTMLCommon td,
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
.jp-RenderedHTMLCommon th,
div.rendered_html th {
  font-weight: bold;
}
.jp-RenderedHTMLCommon tbody tr:nth-child(odd),
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
.jp-RenderedHTMLCommon tbody tr:hover,
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}
</style>
<section id="Sera-concentrations">
<h1>Sera concentrations<a class="headerlink" href="#Sera-concentrations" title="Permalink to this heading">¶</a></h1>
<p>Choosing sera concentrations is an important consideration for DMS selections. We’ll use simulated data to estimate the optimal set of sera concentrations that provides the best performance while minimizing the <strong>number of concentrations</strong> (or selection experiments) required.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">pickle</span>

<span class="kn">import</span> <span class="nn">altair</span> <span class="k">as</span> <span class="nn">alt</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">polyclonal</span>
</pre></div>
</div>
</div>
<p>First, we read in a simulated “noisy” dataset measured at six sera concentrations. The variants in this library were simulated to contain a Poisson-distributed number of mutations, with an average of three mutations per gene. The variants also generally span a wide range of escape fractions across the different sera concentrations.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">noisy_data</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;RBD_variants_escape_noisy.csv&quot;</span><span class="p">,</span> <span class="n">na_filter</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
    <span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s2">&quot;library == &#39;avg3muts&#39;&quot;</span><span class="p">)</span>
    <span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="p">)</span>
<span class="n">noisy_data</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>library</th>
      <th>barcode</th>
      <th>concentration</th>
      <th>prob_escape</th>
      <th>aa_substitutions</th>
      <th>IC90</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>avg3muts</td>
      <td>AAAACTGCTGAGGAGA</td>
      <td>0.125</td>
      <td>0.013540</td>
      <td></td>
      <td>0.08212</td>
    </tr>
    <tr>
      <th>1</th>
      <td>avg3muts</td>
      <td>AAAAGCAGGCTACTCT</td>
      <td>0.125</td>
      <td>0.066420</td>
      <td></td>
      <td>0.08212</td>
    </tr>
    <tr>
      <th>2</th>
      <td>avg3muts</td>
      <td>AAAAGCTATAGGTGCC</td>
      <td>0.125</td>
      <td>0.023120</td>
      <td></td>
      <td>0.08212</td>
    </tr>
    <tr>
      <th>3</th>
      <td>avg3muts</td>
      <td>AAAAGGTATTAGTGGC</td>
      <td>0.125</td>
      <td>0.005456</td>
      <td></td>
      <td>0.08212</td>
    </tr>
    <tr>
      <th>4</th>
      <td>avg3muts</td>
      <td>AAAAGTGCCTTCGTTA</td>
      <td>0.125</td>
      <td>0.054760</td>
      <td></td>
      <td>0.08212</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>239995</th>
      <td>avg3muts</td>
      <td>CTTAAAATAGCTGGTC</td>
      <td>0.250</td>
      <td>0.000000</td>
      <td>Y508W</td>
      <td>0.08212</td>
    </tr>
    <tr>
      <th>239996</th>
      <td>avg3muts</td>
      <td>CTTAAAATAGCTGGTC</td>
      <td>0.500</td>
      <td>0.026480</td>
      <td>Y508W</td>
      <td>0.08212</td>
    </tr>
    <tr>
      <th>239997</th>
      <td>avg3muts</td>
      <td>CTTAAAATAGCTGGTC</td>
      <td>1.000</td>
      <td>0.012260</td>
      <td>Y508W</td>
      <td>0.08212</td>
    </tr>
    <tr>
      <th>239998</th>
      <td>avg3muts</td>
      <td>CTTAAAATAGCTGGTC</td>
      <td>2.000</td>
      <td>0.000000</td>
      <td>Y508W</td>
      <td>0.08212</td>
    </tr>
    <tr>
      <th>239999</th>
      <td>avg3muts</td>
      <td>CTTAAAATAGCTGGTC</td>
      <td>4.000</td>
      <td>0.000000</td>
      <td>Y508W</td>
      <td>0.08212</td>
    </tr>
  </tbody>
</table>
<p>240000 rows × 6 columns</p>
</div></div>
</div>
<p>While the concentrations are in arbitrary units (<code class="docutils literal notranslate"><span class="pre">0.125,</span> <span class="pre">0.25,</span> <span class="pre">0.5,</span> <span class="pre">1,</span> <span class="pre">2,</span> <span class="pre">4</span></code>), we can describe them in terms of their ICXX’s against wildtype virus. So, we can interpret each as a sera concentration that neutralizes XX % of wildtype viruses.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">wt_data_icxx_df</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;RBD_variants_escape_exact.csv&quot;</span><span class="p">,</span> <span class="n">na_filter</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
    <span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s2">&quot;aa_substitutions == &#39;&#39;&quot;</span><span class="p">)</span>
    <span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="s2">&quot;library&quot;</span><span class="p">)</span>
    <span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
    <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">ICxx_against_wt</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">round</span><span class="p">((</span><span class="mi">1</span> <span class="o">-</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;prob_escape&quot;</span><span class="p">])</span> <span class="o">*</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>
    <span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;aa_substitutions&quot;</span><span class="p">,</span> <span class="s2">&quot;prob_escape&quot;</span><span class="p">,</span> <span class="s2">&quot;IC90&quot;</span><span class="p">])</span>
    <span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;concentration&quot;</span><span class="p">)</span>
<span class="p">)</span>

<span class="n">display</span><span class="p">(</span><span class="n">wt_data_icxx_df</span><span class="p">)</span>
<span class="n">wt_data_icxx</span> <span class="o">=</span> <span class="n">wt_data_icxx_df</span><span class="p">[</span><span class="s2">&quot;ICxx_against_wt&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">to_dict</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>barcode</th>
      <th>ICxx_against_wt</th>
    </tr>
    <tr>
      <th>concentration</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0.125</th>
      <td>AAAAAATGTTCTATCC</td>
      <td>95.141</td>
    </tr>
    <tr>
      <th>0.125</th>
      <td>AAAAACAATCCGGACT</td>
      <td>95.141</td>
    </tr>
    <tr>
      <th>0.125</th>
      <td>AAAAACGCGGTCACTT</td>
      <td>95.141</td>
    </tr>
    <tr>
      <th>0.125</th>
      <td>AAAAACTTGGCTAGCT</td>
      <td>95.141</td>
    </tr>
    <tr>
      <th>0.125</th>
      <td>AAAAAGCAAGGCCCAG</td>
      <td>95.141</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>4.000</th>
      <td>TTTGTATGGTCCATAT</td>
      <td>99.999</td>
    </tr>
    <tr>
      <th>4.000</th>
      <td>TTTGTCTCGAATGGTG</td>
      <td>99.999</td>
    </tr>
    <tr>
      <th>4.000</th>
      <td>TTTTAAGCTCATACGC</td>
      <td>99.999</td>
    </tr>
    <tr>
      <th>4.000</th>
      <td>TTTTATCCACCGAACT</td>
      <td>99.999</td>
    </tr>
    <tr>
      <th>4.000</th>
      <td>TTTTCGATCCTTGTCA</td>
      <td>99.999</td>
    </tr>
  </tbody>
</table>
<p>137868 rows × 2 columns</p>
</div></div>
</div>
<p>Now, we’ll fit multiple <code class="docutils literal notranslate"><span class="pre">Polyclonal</span></code> models to data measured at <strong>single</strong> concentrations. We’ll initialize each <code class="docutils literal notranslate"><span class="pre">Polyclonal</span></code> model with the same values. We know from <a class="reference external" href="https://www.nature.com/articles/s41467-021-24435-8">prior work</a> the three most important epitopes and a key mutation in each, so we use this prior knowledge to “seed” initial guesses that assign large escape values to a key site in each epitope:</p>
<ul class="simple">
<li><p>site 417 for class 1 epitope, which is often the least important</p></li>
</ul>
<ul class="simple">
<li><p>site 484 for class 2 epitope, which is often the dominant one</p></li>
</ul>
<ul class="simple">
<li><p>site 444 for class 3 epitope, which is often the second most dominant one</p></li>
</ul>
<p>Additionally, we’ll store fit models as <a class="reference external" href="https://docs.python.org/3/library/pickle.html#module-pickle">pickle</a> files, so that we can conveniently load them in the future without having to fit again.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">conc_sets</span> <span class="o">=</span> <span class="p">[[</span><span class="mf">0.125</span><span class="p">],</span> <span class="p">[</span><span class="mf">0.25</span><span class="p">],</span> <span class="p">[</span><span class="mf">0.5</span><span class="p">],</span> <span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="p">[</span><span class="mi">4</span><span class="p">]]</span>

<span class="c1"># Make a directory to house pickled models</span>
<span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="s2">&quot;fit_polyclonal_models&quot;</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>


<span class="k">def</span> <span class="nf">fit_polyclonal</span><span class="p">(</span><span class="n">conc_set</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Fit `Polyclonal` model with data measured for a specific concentration set.</span>
<span class="sd">    Returns fit `Polyclonal` object.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">poly_abs</span> <span class="o">=</span> <span class="n">polyclonal</span><span class="o">.</span><span class="n">Polyclonal</span><span class="p">(</span>
        <span class="n">data_to_fit</span><span class="o">=</span><span class="n">noisy_data</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;concentration in </span><span class="si">{</span><span class="n">conc_set</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">),</span>
        <span class="n">activity_wt_df</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_records</span><span class="p">(</span>
            <span class="p">[</span>
                <span class="p">(</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">),</span>
                <span class="p">(</span><span class="s2">&quot;2&quot;</span><span class="p">,</span> <span class="mf">3.0</span><span class="p">),</span>
                <span class="p">(</span><span class="s2">&quot;3&quot;</span><span class="p">,</span> <span class="mf">2.0</span><span class="p">),</span>
            <span class="p">],</span>
            <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;epitope&quot;</span><span class="p">,</span> <span class="s2">&quot;activity&quot;</span><span class="p">],</span>
        <span class="p">),</span>
        <span class="n">site_escape_df</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_records</span><span class="p">(</span>
            <span class="p">[</span>
                <span class="p">(</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="mi">417</span><span class="p">,</span> <span class="mf">10.0</span><span class="p">),</span>
                <span class="p">(</span><span class="s2">&quot;2&quot;</span><span class="p">,</span> <span class="mi">484</span><span class="p">,</span> <span class="mf">10.0</span><span class="p">),</span>
                <span class="p">(</span><span class="s2">&quot;3&quot;</span><span class="p">,</span> <span class="mi">444</span><span class="p">,</span> <span class="mf">10.0</span><span class="p">),</span>
            <span class="p">],</span>
            <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;epitope&quot;</span><span class="p">,</span> <span class="s2">&quot;site&quot;</span><span class="p">,</span> <span class="s2">&quot;escape&quot;</span><span class="p">],</span>
        <span class="p">),</span>
        <span class="n">data_mut_escape_overlap</span><span class="o">=</span><span class="s2">&quot;fill_to_data&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">poly_abs</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">reg_escape_weight</span><span class="o">=</span><span class="mf">0.01</span><span class="p">,</span> <span class="n">reg_uniqueness2_weight</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">poly_abs</span>


<span class="c1"># Store all fit models in a dictionary for future lookup</span>
<span class="n">fit_models</span> <span class="o">=</span> <span class="p">{}</span>

<span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">conc_sets</span><span class="p">:</span>
    <span class="c1"># These are the keys for fit models</span>
    <span class="n">model_string</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;noisy_</span><span class="si">{</span><span class="n">s</span><span class="si">}</span><span class="s2">conc_3muts&quot;</span>

    <span class="c1"># If the pickled model exists in fit_polyclonal_models directory,</span>
    <span class="c1"># load it and update fit_models</span>
    <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;fit_polyclonal_models/</span><span class="si">{</span><span class="n">model_string</span><span class="si">}</span><span class="s2">.pkl&quot;</span><span class="p">)</span> <span class="ow">is</span> <span class="kc">True</span><span class="p">:</span>
        <span class="n">model</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;fit_polyclonal_models/</span><span class="si">{</span><span class="n">model_string</span><span class="si">}</span><span class="s2">.pkl&quot;</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">))</span>
        <span class="n">fit_models</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="n">model_string</span><span class="p">:</span> <span class="n">model</span><span class="p">})</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Model with </span><span class="si">{</span><span class="n">s</span><span class="si">}</span><span class="s2"> was already fit.&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># Else, fit a model using fit_polyclonal(), save it to the</span>
        <span class="c1"># fit_polyclonal_models directory, and update fit_models</span>
        <span class="n">model</span> <span class="o">=</span> <span class="n">fit_polyclonal</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
        <span class="n">fit_models</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="n">model_string</span><span class="p">:</span> <span class="n">model</span><span class="p">})</span>
        <span class="n">pickle</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;fit_polyclonal_models/</span><span class="si">{</span><span class="n">model_string</span><span class="si">}</span><span class="s2">.pkl&quot;</span><span class="p">,</span> <span class="s2">&quot;wb&quot;</span><span class="p">))</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Model with </span><span class="si">{</span><span class="n">s</span><span class="si">}</span><span class="s2"> fit and saved.&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model with [0.125] was already fit.
Model with [0.25] was already fit.
Model with [0.5] was already fit.
Model with [1] was already fit.
Model with [2] was already fit.
Model with [4] was already fit.
</pre></div></div>
</div>
<p>We can look at the correlation between the “true” and inferred mutation-escape values, <span class="math notranslate nohighlight">\(\beta_{m,e}\)</span>, for the fit models. These mutation-escape values represent the extent to which mutations mediate escape from specific epitopes.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">all_corrs</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>

<span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">conc_sets</span><span class="p">:</span>
    <span class="n">model</span> <span class="o">=</span> <span class="n">fit_models</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;noisy_</span><span class="si">{</span><span class="n">s</span><span class="si">}</span><span class="s2">conc_3muts&quot;</span><span class="p">]</span>

    <span class="n">mut_escape_pred</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;RBD_mut_escape_df.csv&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span>
        <span class="p">(</span>
            <span class="n">model</span><span class="o">.</span><span class="n">mut_escape_df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
                <span class="n">epitope</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s2">&quot;class &quot;</span> <span class="o">+</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;epitope&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
            <span class="p">)</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;escape&quot;</span><span class="p">:</span> <span class="s2">&quot;predicted escape&quot;</span><span class="p">})</span>
        <span class="p">),</span>
        <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;mutation&quot;</span><span class="p">,</span> <span class="s2">&quot;epitope&quot;</span><span class="p">],</span>
        <span class="n">validate</span><span class="o">=</span><span class="s2">&quot;one_to_one&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">corr</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">mut_escape_pred</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;epitope&quot;</span><span class="p">)</span>
        <span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;escape&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">corr</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s2">&quot;predicted escape&quot;</span><span class="p">])</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span>
        <span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="s2">&quot;correlation (R^2)&quot;</span><span class="p">)</span>
        <span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
    <span class="p">)</span>

    <span class="n">all_corrs</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span>
        <span class="p">[</span>
            <span class="n">all_corrs</span><span class="p">,</span>
            <span class="n">corr</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
                <span class="n">icxx_set</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="sa">f</span><span class="s2">&quot;IC</span><span class="si">{</span><span class="n">wt_data_icxx</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">s</span><span class="p">])]</span>
                <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">corr</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
            <span class="p">),</span>
        <span class="p">]</span>
    <span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>
<span class="n">alt</span><span class="o">.</span><span class="n">Chart</span><span class="p">(</span><span class="n">all_corrs</span><span class="p">)</span><span class="o">.</span><span class="n">mark_circle</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="mi">125</span><span class="p">)</span><span class="o">.</span><span class="n">encode</span><span class="p">(</span>
    <span class="n">x</span><span class="o">=</span><span class="n">alt</span><span class="o">.</span><span class="n">X</span><span class="p">(</span><span class="s2">&quot;icxx_set:O&quot;</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="n">alt</span><span class="o">.</span><span class="n">EncodingSortField</span><span class="p">(</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="n">order</span><span class="o">=</span><span class="s2">&quot;descending&quot;</span><span class="p">)),</span>
    <span class="n">y</span><span class="o">=</span><span class="s2">&quot;correlation (R^2):Q&quot;</span><span class="p">,</span>
    <span class="n">column</span><span class="o">=</span><span class="s2">&quot;epitope:N&quot;</span><span class="p">,</span>
    <span class="n">tooltip</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;icxx_set:O&quot;</span><span class="p">,</span> <span class="n">alt</span><span class="o">.</span><span class="n">Tooltip</span><span class="p">(</span><span class="s2">&quot;correlation (R^2)&quot;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;.3f&quot;</span><span class="p">)],</span>
    <span class="n">color</span><span class="o">=</span><span class="n">alt</span><span class="o">.</span><span class="n">Color</span><span class="p">(</span>
        <span class="s2">&quot;epitope&quot;</span><span class="p">,</span> <span class="n">scale</span><span class="o">=</span><span class="n">alt</span><span class="o">.</span><span class="n">Scale</span><span class="p">(</span><span class="nb">range</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;#0072B2&quot;</span><span class="p">,</span> <span class="s2">&quot;#CC79A7&quot;</span><span class="p">,</span> <span class="s2">&quot;#4C3549&quot;</span><span class="p">]),</span> <span class="n">legend</span><span class="o">=</span><span class="kc">None</span>
    <span class="p">),</span>
<span class="p">)</span><span class="o">.</span><span class="n">properties</span><span class="p">(</span><span class="n">width</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">height</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s2">&quot;inferred vs. true mutation escape values&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/tyu2/.local/lib/python3.8/site-packages/altair/utils/core.py:317: FutureWarning: iteritems is deprecated and will be removed in a future version. Use .items instead.
  for col_name, dtype in df.dtypes.iteritems():
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div id="altair-viz-517cdcd5a6174091994094ab4107e9e5"></div>
<script type="text/javascript">
  var VEGA_DEBUG = (typeof VEGA_DEBUG == "undefined") ? {} : VEGA_DEBUG;
  (function(spec, embedOpt){
    let outputDiv = document.currentScript.previousElementSibling;
    if (outputDiv.id !== "altair-viz-517cdcd5a6174091994094ab4107e9e5") {
      outputDiv = document.getElementById("altair-viz-517cdcd5a6174091994094ab4107e9e5");
    }
    const paths = {
      "vega": "https://cdn.jsdelivr.net/npm//vega@5?noext",
      "vega-lib": "https://cdn.jsdelivr.net/npm//vega-lib?noext",
      "vega-lite": "https://cdn.jsdelivr.net/npm//vega-lite@5.2.0?noext",
      "vega-embed": "https://cdn.jsdelivr.net/npm//vega-embed@6?noext",
    };

    function maybeLoadScript(lib, version) {
      var key = `${lib.replace("-", "")}_version`;
      return (VEGA_DEBUG[key] == version) ?
        Promise.resolve(paths[lib]) :
        new Promise(function(resolve, reject) {
          var s = document.createElement('script');
          document.getElementsByTagName("head")[0].appendChild(s);
          s.async = true;
          s.onload = () => {
            VEGA_DEBUG[key] = version;
            return resolve(paths[lib]);
          };
          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);
          s.src = paths[lib];
        });
    }

    function showError(err) {
      outputDiv.innerHTML = `<div class="error" style="color:red;">${err}</div>`;
      throw err;
    }

    function displayChart(vegaEmbed) {
      vegaEmbed(outputDiv, spec, embedOpt)
        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));
    }

    if(typeof define === "function" && define.amd) {
      requirejs.config({paths});
      require(["vega-embed"], displayChart, err => showError(`Error loading script: ${err.message}`));
    } else {
      maybeLoadScript("vega", "5")
        .then(() => maybeLoadScript("vega-lite", "5.2.0"))
        .then(() => maybeLoadScript("vega-embed", "6"))
        .catch(showError)
        .then(() => displayChart(vegaEmbed));
    }
  })({"config": {"view": {"continuousWidth": 400, "continuousHeight": 300}}, "data": {"name": "data-8cf1de1071b8cf3f72e40839af01a692"}, "mark": {"type": "circle", "size": 125}, "encoding": {"color": {"field": "epitope", "legend": null, "scale": {"range": ["#0072B2", "#CC79A7", "#4C3549"]}, "type": "nominal"}, "column": {"field": "epitope", "type": "nominal"}, "tooltip": [{"field": "icxx_set", "type": "ordinal"}, {"field": "correlation (R^2)", "format": ".3f", "type": "quantitative"}], "x": {"field": "icxx_set", "sort": {"field": "x", "order": "descending"}, "type": "ordinal"}, "y": {"field": "correlation (R^2)", "type": "quantitative"}}, "height": 200, "title": "inferred vs. true mutation escape values", "width": 200, "$schema": "https://vega.github.io/schema/vega-lite/v5.2.0.json", "datasets": {"data-8cf1de1071b8cf3f72e40839af01a692": [{"epitope": "class 1", "correlation (R^2)": 0.6494946791057244, "icxx_set": "IC95.141"}, {"epitope": "class 2", "correlation (R^2)": 0.9411429752536082, "icxx_set": "IC95.141"}, {"epitope": "class 3", "correlation (R^2)": 0.9235798092082146, "icxx_set": "IC95.141"}, {"epitope": "class 1", "correlation (R^2)": 0.633408776422364, "icxx_set": "IC98.854"}, {"epitope": "class 2", "correlation (R^2)": 0.9449715250188195, "icxx_set": "IC98.854"}, {"epitope": "class 3", "correlation (R^2)": 0.9328787988314352, "icxx_set": "IC98.854"}, {"epitope": "class 1", "correlation (R^2)": 0.5855229863399843, "icxx_set": "IC99.79"}, {"epitope": "class 2", "correlation (R^2)": 0.9447264445598186, "icxx_set": "IC99.79"}, {"epitope": "class 3", "correlation (R^2)": 0.9320465960897137, "icxx_set": "IC99.79"}, {"epitope": "class 1", "correlation (R^2)": 0.47277660829732004, "icxx_set": "IC99.968"}, {"epitope": "class 2", "correlation (R^2)": 0.9255322091240006, "icxx_set": "IC99.968"}, {"epitope": "class 3", "correlation (R^2)": 0.9266370719977781, "icxx_set": "IC99.968"}, {"epitope": "class 1", "correlation (R^2)": 0.3998783246881719, "icxx_set": "IC99.995"}, {"epitope": "class 2", "correlation (R^2)": 0.9295250314996651, "icxx_set": "IC99.995"}, {"epitope": "class 3", "correlation (R^2)": 0.9009690071099429, "icxx_set": "IC99.995"}, {"epitope": "class 1", "correlation (R^2)": 0.3722224083975561, "icxx_set": "IC99.999"}, {"epitope": "class 2", "correlation (R^2)": 0.9126328899571241, "icxx_set": "IC99.999"}, {"epitope": "class 3", "correlation (R^2)": 0.8726719283299381, "icxx_set": "IC99.999"}]}}, {"mode": "vega-lite"});
</script></div>
</div>
<p>Additionally, we’ll look at the correlation between “true” and predicted IC90’s for each of the fit models. To do this, we’ll predict the IC90’s of variants in a separate library with a with a different (higher) mutation rate. We therefore read in the “exact” simulated data from a library containing variants with an average of four mutations per gene.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">exact_data</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;RBD_variants_escape_exact.csv&quot;</span><span class="p">,</span> <span class="n">na_filter</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
    <span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">&#39;library == &quot;avg4muts&quot;&#39;</span><span class="p">)</span>
    <span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s2">&quot;concentration in [1]&quot;</span><span class="p">)</span>
    <span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<p>We’ll make the comparison on a log scale, and clip IC90s at values &gt;50 as that is likely to be way outside the dynamic range given the concentrations used.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ic90_corrs</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>

<span class="n">max_ic90</span> <span class="o">=</span> <span class="mi">50</span>
<span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">conc_sets</span><span class="p">:</span>
    <span class="n">model</span> <span class="o">=</span> <span class="n">fit_models</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;noisy_</span><span class="si">{</span><span class="n">s</span><span class="si">}</span><span class="s2">conc_3muts&quot;</span><span class="p">]</span>

    <span class="n">ic90s</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">exact_data</span><span class="p">[[</span><span class="s2">&quot;aa_substitutions&quot;</span><span class="p">,</span> <span class="s2">&quot;IC90&quot;</span><span class="p">]]</span>
        <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">IC90</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;IC90&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">clip</span><span class="p">(</span><span class="n">upper</span><span class="o">=</span><span class="n">max_ic90</span><span class="p">))</span>
        <span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
    <span class="p">)</span>
    <span class="n">ic90s</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">filter_variants_by_seen_muts</span><span class="p">(</span><span class="n">ic90s</span><span class="p">)</span>
    <span class="n">ic90s</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">icXX</span><span class="p">(</span><span class="n">ic90s</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="mf">0.9</span><span class="p">,</span> <span class="n">col</span><span class="o">=</span><span class="s2">&quot;predicted_IC90&quot;</span><span class="p">,</span> <span class="n">max_c</span><span class="o">=</span><span class="n">max_ic90</span><span class="p">)</span>

    <span class="n">ic90s</span> <span class="o">=</span> <span class="n">ic90s</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
        <span class="n">log_IC90</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s2">&quot;IC90&quot;</span><span class="p">]),</span>
        <span class="n">predicted_log_IC90</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s2">&quot;predicted_IC90&quot;</span><span class="p">]),</span>
    <span class="p">)</span>

    <span class="n">corr</span> <span class="o">=</span> <span class="n">ic90s</span><span class="p">[</span><span class="s2">&quot;log_IC90&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">corr</span><span class="p">(</span><span class="n">ic90s</span><span class="p">[</span><span class="s2">&quot;predicted_log_IC90&quot;</span><span class="p">])</span> <span class="o">**</span> <span class="mi">2</span>

    <span class="n">ic90_corrs</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span>
        <span class="p">[</span>
            <span class="n">ic90_corrs</span><span class="p">,</span>
            <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
                <span class="p">{</span>
                    <span class="s2">&quot;correlation (R^2)&quot;</span><span class="p">:</span> <span class="n">corr</span><span class="p">,</span>
                    <span class="s2">&quot;icxx_set&quot;</span><span class="p">:</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;IC</span><span class="si">{</span><span class="n">wt_data_icxx</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">s</span><span class="p">],</span>
                <span class="p">}</span>
            <span class="p">),</span>
        <span class="p">]</span>
    <span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>
<span class="n">alt</span><span class="o">.</span><span class="n">Chart</span><span class="p">(</span><span class="n">ic90_corrs</span><span class="p">)</span><span class="o">.</span><span class="n">mark_circle</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="mi">125</span><span class="p">)</span><span class="o">.</span><span class="n">encode</span><span class="p">(</span>
    <span class="n">x</span><span class="o">=</span><span class="s2">&quot;icxx_set:O&quot;</span><span class="p">,</span>
    <span class="n">y</span><span class="o">=</span><span class="s2">&quot;correlation (R^2):Q&quot;</span><span class="p">,</span>
    <span class="n">tooltip</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;icxx_set&quot;</span><span class="p">,</span> <span class="n">alt</span><span class="o">.</span><span class="n">Tooltip</span><span class="p">(</span><span class="s2">&quot;correlation (R^2)&quot;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;.3f&quot;</span><span class="p">)],</span>
<span class="p">)</span><span class="o">.</span><span class="n">properties</span><span class="p">(</span><span class="n">width</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">height</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s2">&quot;predicted vs. true IC90&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/tyu2/.local/lib/python3.8/site-packages/altair/utils/core.py:317: FutureWarning: iteritems is deprecated and will be removed in a future version. Use .items instead.
  for col_name, dtype in df.dtypes.iteritems():
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div id="altair-viz-f9d08b355c564a1797e593ca476d283c"></div>
<script type="text/javascript">
  var VEGA_DEBUG = (typeof VEGA_DEBUG == "undefined") ? {} : VEGA_DEBUG;
  (function(spec, embedOpt){
    let outputDiv = document.currentScript.previousElementSibling;
    if (outputDiv.id !== "altair-viz-f9d08b355c564a1797e593ca476d283c") {
      outputDiv = document.getElementById("altair-viz-f9d08b355c564a1797e593ca476d283c");
    }
    const paths = {
      "vega": "https://cdn.jsdelivr.net/npm//vega@5?noext",
      "vega-lib": "https://cdn.jsdelivr.net/npm//vega-lib?noext",
      "vega-lite": "https://cdn.jsdelivr.net/npm//vega-lite@5.2.0?noext",
      "vega-embed": "https://cdn.jsdelivr.net/npm//vega-embed@6?noext",
    };

    function maybeLoadScript(lib, version) {
      var key = `${lib.replace("-", "")}_version`;
      return (VEGA_DEBUG[key] == version) ?
        Promise.resolve(paths[lib]) :
        new Promise(function(resolve, reject) {
          var s = document.createElement('script');
          document.getElementsByTagName("head")[0].appendChild(s);
          s.async = true;
          s.onload = () => {
            VEGA_DEBUG[key] = version;
            return resolve(paths[lib]);
          };
          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);
          s.src = paths[lib];
        });
    }

    function showError(err) {
      outputDiv.innerHTML = `<div class="error" style="color:red;">${err}</div>`;
      throw err;
    }

    function displayChart(vegaEmbed) {
      vegaEmbed(outputDiv, spec, embedOpt)
        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));
    }

    if(typeof define === "function" && define.amd) {
      requirejs.config({paths});
      require(["vega-embed"], displayChart, err => showError(`Error loading script: ${err.message}`));
    } else {
      maybeLoadScript("vega", "5")
        .then(() => maybeLoadScript("vega-lite", "5.2.0"))
        .then(() => maybeLoadScript("vega-embed", "6"))
        .catch(showError)
        .then(() => displayChart(vegaEmbed));
    }
  })({"config": {"view": {"continuousWidth": 400, "continuousHeight": 300}}, "data": {"name": "data-20bba568f249810ece7bb6819d46dd5a"}, "mark": {"type": "circle", "size": 125}, "encoding": {"tooltip": [{"field": "icxx_set", "type": "nominal"}, {"field": "correlation (R^2)", "format": ".3f", "type": "quantitative"}], "x": {"field": "icxx_set", "type": "ordinal"}, "y": {"field": "correlation (R^2)", "type": "quantitative"}}, "height": 200, "title": "predicted vs. true IC90", "width": 200, "$schema": "https://vega.github.io/schema/vega-lite/v5.2.0.json", "datasets": {"data-20bba568f249810ece7bb6819d46dd5a": [{"correlation (R^2)": 0.9735547111012681, "icxx_set": "IC95.141"}, {"correlation (R^2)": 0.9706927418536281, "icxx_set": "IC98.854"}, {"correlation (R^2)": 0.9371232495080029, "icxx_set": "IC99.79"}, {"correlation (R^2)": 0.8524044532251411, "icxx_set": "IC99.968"}, {"correlation (R^2)": 0.7156542863450766, "icxx_set": "IC99.995"}, {"correlation (R^2)": 0.5676210971236344, "icxx_set": "IC99.999"}]}}, {"mode": "vega-lite"});
</script></div>
</div>
<p>Across all models fit on data with single concentrations, the correlation was very strong.</p>
<p>Next, we’ll fit a couple more <code class="docutils literal notranslate"><span class="pre">Polyclonal</span></code> models to determine if adding a second or third concentration to the best performing single concentration will improve inference of the mutation-escape values. As a reference, we also fit a <code class="docutils literal notranslate"><span class="pre">Polyclonal</span></code> model on all six concentrations.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">conc_sets</span> <span class="o">=</span> <span class="p">[</span>
    <span class="p">[</span><span class="mf">0.25</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span>
    <span class="p">[</span><span class="mf">0.5</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span>
    <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span>
    <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">],</span>
    <span class="p">[</span><span class="mf">0.5</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span>
    <span class="p">[</span><span class="mf">0.25</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">],</span>
    <span class="p">[</span><span class="mf">0.125</span><span class="p">,</span> <span class="mf">0.25</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">],</span>
<span class="p">]</span>

<span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">conc_sets</span><span class="p">:</span>
    <span class="c1"># These are the keys for fit models</span>
    <span class="n">model_string</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;noisy_</span><span class="si">{</span><span class="n">s</span><span class="si">}</span><span class="s2">conc_3muts&quot;</span>

    <span class="c1"># If the pickled model exists in fit_polyclonal_models directory,</span>
    <span class="c1"># load it and add to fit_models</span>
    <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;fit_polyclonal_models/</span><span class="si">{</span><span class="n">model_string</span><span class="si">}</span><span class="s2">.pkl&quot;</span><span class="p">)</span> <span class="ow">is</span> <span class="kc">True</span><span class="p">:</span>
        <span class="n">model</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;fit_polyclonal_models/</span><span class="si">{</span><span class="n">model_string</span><span class="si">}</span><span class="s2">.pkl&quot;</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">))</span>
        <span class="n">fit_models</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="n">model_string</span><span class="p">:</span> <span class="n">model</span><span class="p">})</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Model with </span><span class="si">{</span><span class="n">s</span><span class="si">}</span><span class="s2"> was already fit.&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># Else, fit a model using fit_polyclonal(), save it to the</span>
        <span class="c1"># fit_polyclonal_models directory, and add to fit_models</span>
        <span class="n">model</span> <span class="o">=</span> <span class="n">fit_polyclonal</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
        <span class="n">fit_models</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="n">model_string</span><span class="p">:</span> <span class="n">model</span><span class="p">})</span>
        <span class="n">pickle</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;fit_polyclonal_models/</span><span class="si">{</span><span class="n">model_string</span><span class="si">}</span><span class="s2">.pkl&quot;</span><span class="p">,</span> <span class="s2">&quot;wb&quot;</span><span class="p">))</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Model with </span><span class="si">{</span><span class="n">s</span><span class="si">}</span><span class="s2"> fit and saved.&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model with [0.25, 1] was already fit.
Model with [0.5, 1] was already fit.
Model with [1, 2] was already fit.
Model with [1, 4] was already fit.
Model with [0.5, 1, 2] was already fit.
Model with [0.25, 1, 4] was already fit.
Model with [0.125, 0.25, 0.5, 1, 2, 4] was already fit.
</pre></div></div>
</div>
<p>Again, lets look at the correlation between “true” and predicted mutation-escape values for each of the fit models.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># add best performing single concentration to concentration sets, so we can plot as reference</span>
<span class="n">conc_sets_to_plot</span> <span class="o">=</span> <span class="p">[[</span><span class="mi">1</span><span class="p">]]</span> <span class="o">+</span> <span class="n">conc_sets</span>

<span class="n">all_corrs</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>

<span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">conc_sets_to_plot</span><span class="p">:</span>
    <span class="n">model</span> <span class="o">=</span> <span class="n">fit_models</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;noisy_</span><span class="si">{</span><span class="n">s</span><span class="si">}</span><span class="s2">conc_3muts&quot;</span><span class="p">]</span>

    <span class="n">mut_escape_pred</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;RBD_mut_escape_df.csv&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span>
        <span class="p">(</span>
            <span class="n">model</span><span class="o">.</span><span class="n">mut_escape_df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
                <span class="n">epitope</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s2">&quot;class &quot;</span> <span class="o">+</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;epitope&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
            <span class="p">)</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;escape&quot;</span><span class="p">:</span> <span class="s2">&quot;predicted escape&quot;</span><span class="p">})</span>
        <span class="p">),</span>
        <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;mutation&quot;</span><span class="p">,</span> <span class="s2">&quot;epitope&quot;</span><span class="p">],</span>
        <span class="n">validate</span><span class="o">=</span><span class="s2">&quot;one_to_one&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">corr</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">mut_escape_pred</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;epitope&quot;</span><span class="p">)</span>
        <span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;escape&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">corr</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s2">&quot;predicted escape&quot;</span><span class="p">])</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span>
        <span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="s2">&quot;correlation (R^2)&quot;</span><span class="p">)</span>
        <span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
    <span class="p">)</span>

    <span class="n">all_corrs</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span>
        <span class="p">[</span>
            <span class="n">all_corrs</span><span class="p">,</span>
            <span class="n">corr</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
                <span class="n">icxx_set</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="sa">f</span><span class="s2">&quot;IC</span><span class="si">{</span><span class="n">wt_data_icxx</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">s</span><span class="p">])]</span>
                <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">corr</span><span class="o">.</span><span class="n">index</span><span class="p">),</span>
                <span class="n">num_concs</span><span class="o">=</span><span class="nb">len</span><span class="p">([</span><span class="n">c</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">s</span><span class="p">]),</span>
            <span class="p">),</span>
        <span class="p">]</span>
    <span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>
<span class="n">alt</span><span class="o">.</span><span class="n">Chart</span><span class="p">(</span><span class="n">all_corrs</span><span class="p">)</span><span class="o">.</span><span class="n">mark_circle</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="mi">125</span><span class="p">)</span><span class="o">.</span><span class="n">encode</span><span class="p">(</span>
    <span class="n">x</span><span class="o">=</span><span class="n">alt</span><span class="o">.</span><span class="n">X</span><span class="p">(</span>
        <span class="s2">&quot;icxx_set:N&quot;</span><span class="p">,</span>
        <span class="n">sort</span><span class="o">=</span><span class="n">alt</span><span class="o">.</span><span class="n">EncodingSortField</span><span class="p">(</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="n">order</span><span class="o">=</span><span class="s2">&quot;descending&quot;</span><span class="p">),</span>
    <span class="p">),</span>
    <span class="n">y</span><span class="o">=</span><span class="s2">&quot;correlation (R^2):Q&quot;</span><span class="p">,</span>
    <span class="n">tooltip</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;icxx_set&quot;</span><span class="p">,</span> <span class="n">alt</span><span class="o">.</span><span class="n">Tooltip</span><span class="p">(</span><span class="s2">&quot;correlation (R^2)&quot;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;.3f&quot;</span><span class="p">)],</span>
    <span class="n">color</span><span class="o">=</span><span class="n">alt</span><span class="o">.</span><span class="n">Color</span><span class="p">(</span>
        <span class="s2">&quot;epitope&quot;</span><span class="p">,</span> <span class="n">scale</span><span class="o">=</span><span class="n">alt</span><span class="o">.</span><span class="n">Scale</span><span class="p">(</span><span class="nb">range</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;#0072B2&quot;</span><span class="p">,</span> <span class="s2">&quot;#CC79A7&quot;</span><span class="p">,</span> <span class="s2">&quot;#4C3549&quot;</span><span class="p">]),</span> <span class="n">legend</span><span class="o">=</span><span class="kc">None</span>
    <span class="p">),</span>
    <span class="n">row</span><span class="o">=</span><span class="s2">&quot;epitope:N&quot;</span><span class="p">,</span>
<span class="p">)</span><span class="o">.</span><span class="n">properties</span><span class="p">(</span>
    <span class="n">width</span><span class="o">=</span><span class="mi">375</span><span class="p">,</span> <span class="n">height</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s2">&quot;inferred vs. true mutation escape values&quot;</span>
<span class="p">)</span><span class="o">.</span><span class="n">configure_axis</span><span class="p">(</span>
    <span class="n">labelLimit</span><span class="o">=</span><span class="mi">300</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/tyu2/.local/lib/python3.8/site-packages/altair/utils/core.py:317: FutureWarning: iteritems is deprecated and will be removed in a future version. Use .items instead.
  for col_name, dtype in df.dtypes.iteritems():
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div id="altair-viz-76de93894ce04dc8950bde4ca99d3450"></div>
<script type="text/javascript">
  var VEGA_DEBUG = (typeof VEGA_DEBUG == "undefined") ? {} : VEGA_DEBUG;
  (function(spec, embedOpt){
    let outputDiv = document.currentScript.previousElementSibling;
    if (outputDiv.id !== "altair-viz-76de93894ce04dc8950bde4ca99d3450") {
      outputDiv = document.getElementById("altair-viz-76de93894ce04dc8950bde4ca99d3450");
    }
    const paths = {
      "vega": "https://cdn.jsdelivr.net/npm//vega@5?noext",
      "vega-lib": "https://cdn.jsdelivr.net/npm//vega-lib?noext",
      "vega-lite": "https://cdn.jsdelivr.net/npm//vega-lite@5.2.0?noext",
      "vega-embed": "https://cdn.jsdelivr.net/npm//vega-embed@6?noext",
    };

    function maybeLoadScript(lib, version) {
      var key = `${lib.replace("-", "")}_version`;
      return (VEGA_DEBUG[key] == version) ?
        Promise.resolve(paths[lib]) :
        new Promise(function(resolve, reject) {
          var s = document.createElement('script');
          document.getElementsByTagName("head")[0].appendChild(s);
          s.async = true;
          s.onload = () => {
            VEGA_DEBUG[key] = version;
            return resolve(paths[lib]);
          };
          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);
          s.src = paths[lib];
        });
    }

    function showError(err) {
      outputDiv.innerHTML = `<div class="error" style="color:red;">${err}</div>`;
      throw err;
    }

    function displayChart(vegaEmbed) {
      vegaEmbed(outputDiv, spec, embedOpt)
        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));
    }

    if(typeof define === "function" && define.amd) {
      requirejs.config({paths});
      require(["vega-embed"], displayChart, err => showError(`Error loading script: ${err.message}`));
    } else {
      maybeLoadScript("vega", "5")
        .then(() => maybeLoadScript("vega-lite", "5.2.0"))
        .then(() => maybeLoadScript("vega-embed", "6"))
        .catch(showError)
        .then(() => displayChart(vegaEmbed));
    }
  })({"config": {"view": {"continuousWidth": 400, "continuousHeight": 300}, "axis": {"labelLimit": 300}}, "data": {"name": "data-a4d697f23857208c41016677ef312adf"}, "mark": {"type": "circle", "size": 125}, "encoding": {"color": {"field": "epitope", "legend": null, "scale": {"range": ["#0072B2", "#CC79A7", "#4C3549"]}, "type": "nominal"}, "row": {"field": "epitope", "type": "nominal"}, "tooltip": [{"field": "icxx_set", "type": "nominal"}, {"field": "correlation (R^2)", "format": ".3f", "type": "quantitative"}], "x": {"field": "icxx_set", "sort": {"field": "x", "order": "descending"}, "type": "nominal"}, "y": {"field": "correlation (R^2)", "type": "quantitative"}}, "height": 200, "title": "inferred vs. true mutation escape values", "width": 375, "$schema": "https://vega.github.io/schema/vega-lite/v5.2.0.json", "datasets": {"data-a4d697f23857208c41016677ef312adf": [{"epitope": "class 1", "correlation (R^2)": 0.47277660829732004, "icxx_set": "IC99.968", "num_concs": 1}, {"epitope": "class 2", "correlation (R^2)": 0.9255322091240006, "icxx_set": "IC99.968", "num_concs": 1}, {"epitope": "class 3", "correlation (R^2)": 0.9266370719977781, "icxx_set": "IC99.968", "num_concs": 1}, {"epitope": "class 1", "correlation (R^2)": 0.6909266584026795, "icxx_set": "IC98.854,IC99.968", "num_concs": 2}, {"epitope": "class 2", "correlation (R^2)": 0.9525609389728366, "icxx_set": "IC98.854,IC99.968", "num_concs": 2}, {"epitope": "class 3", "correlation (R^2)": 0.9404001493279208, "icxx_set": "IC98.854,IC99.968", "num_concs": 2}, {"epitope": "class 1", "correlation (R^2)": 0.6265691352114429, "icxx_set": "IC99.79,IC99.968", "num_concs": 2}, {"epitope": "class 2", "correlation (R^2)": 0.9557568037529991, "icxx_set": "IC99.79,IC99.968", "num_concs": 2}, {"epitope": "class 3", "correlation (R^2)": 0.9387576633420915, "icxx_set": "IC99.79,IC99.968", "num_concs": 2}, {"epitope": "class 1", "correlation (R^2)": 0.4940309374276288, "icxx_set": "IC99.968,IC99.995", "num_concs": 2}, {"epitope": "class 2", "correlation (R^2)": 0.9380210668269039, "icxx_set": "IC99.968,IC99.995", "num_concs": 2}, {"epitope": "class 3", "correlation (R^2)": 0.9271310579812604, "icxx_set": "IC99.968,IC99.995", "num_concs": 2}, {"epitope": "class 1", "correlation (R^2)": 0.5442101528058595, "icxx_set": "IC99.968,IC99.999", "num_concs": 2}, {"epitope": "class 2", "correlation (R^2)": 0.933233971350855, "icxx_set": "IC99.968,IC99.999", "num_concs": 2}, {"epitope": "class 3", "correlation (R^2)": 0.9329567842547906, "icxx_set": "IC99.968,IC99.999", "num_concs": 2}, {"epitope": "class 1", "correlation (R^2)": 0.6672003017549143, "icxx_set": "IC99.79,IC99.968,IC99.995", "num_concs": 3}, {"epitope": "class 2", "correlation (R^2)": 0.9550005721717503, "icxx_set": "IC99.79,IC99.968,IC99.995", "num_concs": 3}, {"epitope": "class 3", "correlation (R^2)": 0.9409394223955386, "icxx_set": "IC99.79,IC99.968,IC99.995", "num_concs": 3}, {"epitope": "class 1", "correlation (R^2)": 0.6903693412396601, "icxx_set": "IC98.854,IC99.968,IC99.999", "num_concs": 3}, {"epitope": "class 2", "correlation (R^2)": 0.973458806521, "icxx_set": "IC98.854,IC99.968,IC99.999", "num_concs": 3}, {"epitope": "class 3", "correlation (R^2)": 0.9496508848872558, "icxx_set": "IC98.854,IC99.968,IC99.999", "num_concs": 3}, {"epitope": "class 1", "correlation (R^2)": 0.7317139483963229, "icxx_set": "IC95.141,IC98.854,IC99.79,IC99.968,IC99.995,IC99.999", "num_concs": 6}, {"epitope": "class 2", "correlation (R^2)": 0.9568631207505862, "icxx_set": "IC95.141,IC98.854,IC99.79,IC99.968,IC99.995,IC99.999", "num_concs": 6}, {"epitope": "class 3", "correlation (R^2)": 0.9549092400778721, "icxx_set": "IC95.141,IC98.854,IC99.79,IC99.968,IC99.995,IC99.999", "num_concs": 6}]}}, {"mode": "vega-lite"});
</script></div>
</div>
<p>We see that the performance of models fit on data with two or three concentrations is nearly indistinguishable from that of the model fit on data with all six concentrations. Additionally, adding two or three concentrations does lead to a modest improvement in the correlation, especially for the class 1 epitope, which is expected to be the hardest to predict since it has the lowest wildtype activity. We can also summarize this by plotting the number of concentrations used for fitting on the
x-axis.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>
<span class="n">alt</span><span class="o">.</span><span class="n">Chart</span><span class="p">(</span><span class="n">all_corrs</span><span class="p">)</span><span class="o">.</span><span class="n">mark_circle</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="mi">125</span><span class="p">)</span><span class="o">.</span><span class="n">encode</span><span class="p">(</span>
    <span class="n">x</span><span class="o">=</span><span class="n">alt</span><span class="o">.</span><span class="n">X</span><span class="p">(</span><span class="s2">&quot;num_concs:O&quot;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="n">alt</span><span class="o">.</span><span class="n">Axis</span><span class="p">(</span><span class="n">labelAngle</span><span class="o">=</span><span class="mi">0</span><span class="p">)),</span>
    <span class="n">y</span><span class="o">=</span><span class="s2">&quot;correlation (R^2):Q&quot;</span><span class="p">,</span>
    <span class="n">column</span><span class="o">=</span><span class="s2">&quot;epitope:N&quot;</span><span class="p">,</span>
    <span class="n">tooltip</span><span class="o">=</span><span class="p">[</span>
        <span class="s2">&quot;icxx_set:O&quot;</span><span class="p">,</span>
        <span class="s2">&quot;num_concs:O&quot;</span><span class="p">,</span>
        <span class="n">alt</span><span class="o">.</span><span class="n">Tooltip</span><span class="p">(</span><span class="s2">&quot;correlation (R^2)&quot;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;.3f&quot;</span><span class="p">),</span>
    <span class="p">],</span>
    <span class="n">color</span><span class="o">=</span><span class="n">alt</span><span class="o">.</span><span class="n">Color</span><span class="p">(</span>
        <span class="s2">&quot;epitope&quot;</span><span class="p">,</span> <span class="n">scale</span><span class="o">=</span><span class="n">alt</span><span class="o">.</span><span class="n">Scale</span><span class="p">(</span><span class="nb">range</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;#0072B2&quot;</span><span class="p">,</span> <span class="s2">&quot;#CC79A7&quot;</span><span class="p">,</span> <span class="s2">&quot;#4C3549&quot;</span><span class="p">]),</span> <span class="n">legend</span><span class="o">=</span><span class="kc">None</span>
    <span class="p">),</span>
<span class="p">)</span><span class="o">.</span><span class="n">properties</span><span class="p">(</span><span class="n">width</span><span class="o">=</span><span class="mi">150</span><span class="p">,</span> <span class="n">height</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s2">&quot;inferred vs. true mutation escape values&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div id="altair-viz-f7fa1cfbcaef410a96500452d34e7882"></div>
<script type="text/javascript">
  var VEGA_DEBUG = (typeof VEGA_DEBUG == "undefined") ? {} : VEGA_DEBUG;
  (function(spec, embedOpt){
    let outputDiv = document.currentScript.previousElementSibling;
    if (outputDiv.id !== "altair-viz-f7fa1cfbcaef410a96500452d34e7882") {
      outputDiv = document.getElementById("altair-viz-f7fa1cfbcaef410a96500452d34e7882");
    }
    const paths = {
      "vega": "https://cdn.jsdelivr.net/npm//vega@5?noext",
      "vega-lib": "https://cdn.jsdelivr.net/npm//vega-lib?noext",
      "vega-lite": "https://cdn.jsdelivr.net/npm//vega-lite@5.2.0?noext",
      "vega-embed": "https://cdn.jsdelivr.net/npm//vega-embed@6?noext",
    };

    function maybeLoadScript(lib, version) {
      var key = `${lib.replace("-", "")}_version`;
      return (VEGA_DEBUG[key] == version) ?
        Promise.resolve(paths[lib]) :
        new Promise(function(resolve, reject) {
          var s = document.createElement('script');
          document.getElementsByTagName("head")[0].appendChild(s);
          s.async = true;
          s.onload = () => {
            VEGA_DEBUG[key] = version;
            return resolve(paths[lib]);
          };
          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);
          s.src = paths[lib];
        });
    }

    function showError(err) {
      outputDiv.innerHTML = `<div class="error" style="color:red;">${err}</div>`;
      throw err;
    }

    function displayChart(vegaEmbed) {
      vegaEmbed(outputDiv, spec, embedOpt)
        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));
    }

    if(typeof define === "function" && define.amd) {
      requirejs.config({paths});
      require(["vega-embed"], displayChart, err => showError(`Error loading script: ${err.message}`));
    } else {
      maybeLoadScript("vega", "5")
        .then(() => maybeLoadScript("vega-lite", "5.2.0"))
        .then(() => maybeLoadScript("vega-embed", "6"))
        .catch(showError)
        .then(() => displayChart(vegaEmbed));
    }
  })({"config": {"view": {"continuousWidth": 400, "continuousHeight": 300}}, "data": {"name": "data-a4d697f23857208c41016677ef312adf"}, "mark": {"type": "circle", "size": 125}, "encoding": {"color": {"field": "epitope", "legend": null, "scale": {"range": ["#0072B2", "#CC79A7", "#4C3549"]}, "type": "nominal"}, "column": {"field": "epitope", "type": "nominal"}, "tooltip": [{"field": "icxx_set", "type": "ordinal"}, {"field": "num_concs", "type": "ordinal"}, {"field": "correlation (R^2)", "format": ".3f", "type": "quantitative"}], "x": {"axis": {"labelAngle": 0}, "field": "num_concs", "type": "ordinal"}, "y": {"field": "correlation (R^2)", "type": "quantitative"}}, "height": 200, "title": "inferred vs. true mutation escape values", "width": 150, "$schema": "https://vega.github.io/schema/vega-lite/v5.2.0.json", "datasets": {"data-a4d697f23857208c41016677ef312adf": [{"epitope": "class 1", "correlation (R^2)": 0.47277660829732004, "icxx_set": "IC99.968", "num_concs": 1}, {"epitope": "class 2", "correlation (R^2)": 0.9255322091240006, "icxx_set": "IC99.968", "num_concs": 1}, {"epitope": "class 3", "correlation (R^2)": 0.9266370719977781, "icxx_set": "IC99.968", "num_concs": 1}, {"epitope": "class 1", "correlation (R^2)": 0.6909266584026795, "icxx_set": "IC98.854,IC99.968", "num_concs": 2}, {"epitope": "class 2", "correlation (R^2)": 0.9525609389728366, "icxx_set": "IC98.854,IC99.968", "num_concs": 2}, {"epitope": "class 3", "correlation (R^2)": 0.9404001493279208, "icxx_set": "IC98.854,IC99.968", "num_concs": 2}, {"epitope": "class 1", "correlation (R^2)": 0.6265691352114429, "icxx_set": "IC99.79,IC99.968", "num_concs": 2}, {"epitope": "class 2", "correlation (R^2)": 0.9557568037529991, "icxx_set": "IC99.79,IC99.968", "num_concs": 2}, {"epitope": "class 3", "correlation (R^2)": 0.9387576633420915, "icxx_set": "IC99.79,IC99.968", "num_concs": 2}, {"epitope": "class 1", "correlation (R^2)": 0.4940309374276288, "icxx_set": "IC99.968,IC99.995", "num_concs": 2}, {"epitope": "class 2", "correlation (R^2)": 0.9380210668269039, "icxx_set": "IC99.968,IC99.995", "num_concs": 2}, {"epitope": "class 3", "correlation (R^2)": 0.9271310579812604, "icxx_set": "IC99.968,IC99.995", "num_concs": 2}, {"epitope": "class 1", "correlation (R^2)": 0.5442101528058595, "icxx_set": "IC99.968,IC99.999", "num_concs": 2}, {"epitope": "class 2", "correlation (R^2)": 0.933233971350855, "icxx_set": "IC99.968,IC99.999", "num_concs": 2}, {"epitope": "class 3", "correlation (R^2)": 0.9329567842547906, "icxx_set": "IC99.968,IC99.999", "num_concs": 2}, {"epitope": "class 1", "correlation (R^2)": 0.6672003017549143, "icxx_set": "IC99.79,IC99.968,IC99.995", "num_concs": 3}, {"epitope": "class 2", "correlation (R^2)": 0.9550005721717503, "icxx_set": "IC99.79,IC99.968,IC99.995", "num_concs": 3}, {"epitope": "class 3", "correlation (R^2)": 0.9409394223955386, "icxx_set": "IC99.79,IC99.968,IC99.995", "num_concs": 3}, {"epitope": "class 1", "correlation (R^2)": 0.6903693412396601, "icxx_set": "IC98.854,IC99.968,IC99.999", "num_concs": 3}, {"epitope": "class 2", "correlation (R^2)": 0.973458806521, "icxx_set": "IC98.854,IC99.968,IC99.999", "num_concs": 3}, {"epitope": "class 3", "correlation (R^2)": 0.9496508848872558, "icxx_set": "IC98.854,IC99.968,IC99.999", "num_concs": 3}, {"epitope": "class 1", "correlation (R^2)": 0.7317139483963229, "icxx_set": "IC95.141,IC98.854,IC99.79,IC99.968,IC99.995,IC99.999", "num_concs": 6}, {"epitope": "class 2", "correlation (R^2)": 0.9568631207505862, "icxx_set": "IC95.141,IC98.854,IC99.79,IC99.968,IC99.995,IC99.999", "num_concs": 6}, {"epitope": "class 3", "correlation (R^2)": 0.9549092400778721, "icxx_set": "IC95.141,IC98.854,IC99.79,IC99.968,IC99.995,IC99.999", "num_concs": 6}]}}, {"mode": "vega-lite"});
</script></div>
</div>
<p>Again, lets look at the correlation between “true” and predicted IC90’s for each of the fit models.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ic90_corrs</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>

<span class="n">max_ic90</span> <span class="o">=</span> <span class="mi">50</span>
<span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">conc_sets_to_plot</span><span class="p">:</span>
    <span class="n">model</span> <span class="o">=</span> <span class="n">fit_models</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;noisy_</span><span class="si">{</span><span class="n">s</span><span class="si">}</span><span class="s2">conc_3muts&quot;</span><span class="p">]</span>

    <span class="n">ic90s</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">exact_data</span><span class="p">[[</span><span class="s2">&quot;aa_substitutions&quot;</span><span class="p">,</span> <span class="s2">&quot;IC90&quot;</span><span class="p">]]</span>
        <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">IC90</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;IC90&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">clip</span><span class="p">(</span><span class="n">upper</span><span class="o">=</span><span class="n">max_ic90</span><span class="p">))</span>
        <span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
    <span class="p">)</span>
    <span class="n">ic90s</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">filter_variants_by_seen_muts</span><span class="p">(</span><span class="n">ic90s</span><span class="p">)</span>
    <span class="n">ic90s</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">icXX</span><span class="p">(</span><span class="n">ic90s</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="mf">0.9</span><span class="p">,</span> <span class="n">col</span><span class="o">=</span><span class="s2">&quot;predicted_IC90&quot;</span><span class="p">,</span> <span class="n">max_c</span><span class="o">=</span><span class="n">max_ic90</span><span class="p">)</span>

    <span class="n">ic90s</span> <span class="o">=</span> <span class="n">ic90s</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
        <span class="n">log_IC90</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s2">&quot;IC90&quot;</span><span class="p">]),</span>
        <span class="n">predicted_log_IC90</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s2">&quot;predicted_IC90&quot;</span><span class="p">]),</span>
    <span class="p">)</span>

    <span class="n">corr</span> <span class="o">=</span> <span class="n">ic90s</span><span class="p">[</span><span class="s2">&quot;log_IC90&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">corr</span><span class="p">(</span><span class="n">ic90s</span><span class="p">[</span><span class="s2">&quot;predicted_log_IC90&quot;</span><span class="p">])</span> <span class="o">**</span> <span class="mi">2</span>

    <span class="n">ic90_corrs</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span>
        <span class="p">[</span>
            <span class="n">ic90_corrs</span><span class="p">,</span>
            <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
                <span class="p">{</span>
                    <span class="s2">&quot;correlation (R^2)&quot;</span><span class="p">:</span> <span class="n">corr</span><span class="p">,</span>
                    <span class="s2">&quot;icxx_set&quot;</span><span class="p">:</span> <span class="p">[[</span><span class="sa">f</span><span class="s2">&quot;IC</span><span class="si">{</span><span class="n">wt_data_icxx</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">s</span><span class="p">]],</span>
                <span class="p">}</span>
            <span class="p">),</span>
        <span class="p">]</span>
    <span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>
<span class="n">alt</span><span class="o">.</span><span class="n">Chart</span><span class="p">(</span><span class="n">ic90_corrs</span><span class="p">)</span><span class="o">.</span><span class="n">mark_circle</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="mi">125</span><span class="p">)</span><span class="o">.</span><span class="n">encode</span><span class="p">(</span>
    <span class="n">x</span><span class="o">=</span><span class="n">alt</span><span class="o">.</span><span class="n">X</span><span class="p">(</span>
        <span class="s2">&quot;icxx_set:O&quot;</span><span class="p">,</span>
        <span class="n">sort</span><span class="o">=</span><span class="n">alt</span><span class="o">.</span><span class="n">EncodingSortField</span><span class="p">(</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="n">order</span><span class="o">=</span><span class="s2">&quot;descending&quot;</span><span class="p">),</span>
        <span class="n">axis</span><span class="o">=</span><span class="n">alt</span><span class="o">.</span><span class="n">Axis</span><span class="p">(</span><span class="n">labelAngle</span><span class="o">=</span><span class="mi">0</span><span class="p">),</span>
    <span class="p">),</span>
    <span class="n">y</span><span class="o">=</span><span class="s2">&quot;correlation (R^2):Q&quot;</span><span class="p">,</span>
    <span class="n">tooltip</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;icxx_set&quot;</span><span class="p">,</span> <span class="n">alt</span><span class="o">.</span><span class="n">Tooltip</span><span class="p">(</span><span class="s2">&quot;correlation (R^2)&quot;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;.3f&quot;</span><span class="p">)],</span>
<span class="p">)</span><span class="o">.</span><span class="n">properties</span><span class="p">(</span><span class="n">width</span><span class="o">=</span><span class="mi">375</span><span class="p">,</span> <span class="n">height</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s2">&quot;predicted vs. true IC90&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/tyu2/.local/lib/python3.8/site-packages/altair/utils/core.py:317: FutureWarning: iteritems is deprecated and will be removed in a future version. Use .items instead.
  for col_name, dtype in df.dtypes.iteritems():
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div id="altair-viz-77fd88ba83ec4c418398a88c2dd39a86"></div>
<script type="text/javascript">
  var VEGA_DEBUG = (typeof VEGA_DEBUG == "undefined") ? {} : VEGA_DEBUG;
  (function(spec, embedOpt){
    let outputDiv = document.currentScript.previousElementSibling;
    if (outputDiv.id !== "altair-viz-77fd88ba83ec4c418398a88c2dd39a86") {
      outputDiv = document.getElementById("altair-viz-77fd88ba83ec4c418398a88c2dd39a86");
    }
    const paths = {
      "vega": "https://cdn.jsdelivr.net/npm//vega@5?noext",
      "vega-lib": "https://cdn.jsdelivr.net/npm//vega-lib?noext",
      "vega-lite": "https://cdn.jsdelivr.net/npm//vega-lite@5.2.0?noext",
      "vega-embed": "https://cdn.jsdelivr.net/npm//vega-embed@6?noext",
    };

    function maybeLoadScript(lib, version) {
      var key = `${lib.replace("-", "")}_version`;
      return (VEGA_DEBUG[key] == version) ?
        Promise.resolve(paths[lib]) :
        new Promise(function(resolve, reject) {
          var s = document.createElement('script');
          document.getElementsByTagName("head")[0].appendChild(s);
          s.async = true;
          s.onload = () => {
            VEGA_DEBUG[key] = version;
            return resolve(paths[lib]);
          };
          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);
          s.src = paths[lib];
        });
    }

    function showError(err) {
      outputDiv.innerHTML = `<div class="error" style="color:red;">${err}</div>`;
      throw err;
    }

    function displayChart(vegaEmbed) {
      vegaEmbed(outputDiv, spec, embedOpt)
        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));
    }

    if(typeof define === "function" && define.amd) {
      requirejs.config({paths});
      require(["vega-embed"], displayChart, err => showError(`Error loading script: ${err.message}`));
    } else {
      maybeLoadScript("vega", "5")
        .then(() => maybeLoadScript("vega-lite", "5.2.0"))
        .then(() => maybeLoadScript("vega-embed", "6"))
        .catch(showError)
        .then(() => displayChart(vegaEmbed));
    }
  })({"config": {"view": {"continuousWidth": 400, "continuousHeight": 300}}, "data": {"name": "data-dbc277a6a2b7f964ce12499caf84a24e"}, "mark": {"type": "circle", "size": 125}, "encoding": {"tooltip": [{"field": "icxx_set", "type": "nominal"}, {"field": "correlation (R^2)", "format": ".3f", "type": "quantitative"}], "x": {"axis": {"labelAngle": 0}, "field": "icxx_set", "sort": {"field": "x", "order": "descending"}, "type": "ordinal"}, "y": {"field": "correlation (R^2)", "type": "quantitative"}}, "height": 200, "title": "predicted vs. true IC90", "width": 375, "$schema": "https://vega.github.io/schema/vega-lite/v5.2.0.json", "datasets": {"data-dbc277a6a2b7f964ce12499caf84a24e": [{"correlation (R^2)": 0.8524044532251411, "icxx_set": ["IC99.968"]}, {"correlation (R^2)": 0.96607646947196, "icxx_set": ["IC98.854", "IC99.968"]}, {"correlation (R^2)": 0.9425893159550646, "icxx_set": ["IC99.79", "IC99.968"]}, {"correlation (R^2)": 0.8242406059856914, "icxx_set": ["IC99.968", "IC99.995"]}, {"correlation (R^2)": 0.8256362539344192, "icxx_set": ["IC99.968", "IC99.999"]}, {"correlation (R^2)": 0.9199170200617324, "icxx_set": ["IC99.79", "IC99.968", "IC99.995"]}, {"correlation (R^2)": 0.9647391925377282, "icxx_set": ["IC98.854", "IC99.968", "IC99.999"]}, {"correlation (R^2)": 0.9096933604307439, "icxx_set": ["IC95.141", "IC98.854", "IC99.79", "IC99.968", "IC99.995", "IC99.999"]}]}}, {"mode": "vega-lite"});
</script></div>
</div>
<p>Fitting with more concentrations maintained the excellent IC90 prediction exhibited by models fit on data with a single concentration.</p>
<section id="Summary">
<h2>Summary<a class="headerlink" href="#Summary" title="Permalink to this heading">¶</a></h2>
<p>Based on these simulation experiments:</p>
<ol class="arabic simple">
<li><p>Choose a single concentration that is potent enough (ex. <code class="docutils literal notranslate"><span class="pre">IC99.9</span></code>) to neutralize all wildtype viruses, but not potent enough to neutralize all the variants.</p></li>
</ol>
<ol class="arabic simple" start="2">
<li><p>Adding a second or third concentration can help refine inference of true mutation-escape values, especially for the most subdominant epitopes. Due to technical difficulties in precisely achieving the <code class="docutils literal notranslate"><span class="pre">IC99.9</span></code> in a given experiment, we suggest using concentrations that are 2-4 fold higher or lower than <code class="docutils literal notranslate"><span class="pre">IC99.9</span></code> in order to more likely span the correct dynamic range.</p></li>
</ol>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="index.html">
    <img class="logo" src="_static/BloomLogo.jpg" alt="Logo"/>
    
    <h1 class="logo logo-name">polyclonal</h1>
    
  </a>
</p>



<p class="blurb">Model mutational escape from polyclonal antibodies.</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=jbloomlab&repo=polyclonal&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





<h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="index.html">polyclonal documentation</a></li>
</ul>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="biophysical_model.html">Biophysical model</a></li>
<li class="toctree-l1"><a class="reference internal" href="visualize_RBD.html">Visualize antibody mix</a></li>
<li class="toctree-l1"><a class="reference internal" href="simulate_RBD.html">Simulate experiment</a></li>
<li class="toctree-l1"><a class="reference internal" href="fit_RBD.html">Fit model to data</a></li>
<li class="toctree-l1"><a class="reference internal" href="fit_2epitope.html">Fit more complex curves</a></li>
<li class="toctree-l1"><a class="reference internal" href="regularization.html">Regularization</a></li>
<li class="toctree-l1"><a class="reference internal" href="RBD_average.html">Averaging models</a></li>
<li class="toctree-l1"><a class="reference internal" href="real_LyCoV1404.html">Real antibody example</a></li>
<li class="toctree-l1"><a class="reference internal" href="reference_site_numbering.html">Reference site numbering</a></li>
<li class="toctree-l1"><a class="reference internal" href="real_cocktail.html">Real antibody cocktail</a></li>
<li class="toctree-l1"><a class="reference internal" href="specify_epitopes.html">Specifying epitopes</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="expt_design.html">Experimental design</a></li>
<li class="toctree-l1"><a class="reference internal" href="optimization.html">Details on fitting</a></li>
<li class="toctree-l1"><a class="reference internal" href="partition_function.html">Distinct epitopes assumption</a></li>
<li class="toctree-l1"><a class="reference internal" href="polyclonal.html">polyclonal package</a></li>
<li class="toctree-l1"><a class="reference internal" href="package_index.html">package index</a></li>
<li class="toctree-l1"><a class="reference internal" href="acknowledgments.html">Acknowledgements</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="expt_design.html">Experimental design</a><ul>
      <li>Previous: <a href="mutation_rate.html" title="previous chapter">Library mutation rate</a></li>
      <li>Next: <a href="optimization.html" title="next chapter">Details on fitting</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2019--2023.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 6.1.3</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.13</a>
      
      |
      <a href="_sources/concentration_set.nblink.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/jbloomlab/polyclonal" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>