
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Library mutation rate &#8212; polyclonal 2.2 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css" />
    <link rel="stylesheet" type="text/css" href="_static/plot_directive.css" />
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Sera concentrations" href="concentration_set.html" />
    <link rel="prev" title="Library size" href="library_size.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt a.copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}
</style>
<section id="Library-mutation-rate">
<h1>Library mutation rate<a class="headerlink" href="#Library-mutation-rate" title="Permalink to this headline">¶</a></h1>
<p>We’ll use simulated data to show how the average mutation rate of variants in a DMS library affects the performance of <code class="docutils literal notranslate"><span class="pre">Polyclonal</span></code> models.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">pickle</span>

<span class="kn">import</span> <span class="nn">altair</span> <span class="k">as</span> <span class="nn">alt</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">polyclonal</span>
</pre></div>
</div>
</div>
<p>First, we read in a four simulated “noisy” libraries, each measured at three sera concentrations. The variants in these libraries were simulated to contain a Poisson-distributed number of mutations. The libraries differ in their average number of mutations (1, 2, 3, or 4) per gene, and are named accordingly.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">noisy_data</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;RBD_variants_escape_noisy.csv&quot;</span><span class="p">,</span> <span class="n">na_filter</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
    <span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s2">&quot;concentration in [0.25, 1, 4]&quot;</span><span class="p">)</span>
    <span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="p">)</span>

<span class="n">noisy_data</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>library</th>
      <th>barcode</th>
      <th>concentration</th>
      <th>prob_escape</th>
      <th>aa_substitutions</th>
      <th>IC90</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>avg1muts</td>
      <td>AAAAACGCGGTCACTT</td>
      <td>0.25</td>
      <td>0.08748</td>
      <td></td>
      <td>0.1128</td>
    </tr>
    <tr>
      <th>1</th>
      <td>avg1muts</td>
      <td>AAAAAGCACCATAACG</td>
      <td>0.25</td>
      <td>0.03424</td>
      <td></td>
      <td>0.1128</td>
    </tr>
    <tr>
      <th>2</th>
      <td>avg1muts</td>
      <td>AAAAAGGGAGCTAAAC</td>
      <td>0.25</td>
      <td>0.03788</td>
      <td></td>
      <td>0.1128</td>
    </tr>
    <tr>
      <th>3</th>
      <td>avg1muts</td>
      <td>AAAAAGTCGGATGGGC</td>
      <td>0.25</td>
      <td>0.03573</td>
      <td></td>
      <td>0.1128</td>
    </tr>
    <tr>
      <th>4</th>
      <td>avg1muts</td>
      <td>AAAAAGTCTTAGGGAA</td>
      <td>0.25</td>
      <td>0.00000</td>
      <td></td>
      <td>0.1128</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>359995</th>
      <td>avg4muts</td>
      <td>GGGAAATTAGACTTTC</td>
      <td>1.00</td>
      <td>0.03703</td>
      <td>Y508W C525F</td>
      <td>0.4073</td>
    </tr>
    <tr>
      <th>359996</th>
      <td>avg4muts</td>
      <td>GGGAAATTAGACTTTC</td>
      <td>4.00</td>
      <td>0.03618</td>
      <td>Y508W C525F</td>
      <td>0.4073</td>
    </tr>
    <tr>
      <th>359997</th>
      <td>avg4muts</td>
      <td>GCCACGTTGCAATTCA</td>
      <td>0.25</td>
      <td>0.25930</td>
      <td>Y508W G526L</td>
      <td>0.4122</td>
    </tr>
    <tr>
      <th>359998</th>
      <td>avg4muts</td>
      <td>GCCACGTTGCAATTCA</td>
      <td>1.00</td>
      <td>0.00000</td>
      <td>Y508W G526L</td>
      <td>0.4122</td>
    </tr>
    <tr>
      <th>359999</th>
      <td>avg4muts</td>
      <td>GCCACGTTGCAATTCA</td>
      <td>4.00</td>
      <td>0.00000</td>
      <td>Y508W G526L</td>
      <td>0.4122</td>
    </tr>
  </tbody>
</table>
<p>360000 rows × 6 columns</p>
</div></div>
</div>
<p>Now, we’ll fit a <code class="docutils literal notranslate"><span class="pre">Polyclonal</span></code> model to data in each library. We’ll initialize each <code class="docutils literal notranslate"><span class="pre">Polyclonal</span></code> model with the same values. We know from <a class="reference external" href="https://www.nature.com/articles/s41467-021-24435-8">prior work</a> the three most important epitopes and a key mutation in each, so we use this prior knowledge to “seed” initial guesses that assign large escape values to a key site in each epitope:</p>
<ul class="simple">
<li><p>site 417 for class 1 epitope, which is often the least important</p></li>
</ul>
<ul class="simple">
<li><p>site 484 for class 2 epitope, which is often the dominant one</p></li>
</ul>
<ul class="simple">
<li><p>site 444 for class 3 epitope, which is often the second most dominant one</p></li>
</ul>
<p>Additionally, we’ll store fit models as <a class="reference external" href="https://docs.python.org/3/library/pickle.html#module-pickle">pickle</a> files, so that we can conveniently load them in the future without having to fit again.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">avg_mut_rates</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">]</span>

<span class="c1"># Make a directory to house pickled models</span>
<span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="s2">&quot;fit_polyclonal_models&quot;</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>


<span class="k">def</span> <span class="nf">fit_polyclonal</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Fit `Polyclonal` model with data with a specific average mutation rate.</span>
<span class="sd">    Returns fit `Polyclonal` object.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">poly_abs</span> <span class="o">=</span> <span class="n">polyclonal</span><span class="o">.</span><span class="n">Polyclonal</span><span class="p">(</span>
        <span class="n">data_to_fit</span><span class="o">=</span><span class="n">noisy_data</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;library == &#39;avg</span><span class="si">{</span><span class="n">n</span><span class="si">}</span><span class="s2">muts&#39;&quot;</span><span class="p">),</span>
        <span class="n">activity_wt_df</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_records</span><span class="p">(</span>
            <span class="p">[</span>
                <span class="p">(</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">),</span>
                <span class="p">(</span><span class="s2">&quot;2&quot;</span><span class="p">,</span> <span class="mf">3.0</span><span class="p">),</span>
                <span class="p">(</span><span class="s2">&quot;3&quot;</span><span class="p">,</span> <span class="mf">2.0</span><span class="p">),</span>
            <span class="p">],</span>
            <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;epitope&quot;</span><span class="p">,</span> <span class="s2">&quot;activity&quot;</span><span class="p">],</span>
        <span class="p">),</span>
        <span class="n">site_escape_df</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_records</span><span class="p">(</span>
            <span class="p">[</span>
                <span class="p">(</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="mi">417</span><span class="p">,</span> <span class="mf">10.0</span><span class="p">),</span>
                <span class="p">(</span><span class="s2">&quot;2&quot;</span><span class="p">,</span> <span class="mi">484</span><span class="p">,</span> <span class="mf">10.0</span><span class="p">),</span>
                <span class="p">(</span><span class="s2">&quot;3&quot;</span><span class="p">,</span> <span class="mi">444</span><span class="p">,</span> <span class="mf">10.0</span><span class="p">),</span>
            <span class="p">],</span>
            <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;epitope&quot;</span><span class="p">,</span> <span class="s2">&quot;site&quot;</span><span class="p">,</span> <span class="s2">&quot;escape&quot;</span><span class="p">],</span>
        <span class="p">),</span>
        <span class="n">data_mut_escape_overlap</span><span class="o">=</span><span class="s2">&quot;fill_to_data&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">poly_abs</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">reg_escape_weight</span><span class="o">=</span><span class="mf">0.01</span><span class="p">,</span> <span class="n">reg_activity_weight</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">poly_abs</span>


<span class="c1"># Store all fit models in a dictionary for future lookup</span>
<span class="n">fit_models</span> <span class="o">=</span> <span class="p">{}</span>

<span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">avg_mut_rates</span><span class="p">:</span>
    <span class="c1"># These are the keys for fit models</span>
    <span class="n">model_string</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;noisy_[0.25, 1, 4]conc_</span><span class="si">{</span><span class="n">n</span><span class="si">}</span><span class="s2">muts&quot;</span>

    <span class="c1"># If the pickled model exists in fit_polyclonal_models directory,</span>
    <span class="c1"># load it and update fit_models</span>
    <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;fit_polyclonal_models/</span><span class="si">{</span><span class="n">model_string</span><span class="si">}</span><span class="s2">.pkl&quot;</span><span class="p">)</span> <span class="ow">is</span> <span class="kc">True</span><span class="p">:</span>
        <span class="n">model</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;fit_polyclonal_models/</span><span class="si">{</span><span class="n">model_string</span><span class="si">}</span><span class="s2">.pkl&quot;</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">))</span>
        <span class="n">fit_models</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="n">model_string</span><span class="p">:</span> <span class="n">model</span><span class="p">})</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Model on data with </span><span class="si">{</span><span class="n">n</span><span class="si">}</span><span class="s2"> average mutations was already fit.&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># Else, fit a model using fit_polyclonal(), save it to the</span>
        <span class="c1"># fit_polyclonal_models directory, and update fit_models</span>
        <span class="n">model</span> <span class="o">=</span> <span class="n">fit_polyclonal</span><span class="p">(</span><span class="n">n</span><span class="p">)</span>
        <span class="n">fit_models</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="n">model_string</span><span class="p">:</span> <span class="n">model</span><span class="p">})</span>
        <span class="n">pickle</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;fit_polyclonal_models/</span><span class="si">{</span><span class="n">model_string</span><span class="si">}</span><span class="s2">.pkl&quot;</span><span class="p">,</span> <span class="s2">&quot;wb&quot;</span><span class="p">))</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Model on data with </span><span class="si">{</span><span class="n">n</span><span class="si">}</span><span class="s2"> average mutations fit and saved.&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model on data with 1 average mutations was already fit.
Model on data with 2 average mutations was already fit.
Model on data with 3 average mutations was already fit.
Model on data with 4 average mutations was already fit.
</pre></div></div>
</div>
<p>We can look at the correlation between the “true” and inferred mutation-escape values, <span class="math notranslate nohighlight">\(\beta_{m,e}\)</span>, for the fit models. These mutation-escape values represent the extent to which mutations mediate escape from specific epitopes.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">all_corrs</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;epitope&quot;</span><span class="p">:</span> <span class="p">[],</span> <span class="s2">&quot;correlation (R^2)&quot;</span><span class="p">:</span> <span class="p">[],</span> <span class="s2">&quot;mutation_rate&quot;</span><span class="p">:</span> <span class="p">[]})</span>

<span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">avg_mut_rates</span><span class="p">:</span>
    <span class="n">model</span> <span class="o">=</span> <span class="n">fit_models</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;noisy_[0.25, 1, 4]conc_</span><span class="si">{</span><span class="n">n</span><span class="si">}</span><span class="s2">muts&quot;</span><span class="p">]</span>

    <span class="n">mut_escape_pred</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;RBD_mut_escape_df.csv&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span>
        <span class="p">(</span>
            <span class="n">model</span><span class="o">.</span><span class="n">mut_escape_df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
                <span class="n">epitope</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s2">&quot;class &quot;</span> <span class="o">+</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;epitope&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
            <span class="p">)</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;escape&quot;</span><span class="p">:</span> <span class="s2">&quot;predicted escape&quot;</span><span class="p">})</span>
        <span class="p">),</span>
        <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;mutation&quot;</span><span class="p">,</span> <span class="s2">&quot;epitope&quot;</span><span class="p">],</span>
        <span class="n">validate</span><span class="o">=</span><span class="s2">&quot;one_to_one&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">corr</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">mut_escape_pred</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;epitope&quot;</span><span class="p">)</span>
        <span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;escape&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">corr</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s2">&quot;predicted escape&quot;</span><span class="p">])</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span>
        <span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="s2">&quot;correlation (R^2)&quot;</span><span class="p">)</span>
        <span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
    <span class="p">)</span>

    <span class="n">all_corrs</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span>
        <span class="p">[</span><span class="n">all_corrs</span><span class="p">,</span> <span class="n">corr</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">mutation_rate</span><span class="o">=</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;avg</span><span class="si">{</span><span class="n">n</span><span class="si">}</span><span class="s2">muts&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">corr</span><span class="o">.</span><span class="n">index</span><span class="p">))]</span>
    <span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>
<span class="n">alt</span><span class="o">.</span><span class="n">Chart</span><span class="p">(</span><span class="n">all_corrs</span><span class="p">)</span><span class="o">.</span><span class="n">mark_circle</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="mi">125</span><span class="p">)</span><span class="o">.</span><span class="n">encode</span><span class="p">(</span>
    <span class="n">x</span><span class="o">=</span><span class="n">alt</span><span class="o">.</span><span class="n">X</span><span class="p">(</span><span class="s2">&quot;mutation_rate:O&quot;</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="n">alt</span><span class="o">.</span><span class="n">EncodingSortField</span><span class="p">(</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="n">order</span><span class="o">=</span><span class="s2">&quot;descending&quot;</span><span class="p">)),</span>
    <span class="n">y</span><span class="o">=</span><span class="s2">&quot;correlation (R^2):Q&quot;</span><span class="p">,</span>
    <span class="n">column</span><span class="o">=</span><span class="s2">&quot;epitope:N&quot;</span><span class="p">,</span>
    <span class="n">tooltip</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;mutation_rate&quot;</span><span class="p">,</span> <span class="n">alt</span><span class="o">.</span><span class="n">Tooltip</span><span class="p">(</span><span class="s2">&quot;correlation (R^2)&quot;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;.3f&quot;</span><span class="p">)],</span>
    <span class="n">color</span><span class="o">=</span><span class="n">alt</span><span class="o">.</span><span class="n">Color</span><span class="p">(</span>
        <span class="s2">&quot;epitope&quot;</span><span class="p">,</span> <span class="n">scale</span><span class="o">=</span><span class="n">alt</span><span class="o">.</span><span class="n">Scale</span><span class="p">(</span><span class="nb">range</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;#0072B2&quot;</span><span class="p">,</span> <span class="s2">&quot;#009E73&quot;</span><span class="p">,</span> <span class="s2">&quot;#CC79A7&quot;</span><span class="p">]),</span> <span class="n">legend</span><span class="o">=</span><span class="kc">None</span>
    <span class="p">),</span>
<span class="p">)</span><span class="o">.</span><span class="n">properties</span><span class="p">(</span><span class="n">width</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">height</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s2">&quot;inferred vs. true mutation escape values&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div id="altair-viz-6416e22e389b4291b6e4eca09d47bb0d"></div>
<script type="text/javascript">
  var VEGA_DEBUG = (typeof VEGA_DEBUG == "undefined") ? {} : VEGA_DEBUG;
  (function(spec, embedOpt){
    let outputDiv = document.currentScript.previousElementSibling;
    if (outputDiv.id !== "altair-viz-6416e22e389b4291b6e4eca09d47bb0d") {
      outputDiv = document.getElementById("altair-viz-6416e22e389b4291b6e4eca09d47bb0d");
    }
    const paths = {
      "vega": "https://cdn.jsdelivr.net/npm//vega@5?noext",
      "vega-lib": "https://cdn.jsdelivr.net/npm//vega-lib?noext",
      "vega-lite": "https://cdn.jsdelivr.net/npm//vega-lite@5.2.0?noext",
      "vega-embed": "https://cdn.jsdelivr.net/npm//vega-embed@6?noext",
    };

    function maybeLoadScript(lib, version) {
      var key = `${lib.replace("-", "")}_version`;
      return (VEGA_DEBUG[key] == version) ?
        Promise.resolve(paths[lib]) :
        new Promise(function(resolve, reject) {
          var s = document.createElement('script');
          document.getElementsByTagName("head")[0].appendChild(s);
          s.async = true;
          s.onload = () => {
            VEGA_DEBUG[key] = version;
            return resolve(paths[lib]);
          };
          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);
          s.src = paths[lib];
        });
    }

    function showError(err) {
      outputDiv.innerHTML = `<div class="error" style="color:red;">${err}</div>`;
      throw err;
    }

    function displayChart(vegaEmbed) {
      vegaEmbed(outputDiv, spec, embedOpt)
        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));
    }

    if(typeof define === "function" && define.amd) {
      requirejs.config({paths});
      require(["vega-embed"], displayChart, err => showError(`Error loading script: ${err.message}`));
    } else {
      maybeLoadScript("vega", "5")
        .then(() => maybeLoadScript("vega-lite", "5.2.0"))
        .then(() => maybeLoadScript("vega-embed", "6"))
        .catch(showError)
        .then(() => displayChart(vegaEmbed));
    }
  })({"config": {"view": {"continuousWidth": 400, "continuousHeight": 300}}, "data": {"name": "data-2df929f67f8a87bcedc65c2397a19805"}, "mark": {"type": "circle", "size": 125}, "encoding": {"color": {"field": "epitope", "legend": null, "scale": {"range": ["#0072B2", "#009E73", "#CC79A7"]}, "type": "nominal"}, "column": {"field": "epitope", "type": "nominal"}, "tooltip": [{"field": "mutation_rate", "type": "nominal"}, {"field": "correlation (R^2)", "format": ".3f", "type": "quantitative"}], "x": {"field": "mutation_rate", "sort": {"field": "x", "order": "descending"}, "type": "ordinal"}, "y": {"field": "correlation (R^2)", "type": "quantitative"}}, "height": 200, "title": "inferred vs. true mutation escape values", "width": 200, "$schema": "https://vega.github.io/schema/vega-lite/v5.2.0.json", "datasets": {"data-2df929f67f8a87bcedc65c2397a19805": [{"epitope": "class 1", "correlation (R^2)": 0.005858759829229752, "mutation_rate": "avg1muts"}, {"epitope": "class 2", "correlation (R^2)": 0.5680807846818832, "mutation_rate": "avg1muts"}, {"epitope": "class 3", "correlation (R^2)": 0.0017380628946666657, "mutation_rate": "avg1muts"}, {"epitope": "class 1", "correlation (R^2)": 0.6538360924023249, "mutation_rate": "avg2muts"}, {"epitope": "class 2", "correlation (R^2)": 0.9126253026896349, "mutation_rate": "avg2muts"}, {"epitope": "class 3", "correlation (R^2)": 0.8611996177323354, "mutation_rate": "avg2muts"}, {"epitope": "class 1", "correlation (R^2)": 0.7653930008850555, "mutation_rate": "avg3muts"}, {"epitope": "class 2", "correlation (R^2)": 0.946361669262305, "mutation_rate": "avg3muts"}, {"epitope": "class 3", "correlation (R^2)": 0.9064549318480574, "mutation_rate": "avg3muts"}, {"epitope": "class 1", "correlation (R^2)": 0.7837810353244848, "mutation_rate": "avg4muts"}, {"epitope": "class 2", "correlation (R^2)": 0.9414944910250483, "mutation_rate": "avg4muts"}, {"epitope": "class 3", "correlation (R^2)": 0.9140995655059172, "mutation_rate": "avg4muts"}]}}, {"mode": "vega-lite"});
</script></div>
</div>
<section id="Summary">
<h2>Summary<a class="headerlink" href="#Summary" title="Permalink to this headline">¶</a></h2>
<p>With prior knowledge, an average of at least 2 mutations per gene is needed to infer the true mutation-escape values for all epitopes. When there is an average of 1 mutation per gene, the correlation is highest for the most immunodominant epitope 2 and lowest for the most subdominant epitope 1. This is expected, as we should not observe escape for variants with a single mutation in a subdominant epitope.</p>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="index.html">
    <img class="logo" src="_static/BloomLogo.jpg" alt="Logo"/>
    
    <h1 class="logo logo-name">polyclonal</h1>
    
  </a>
</p>



<p class="blurb">Model mutational escape from polyclonal antibodies.</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=jbloomlab&repo=polyclonal&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





<h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="index.html">polyclonal documentation</a></li>
</ul>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="biophysical_model.html">Biophysical model</a></li>
<li class="toctree-l1"><a class="reference internal" href="visualize_RBD.html">Visualize antibody mix</a></li>
<li class="toctree-l1"><a class="reference internal" href="simulate_RBD.html">Simulate experiment</a></li>
<li class="toctree-l1"><a class="reference internal" href="fit_RBD.html">Fit model to data</a></li>
<li class="toctree-l1"><a class="reference internal" href="RBD_bootstrap.html">Bootstrapping model fits</a></li>
<li class="toctree-l1"><a class="reference internal" href="RBD_average.html">Averaging models</a></li>
<li class="toctree-l1"><a class="reference internal" href="real_LyCoV1404.html">Real antibody data example</a></li>
<li class="toctree-l1"><a class="reference internal" href="reference_site_numbering.html">Reference site numbering</a></li>
<li class="toctree-l1"><a class="reference internal" href="specify_epitopes.html">Specifying epitopes</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="expt_design.html">Experimental design</a></li>
<li class="toctree-l1"><a class="reference internal" href="optimization.html">Details on fitting</a></li>
<li class="toctree-l1"><a class="reference internal" href="partition_function.html">Distinct epitopes assumption</a></li>
<li class="toctree-l1"><a class="reference internal" href="polyclonal.html">polyclonal package</a></li>
<li class="toctree-l1"><a class="reference internal" href="package_index.html">package index</a></li>
<li class="toctree-l1"><a class="reference internal" href="acknowledgments.html">Acknowledgements</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="expt_design.html">Experimental design</a><ul>
      <li>Previous: <a href="library_size.html" title="previous chapter">Library size</a></li>
      <li>Next: <a href="concentration_set.html" title="next chapter">Sera concentrations</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2019--2022.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 4.2.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/mutation_rate.nblink.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/jbloomlab/polyclonal" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>