
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>polyclonal &#8212; polyclonal 3.1 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css" />
    <link rel="stylesheet" type="text/css" href="_static/plot_directive.css" />
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="polyclonal_collection" href="polyclonal.polyclonal_collection.html" />
    <link rel="prev" title="plot" href="polyclonal.plot.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <span class="target" id="module-polyclonal.polyclonal"></span><section id="polyclonal">
<h1>polyclonal<a class="headerlink" href="#polyclonal" title="Permalink to this headline">¶</a></h1>
<p>Defines <a class="reference internal" href="#polyclonal.polyclonal.Polyclonal" title="polyclonal.polyclonal.Polyclonal"><code class="xref py py-class docutils literal notranslate"><span class="pre">Polyclonal</span></code></a> objects for handling antibody mixtures.</p>
<dl class="py class">
<dt class="sig sig-object py" id="polyclonal.polyclonal.Polyclonal">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">polyclonal.polyclonal.</span></span><span class="sig-name descname"><span class="pre">Polyclonal</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">activity_wt_df</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mut_escape_df</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data_to_fit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">site_escape_df</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_epitopes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spatial_distances</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">collapse_identical_variants</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alphabet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">('A',</span> <span class="pre">'C',</span> <span class="pre">'D',</span> <span class="pre">'E',</span> <span class="pre">'F',</span> <span class="pre">'G',</span> <span class="pre">'H',</span> <span class="pre">'I',</span> <span class="pre">'K',</span> <span class="pre">'L',</span> <span class="pre">'M',</span> <span class="pre">'N',</span> <span class="pre">'P',</span> <span class="pre">'Q',</span> <span class="pre">'R',</span> <span class="pre">'S',</span> <span class="pre">'T',</span> <span class="pre">'V',</span> <span class="pre">'W',</span> <span class="pre">'Y')</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sites</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">epitope_colors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">('#0072B2',</span> <span class="pre">'#CC79A7',</span> <span class="pre">'#4C3549',</span> <span class="pre">'#009E73')</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">init_missing</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'zero'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data_mut_escape_overlap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'exact_match'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/polyclonal/polyclonal.html#Polyclonal"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polyclonal.polyclonal.Polyclonal" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Represent polyclonal antibody mixtures targeting multiple epitopes.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>At several concentrations <span class="math notranslate nohighlight">\(c\)</span> of an antibody mixture, we measure
<span class="math notranslate nohighlight">\(p_v\left(c\right)\)</span>, the probability that variant <span class="math notranslate nohighlight">\(v\)</span> is
<strong>not</strong> bound (or neutralized). We assume antibodies act independently on
one of <span class="math notranslate nohighlight">\(E\)</span> epitopes, so the probability <span class="math notranslate nohighlight">\(U_e\left(v, c\right)\)</span>
that <span class="math notranslate nohighlight">\(v\)</span> is unbound at concentration <span class="math notranslate nohighlight">\(c\)</span> is related to the
probability that epitope <span class="math notranslate nohighlight">\(e\)</span> is unbound by</p>
<div class="math notranslate nohighlight" id="equation-p-v">
<span class="eqno">(1)<a class="headerlink" href="#equation-p-v" title="Permalink to this equation">¶</a></span>\[p_v\left(c\right) = \prod_{e=1}^E U_e\left(v, c\right).\]</div>
<p>We furthermore assume that <span class="math notranslate nohighlight">\(U_e\left(v, c\right)\)</span> is related to the
total binding activity <span class="math notranslate nohighlight">\(\phi_e\left(v\right)\)</span> of antibodies targeting
epitope <span class="math notranslate nohighlight">\(e\)</span> on variant <span class="math notranslate nohighlight">\(v\)</span> by</p>
<div class="math notranslate nohighlight" id="equation-u-e">
<span class="eqno">(2)<a class="headerlink" href="#equation-u-e" title="Permalink to this equation">¶</a></span>\[U_e\left(v,c\right)=\frac{1}{1+c\exp\left(-\phi_e\left(v\right)\right)}\]</div>
<p>where smaller (more negative) values of <span class="math notranslate nohighlight">\(\phi_e\left(v\right)\)</span>
correspond to higher overall binding activity against epitope <span class="math notranslate nohighlight">\(e\)</span>
variant <span class="math notranslate nohighlight">\(v\)</span>.</p>
<p>We define <span class="math notranslate nohighlight">\(\phi_e\left(v\right)\)</span> in terms of the underlying
quantities of biological interest as</p>
<div class="math notranslate nohighlight" id="equation-phi">
<span class="eqno">(3)<a class="headerlink" href="#equation-phi" title="Permalink to this equation">¶</a></span>\[\phi_e\left(v\right) = -a_{\rm{wt}, e} +
                       \sum_{m=1}^M \beta_{m,e} b\left(v\right)_m,\]</div>
<p>where <span class="math notranslate nohighlight">\(a_{\rm{wt}, e}\)</span> is the activity of the serum against
epitope <span class="math notranslate nohighlight">\(e\)</span> for the “wildtype” (unmutated) protein (larger values
indicate higher activity against this epitope), <span class="math notranslate nohighlight">\(\beta_{m,e}\)</span>
is the extent to which mutation <span class="math notranslate nohighlight">\(m\)</span> (where <span class="math notranslate nohighlight">\(1 \le m \le M\)</span>)
escapes binding from antibodies targeting epitope <span class="math notranslate nohighlight">\(e\)</span> (larger
values indicate more escape by this mutation), and
<span class="math notranslate nohighlight">\(b\left(v\right)_m\)</span> is 1 if variant <span class="math notranslate nohighlight">\(v\)</span> has mutation <span class="math notranslate nohighlight">\(m\)</span>
and 0 otherwise.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>You can initialize a <a class="reference internal" href="#polyclonal.polyclonal.Polyclonal" title="polyclonal.polyclonal.Polyclonal"><code class="xref py py-class docutils literal notranslate"><span class="pre">Polyclonal</span></code></a> object in three ways:</p>
<ol class="arabic simple">
<li><p>With known epitope activities <span class="math notranslate nohighlight">\(a_{\rm{wt}, e}\)</span> and mutation-escape
values <span class="math notranslate nohighlight">\(\beta_{m,e}\)</span>, and no data. Use this approach if you
already know these values and just want to visualize the polyclonal
antibody mixture properties or predict escape of variants. To do this,
initialize with <code class="docutils literal notranslate"><span class="pre">activity_wt_df</span></code> and <code class="docutils literal notranslate"><span class="pre">mut_escape_df</span></code> storing the
known values, and <code class="docutils literal notranslate"><span class="pre">data_to_fit=None</span></code>.</p></li>
<li><p>With data to fit the epitope activities and mutation-escape values,
and initial guesses for the epitope activities and mutation-escape
values. To do this, initialize with <code class="docutils literal notranslate"><span class="pre">data_to_fit</span></code> holding the data,
<code class="docutils literal notranslate"><span class="pre">activity_wt_df</span></code> holding initial guesses of activities, and
<code class="docutils literal notranslate"><span class="pre">mut_escape_df</span></code> or <code class="docutils literal notranslate"><span class="pre">site_escape_df</span></code> holding initial guesses for
mutation escapes (see also <code class="docutils literal notranslate"><span class="pre">init_missing</span></code> and
<code class="docutils literal notranslate"><span class="pre">data_mut_escape_overlap</span></code>). Then call <a class="reference internal" href="#polyclonal.polyclonal.Polyclonal.fit" title="polyclonal.polyclonal.Polyclonal.fit"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Polyclonal.fit()</span></code></a>.</p></li>
<li><p>With data to fit the epitope activities and mutation-escape values,
but no initial guesses. To do this, initialize with <code class="docutils literal notranslate"><span class="pre">data_to_fit</span></code>
holding the data, <code class="docutils literal notranslate"><span class="pre">activity_wt_df=None</span></code>, <code class="docutils literal notranslate"><span class="pre">mut_escape_df=None</span></code>,
and <code class="docutils literal notranslate"><span class="pre">n_epitopes</span></code> holding the number of epitopes. Then call
<a class="reference internal" href="#polyclonal.polyclonal.Polyclonal.fit" title="polyclonal.polyclonal.Polyclonal.fit"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Polyclonal.fit()</span></code></a>.</p></li>
</ol>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_to_fit</strong> (<em>pandas.DataFrame</em><em> or </em><em>None</em>) – Should have columns named ‘aa_substitutions’, ‘concentration’, and
‘prob_escape’. The ‘aa_substitutions’ column defines each variant
<span class="math notranslate nohighlight">\(v\)</span> as a string of substitutions (e.g., ‘M3A K5G’). The
‘prob_escape’ column gives the <span class="math notranslate nohighlight">\(p_v\left(c\right)\)</span> value for
each variant at each concentration <span class="math notranslate nohighlight">\(c\)</span>.</p></li>
<li><p><strong>activity_wt_df</strong> (<em>pandas.DataFrame</em><em> or </em><em>None</em>) – Should have columns named ‘epitope’ and ‘activity’, giving the names
of the epitopes and the activity against epitope in the wildtype
protein, <span class="math notranslate nohighlight">\(a_{\rm{wt}, e}\)</span>.</p></li>
<li><p><strong>mut_escape_df</strong> (<em>pandas.DataFrame</em><em> or </em><em>None</em>) – Should have columns named ‘mutation’, ‘epitope’, and ‘escape’ that
give the <span class="math notranslate nohighlight">\(\beta_{m,e}\)</span> values (in the ‘escape’ column), with
mutations written like “G7M”.</p></li>
<li><p><strong>site_escape_df</strong> (<em>pandas.DataFrame</em><em> or </em><em>None</em>) – Use if you want to initialize all mutations at a given site to have
the same <span class="math notranslate nohighlight">\(\beta_{m,e}\)</span> values. In this case, columns should be
‘site’, ‘epitope’, and ‘escape’. This option is mutually exclusive
with <code class="docutils literal notranslate"><span class="pre">mut_escape_df</span></code>.</p></li>
<li><p><strong>n_epitopes</strong> (<em>int</em><em> or </em><em>None</em>) – If initializing with <code class="docutils literal notranslate"><span class="pre">activity_wt_df=None</span></code>, specifies number
of epitopes.</p></li>
<li><p><strong>spatial_distances</strong> (<em>pandas.DataFrame</em>) – Gives inter-residue distances. Must have columns “site_1”, “site_2”,
and “distance” where the distance is typically in angstroms.
Distances may be missing for some pairs. Used for the spatial regularization.</p></li>
<li><p><strong>collapse_identical_variants</strong> (<em>{'mean'</em><em>, </em><em>'median'</em><em>, </em><em>False}</em>) – If identical variants in <code class="docutils literal notranslate"><span class="pre">data_to_fit</span></code> (same ‘aa_substitutions’),
collapse them and make weight proportional to number of collapsed
variants? Collapse by taking mean or median of ‘prob_escape’, or
(if <cite>False</cite>) do not collapse at all. Collapsing will make fitting faster,
but <em>not</em> a good idea if you are doing bootstrapping.</p></li>
<li><p><strong>alphabet</strong> (<em>array-like</em>) – Allowed characters in mutation strings.</p></li>
<li><p><strong>sites</strong> (<em>array-like</em><em> or </em><em>None</em>) – By default, sites are assumed to be sequential integer values are and inferred
from <code class="docutils literal notranslate"><span class="pre">data_to_fit</span></code> or <code class="docutils literal notranslate"><span class="pre">mut_escape_df</span></code>. However, you can also have
non-sequential integer sites, or sites with lower-case letter suffixes
(eg, <cite>214a</cite>) if your protein is numbered against a reference that it has
indels relative to. In that case, provide list of all expected in order
here; we require that order to be natsorted.</p></li>
<li><p><strong>epitope_colors</strong> (<em>array-like</em><em> or </em><em>dict</em>) – Maps each epitope to the color used for plotting. Either a dict keyed
by each epitope, or an array of colors that are sequentially assigned
to the epitopes.</p></li>
<li><p><strong>init_missing</strong> (<em>'zero'</em><em> or </em><em>int</em>) – How to initialize activities or mutation-escape values not specified in
<code class="docutils literal notranslate"><span class="pre">activity_wt_df</span></code> or <code class="docutils literal notranslate"><span class="pre">mut_escape_df</span></code> / <code class="docutils literal notranslate"><span class="pre">site_escape_df</span></code>. If
‘zero’, set mutation-escapes to zero and activities uniformly spaced
from 1 to 0. Otherwise draw uniformly from between 0 and 1 using
specified random number seed.</p></li>
<li><p><strong>data_mut_escape_overlap</strong> (<em>{'exact_match'</em><em>, </em><em>'fill_to_data'</em><em>, </em><em>'prune_to_data'}</em>) – If <code class="docutils literal notranslate"><span class="pre">data_to_fit</span></code> and <code class="docutils literal notranslate"><span class="pre">mut_escape_df</span></code> (or <code class="docutils literal notranslate"><span class="pre">site_escape_df</span></code>) both
specified, what if they don’t specify same mutations.
If ‘exact_match’, raise error. If ‘fill_to_data’, then take
sites / wildtypes / mutations from <code class="docutils literal notranslate"><span class="pre">data_to_fit</span></code> and fill init
values from any not specified in <code class="docutils literal notranslate"><span class="pre">mut_escape_df</span></code> as indicated by
<code class="docutils literal notranslate"><span class="pre">init_missing</span></code>. If ‘prune_to_data’, remove any extra mutations
from <code class="docutils literal notranslate"><span class="pre">mut_escape_df</span></code> that are not in <code class="docutils literal notranslate"><span class="pre">data_to_fit</span></code>.</p></li>
</ul>
</dd>
</dl>
<dl class="py attribute">
<dt class="sig sig-object py" id="polyclonal.polyclonal.Polyclonal.epitopes">
<span class="sig-name descname"><span class="pre">epitopes</span></span><a class="headerlink" href="#polyclonal.polyclonal.Polyclonal.epitopes" title="Permalink to this definition">¶</a></dt>
<dd><p>Names of all epitopes.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>tuple</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="polyclonal.polyclonal.Polyclonal.mutations">
<span class="sig-name descname"><span class="pre">mutations</span></span><a class="headerlink" href="#polyclonal.polyclonal.Polyclonal.mutations" title="Permalink to this definition">¶</a></dt>
<dd><p>All mutations for which we have escape values.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>tuple</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="polyclonal.polyclonal.Polyclonal.alphabet">
<span class="sig-name descname"><span class="pre">alphabet</span></span><a class="headerlink" href="#polyclonal.polyclonal.Polyclonal.alphabet" title="Permalink to this definition">¶</a></dt>
<dd><p>Allowed characters in mutation strings.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>tuple</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="polyclonal.polyclonal.Polyclonal.sites">
<span class="sig-name descname"><span class="pre">sites</span></span><a class="headerlink" href="#polyclonal.polyclonal.Polyclonal.sites" title="Permalink to this definition">¶</a></dt>
<dd><p>List of all sites. These are the sites provided via the <code class="docutils literal notranslate"><span class="pre">sites</span></code> parameter,
or inferred from <code class="docutils literal notranslate"><span class="pre">data_to_fit</span></code> or <code class="docutils literal notranslate"><span class="pre">mut_escape_df</span></code> if that isn’t provided.
If <cite>sequential_integer_sites</cite> is <cite>False</cite>, these are str, otherwise int.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>tuple</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="polyclonal.polyclonal.Polyclonal.sequential_integer_sites">
<span class="sig-name descname"><span class="pre">sequential_integer_sites</span></span><a class="headerlink" href="#polyclonal.polyclonal.Polyclonal.sequential_integer_sites" title="Permalink to this definition">¶</a></dt>
<dd><p>True if sites are sequential and integer, False otherwise.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="polyclonal.polyclonal.Polyclonal.wts">
<span class="sig-name descname"><span class="pre">wts</span></span><a class="headerlink" href="#polyclonal.polyclonal.Polyclonal.wts" title="Permalink to this definition">¶</a></dt>
<dd><p>Keyed by site, value is wildtype at that site.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="polyclonal.polyclonal.Polyclonal.epitope_colors">
<span class="sig-name descname"><span class="pre">epitope_colors</span></span><a class="headerlink" href="#polyclonal.polyclonal.Polyclonal.epitope_colors" title="Permalink to this definition">¶</a></dt>
<dd><p>Maps each epitope to its color.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="polyclonal.polyclonal.Polyclonal.data_to_fit">
<span class="sig-name descname"><span class="pre">data_to_fit</span></span><a class="headerlink" href="#polyclonal.polyclonal.Polyclonal.data_to_fit" title="Permalink to this definition">¶</a></dt>
<dd><p>Data to fit as passed when initializing this <a class="reference internal" href="#polyclonal.polyclonal.Polyclonal" title="polyclonal.polyclonal.Polyclonal"><code class="xref py py-class docutils literal notranslate"><span class="pre">Polyclonal</span></code></a> object.
If using <code class="docutils literal notranslate"><span class="pre">collapse_identical_variants</span></code>, then identical variants
are collapsed on columns ‘concentration’, ‘aa_substitutions’,
and ‘prob_escape’, and a column ‘weight’ is added to represent number
of collapsed variants. Also, row-order may be changed.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>pandas.DataFrame or None</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="polyclonal.polyclonal.Polyclonal.spatial_distances">
<span class="sig-name descname"><span class="pre">spatial_distances</span></span><a class="headerlink" href="#polyclonal.polyclonal.Polyclonal.spatial_distances" title="Permalink to this definition">¶</a></dt>
<dd><p>Spatial distances passed when initializing this <a class="reference internal" href="#polyclonal.polyclonal.Polyclonal" title="polyclonal.polyclonal.Polyclonal"><code class="xref py py-class docutils literal notranslate"><span class="pre">Polyclonal</span></code></a> object.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>pandas.DataFrame or None</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="polyclonal.polyclonal.Polyclonal.mutations_times_seen">
<span class="sig-name descname"><span class="pre">mutations_times_seen</span></span><a class="headerlink" href="#polyclonal.polyclonal.Polyclonal.mutations_times_seen" title="Permalink to this definition">¶</a></dt>
<dd><p>If <cite>data_to_fit</cite> is not <cite>None</cite>, keyed by all mutations with escape values
and values are number of variants in which the mutation is seen. It is formally
calculated as the number of variants with mutation across all concentrations
divided by the number of concentrations, so can have non-integer values if
there are variants only observed at some concentrations.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>frozendict.frozendict or None</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="polyclonal.polyclonal.Polyclonal.distance_matrix">
<span class="sig-name descname"><span class="pre">distance_matrix</span></span><a class="headerlink" href="#polyclonal.polyclonal.Polyclonal.distance_matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>If <code class="docutils literal notranslate"><span class="pre">data_to_fit</span></code> is not <cite>None</cite> and <code class="docutils literal notranslate"><span class="pre">spatial_distances</span></code> is not <cite>None</cite>,
then this is a matrix giving the pairwise distances between residues.
Pairs with no distance have values of <cite>pd.NA</cite> and diagonal values are 0.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>pandas.DataFrame or <cite>None</cite></p>
</dd>
</dl>
</dd></dl>

<p class="rubric">Example</p>
<p>Simple example with two epitopes (<cite>e1</cite> and <cite>e2</cite>) and a few mutations where
we know the activities and mutation-level escape values ahead of time:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">activity_wt_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;epitope&#39;</span><span class="p">:</span>  <span class="p">[</span><span class="s1">&#39;e1&#39;</span><span class="p">,</span> <span class="s1">&#39;e2&#39;</span><span class="p">],</span>
<span class="gp">... </span>                               <span class="s1">&#39;activity&#39;</span><span class="p">:</span> <span class="p">[</span> <span class="mf">2.0</span><span class="p">,</span>  <span class="mf">1.0</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mut_escape_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
<span class="gp">... </span>  <span class="s1">&#39;mutation&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;M1C&#39;</span><span class="p">,</span> <span class="s1">&#39;M1C&#39;</span><span class="p">,</span> <span class="s1">&#39;G2A&#39;</span><span class="p">,</span> <span class="s1">&#39;G2A&#39;</span><span class="p">,</span> <span class="s1">&#39;A4K&#39;</span><span class="p">,</span> <span class="s1">&#39;A4K&#39;</span><span class="p">,</span> <span class="s1">&#39;A4L&#39;</span><span class="p">,</span> <span class="s1">&#39;A4L&#39;</span><span class="p">],</span>
<span class="gp">... </span>  <span class="s1">&#39;epitope&#39;</span><span class="p">:</span>  <span class="p">[</span> <span class="s1">&#39;e1&#39;</span><span class="p">,</span>  <span class="s1">&#39;e2&#39;</span><span class="p">,</span>  <span class="s1">&#39;e1&#39;</span><span class="p">,</span>  <span class="s1">&#39;e2&#39;</span><span class="p">,</span>  <span class="s1">&#39;e1&#39;</span><span class="p">,</span>  <span class="s1">&#39;e2&#39;</span><span class="p">,</span>  <span class="s1">&#39;e1&#39;</span><span class="p">,</span>  <span class="s1">&#39;e2&#39;</span><span class="p">],</span>
<span class="gp">... </span>  <span class="s1">&#39;escape&#39;</span><span class="p">:</span>   <span class="p">[</span>  <span class="mf">2.0</span><span class="p">,</span>   <span class="mf">0.0</span><span class="p">,</span>   <span class="mf">3.0</span><span class="p">,</span>   <span class="mf">0.0</span><span class="p">,</span>  <span class="mf">0.0</span><span class="p">,</span>    <span class="mf">2.5</span><span class="p">,</span>   <span class="mf">0.0</span><span class="p">,</span>   <span class="mf">1.5</span><span class="p">],</span>
<span class="gp">... </span>  <span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">model</span> <span class="o">=</span> <span class="n">Polyclonal</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">activity_wt_df</span><span class="o">=</span><span class="n">activity_wt_df</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">mut_escape_df</span><span class="o">=</span><span class="n">mut_escape_df</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">collapse_identical_variants</span><span class="o">=</span><span class="s2">&quot;mean&quot;</span><span class="p">,</span>
<span class="gp">... </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">model</span><span class="o">.</span><span class="n">epitopes</span>
<span class="go">(&#39;e1&#39;, &#39;e2&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">model</span><span class="o">.</span><span class="n">mutations</span>
<span class="go">(&#39;M1C&#39;, &#39;G2A&#39;, &#39;A4K&#39;, &#39;A4L&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">model</span><span class="o">.</span><span class="n">mutations_times_seen</span> <span class="ow">is</span> <span class="kc">None</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">model</span><span class="o">.</span><span class="n">sites</span>
<span class="go">(1, 2, 4)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">model</span><span class="o">.</span><span class="n">wts</span>
<span class="go">{1: &#39;M&#39;, 2: &#39;G&#39;, 4: &#39;A&#39;}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">model</span><span class="o">.</span><span class="n">activity_wt_df</span>
<span class="go">  epitope  activity</span>
<span class="go">0      e1       2.0</span>
<span class="go">1      e2       1.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">model</span><span class="o">.</span><span class="n">mut_escape_df</span>
<span class="go">  epitope  site wildtype mutant mutation  escape</span>
<span class="go">0      e1     1        M      C      M1C     2.0</span>
<span class="go">1      e1     2        G      A      G2A     3.0</span>
<span class="go">2      e1     4        A      K      A4K     0.0</span>
<span class="go">3      e1     4        A      L      A4L     0.0</span>
<span class="go">4      e2     1        M      C      M1C     0.0</span>
<span class="go">5      e2     2        G      A      G2A     0.0</span>
<span class="go">6      e2     4        A      K      A4K     2.5</span>
<span class="go">7      e2     4        A      L      A4L     1.5</span>
</pre></div>
</div>
<p>We can also summarize the mutation-level escape at the site level:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">pd</span><span class="o">.</span><span class="n">set_option</span><span class="p">(</span><span class="s2">&quot;display.max_columns&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pd</span><span class="o">.</span><span class="n">set_option</span><span class="p">(</span><span class="s2">&quot;display.width&quot;</span><span class="p">,</span> <span class="mi">89</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">model</span><span class="o">.</span><span class="n">mut_escape_site_summary_df</span><span class="p">()</span>
<span class="go">  epitope  site wildtype  mean  total positive  max  min  total negative  n mutations</span>
<span class="go">0      e1     1        M   2.0             2.0  2.0  2.0             0.0            1</span>
<span class="go">1      e1     2        G   3.0             3.0  3.0  3.0             0.0            1</span>
<span class="go">2      e1     4        A   0.0             0.0  0.0  0.0             0.0            2</span>
<span class="go">3      e2     1        M   0.0             0.0  0.0  0.0             0.0            1</span>
<span class="go">4      e2     2        G   0.0             0.0  0.0  0.0             0.0            1</span>
<span class="go">5      e2     4        A   2.0             4.0  2.5  1.5             0.0            2</span>
</pre></div>
</div>
<p>Note that we can <strong>not</strong> initialize a <a class="reference internal" href="#polyclonal.polyclonal.Polyclonal" title="polyclonal.polyclonal.Polyclonal"><code class="xref py py-class docutils literal notranslate"><span class="pre">Polyclonal</span></code></a> object if we are
missing escape estimates for any mutations for any epitopes:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Polyclonal</span><span class="p">(</span><span class="n">activity_wt_df</span><span class="o">=</span><span class="n">activity_wt_df</span><span class="p">,</span>
<span class="gp">... </span>           <span class="n">mut_escape_df</span><span class="o">=</span><span class="n">mut_escape_df</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="n">n</span><span class="o">=</span><span class="mi">5</span><span class="p">))</span>
<span class="gt">Traceback (most recent call last):</span>
  <span class="c">...</span>
<span class="gr">ValueError</span>: <span class="n">invalid set of mutations for epitope=&#39;e2&#39;</span>
</pre></div>
</div>
<p>Now make a data frame with some variants:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">variants_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_records</span><span class="p">(</span>
<span class="gp">... </span>        <span class="p">[(</span><span class="s1">&#39;AA&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">),</span>
<span class="gp">... </span>         <span class="p">(</span><span class="s1">&#39;AC&#39;</span><span class="p">,</span> <span class="s1">&#39;M1C&#39;</span><span class="p">),</span>
<span class="gp">... </span>         <span class="p">(</span><span class="s1">&#39;AG&#39;</span><span class="p">,</span> <span class="s1">&#39;G2A&#39;</span><span class="p">),</span>
<span class="gp">... </span>         <span class="p">(</span><span class="s1">&#39;AT&#39;</span><span class="p">,</span> <span class="s1">&#39;A4K&#39;</span><span class="p">),</span>
<span class="gp">... </span>         <span class="p">(</span><span class="s1">&#39;TA&#39;</span><span class="p">,</span> <span class="s1">&#39;A4L&#39;</span><span class="p">),</span>
<span class="gp">... </span>         <span class="p">(</span><span class="s1">&#39;CA&#39;</span><span class="p">,</span> <span class="s1">&#39;M1C G2A&#39;</span><span class="p">),</span>
<span class="gp">... </span>         <span class="p">(</span><span class="s1">&#39;CG&#39;</span><span class="p">,</span> <span class="s1">&#39;M1C A4K&#39;</span><span class="p">),</span>
<span class="gp">... </span>         <span class="p">(</span><span class="s1">&#39;CC&#39;</span><span class="p">,</span> <span class="s1">&#39;G2A A4K&#39;</span><span class="p">),</span>
<span class="gp">... </span>         <span class="p">(</span><span class="s1">&#39;TC&#39;</span><span class="p">,</span> <span class="s1">&#39;G2A A4L&#39;</span><span class="p">),</span>
<span class="gp">... </span>         <span class="p">(</span><span class="s1">&#39;CT&#39;</span><span class="p">,</span> <span class="s1">&#39;M1C G2A A4K&#39;</span><span class="p">),</span>
<span class="gp">... </span>         <span class="p">(</span><span class="s1">&#39;TG&#39;</span><span class="p">,</span> <span class="s1">&#39;M1C G2A A4L&#39;</span><span class="p">),</span>
<span class="gp">... </span>         <span class="p">(</span><span class="s1">&#39;GA&#39;</span><span class="p">,</span> <span class="s1">&#39;M1C&#39;</span><span class="p">),</span>
<span class="gp">... </span>         <span class="p">],</span>
<span class="gp">... </span>        <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;barcode&#39;</span><span class="p">,</span> <span class="s1">&#39;aa_substitutions&#39;</span><span class="p">])</span>
</pre></div>
</div>
<p>Get the escape probabilities predicted on these variants from
the values in the <a class="reference internal" href="#polyclonal.polyclonal.Polyclonal" title="polyclonal.polyclonal.Polyclonal"><code class="xref py py-class docutils literal notranslate"><span class="pre">Polyclonal</span></code></a> object:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">escape_probs</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">prob_escape</span><span class="p">(</span><span class="n">variants_df</span><span class="o">=</span><span class="n">variants_df</span><span class="p">,</span>
<span class="gp">... </span>                                 <span class="n">concentrations</span><span class="o">=</span><span class="p">[</span><span class="mf">1.0</span><span class="p">,</span> <span class="mf">2.0</span><span class="p">,</span> <span class="mf">4.0</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">escape_probs</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
<span class="go">   barcode aa_substitutions  concentration  predicted_prob_escape</span>
<span class="go">0       AA                             1.0                  0.032</span>
<span class="go">1       AT              A4K            1.0                  0.097</span>
<span class="go">2       TA              A4L            1.0                  0.074</span>
<span class="go">3       AG              G2A            1.0                  0.197</span>
<span class="go">4       CC          G2A A4K            1.0                  0.598</span>
<span class="go">5       TC          G2A A4L            1.0                  0.455</span>
<span class="go">6       AC              M1C            1.0                  0.134</span>
<span class="go">7       GA              M1C            1.0                  0.134</span>
<span class="go">8       CG          M1C A4K            1.0                  0.409</span>
<span class="go">9       CA          M1C G2A            1.0                  0.256</span>
<span class="go">10      CT      M1C G2A A4K            1.0                  0.779</span>
<span class="go">11      TG      M1C G2A A4L            1.0                  0.593</span>
<span class="go">12      AA                             2.0                  0.010</span>
<span class="go">13      AT              A4K            2.0                  0.044</span>
<span class="go">14      TA              A4L            2.0                  0.029</span>
<span class="go">15      AG              G2A            2.0                  0.090</span>
<span class="go">16      CC          G2A A4K            2.0                  0.398</span>
<span class="go">17      TC          G2A A4L            2.0                  0.260</span>
<span class="go">18      AC              M1C            2.0                  0.052</span>
<span class="go">19      GA              M1C            2.0                  0.052</span>
<span class="go">20      CG          M1C A4K            2.0                  0.230</span>
<span class="go">21      CA          M1C G2A            2.0                  0.141</span>
<span class="go">22      CT      M1C G2A A4K            2.0                  0.629</span>
<span class="go">23      TG      M1C G2A A4L            2.0                  0.411</span>
<span class="go">24      AA                             4.0                  0.003</span>
<span class="go">25      AT              A4K            4.0                  0.017</span>
<span class="go">26      TA              A4L            4.0                  0.010</span>
<span class="go">27      AG              G2A            4.0                  0.034</span>
<span class="go">28      CC          G2A A4K            4.0                  0.214</span>
<span class="go">29      TC          G2A A4L            4.0                  0.118</span>
<span class="go">30      AC              M1C            4.0                  0.017</span>
<span class="go">31      GA              M1C            4.0                  0.017</span>
<span class="go">32      CG          M1C A4K            4.0                  0.106</span>
<span class="go">33      CA          M1C G2A            4.0                  0.070</span>
<span class="go">34      CT      M1C G2A A4K            4.0                  0.441</span>
<span class="go">35      TG      M1C G2A A4L            4.0                  0.243</span>
</pre></div>
</div>
<p>We can also get predicted escape probabilities by including concentrations
in the data frame passed to <a class="reference internal" href="#polyclonal.polyclonal.Polyclonal.prob_escape" title="polyclonal.polyclonal.Polyclonal.prob_escape"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Polyclonal.prob_escape()</span></code></a>:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">model</span><span class="o">.</span><span class="n">prob_escape</span><span class="p">(</span>
<span class="gp">... </span>        <span class="n">variants_df</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">variants_df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">concentration</span><span class="o">=</span><span class="n">c</span><span class="p">)</span>
<span class="gp">... </span>                               <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="p">[</span><span class="mf">1.0</span><span class="p">,</span> <span class="mf">2.0</span><span class="p">,</span> <span class="mf">4.0</span><span class="p">]])</span>
<span class="gp">... </span>        <span class="p">)</span><span class="o">.</span><span class="n">equals</span><span class="p">(</span><span class="n">escape_probs</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
<p>We can also compute the IC50s:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">model</span><span class="o">.</span><span class="n">icXX</span><span class="p">(</span><span class="n">variants_df</span><span class="p">)</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
<span class="go">   barcode aa_substitutions   IC50</span>
<span class="go">0       AA                   0.085</span>
<span class="go">1       AC              M1C  0.230</span>
<span class="go">2       GA              M1C  0.230</span>
<span class="go">3       AG              G2A  0.296</span>
<span class="go">4       AT              A4K  0.128</span>
<span class="go">5       TA              A4L  0.117</span>
<span class="go">6       CA          M1C G2A  0.355</span>
<span class="go">7       CG          M1C A4K  0.722</span>
<span class="go">8       CC          G2A A4K  1.414</span>
<span class="go">9       TC          G2A A4L  0.858</span>
<span class="go">10      CT      M1C G2A A4K  3.237</span>
<span class="go">11      TG      M1C G2A A4L  1.430</span>
</pre></div>
</div>
<p>Or the IC90s:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">model</span><span class="o">.</span><span class="n">icXX</span><span class="p">(</span><span class="n">variants_df</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="mf">0.9</span><span class="p">,</span> <span class="n">col</span><span class="o">=</span><span class="s1">&#39;IC90&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
<span class="go">   barcode aa_substitutions    IC90</span>
<span class="go">0       AA                    0.464</span>
<span class="go">1       AC              M1C   1.260</span>
<span class="go">2       GA              M1C   1.260</span>
<span class="go">3       AG              G2A   1.831</span>
<span class="go">4       AT              A4K   0.976</span>
<span class="go">5       TA              A4L   0.782</span>
<span class="go">6       CA          M1C G2A   2.853</span>
<span class="go">7       CG          M1C A4K   4.176</span>
<span class="go">8       CC          G2A A4K   7.473</span>
<span class="go">9       TC          G2A A4L   4.532</span>
<span class="go">10      CT      M1C G2A A4K  18.717</span>
<span class="go">11      TG      M1C G2A A4L   9.532</span>
</pre></div>
</div>
<p class="rubric">Example</p>
<p>Initialize with <code class="docutils literal notranslate"><span class="pre">escape_probs</span></code> created above as data to fit. In order
to do this, we need to change the name of the column with the
predicted escape probs to just be escape probs as we are now assuming
these are the real values:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">data_to_fit</span> <span class="o">=</span> <span class="p">(</span>
<span class="gp">... </span>        <span class="n">escape_probs</span>
<span class="gp">... </span>        <span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;predicted_prob_escape&#39;</span><span class="p">:</span> <span class="s1">&#39;prob_escape&#39;</span><span class="p">})</span>
<span class="gp">... </span>        <span class="p">)</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">model_data</span> <span class="o">=</span> <span class="n">Polyclonal</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">data_to_fit</span><span class="o">=</span><span class="n">data_to_fit</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">n_epitopes</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">collapse_identical_variants</span><span class="o">=</span><span class="s2">&quot;mean&quot;</span><span class="p">,</span>
<span class="gp">... </span><span class="p">)</span>
</pre></div>
</div>
<p>The mutations are those in <code class="docutils literal notranslate"><span class="pre">data_to_fit</span></code>:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">model_data</span><span class="o">.</span><span class="n">mutations</span>
<span class="go">(&#39;M1C&#39;, &#39;G2A&#39;, &#39;A4K&#39;, &#39;A4L&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">dict</span><span class="p">(</span><span class="n">model_data</span><span class="o">.</span><span class="n">mutations_times_seen</span><span class="p">)</span>
<span class="go">{&#39;G2A&#39;: 6, &#39;M1C&#39;: 6, &#39;A4K&#39;: 4, &#39;A4L&#39;: 3}</span>
</pre></div>
</div>
<p>The activities are evenly spaced from 1 to 0, while the mutation escapes
are all initialized to zero:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">model_data</span><span class="o">.</span><span class="n">activity_wt_df</span>
<span class="go">  epitope  activity</span>
<span class="go">0       1       1.0</span>
<span class="go">1       2       0.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">model_data</span><span class="o">.</span><span class="n">mut_escape_df</span>
<span class="go">  epitope  site wildtype mutant mutation  escape  times_seen</span>
<span class="go">0       1     1        M      C      M1C     0.0           6</span>
<span class="go">1       1     2        G      A      G2A     0.0           6</span>
<span class="go">2       1     4        A      K      A4K     0.0           4</span>
<span class="go">3       1     4        A      L      A4L     0.0           3</span>
<span class="go">4       2     1        M      C      M1C     0.0           6</span>
<span class="go">5       2     2        G      A      G2A     0.0           6</span>
<span class="go">6       2     4        A      K      A4K     0.0           4</span>
<span class="go">7       2     4        A      L      A4L     0.0           3</span>
</pre></div>
</div>
<p>You can initialize to random numbers by setting <code class="docutils literal notranslate"><span class="pre">init_missing</span></code> to seed
(in this example we also don’t include all variants for one concentration):</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">model_data2</span> <span class="o">=</span> <span class="n">Polyclonal</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">data_to_fit</span><span class="o">=</span><span class="n">data_to_fit</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">30</span><span class="p">),</span>
<span class="gp">... </span>    <span class="n">n_epitopes</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">init_missing</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">collapse_identical_variants</span><span class="o">=</span><span class="s2">&quot;mean&quot;</span><span class="p">,</span>
<span class="gp">... </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">model_data2</span><span class="o">.</span><span class="n">activity_wt_df</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
<span class="go">  epitope  activity</span>
<span class="go">0       1     0.417</span>
<span class="go">1       2     0.720</span>
</pre></div>
</div>
<p>You can set some or all mutation escapes to initial values:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">model_data3</span> <span class="o">=</span> <span class="n">Polyclonal</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">data_to_fit</span><span class="o">=</span><span class="n">data_to_fit</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">activity_wt_df</span><span class="o">=</span><span class="n">activity_wt_df</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">mut_escape_df</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;epitope&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;e1&#39;</span><span class="p">],</span>
<span class="gp">... </span>                                <span class="s1">&#39;mutation&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;M1C&#39;</span><span class="p">],</span>
<span class="gp">... </span>                                <span class="s1">&#39;escape&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">4</span><span class="p">]}),</span>
<span class="gp">... </span>    <span class="n">data_mut_escape_overlap</span><span class="o">=</span><span class="s1">&#39;fill_to_data&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">collapse_identical_variants</span><span class="o">=</span><span class="s2">&quot;mean&quot;</span><span class="p">,</span>
<span class="gp">... </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">model_data3</span><span class="o">.</span><span class="n">mut_escape_df</span>
<span class="go">  epitope  site wildtype mutant mutation  escape  times_seen</span>
<span class="go">0      e1     1        M      C      M1C     4.0           6</span>
<span class="go">1      e1     2        G      A      G2A     0.0           6</span>
<span class="go">2      e1     4        A      K      A4K     0.0           4</span>
<span class="go">3      e1     4        A      L      A4L     0.0           3</span>
<span class="go">4      e2     1        M      C      M1C     0.0           6</span>
<span class="go">5      e2     2        G      A      G2A     0.0           6</span>
<span class="go">6      e2     4        A      K      A4K     0.0           4</span>
<span class="go">7      e2     4        A      L      A4L     0.0           3</span>
</pre></div>
</div>
<p>You can initialize <strong>sites</strong> to escape values via <code class="docutils literal notranslate"><span class="pre">site_activity_df</span></code>:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">model_data4</span> <span class="o">=</span> <span class="n">Polyclonal</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">data_to_fit</span><span class="o">=</span><span class="n">data_to_fit</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">activity_wt_df</span><span class="o">=</span><span class="n">activity_wt_df</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">site_escape_df</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_records</span><span class="p">(</span>
<span class="gp">... </span>        <span class="p">[(</span><span class="s1">&#39;e1&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">),</span> <span class="p">(</span><span class="s1">&#39;e1&#39;</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">),</span>
<span class="gp">... </span>         <span class="p">(</span><span class="s1">&#39;e2&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">),</span> <span class="p">(</span><span class="s1">&#39;e2&#39;</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mf">2.0</span><span class="p">)],</span>
<span class="gp">... </span>        <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;epitope&#39;</span><span class="p">,</span> <span class="s1">&#39;site&#39;</span><span class="p">,</span> <span class="s1">&#39;escape&#39;</span><span class="p">],</span>
<span class="gp">... </span>    <span class="p">),</span>
<span class="gp">... </span>    <span class="n">data_mut_escape_overlap</span><span class="o">=</span><span class="s1">&#39;fill_to_data&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">collapse_identical_variants</span><span class="o">=</span><span class="s2">&quot;mean&quot;</span><span class="p">,</span>
<span class="gp">... </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">model_data4</span><span class="o">.</span><span class="n">mut_escape_df</span>
<span class="go">  epitope  site wildtype mutant mutation  escape  times_seen</span>
<span class="go">0      e1     1        M      C      M1C     1.0           6</span>
<span class="go">1      e1     2        G      A      G2A     0.0           6</span>
<span class="go">2      e1     4        A      K      A4K     0.0           4</span>
<span class="go">3      e1     4        A      L      A4L     0.0           3</span>
<span class="go">4      e2     1        M      C      M1C     0.0           6</span>
<span class="go">5      e2     2        G      A      G2A     0.0           6</span>
<span class="go">6      e2     4        A      K      A4K     2.0           4</span>
<span class="go">7      e2     4        A      L      A4L     2.0           3</span>
</pre></div>
</div>
<p>Fit the data using <a class="reference internal" href="#polyclonal.polyclonal.Polyclonal.fit" title="polyclonal.polyclonal.Polyclonal.fit"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Polyclonal.fit()</span></code></a>, and make sure the new
predicted escape probabilities are close to the real ones being fit.
Reduce weight on regularization since there is so little data in this
toy example:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="p">[</span><span class="n">model_data</span><span class="p">,</span> <span class="n">model_data2</span><span class="p">,</span> <span class="n">model_data3</span><span class="p">,</span> <span class="n">model_data4</span><span class="p">]:</span>
<span class="gp">... </span>    <span class="n">opt_res</span> <span class="o">=</span> <span class="n">m</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span>
<span class="gp">... </span>        <span class="n">reg_escape_weight</span><span class="o">=</span><span class="mf">0.001</span><span class="p">,</span>
<span class="gp">... </span>        <span class="n">reg_spread_weight</span><span class="o">=</span><span class="mf">0.001</span><span class="p">,</span>
<span class="gp">... </span>        <span class="n">reg_activity_weight</span><span class="o">=</span><span class="mf">0.0001</span><span class="p">,</span>
<span class="gp">... </span>        <span class="n">reg_uniqueness2_weight</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
<span class="gp">... </span>    <span class="p">)</span>
<span class="gp">... </span>    <span class="n">pred_df</span> <span class="o">=</span> <span class="n">m</span><span class="o">.</span><span class="n">prob_escape</span><span class="p">(</span><span class="n">variants_df</span><span class="o">=</span><span class="n">data_to_fit</span><span class="p">)</span>
<span class="gp">... </span>    <span class="k">if</span> <span class="ow">not</span> <span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">pred_df</span><span class="p">[</span><span class="s1">&#39;prob_escape&#39;</span><span class="p">],</span>
<span class="gp">... </span>                          <span class="n">pred_df</span><span class="p">[</span><span class="s1">&#39;predicted_prob_escape&#39;</span><span class="p">],</span>
<span class="gp">... </span>                          <span class="n">atol</span><span class="o">=</span><span class="mf">0.01</span><span class="p">):</span>
<span class="gp">... </span>         <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;wrong predictions</span><span class="se">\n</span><span class="si">{</span><span class="n">pred_df</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="gp">... </span>    <span class="k">if</span> <span class="ow">not</span> <span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span>
<span class="gp">... </span>             <span class="n">activity_wt_df</span><span class="p">[</span><span class="s1">&#39;activity&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(),</span>
<span class="gp">... </span>             <span class="n">m</span><span class="o">.</span><span class="n">activity_wt_df</span><span class="p">[</span><span class="s1">&#39;activity&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(),</span>
<span class="gp">... </span>             <span class="n">atol</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>
<span class="gp">... </span>             <span class="p">):</span>
<span class="gp">... </span>         <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;wrong activities</span><span class="se">\n</span><span class="si">{</span><span class="n">m</span><span class="o">.</span><span class="n">activity_wt_df</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="gp">... </span>    <span class="k">if</span> <span class="ow">not</span> <span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span>
<span class="gp">... </span>             <span class="n">mut_escape_df</span><span class="p">[</span><span class="s1">&#39;escape&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(),</span>
<span class="gp">... </span>             <span class="n">m</span><span class="o">.</span><span class="n">mut_escape_df</span><span class="p">[</span><span class="s1">&#39;escape&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(),</span>
<span class="gp">... </span>             <span class="n">atol</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span>
<span class="gp">... </span>             <span class="p">):</span>
<span class="gp">... </span>         <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;wrong escapes</span><span class="se">\n</span><span class="si">{</span><span class="n">m</span><span class="o">.</span><span class="n">mut_escape_df</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">model_data</span><span class="o">.</span><span class="n">mut_escape_site_summary_df</span><span class="p">()</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
<span class="go">  epitope  site wildtype  mean  total positive  max  min  total negative  n mutations</span>
<span class="go">0       1     1        M  -0.0             0.0 -0.0 -0.0            -0.0            1</span>
<span class="go">1       1     2        G   0.0             0.0  0.0  0.0             0.0            1</span>
<span class="go">2       1     4        A   2.0             4.0  2.5  1.5             0.0            2</span>
<span class="go">3       2     1        M   2.0             2.0  2.0  2.0             0.0            1</span>
<span class="go">4       2     2        G   3.0             3.0  3.0  3.0             0.0            1</span>
<span class="go">5       2     4        A   0.0             0.0  0.0  0.0             0.0            2</span>
</pre></div>
</div>
<p>You can also exclude mutations to specific characters (typically you would want to
do this for stop codons and/or gaps):</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">model_data</span><span class="o">.</span><span class="n">mut_escape_site_summary_df</span><span class="p">(</span><span class="n">exclude_chars</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="s2">&quot;K&quot;</span><span class="p">})</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
<span class="go">  epitope  site wildtype  mean  total positive  max  min  total negative  n mutations</span>
<span class="go">0       1     2        G   0.0             0.0  0.0  0.0             0.0            1</span>
<span class="go">1       1     4        A   1.5             1.5  1.5  1.5             0.0            1</span>
<span class="go">2       2     2        G   3.0             3.0  3.0  3.0             0.0            1</span>
<span class="go">3       2     4        A   0.0             0.0  0.0  0.0             0.0            1</span>
</pre></div>
</div>
<p class="rubric">Example</p>
<p>You can convert a <a class="reference internal" href="#polyclonal.polyclonal.Polyclonal" title="polyclonal.polyclonal.Polyclonal"><code class="xref py py-class docutils literal notranslate"><span class="pre">Polyclonal</span></code></a> model into a site-level model via
the transformation of <a class="reference internal" href="polyclonal.utils.html#polyclonal.utils.site_level_variants" title="polyclonal.utils.site_level_variants"><code class="xref py py-func docutils literal notranslate"><span class="pre">polyclonal.utils.site_level_variants()</span></code></a>. The
site-level model is another <a class="reference internal" href="#polyclonal.polyclonal.Polyclonal" title="polyclonal.polyclonal.Polyclonal"><code class="xref py py-class docutils literal notranslate"><span class="pre">Polyclonal</span></code></a> model that just keeps
track of whether or not sites are mutated using a 2-letter wildtype/mutant
alphabet, and is generated using <a class="reference internal" href="#polyclonal.polyclonal.Polyclonal.site_level_model" title="polyclonal.polyclonal.Polyclonal.site_level_model"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Polyclonal.site_level_model()</span></code></a>:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">model_site</span> <span class="o">=</span> <span class="n">model_data4</span><span class="o">.</span><span class="n">site_level_model</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">model_site</span><span class="o">.</span><span class="n">alphabet</span>
<span class="go">(&#39;w&#39;, &#39;m&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">(</span><span class="n">model_site</span><span class="o">.</span><span class="n">mut_escape_df</span>
<span class="gp">... </span> <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">escape</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;escape&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">abs</span><span class="p">())</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">1</span><span class="p">))</span>
<span class="go">  epitope  site wildtype mutant mutation  escape  times_seen</span>
<span class="go">0      e1     1        w      m      w1m     2.0           5</span>
<span class="go">1      e1     2        w      m      w2m     3.0           6</span>
<span class="go">2      e1     4        w      m      w4m     0.0           7</span>
<span class="go">3      e2     1        w      m      w1m     0.0           5</span>
<span class="go">4      e2     2        w      m      w2m     0.0           6</span>
<span class="go">5      e2     4        w      m      w4m     2.0           7</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">model_site</span><span class="o">.</span><span class="n">data_to_fit</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="n">n</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
<span class="go">   concentration aa_substitutions  weight  prob_escape</span>
<span class="go">0            1.0                        1        0.032</span>
<span class="go">1            1.0              w1m       1        0.134</span>
<span class="go">2            1.0          w1m w2m       1        0.256</span>
<span class="go">3            1.0      w1m w2m w4m       2        0.686</span>
<span class="go">4            1.0          w1m w4m       1        0.409</span>
</pre></div>
</div>
<p class="rubric">Example</p>
<p>Epitope assignments are arbitrary, so you can calculate correlations
among their mutation-escape values and create harmonized models that use
the same label to refer to epitopes with similar mutation-escape values.</p>
<p>First, correlations of a model to itself:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ref_model</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">model</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">model</span><span class="o">.</span><span class="n">mut_escape_corr</span><span class="p">(</span><span class="n">ref_model</span><span class="p">)</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
<span class="go">  ref_epitope self_epitope  correlation</span>
<span class="go">0          e1           e1        1.000</span>
<span class="go">1          e1           e2       -0.907</span>
<span class="go">2          e2           e1       -0.907</span>
<span class="go">3          e2           e2        1.000</span>
</pre></div>
</div>
<p>Make another model with epitope assignments inverted and a few mutations missing:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">inverted_model</span> <span class="o">=</span> <span class="n">Polyclonal</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">activity_wt_df</span><span class="o">=</span><span class="p">(</span>
<span class="gp">... </span>        <span class="n">activity_wt_df</span>
<span class="gp">... </span>        <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">epitope</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;epitope&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">({</span><span class="s2">&quot;e1&quot;</span><span class="p">:</span> <span class="s2">&quot;e2&quot;</span><span class="p">,</span> <span class="s2">&quot;e2&quot;</span><span class="p">:</span> <span class="s2">&quot;e1&quot;</span><span class="p">}))</span>
<span class="gp">... </span>    <span class="p">),</span>
<span class="gp">... </span>    <span class="n">mut_escape_df</span><span class="o">=</span><span class="p">(</span>
<span class="gp">... </span>        <span class="n">mut_escape_df</span>
<span class="gp">... </span>        <span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s2">&quot;mutation != &#39;A4K&#39;&quot;</span><span class="p">)</span>
<span class="gp">... </span>        <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">epitope</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;epitope&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">({</span><span class="s2">&quot;e1&quot;</span><span class="p">:</span> <span class="s2">&quot;e2&quot;</span><span class="p">,</span> <span class="s2">&quot;e2&quot;</span><span class="p">:</span> <span class="s2">&quot;e1&quot;</span><span class="p">}))</span>
<span class="gp">... </span>    <span class="p">),</span>
<span class="gp">... </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">inverted_model</span><span class="o">.</span><span class="n">mut_escape_corr</span><span class="p">(</span><span class="n">ref_model</span><span class="p">)</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
<span class="go">  ref_epitope self_epitope  correlation</span>
<span class="go">0          e1           e1       -0.945</span>
<span class="go">1          e1           e2        1.000</span>
<span class="go">2          e2           e1        1.000</span>
<span class="go">3          e2           e2       -0.945</span>
</pre></div>
</div>
<p>Now actually get epitope-harmonized models:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">model_harmonized</span><span class="p">,</span> <span class="n">harmonize_df</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">epitope_harmonized_model</span><span class="p">(</span><span class="n">ref_model</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">harmonize_df</span>
<span class="go">  self_initial_epitope self_harmonized_epitope ref_epitope  correlation</span>
<span class="go">0                   e1                      e1          e1          1.0</span>
<span class="go">1                   e2                      e2          e2          1.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">assert</span> <span class="n">model</span><span class="o">.</span><span class="n">mut_escape_df</span><span class="o">.</span><span class="n">equals</span><span class="p">(</span><span class="n">model_harmonized</span><span class="o">.</span><span class="n">mut_escape_df</span><span class="p">)</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">inverted_harmonized</span><span class="p">,</span> <span class="n">harmonize_df</span> <span class="o">=</span> <span class="n">inverted_model</span><span class="o">.</span><span class="n">epitope_harmonized_model</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">ref_model</span>
<span class="gp">... </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">harmonize_df</span>
<span class="go">  self_initial_epitope self_harmonized_epitope ref_epitope  correlation</span>
<span class="go">0                   e1                      e2          e2          1.0</span>
<span class="go">1                   e2                      e1          e1          1.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">inverted_harmonized</span><span class="o">.</span><span class="n">mut_escape_df</span>
<span class="go">  epitope  site wildtype mutant mutation  escape</span>
<span class="go">0      e1     1        M      C      M1C     2.0</span>
<span class="go">1      e1     2        G      A      G2A     3.0</span>
<span class="go">2      e1     4        A      L      A4L     0.0</span>
<span class="go">3      e2     1        M      C      M1C     0.0</span>
<span class="go">4      e2     2        G      A      G2A     0.0</span>
<span class="go">5      e2     4        A      L      A4L     1.5</span>
</pre></div>
</div>
<p class="rubric">Example</p>
<p>Filter variants by how often they are seen in data:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">model_data</span><span class="o">.</span><span class="n">filter_variants_by_seen_muts</span><span class="p">(</span><span class="n">variants_df</span><span class="p">)</span>
<span class="gp">... </span>
<span class="go">   barcode aa_substitutions</span>
<span class="go">0       AA</span>
<span class="go">1       AC              M1C</span>
<span class="go">2       AG              G2A</span>
<span class="go">3       AT              A4K</span>
<span class="go">4       TA              A4L</span>
<span class="go">5       CA          M1C G2A</span>
<span class="go">6       CG          M1C A4K</span>
<span class="go">7       CC          G2A A4K</span>
<span class="go">8       TC          G2A A4L</span>
<span class="go">9       CT      M1C G2A A4K</span>
<span class="go">10      TG      M1C G2A A4L</span>
<span class="go">11      GA              M1C</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">model_data</span><span class="o">.</span><span class="n">filter_variants_by_seen_muts</span><span class="p">(</span><span class="n">variants_df</span><span class="p">,</span> <span class="n">min_times_seen</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
<span class="gp">... </span>
<span class="go">  barcode aa_substitutions</span>
<span class="go">0      AA</span>
<span class="go">1      AC              M1C</span>
<span class="go">2      AG              G2A</span>
<span class="go">3      CA          M1C G2A</span>
<span class="go">4      GA              M1C</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">model_data</span><span class="o">.</span><span class="n">filter_variants_by_seen_muts</span><span class="p">(</span><span class="n">variants_df</span><span class="p">,</span> <span class="n">min_times_seen</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>
<span class="gp">... </span>
<span class="go">  barcode aa_substitutions</span>
<span class="go">0      AA</span>
<span class="go">1      AC              M1C</span>
<span class="go">2      AG              G2A</span>
<span class="go">3      AT              A4K</span>
<span class="go">4      CA          M1C G2A</span>
<span class="go">5      CG          M1C A4K</span>
<span class="go">6      CC          G2A A4K</span>
<span class="go">7      CT      M1C G2A A4K</span>
<span class="go">8      GA              M1C</span>
</pre></div>
</div>
<p class="rubric">Example</p>
<p>Add spatial distances:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">spatial_distances</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
<span class="gp">... </span>    <span class="p">[(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mf">1.5</span><span class="p">),</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mf">7.5</span><span class="p">)],</span>
<span class="gp">... </span>    <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;site_1&quot;</span><span class="p">,</span> <span class="s2">&quot;site_2&quot;</span><span class="p">,</span> <span class="s2">&quot;distance&quot;</span><span class="p">],</span>
<span class="gp">... </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">model_spatial</span> <span class="o">=</span> <span class="n">Polyclonal</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">data_to_fit</span><span class="o">=</span><span class="n">data_to_fit</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">n_epitopes</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">spatial_distances</span><span class="o">=</span><span class="n">spatial_distances</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">collapse_identical_variants</span><span class="o">=</span><span class="s2">&quot;mean&quot;</span><span class="p">,</span>
<span class="gp">... </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">model_spatial</span><span class="o">.</span><span class="n">distance_matrix</span>
<span class="go">     1     2     4</span>
<span class="go">1  0.0   1.5   7.5</span>
<span class="go">2  1.5   0.0  &lt;NA&gt;</span>
<span class="go">4  7.5  &lt;NA&gt;   0.0</span>
</pre></div>
</div>
<p>Fit with lower regularization as data so small</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">opt_res</span> <span class="o">=</span> <span class="n">model_spatial</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">reg_escape_weight</span><span class="o">=</span><span class="mf">0.001</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">reg_spread_weight</span><span class="o">=</span><span class="mf">0.001</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">reg_activity_weight</span><span class="o">=</span><span class="mf">0.0001</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">reg_spatial_weight</span><span class="o">=</span><span class="mf">0.001</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">reg_spatial2_weight</span><span class="o">=</span><span class="mf">0.0001</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">reg_uniqueness2_weight</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
<span class="gp">... </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pred_df</span> <span class="o">=</span> <span class="n">model_spatial</span><span class="o">.</span><span class="n">prob_escape</span><span class="p">(</span><span class="n">variants_df</span><span class="o">=</span><span class="n">data_to_fit</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">if</span> <span class="ow">not</span> <span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">pred_df</span><span class="p">[</span><span class="s1">&#39;prob_escape&#39;</span><span class="p">],</span>
<span class="gp">... </span>    <span class="n">pred_df</span><span class="p">[</span><span class="s1">&#39;predicted_prob_escape&#39;</span><span class="p">],</span>
<span class="gp">... </span>    <span class="n">atol</span><span class="o">=</span><span class="mf">0.01</span><span class="p">,</span>
<span class="gp">... </span><span class="p">):</span>
<span class="gp">... </span>    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;wrong predictions</span><span class="se">\n</span><span class="si">{</span><span class="n">pred_df</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">if</span> <span class="ow">not</span> <span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span>
<span class="gp">... </span>     <span class="n">activity_wt_df</span><span class="p">[</span><span class="s1">&#39;activity&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(),</span>
<span class="gp">... </span>     <span class="n">m</span><span class="o">.</span><span class="n">activity_wt_df</span><span class="p">[</span><span class="s1">&#39;activity&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(),</span>
<span class="gp">... </span>     <span class="n">atol</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>
<span class="gp">... </span><span class="p">):</span>
<span class="gp">... </span>    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;wrong activities</span><span class="se">\n</span><span class="si">{</span><span class="n">m</span><span class="o">.</span><span class="n">activity_wt_df</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">if</span> <span class="ow">not</span> <span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span>
<span class="gp">... </span>     <span class="n">mut_escape_df</span><span class="p">[</span><span class="s1">&#39;escape&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(),</span>
<span class="gp">... </span>     <span class="n">m</span><span class="o">.</span><span class="n">mut_escape_df</span><span class="p">[</span><span class="s1">&#39;escape&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(),</span>
<span class="gp">... </span>     <span class="n">atol</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span>
<span class="gp">... </span><span class="p">):</span>
<span class="gp">... </span>    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;wrong escapes</span><span class="se">\n</span><span class="si">{</span><span class="n">m</span><span class="o">.</span><span class="n">mut_escape_df</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
<dl class="py attribute">
<dt class="sig sig-object py" id="polyclonal.polyclonal.Polyclonal.DEFAULT_SCIPY_MINIMIZE_KWARGS">
<span class="sig-name descname"><span class="pre">DEFAULT_SCIPY_MINIMIZE_KWARGS</span></span><em class="property"> <span class="pre">=</span> <span class="pre">{'method':</span> <span class="pre">'L-BFGS-B',</span> <span class="pre">'options':</span> <span class="pre">{'ftol':</span> <span class="pre">1e-07,</span> <span class="pre">'maxfun':</span> <span class="pre">10000000.0,</span> <span class="pre">'maxiter':</span> <span class="pre">1000000.0}}</span></em><a class="headerlink" href="#polyclonal.polyclonal.Polyclonal.DEFAULT_SCIPY_MINIMIZE_KWARGS" title="Permalink to this definition">¶</a></dt>
<dd><p>default <code class="docutils literal notranslate"><span class="pre">scipy_minimize_kwargs</span></code> to <code class="docutils literal notranslate"><span class="pre">fit</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>frozendict.frozendict</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="polyclonal.polyclonal.Polyclonal.activity_wt_barplot">
<span class="sig-name descname"><span class="pre">activity_wt_barplot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/polyclonal/polyclonal.html#Polyclonal.activity_wt_barplot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polyclonal.polyclonal.Polyclonal.activity_wt_barplot" title="Permalink to this definition">¶</a></dt>
<dd><p>Bar plot of activity against each epitope, <span class="math notranslate nohighlight">\(a_{\rm{wt},e}\)</span>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>**kwargs</strong> – Keyword args for <a class="reference internal" href="polyclonal.plot.html#polyclonal.plot.activity_wt_barplot" title="polyclonal.plot.activity_wt_barplot"><code class="xref py py-func docutils literal notranslate"><span class="pre">polyclonal.plot.activity_wt_barplot()</span></code></a>.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Interactive plot.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>altair.Chart</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="polyclonal.polyclonal.Polyclonal.activity_wt_df">
<em class="property"><span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">activity_wt_df</span></span><a class="headerlink" href="#polyclonal.polyclonal.Polyclonal.activity_wt_df" title="Permalink to this definition">¶</a></dt>
<dd><p>Activities <span class="math notranslate nohighlight">\(a_{\rm{wt,e}}\)</span> for epitopes.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="polyclonal.polyclonal.Polyclonal.epitope_harmonized_model">
<span class="sig-name descname"><span class="pre">epitope_harmonized_model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ref_poly</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/polyclonal/polyclonal.html#Polyclonal.epitope_harmonized_model"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polyclonal.polyclonal.Polyclonal.epitope_harmonized_model" title="Permalink to this definition">¶</a></dt>
<dd><p>Get a copy of model with epitopes “harmonized” with a reference model.</p>
<p>Epitopes are unidentifiable, meaning there is no guarantee that we will assign
the same epitope labels across multiple models fit to similar data. This
function returns a copy of the current model where the epitope labels are
harmonized with those of another model with the same epitope labels. The
harmonization is done by putting into correspondence the epitopes with
the best correlated mutation-escape values.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>ref_poly</strong> (<a class="reference internal" href="#polyclonal.polyclonal.Polyclonal" title="polyclonal.polyclonal.Polyclonal"><code class="xref py py-class docutils literal notranslate"><span class="pre">Polyclonal</span></code></a>) – The reference polyclonal object to harmonize epitope labels with.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>harmonized_model, harmonize_df</strong> – <cite>harmonized_model</cite> is a <a class="reference internal" href="#polyclonal.polyclonal.Polyclonal" title="polyclonal.polyclonal.Polyclonal"><code class="xref py py-class docutils literal notranslate"><span class="pre">Polyclonal</span></code></a> object that is a copy of
the current model but with epitopes harmonized, and
<a class="reference internal" href="#polyclonal.polyclonal.Polyclonal.epitope_colors" title="polyclonal.polyclonal.Polyclonal.epitope_colors"><code class="xref py py-attr docutils literal notranslate"><span class="pre">Polyclonal.epitope_colors</span></code></a> taken from <cite>ref_poly</cite>. <cite>harmonize_df</cite>
is a <cite>pandas.DataFrame</cite> that shows how the epitopes were re-assigned
and the correlations in their escape values.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>tuple</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><a class="reference internal" href="#polyclonal.polyclonal.PolyclonalHarmonizeError" title="polyclonal.polyclonal.PolyclonalHarmonizeError"><strong>PolyclonalHarmonizeError</strong></a> – Raise this error if epitopes cannot be harmonized one-to-one.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="polyclonal.polyclonal.Polyclonal.filter_variants_by_seen_muts">
<span class="sig-name descname"><span class="pre">filter_variants_by_seen_muts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">variants_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_times_seen</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subs_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'aa_substitutions'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/polyclonal/polyclonal.html#Polyclonal.filter_variants_by_seen_muts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polyclonal.polyclonal.Polyclonal.filter_variants_by_seen_muts" title="Permalink to this definition">¶</a></dt>
<dd><p>Remove variants that contain mutations not seen during model fitting.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>variants_df</strong> (<em>pandas.DataFrame</em>) – Contains variants as rows.</p></li>
<li><p><strong>min_times_seen</strong> (<em>int</em>) – Require mutations to be seen &gt;= this many times in data used to fit model.</p></li>
<li><p><strong>subs_col</strong> (<em>str</em>) – Column in <cite>variants_df</cite> with mutations in each variant.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>variants_df</strong> – Copy of input dataframe, with rows of variants
that have unseen mutations removed.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="polyclonal.polyclonal.Polyclonal.fit">
<span class="sig-name descname"><span class="pre">fit</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">loss_delta</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reg_escape_weight</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.02</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reg_escape_delta</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reg_spread_weight</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">site_avg_abs_escape_epsilon</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reg_spatial_weight</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reg_spatial2_weight</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.0001</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reg_uniqueness_weight</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reg_uniqueness2_weight</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reg_activity_weight</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reg_activity_delta</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fit_site_level_first</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scipy_minimize_kwargs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{'method':</span> <span class="pre">'L-BFGS-B',</span> <span class="pre">'options':</span> <span class="pre">{'ftol':</span> <span class="pre">1e-07,</span> <span class="pre">'maxfun':</span> <span class="pre">10000000.0,</span> <span class="pre">'maxiter':</span> <span class="pre">1000000.0}}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">logfreq</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/polyclonal/polyclonal.html#Polyclonal.fit"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polyclonal.polyclonal.Polyclonal.fit" title="Permalink to this definition">¶</a></dt>
<dd><p>Fit parameters (activities and mutation escapes) to the data.</p>
<p>Requires <a class="reference internal" href="#polyclonal.polyclonal.Polyclonal.data_to_fit" title="polyclonal.polyclonal.Polyclonal.data_to_fit"><code class="xref py py-attr docutils literal notranslate"><span class="pre">Polyclonal.data_to_fit</span></code></a> be set at initialization of
this <a class="reference internal" href="#polyclonal.polyclonal.Polyclonal" title="polyclonal.polyclonal.Polyclonal"><code class="xref py py-class docutils literal notranslate"><span class="pre">Polyclonal</span></code></a> object. After calling this method, the
<span class="math notranslate nohighlight">\(a_{\rm{wt},e}\)</span> and <span class="math notranslate nohighlight">\(\beta_{m,e}\)</span> have been optimized, and
can be accessed using other methods of the <a class="reference internal" href="#polyclonal.polyclonal.Polyclonal" title="polyclonal.polyclonal.Polyclonal"><code class="xref py py-class docutils literal notranslate"><span class="pre">Polyclonal</span></code></a> object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>loss_delta</strong> (<em>float</em>) – Pseudo-Huber <span class="math notranslate nohighlight">\(\delta\)</span> parameter for loss on
<span class="math notranslate nohighlight">\(p_v\left(c\right)\)</span> fitting.</p></li>
<li><p><strong>reg_escape_weight</strong> (<em>float</em>) – Strength of Pseudo-Huber regularization on <span class="math notranslate nohighlight">\(\beta_{m,e}\)</span>.</p></li>
<li><p><strong>reg_escape_delta</strong> (<em>float</em>) – Pseudo-Huber <span class="math notranslate nohighlight">\(\delta\)</span> for regularizing <span class="math notranslate nohighlight">\(\beta_{m,e}\)</span>.</p></li>
<li><p><strong>reg_spread_weight</strong> (<em>float</em>) – Strength of regularization on variance of <span class="math notranslate nohighlight">\(\beta_{m,e}\)</span>
values at each site.</p></li>
<li><p><strong>site_avg_abs_escape_epsilon</strong> (<em>float</em>) – The epsilon value used when computing a differentiable measure of the
average absolute value of escape at a site for each epitope.</p></li>
<li><p><strong>reg_spatial_weight</strong> (<em>float</em>) – Strength of regularization of spatial distance between <span class="math notranslate nohighlight">\(\beta_{m,e}\)</span>
values at each site. Only meaningful if <a class="reference internal" href="#polyclonal.polyclonal.Polyclonal.distance_matrix" title="polyclonal.polyclonal.Polyclonal.distance_matrix"><code class="xref py py-attr docutils literal notranslate"><span class="pre">Polyclonal.distance_matrix</span></code></a>
is not <cite>None</cite>.</p></li>
<li><p><strong>reg_spatial2_weight</strong> (<em>float</em>) – Strength of regularization of squared spatial distance between
<span class="math notranslate nohighlight">\(\beta_{m,e}\)</span> values at each site. Only meaningful if
<a class="reference internal" href="#polyclonal.polyclonal.Polyclonal.distance_matrix" title="polyclonal.polyclonal.Polyclonal.distance_matrix"><code class="xref py py-attr docutils literal notranslate"><span class="pre">Polyclonal.distance_matrix</span></code></a> is not <cite>None</cite>.</p></li>
<li><p><strong>reg_uniqueness_weight</strong> (<em>float</em>) – Strength of regularization on epitope uniqueness.</p></li>
<li><p><strong>reg_uniqueness2_weight</strong> (<em>float</em>) – Strength of regularization on uniqueness of squared escape.
values at each site across epitopes.</p></li>
<li><p><strong>site_avg_abs_escape_epsilon</strong> – Epsilon value for caclulating site average absolute escape.</p></li>
<li><p><strong>reg_activity_weight</strong> (<em>float</em>) – Strength of Pseudo-Huber regularization on <span class="math notranslate nohighlight">\(a_{\rm{wt},e}\)</span>.</p></li>
<li><p><strong>reg_activity_delta</strong> (<em>float</em>) – Pseudo-Huber <span class="math notranslate nohighlight">\(\delta\)</span> for regularizing <span class="math notranslate nohighlight">\(a_{\rm{wt},e}\)</span>.</p></li>
<li><p><strong>fit_site_level_first</strong> (<em>bool</em>) – First fit a site-level model, then use those activities /
escapes to initialize fit of this model. Generally works better.</p></li>
<li><p><strong>scipy_minimize_kwargs</strong> (<em>dict</em>) – Keyword arguments passed to <code class="docutils literal notranslate"><span class="pre">scipy.optimize.minimize</span></code>.</p></li>
<li><p><strong>log</strong> (<em>None</em><em> or </em><em>writable file-like object</em>) – Where to log output. If <code class="docutils literal notranslate"><span class="pre">None</span></code>, use <code class="docutils literal notranslate"><span class="pre">sys.stdout</span></code>.</p></li>
<li><p><strong>logfreq</strong> (<em>None</em><em> or </em><em>int</em>) – How frequently to write updates on fitting to <code class="docutils literal notranslate"><span class="pre">log</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>scipy.optimize.OptimizeResult</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="polyclonal.polyclonal.Polyclonal.icXX">
<span class="sig-name descname"><span class="pre">icXX</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">variants_df</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'IC50'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_c</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1e-05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_c</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100000.0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/polyclonal/polyclonal.html#Polyclonal.icXX"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polyclonal.polyclonal.Polyclonal.icXX" title="Permalink to this definition">¶</a></dt>
<dd><p>Concentration at which a given fraction is neutralized (eg, IC50).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>variants_df</strong> (<em>pandas.DataFrame</em>) – Data frame defining variants. Should have column named
‘aa_substitutions’ that defines variants as space-delimited
strings of substitutions (e.g., ‘M1A K3T’).</p></li>
<li><p><strong>x</strong> (<em>float</em>) – Compute concentration at which this fraction is neutralized for
each variant. So set to 0.5 for IC50, and 0.9 for IC90.</p></li>
<li><p><strong>col</strong> (<em>str</em>) – Name of column in returned data frame with the ICXX value.</p></li>
<li><p><strong>min_c</strong> (<em>float</em>) – Minimum allowed icXX, truncate values &lt; this at this.</p></li>
<li><p><strong>max_c</strong> (<em>float</em>) – Maximum allowed icXX, truncate values &gt; this at this.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Copy of <code class="docutils literal notranslate"><span class="pre">variants_df</span></code> with added column <code class="docutils literal notranslate"><span class="pre">col</span></code> containing icXX.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="polyclonal.polyclonal.Polyclonal.mut_escape_corr">
<span class="sig-name descname"><span class="pre">mut_escape_corr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ref_poly</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/polyclonal/polyclonal.html#Polyclonal.mut_escape_corr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polyclonal.polyclonal.Polyclonal.mut_escape_corr" title="Permalink to this definition">¶</a></dt>
<dd><p>Correlation of mutation-escape values with another model.</p>
<p>For each epitope, how well is this model’s mutation-escape values
correlation with another model?</p>
<p>Mutations present in only one model are ignored.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>ref_poly</strong> (<a class="reference internal" href="#polyclonal.polyclonal.Polyclonal" title="polyclonal.polyclonal.Polyclonal"><code class="xref py py-class docutils literal notranslate"><span class="pre">Polyclonal</span></code></a>) – Other (reference) polyclonal model with which we calculate correlations.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>corr_df</strong> – Pairwise epitope correlations for escape.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="polyclonal.polyclonal.Polyclonal.mut_escape_df">
<em class="property"><span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">mut_escape_df</span></span><a class="headerlink" href="#polyclonal.polyclonal.Polyclonal.mut_escape_df" title="Permalink to this definition">¶</a></dt>
<dd><p>Escape <span class="math notranslate nohighlight">\(\beta_{m,e}\)</span> for each mutation.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="polyclonal.polyclonal.Polyclonal.mut_escape_pdb_b_factor">
<span class="sig-name descname"><span class="pre">mut_escape_pdb_b_factor</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_pdbfile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chains</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metric</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outdir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'{metric}-{epitope}.pdb'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">missing_metric</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_times_seen</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/polyclonal/polyclonal.html#Polyclonal.mut_escape_pdb_b_factor"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polyclonal.polyclonal.Polyclonal.mut_escape_pdb_b_factor" title="Permalink to this definition">¶</a></dt>
<dd><p>Create PDB files with B factors from a site’s mutation escape.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>input_pdbfile</strong> (<em>str</em>) – Path to input PDB file.</p></li>
<li><p><strong>chains</strong> (<em>str</em><em> or </em><em>array-like</em>) – Single chain or list of them to re-color.</p></li>
<li><p><strong>metric</strong> (<em>str</em>) – Which site-level summary metric to use. Can be any metric in
<a class="reference internal" href="#polyclonal.polyclonal.Polyclonal.mut_escape_site_summary_df" title="polyclonal.polyclonal.Polyclonal.mut_escape_site_summary_df"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Polyclonal.mut_escape_site_summary_df()</span></code></a>.</p></li>
<li><p><strong>outdir</strong> (<em>str</em>) – Output directory for created PDB files.</p></li>
<li><p><strong>outfile</strong> (<em>str</em>) – Output file name, with formatting used to replace metric and
epitope in curly brackets.</p></li>
<li><p><strong>missing_metric</strong> (<em>float</em><em> or </em><em>dict</em>) – How do we handle sites in PDB that are missing in escape metric?
If a float, reassign B factors for all missing sites to this value.
If a dict, should be keyed by chain and assign all missing sites in
each chain to indicated value.</p></li>
<li><p><strong>min_times_seen</strong> (<em>int</em>) – Value passed to <a class="reference internal" href="#polyclonal.polyclonal.Polyclonal.mut_escape_site_summary_df" title="polyclonal.polyclonal.Polyclonal.mut_escape_site_summary_df"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Polyclonal.mut_escape_site_summary_df()</span></code></a>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Gives name of created B-factor re-colored PDB for each epitope.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="polyclonal.polyclonal.Polyclonal.mut_escape_plot">
<span class="sig-name descname"><span class="pre">mut_escape_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">biochem_order_aas</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix_epitope</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df_to_merge</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/polyclonal/polyclonal.html#Polyclonal.mut_escape_plot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polyclonal.polyclonal.Polyclonal.mut_escape_plot" title="Permalink to this definition">¶</a></dt>
<dd><p>Make plot of the mutation escape values, <span class="math notranslate nohighlight">\(\beta_{m,e}\)</span>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>biochem_order_aas</strong> (<em>bool</em>) – Biochemically order the amino-acid alphabet in <a class="reference internal" href="#polyclonal.polyclonal.Polyclonal.alphabet" title="polyclonal.polyclonal.Polyclonal.alphabet"><code class="xref py py-attr docutils literal notranslate"><span class="pre">Polyclonal.alphabet</span></code></a>
by passing it through <a class="reference internal" href="polyclonal.alphabets.html#polyclonal.alphabets.biochem_order_aas" title="polyclonal.alphabets.biochem_order_aas"><code class="xref py py-func docutils literal notranslate"><span class="pre">polyclonal.alphabets.biochem_order_aas()</span></code></a>.</p></li>
<li><p><strong>prefix_epitope</strong> (<em>bool</em><em> or </em><em>None</em>) – Do we add the prefix “epitope ” to the epitope labels? If <cite>None</cite>, do
only if epitope is integer.</p></li>
<li><p><strong>df_to_merge</strong> (<em>None</em><em> or </em><em>pandas.DataFrame</em><em> or </em><em>list</em>) – To include additional properties, specify data frame or list of them which
are merged with <a class="reference internal" href="#polyclonal.polyclonal.Polyclonal.mut_escape_df" title="polyclonal.polyclonal.Polyclonal.mut_escape_df"><code class="xref py py-attr docutils literal notranslate"><span class="pre">Polyclonal.mut_escape_df</span></code></a> before being passed
to <a class="reference internal" href="polyclonal.plot.html#polyclonal.plot.lineplot_and_heatmap" title="polyclonal.plot.lineplot_and_heatmap"><code class="xref py py-func docutils literal notranslate"><span class="pre">polyclonal.plot.lineplot_and_heatmap()</span></code></a>. Properties will
only be included in plot if relevant columns are passed to
<a class="reference internal" href="polyclonal.plot.html#polyclonal.plot.lineplot_and_heatmap" title="polyclonal.plot.lineplot_and_heatmap"><code class="xref py py-func docutils literal notranslate"><span class="pre">polyclonal.plot.lineplot_and_heatmap()</span></code></a> via <cite>addtl_slider_stats</cite>,
<cite>addtl_tooltip_stats</cite>, or <cite>site_zoom_bar_color_col</cite>.</p></li>
<li><p><strong>**kwargs</strong> – Keyword args for <a class="reference internal" href="polyclonal.plot.html#polyclonal.plot.lineplot_and_heatmap" title="polyclonal.plot.lineplot_and_heatmap"><code class="xref py py-func docutils literal notranslate"><span class="pre">polyclonal.plot.lineplot_and_heatmap()</span></code></a>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Interactive line plot and heatmap.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>altair.Chart</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="polyclonal.polyclonal.Polyclonal.mut_escape_site_summary_df">
<span class="sig-name descname"><span class="pre">mut_escape_site_summary_df</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_times_seen</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutation_whitelist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_chars</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">frozenset({'*'})</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/polyclonal/polyclonal.html#Polyclonal.mut_escape_site_summary_df"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polyclonal.polyclonal.Polyclonal.mut_escape_site_summary_df" title="Permalink to this definition">¶</a></dt>
<dd><p>Site-level summaries of mutation escape.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>min_times_seen</strong> (<em>int</em>) – Only include in summaries mutations seen in at least this many variants.</p></li>
<li><p><strong>mutation_whitelist</strong> (<em>None</em><em> or </em><em>set</em>) – Only include in summaries these mutations.</p></li>
<li><p><strong>exclude_chars</strong> (<em>set</em><em> or </em><em>list</em>) – Exclude mutations to these characters when calculating site summaries.
Useful if you want to ignore stop codons (<code class="docutils literal notranslate"><span class="pre">*</span></code>), and perhaps in some
cases also gaps (<code class="docutils literal notranslate"><span class="pre">-</span></code>).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="polyclonal.polyclonal.Polyclonal.prob_escape">
<span class="sig-name descname"><span class="pre">prob_escape</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variants_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">concentrations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/polyclonal/polyclonal.html#Polyclonal.prob_escape"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polyclonal.polyclonal.Polyclonal.prob_escape" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute predicted probability of escape <span class="math notranslate nohighlight">\(p_v\left(c\right)\)</span>.</p>
<p>Computed using current mutation-escape values <span class="math notranslate nohighlight">\(\beta_{m,e}\)</span> and
epitope activities <span class="math notranslate nohighlight">\(a_{\rm{wt},e}\)</span> stored in this
<a class="reference internal" href="#polyclonal.polyclonal.Polyclonal" title="polyclonal.polyclonal.Polyclonal"><code class="xref py py-class docutils literal notranslate"><span class="pre">Polyclonal</span></code></a> object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>variants_df</strong> (<em>pandas.DataFrame</em>) – Input data frame defining variants. Should have a column
named ‘aa_substitutions’ that defines variants as space-delimited
strings of substitutions (e.g., ‘M1A K3T’). Should also have a
column ‘concentration’ if <code class="docutils literal notranslate"><span class="pre">concentrations=None</span></code>.</p></li>
<li><p><strong>concentrations</strong> (<em>array-like</em><em> or </em><em>None</em>) – Concentrations at which we compute probability of escape.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Version of <code class="docutils literal notranslate"><span class="pre">variants_df</span></code> with columns named ‘concentration’
and ‘predicted_prob_escape’ giving predicted probability of escape
<span class="math notranslate nohighlight">\(p_v\left(c\right)\)</span> for each variant at each concentration.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="polyclonal.polyclonal.Polyclonal.site_level_model">
<span class="sig-name descname"><span class="pre">site_level_model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aggregate_mut_escapes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'mean'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">collapse_identical_variants</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'mean'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/polyclonal/polyclonal.html#Polyclonal.site_level_model"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polyclonal.polyclonal.Polyclonal.site_level_model" title="Permalink to this definition">¶</a></dt>
<dd><p>Model with mutations collapsed at site level.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>aggregate_mut_escapes</strong> (<em>{'mean'}</em>) – How to aggregate mutation-level escape values to site-level
ones in <code class="docutils literal notranslate"><span class="pre">mut_escape_df</span></code>.</p></li>
<li><p><strong>collapse_identical_variants</strong> (<em>{&quot;mean&quot;</em><em>, </em><em>&quot;median&quot;</em><em>, </em><em>False}</em>) – Same meaning as for <a class="reference internal" href="#polyclonal.polyclonal.Polyclonal" title="polyclonal.polyclonal.Polyclonal"><code class="xref py py-class docutils literal notranslate"><span class="pre">Polyclonal</span></code></a> initialization.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#polyclonal.polyclonal.Polyclonal" title="polyclonal.polyclonal.Polyclonal"><code class="xref py py-class docutils literal notranslate"><span class="pre">Polyclonal</span></code></a></p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py exception">
<dt class="sig sig-object py" id="polyclonal.polyclonal.PolyclonalFitError">
<em class="property"><span class="pre">exception</span> </em><span class="sig-prename descclassname"><span class="pre">polyclonal.polyclonal.</span></span><span class="sig-name descname"><span class="pre">PolyclonalFitError</span></span><a class="reference internal" href="_modules/polyclonal/polyclonal.html#PolyclonalFitError"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polyclonal.polyclonal.PolyclonalFitError" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Exception</span></code></p>
<p>Error fitting in <a class="reference internal" href="#polyclonal.polyclonal.Polyclonal.fit" title="polyclonal.polyclonal.Polyclonal.fit"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Polyclonal.fit()</span></code></a>.</p>
</dd></dl>

<dl class="py exception">
<dt class="sig sig-object py" id="polyclonal.polyclonal.PolyclonalHarmonizeError">
<em class="property"><span class="pre">exception</span> </em><span class="sig-prename descclassname"><span class="pre">polyclonal.polyclonal.</span></span><span class="sig-name descname"><span class="pre">PolyclonalHarmonizeError</span></span><a class="reference internal" href="_modules/polyclonal/polyclonal.html#PolyclonalHarmonizeError"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polyclonal.polyclonal.PolyclonalHarmonizeError" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Exception</span></code></p>
<p>Error harmonizing epitopes in <a class="reference internal" href="#polyclonal.polyclonal.Polyclonal.epitope_harmonized_model" title="polyclonal.polyclonal.Polyclonal.epitope_harmonized_model"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Polyclonal.epitope_harmonized_model()</span></code></a>.</p>
</dd></dl>

</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="index.html">
    <img class="logo" src="_static/BloomLogo.jpg" alt="Logo"/>
    
    <h1 class="logo logo-name">polyclonal</h1>
    
  </a>
</p>



<p class="blurb">Model mutational escape from polyclonal antibodies.</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=jbloomlab&repo=polyclonal&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





<h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="index.html">polyclonal documentation</a></li>
</ul>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="biophysical_model.html">Biophysical model</a></li>
<li class="toctree-l1"><a class="reference internal" href="visualize_RBD.html">Visualize antibody mix</a></li>
<li class="toctree-l1"><a class="reference internal" href="simulate_RBD.html">Simulate experiment</a></li>
<li class="toctree-l1"><a class="reference internal" href="fit_RBD.html">Fit model to data</a></li>
<li class="toctree-l1"><a class="reference internal" href="regularization.html">Regularization</a></li>
<li class="toctree-l1"><a class="reference internal" href="RBD_bootstrap.html">Bootstrapping model fits</a></li>
<li class="toctree-l1"><a class="reference internal" href="RBD_average.html">Averaging models</a></li>
<li class="toctree-l1"><a class="reference internal" href="real_LyCoV1404.html">Real antibody data example</a></li>
<li class="toctree-l1"><a class="reference internal" href="reference_site_numbering.html">Reference site numbering</a></li>
<li class="toctree-l1"><a class="reference internal" href="real_cocktail.html">Real antibody cocktail example</a></li>
<li class="toctree-l1"><a class="reference internal" href="real_serum.html">Real serum example</a></li>
<li class="toctree-l1"><a class="reference internal" href="specify_epitopes.html">Specifying epitopes</a></li>
<li class="toctree-l1"><a class="reference internal" href="expt_design.html">Experimental design</a></li>
<li class="toctree-l1"><a class="reference internal" href="optimization.html">Details on fitting</a></li>
<li class="toctree-l1"><a class="reference internal" href="partition_function.html">Distinct epitopes assumption</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="polyclonal.html">polyclonal package</a></li>
<li class="toctree-l1"><a class="reference internal" href="package_index.html">package index</a></li>
<li class="toctree-l1"><a class="reference internal" href="acknowledgments.html">Acknowledgements</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="polyclonal.html">polyclonal package</a><ul>
      <li>Previous: <a href="polyclonal.plot.html" title="previous chapter">plot</a></li>
      <li>Next: <a href="polyclonal.polyclonal_collection.html" title="next chapter">polyclonal_collection</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2019--2022.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 4.2.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/polyclonal.polyclonal.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/jbloomlab/polyclonal" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>