
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

    <title>Simulate experiment &#8212; polyclonal 5.1 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css" />
    <link rel="stylesheet" type="text/css" href="_static/plot_directive.css" />
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/sphinx_highlight.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Fit model to data" href="fit_RBD.html" />
    <link rel="prev" title="Visualize antibody mix" href="visualize_RBD.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars and line breaks on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
    white-space: pre;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt .copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
.jp-RenderedHTMLCommon table,
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
.jp-RenderedHTMLCommon thead,
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
.jp-RenderedHTMLCommon tr,
.jp-RenderedHTMLCommon th,
.jp-RenderedHTMLCommon td,
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
.jp-RenderedHTMLCommon th,
div.rendered_html th {
  font-weight: bold;
}
.jp-RenderedHTMLCommon tbody tr:nth-child(odd),
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
.jp-RenderedHTMLCommon tbody tr:hover,
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}
</style>
<section id="Simulate-experiment">
<h1>Simulate experiment<a class="headerlink" href="#Simulate-experiment" title="Permalink to this heading">¶</a></h1>
<p>We will simulate data from a deep mutational scanning experiment using the RBD antibody mix in order to provide hypothetical data on which to fit a <code class="docutils literal notranslate"><span class="pre">Polyclonal</span></code> model.</p>
<section id="Simulate-three-epitopes">
<h2>Simulate three epitopes<a class="headerlink" href="#Simulate-three-epitopes" title="Permalink to this heading">¶</a></h2>
<p>We first define a <code class="docutils literal notranslate"><span class="pre">Polyclonal</span></code> object that represents the antibody mix we want to simulate. This mix is again based on the three RBD antibodies targeting class 1, 2, and 3 epitopes.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="kn">import</span> <span class="nn">polyclonal</span>

<span class="c1"># read data needed to initialize `Polyclonal` object</span>
<span class="n">activity_wt_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;RBD_activity_wt_df.csv&quot;</span><span class="p">)</span>
<span class="n">mut_escape_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;RBD_mut_escape_df.csv&quot;</span><span class="p">)</span>

<span class="c1"># create `Polyclonal` object with actual antibody mix</span>
<span class="n">model_3ep</span> <span class="o">=</span> <span class="n">polyclonal</span><span class="o">.</span><span class="n">Polyclonal</span><span class="p">(</span>
    <span class="n">activity_wt_df</span><span class="o">=</span><span class="n">activity_wt_df</span><span class="p">,</span> <span class="n">mut_escape_df</span><span class="o">=</span><span class="n">mut_escape_df</span>
<span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Epitopes: </span><span class="si">{</span><span class="n">model_3ep</span><span class="o">.</span><span class="n">epitopes</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Number of mutations: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">model_3ep</span><span class="o">.</span><span class="n">mutations</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Number of sites: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">model_3ep</span><span class="o">.</span><span class="n">sites</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Epitopes: (&#39;class 1&#39;, &#39;class 2&#39;, &#39;class 3&#39;)
Number of mutations: 1932
Number of sites: 173
</pre></div></div>
</div>
<p>Now we simulate libraries of variants with an average of 1, 2, 3, or 4 mutations per gene. These libraries only contain mutations for which we defined mutation escape above. We a <a class="reference external" href="https://jbloomlab.github.io/dms_variants/">dms_variants</a> <code class="docutils literal notranslate"><span class="pre">CodonVariantTable</span></code> for the RBD with these mutations. Note that because the <code class="docutils literal notranslate"><span class="pre">CodonVariantTable</span></code> requires sequential 1, 2, … numbering, we have to do some shifting of sites back and forth with an offset of 331 since that is where the RBD starts.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">Bio.SeqIO</span>

<span class="kn">import</span> <span class="nn">dms_variants.simulate</span>

<span class="kn">import</span> <span class="nn">polyclonal.utils</span>

<span class="c1"># read coding sequence, then slice to just region of interest</span>
<span class="c1"># noting that RBD starts at residue 331</span>
<span class="n">geneseq</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">Bio</span><span class="o">.</span><span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;RBD_seq.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>

<span class="n">allowed_aa_muts</span> <span class="o">=</span> <span class="n">model_3ep</span><span class="o">.</span><span class="n">mut_escape_df</span><span class="p">[</span><span class="s2">&quot;mutation&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;There are </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">allowed_aa_muts</span><span class="p">)</span><span class="si">}</span><span class="s2"> allowed amino-acid mutations.&quot;</span><span class="p">)</span>

<span class="n">variants</span> <span class="o">=</span> <span class="n">dms_variants</span><span class="o">.</span><span class="n">simulate</span><span class="o">.</span><span class="n">simulate_CodonVariantTable</span><span class="p">(</span>
    <span class="n">geneseq</span><span class="o">=</span><span class="n">geneseq</span><span class="p">,</span>
    <span class="n">bclen</span><span class="o">=</span><span class="mi">16</span><span class="p">,</span>
    <span class="n">library_specs</span><span class="o">=</span><span class="p">{</span>
        <span class="sa">f</span><span class="s2">&quot;avg</span><span class="si">{</span><span class="n">m</span><span class="si">}</span><span class="s2">muts&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;avgmuts&quot;</span><span class="p">:</span> <span class="n">m</span><span class="p">,</span> <span class="s2">&quot;nvariants&quot;</span><span class="p">:</span> <span class="mi">40000</span><span class="p">}</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">]</span>
    <span class="p">},</span>
    <span class="n">allowed_aa_muts</span><span class="o">=</span><span class="p">[</span><span class="n">polyclonal</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">shift_mut_site</span><span class="p">(</span><span class="n">m</span><span class="p">,</span> <span class="o">-</span><span class="mi">330</span><span class="p">)</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">allowed_aa_muts</span><span class="p">],</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
There are 1932 allowed amino-acid mutations.
</pre></div></div>
</div>
<p>Now look at the distribution of the number of amino-acid mutations per-variant in the library. The mutation rate is relatively high as we pre-screened for tolerated mutations and need multiple mutants to decompose the sera mix:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>

<span class="n">p</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">plotNumMutsHistogram</span><span class="p">(</span>
    <span class="n">mut_type</span><span class="o">=</span><span class="s2">&quot;aa&quot;</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">libraries</span><span class="o">=</span><span class="n">variants</span><span class="o">.</span><span class="n">libraries</span>
<span class="p">)</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<p>We also look at the mutation rate across the gene. Note that this is sequential numbering of the sequence. Also, the per-site mutation frequency is uneven because we only include tolerated mutations, and there are different number of tolerated mutations per site:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>

<span class="n">p</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">plotMutFreqs</span><span class="p">(</span>
    <span class="n">variant_type</span><span class="o">=</span><span class="s2">&quot;all&quot;</span><span class="p">,</span> <span class="n">mut_type</span><span class="o">=</span><span class="s2">&quot;aa&quot;</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">libraries</span><span class="o">=</span><span class="n">variants</span><span class="o">.</span><span class="n">libraries</span>
<span class="p">)</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<p>Now we get the data frame with the amino-acid substitutions for each variant, re-adding site offset to get back to RBD numbering:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">variants_df</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">barcode_variant_df</span><span class="p">[</span>
    <span class="p">[</span><span class="s2">&quot;library&quot;</span><span class="p">,</span> <span class="s2">&quot;barcode&quot;</span><span class="p">,</span> <span class="s2">&quot;aa_substitutions&quot;</span><span class="p">,</span> <span class="s2">&quot;n_aa_substitutions&quot;</span><span class="p">]</span>
<span class="p">]</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
    <span class="n">aa_substitutions</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;aa_substitutions&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
        <span class="n">polyclonal</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">shift_mut_site</span><span class="p">,</span> <span class="n">shift</span><span class="o">=</span><span class="mi">330</span>
    <span class="p">)</span>
<span class="p">)</span>

<span class="n">variants_df</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>library</th>
      <th>barcode</th>
      <th>aa_substitutions</th>
      <th>n_aa_substitutions</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>avg1muts</td>
      <td>AAAAAAAATTTACGCG</td>
      <td>F486P</td>
      <td>1</td>
    </tr>
    <tr>
      <th>1</th>
      <td>avg1muts</td>
      <td>AAAAAAATCCTCAATT</td>
      <td>T385A</td>
      <td>1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>avg1muts</td>
      <td>AAAAAACTATGCACCT</td>
      <td>Q498N</td>
      <td>1</td>
    </tr>
    <tr>
      <th>3</th>
      <td>avg1muts</td>
      <td>AAAAAAGACGCTGTGC</td>
      <td>P337S</td>
      <td>1</td>
    </tr>
    <tr>
      <th>4</th>
      <td>avg1muts</td>
      <td>AAAAAATAGTTCTGAT</td>
      <td>S371F R408V</td>
      <td>2</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>159995</th>
      <td>avg4muts</td>
      <td>TTTTTTGCGTTTATCT</td>
      <td>V367T</td>
      <td>1</td>
    </tr>
    <tr>
      <th>159996</th>
      <td>avg4muts</td>
      <td>TTTTTTTCCGCTCTAA</td>
      <td>Y369R C391T N440K G447C K462G</td>
      <td>5</td>
    </tr>
    <tr>
      <th>159997</th>
      <td>avg4muts</td>
      <td>TTTTTTTGTAAGCGAG</td>
      <td>N334L A348S K386S S443G Q493N V503S</td>
      <td>6</td>
    </tr>
    <tr>
      <th>159998</th>
      <td>avg4muts</td>
      <td>TTTTTTTGTTTTCAGC</td>
      <td>L335A P337S R357E Y396H E406W G485R T500E</td>
      <td>7</td>
    </tr>
    <tr>
      <th>159999</th>
      <td>avg4muts</td>
      <td>TTTTTTTTGCGGCCGT</td>
      <td>K444A P527S</td>
      <td>2</td>
    </tr>
  </tbody>
</table>
<p>160000 rows × 4 columns</p>
</div></div>
</div>
<p>Now we use our <code class="docutils literal notranslate"><span class="pre">Polyclonal</span></code> object to compute the predicted probability of escape (<span class="math notranslate nohighlight">\(p_v\left(c\right)\)</span>) at multiple serum concentrations given the mutation escape values <span class="math notranslate nohighlight">\(\beta_{m,e}\)</span> and activities <span class="math notranslate nohighlight">\(a_{\rm{wt},e}\)</span> stored in the <code class="docutils literal notranslate"><span class="pre">Polyclonal</span></code> object. Since we will use these as the “true” values in our our simulation, we then rename the column with the predicted probabilities of escape to just be the probabilities of escape:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">variants_escape</span> <span class="o">=</span> <span class="n">model_3ep</span><span class="o">.</span><span class="n">prob_escape</span><span class="p">(</span>
    <span class="n">variants_df</span><span class="o">=</span><span class="n">variants_df</span><span class="p">,</span> <span class="n">concentrations</span><span class="o">=</span><span class="p">[</span><span class="mf">0.125</span><span class="p">,</span> <span class="mf">0.25</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">]</span>
<span class="p">)</span>

<span class="n">variants_escape</span> <span class="o">=</span> <span class="n">variants_escape</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span>
    <span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;predicted_prob_escape&quot;</span><span class="p">:</span> <span class="s2">&quot;prob_escape&quot;</span><span class="p">}</span>
<span class="p">)</span>

<span class="n">variants_escape</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>library</th>
      <th>barcode</th>
      <th>aa_substitutions</th>
      <th>n_aa_substitutions</th>
      <th>concentration</th>
      <th>prob_escape</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>avg1muts</td>
      <td>AAAAAATGTTCTATCC</td>
      <td></td>
      <td>0</td>
      <td>0.125</td>
      <td>0.048594</td>
    </tr>
    <tr>
      <th>1</th>
      <td>avg1muts</td>
      <td>AAAAACAATCCGGACT</td>
      <td></td>
      <td>0</td>
      <td>0.125</td>
      <td>0.048594</td>
    </tr>
    <tr>
      <th>2</th>
      <td>avg1muts</td>
      <td>AAAAACGCGGTCACTT</td>
      <td></td>
      <td>0</td>
      <td>0.125</td>
      <td>0.048594</td>
    </tr>
    <tr>
      <th>3</th>
      <td>avg1muts</td>
      <td>AAAAACTTGGCTAGCT</td>
      <td></td>
      <td>0</td>
      <td>0.125</td>
      <td>0.048594</td>
    </tr>
    <tr>
      <th>4</th>
      <td>avg1muts</td>
      <td>AAAAAGCAAGGCCCAG</td>
      <td></td>
      <td>0</td>
      <td>0.125</td>
      <td>0.048594</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>959995</th>
      <td>avg2muts</td>
      <td>ATTCACCACCAGGTAG</td>
      <td>Y508W</td>
      <td>1</td>
      <td>4.000</td>
      <td>0.000006</td>
    </tr>
    <tr>
      <th>959996</th>
      <td>avg2muts</td>
      <td>TAGAATGTATTATTCT</td>
      <td>Y508W</td>
      <td>1</td>
      <td>4.000</td>
      <td>0.000006</td>
    </tr>
    <tr>
      <th>959997</th>
      <td>avg3muts</td>
      <td>CTTAAAATAGCTGGTC</td>
      <td>Y508W</td>
      <td>1</td>
      <td>4.000</td>
      <td>0.000006</td>
    </tr>
    <tr>
      <th>959998</th>
      <td>avg4muts</td>
      <td>GGTCAATTATGTCGGG</td>
      <td>Y508W</td>
      <td>1</td>
      <td>4.000</td>
      <td>0.000006</td>
    </tr>
    <tr>
      <th>959999</th>
      <td>avg1muts</td>
      <td>GGAACGACAGTGATCG</td>
      <td>Y508W T531H</td>
      <td>2</td>
      <td>4.000</td>
      <td>0.000006</td>
    </tr>
  </tbody>
</table>
<p>960000 rows × 6 columns</p>
</div></div>
</div>
<p>Plot some features of the variants, namely the distribution of probability of escape for each number of mutations. As expected, variants with many mutations are more likely to escape the serum:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>

<span class="kn">from</span> <span class="nn">plotnine</span> <span class="kn">import</span> <span class="o">*</span>

<span class="n">p</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">ggplot</span><span class="p">(</span>
        <span class="n">variants_escape</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
            <span class="n">n_aa_substitutions</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;n_aa_substitutions&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
                <span class="k">lambda</span> <span class="n">n</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">n</span><span class="p">)</span> <span class="k">if</span> <span class="n">n</span> <span class="o">&lt;=</span> <span class="mi">6</span> <span class="k">else</span> <span class="s2">&quot;&gt;6&quot;</span>
            <span class="p">),</span>
            <span class="n">concentration</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s2">&quot;concentration&quot;</span><span class="p">]),</span>
        <span class="p">)</span>
    <span class="p">)</span>
    <span class="o">+</span> <span class="n">aes</span><span class="p">(</span><span class="s2">&quot;n_aa_substitutions&quot;</span><span class="p">,</span> <span class="s2">&quot;prob_escape&quot;</span><span class="p">,</span> <span class="n">fill</span><span class="o">=</span><span class="s2">&quot;concentration&quot;</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">geom_boxplot</span><span class="p">(</span><span class="n">outlier_size</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">outlier_alpha</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">facet_wrap</span><span class="p">(</span><span class="s2">&quot;~ library&quot;</span><span class="p">,</span> <span class="n">ncol</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">theme_classic</span><span class="p">()</span>
    <span class="o">+</span> <span class="n">theme</span><span class="p">(</span><span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mi">9</span><span class="p">,</span> <span class="mi">7</span><span class="p">))</span>
<span class="p">)</span>

<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<p>Also, just compute the overall average probability of escape for each library:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="p">(</span>
    <span class="n">variants_escape</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s2">&quot;library&quot;</span><span class="p">,</span> <span class="s2">&quot;concentration&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span>
        <span class="p">{</span><span class="s2">&quot;prob_escape&quot;</span><span class="p">:</span> <span class="s2">&quot;mean&quot;</span><span class="p">}</span>
    <span class="p">)</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th></th>
      <th>prob_escape</th>
    </tr>
    <tr>
      <th>library</th>
      <th>concentration</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th rowspan="6" valign="top">avg1muts</th>
      <th>0.125</th>
      <td>0.086326</td>
    </tr>
    <tr>
      <th>0.250</th>
      <td>0.033071</td>
    </tr>
    <tr>
      <th>0.500</th>
      <td>0.012991</td>
    </tr>
    <tr>
      <th>1.000</th>
      <td>0.005777</td>
    </tr>
    <tr>
      <th>2.000</th>
      <td>0.003006</td>
    </tr>
    <tr>
      <th>4.000</th>
      <td>0.001809</td>
    </tr>
    <tr>
      <th rowspan="6" valign="top">avg2muts</th>
      <th>0.125</th>
      <td>0.125877</td>
    </tr>
    <tr>
      <th>0.250</th>
      <td>0.060015</td>
    </tr>
    <tr>
      <th>0.500</th>
      <td>0.030075</td>
    </tr>
    <tr>
      <th>1.000</th>
      <td>0.016740</td>
    </tr>
    <tr>
      <th>2.000</th>
      <td>0.010386</td>
    </tr>
    <tr>
      <th>4.000</th>
      <td>0.007009</td>
    </tr>
    <tr>
      <th rowspan="6" valign="top">avg3muts</th>
      <th>0.125</th>
      <td>0.165698</td>
    </tr>
    <tr>
      <th>0.250</th>
      <td>0.090470</td>
    </tr>
    <tr>
      <th>0.500</th>
      <td>0.052089</td>
    </tr>
    <tr>
      <th>1.000</th>
      <td>0.032763</td>
    </tr>
    <tr>
      <th>2.000</th>
      <td>0.022442</td>
    </tr>
    <tr>
      <th>4.000</th>
      <td>0.016411</td>
    </tr>
    <tr>
      <th rowspan="6" valign="top">avg4muts</th>
      <th>0.125</th>
      <td>0.204486</td>
    </tr>
    <tr>
      <th>0.250</th>
      <td>0.122522</td>
    </tr>
    <tr>
      <th>0.500</th>
      <td>0.077355</td>
    </tr>
    <tr>
      <th>1.000</th>
      <td>0.052786</td>
    </tr>
    <tr>
      <th>2.000</th>
      <td>0.038664</td>
    </tr>
    <tr>
      <th>4.000</th>
      <td>0.029815</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>We also want to acknowledge the fact that a real experiment will have some noise in the data. This noise is likely to come in two forms:</p>
<ol class="arabic simple">
<li><p>Inaccuracies in the measurements of the probabilities of escape, <span class="math notranslate nohighlight">\(p_v\left(c\right)\)</span>.</p></li>
<li><p>Occassional mis-assignment of which mutations are in a variant due to sequencing errors.</p></li>
</ol>
<p>We therefore will make a “noisy” version of <code class="docutils literal notranslate"><span class="pre">variants_df</span></code> where we have incorporated both of these sources of noise by adding Gaussian measurement error to the escape probabilities (but then truncating them to be between 0 and 1), and by adding or subtracting a mutation to occassional variants to represent sequencing errors.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">variants_escape</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>library</th>
      <th>barcode</th>
      <th>aa_substitutions</th>
      <th>n_aa_substitutions</th>
      <th>concentration</th>
      <th>prob_escape</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>avg1muts</td>
      <td>AAAAAATGTTCTATCC</td>
      <td></td>
      <td>0</td>
      <td>0.125</td>
      <td>0.048594</td>
    </tr>
    <tr>
      <th>1</th>
      <td>avg1muts</td>
      <td>AAAAACAATCCGGACT</td>
      <td></td>
      <td>0</td>
      <td>0.125</td>
      <td>0.048594</td>
    </tr>
    <tr>
      <th>2</th>
      <td>avg1muts</td>
      <td>AAAAACGCGGTCACTT</td>
      <td></td>
      <td>0</td>
      <td>0.125</td>
      <td>0.048594</td>
    </tr>
    <tr>
      <th>3</th>
      <td>avg1muts</td>
      <td>AAAAACTTGGCTAGCT</td>
      <td></td>
      <td>0</td>
      <td>0.125</td>
      <td>0.048594</td>
    </tr>
    <tr>
      <th>4</th>
      <td>avg1muts</td>
      <td>AAAAAGCAAGGCCCAG</td>
      <td></td>
      <td>0</td>
      <td>0.125</td>
      <td>0.048594</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>959995</th>
      <td>avg2muts</td>
      <td>ATTCACCACCAGGTAG</td>
      <td>Y508W</td>
      <td>1</td>
      <td>4.000</td>
      <td>0.000006</td>
    </tr>
    <tr>
      <th>959996</th>
      <td>avg2muts</td>
      <td>TAGAATGTATTATTCT</td>
      <td>Y508W</td>
      <td>1</td>
      <td>4.000</td>
      <td>0.000006</td>
    </tr>
    <tr>
      <th>959997</th>
      <td>avg3muts</td>
      <td>CTTAAAATAGCTGGTC</td>
      <td>Y508W</td>
      <td>1</td>
      <td>4.000</td>
      <td>0.000006</td>
    </tr>
    <tr>
      <th>959998</th>
      <td>avg4muts</td>
      <td>GGTCAATTATGTCGGG</td>
      <td>Y508W</td>
      <td>1</td>
      <td>4.000</td>
      <td>0.000006</td>
    </tr>
    <tr>
      <th>959999</th>
      <td>avg1muts</td>
      <td>GGAACGACAGTGATCG</td>
      <td>Y508W T531H</td>
      <td>2</td>
      <td>4.000</td>
      <td>0.000006</td>
    </tr>
  </tbody>
</table>
<p>960000 rows × 6 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">numpy</span>

<span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>  <span class="c1"># seed for reproducible output</span>


<span class="k">def</span> <span class="nf">add_subtract_mutation</span><span class="p">(</span><span class="n">subs</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Sometimes add or subtract a mutation.&quot;&quot;&quot;</span>
    <span class="n">subs</span> <span class="o">=</span> <span class="n">subs</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
    <span class="n">sub_sites</span> <span class="o">=</span> <span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">sub</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span> <span class="k">for</span> <span class="n">sub</span> <span class="ow">in</span> <span class="n">subs</span><span class="p">]</span>
    <span class="n">rand</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">random</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">rand</span> <span class="o">&lt;</span> <span class="mf">0.01</span><span class="p">:</span>  <span class="c1"># add mutation with 1% probability</span>
        <span class="n">mut</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">model_3ep</span><span class="o">.</span><span class="n">mutations</span><span class="p">)</span>
        <span class="k">while</span> <span class="nb">int</span><span class="p">(</span><span class="n">mut</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span> <span class="ow">in</span> <span class="n">sub_sites</span><span class="p">:</span>
            <span class="n">mut</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">model_3ep</span><span class="o">.</span><span class="n">mutations</span><span class="p">)</span>
        <span class="n">subs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mut</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">rand</span> <span class="o">&gt;</span> <span class="mf">0.99</span> <span class="ow">and</span> <span class="n">subs</span><span class="p">:</span>  <span class="c1"># subtract mutation with 1% probability</span>
        <span class="n">subs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">subs</span><span class="p">)))</span>
    <span class="k">return</span> <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">subs</span><span class="p">)</span>


<span class="c1"># only keep needed columns in variants_escape</span>
<span class="n">variants_escape</span> <span class="o">=</span> <span class="n">variants_escape</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="s2">&quot;n_aa_substitutions&quot;</span><span class="p">)</span>

<span class="c1"># simulate noisy variants</span>
<span class="n">noisy_variants_escape</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">variants_escape</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
        <span class="c1"># add Gaussian noise with standard deviation of 0.03</span>
        <span class="n">prob_escape</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="p">(</span>
            <span class="n">x</span><span class="p">[</span><span class="s2">&quot;prob_escape&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="n">scale</span><span class="o">=</span><span class="mf">0.03</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="p">))</span>
        <span class="p">)</span><span class="o">.</span><span class="n">clip</span><span class="p">(</span><span class="n">lower</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">upper</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span>
    <span class="p">)</span>
    <span class="c1"># add rare sequencing errors to variants</span>
    <span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="s2">&quot;aa_substitutions&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span>
        <span class="n">variants_df</span><span class="p">[[</span><span class="s2">&quot;library&quot;</span><span class="p">,</span> <span class="s2">&quot;barcode&quot;</span><span class="p">,</span> <span class="s2">&quot;aa_substitutions&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
            <span class="n">aa_substitutions</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;aa_substitutions&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">add_subtract_mutation</span><span class="p">)</span>
        <span class="p">),</span>
        <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">,</span>
        <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;library&quot;</span><span class="p">,</span> <span class="s2">&quot;barcode&quot;</span><span class="p">],</span>
        <span class="n">validate</span><span class="o">=</span><span class="s2">&quot;many_to_one&quot;</span><span class="p">,</span>
    <span class="p">)</span>
<span class="p">)</span>

<span class="n">noisy_variants_escape</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>library</th>
      <th>barcode</th>
      <th>concentration</th>
      <th>prob_escape</th>
      <th>aa_substitutions</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>avg1muts</td>
      <td>AAAAAATGTTCTATCC</td>
      <td>0.125</td>
      <td>0.097324</td>
      <td></td>
    </tr>
    <tr>
      <th>1</th>
      <td>avg1muts</td>
      <td>AAAAACAATCCGGACT</td>
      <td>0.125</td>
      <td>0.030241</td>
      <td></td>
    </tr>
    <tr>
      <th>2</th>
      <td>avg1muts</td>
      <td>AAAAACGCGGTCACTT</td>
      <td>0.125</td>
      <td>0.032749</td>
      <td></td>
    </tr>
    <tr>
      <th>3</th>
      <td>avg1muts</td>
      <td>AAAAACTTGGCTAGCT</td>
      <td>0.125</td>
      <td>0.016405</td>
      <td></td>
    </tr>
    <tr>
      <th>4</th>
      <td>avg1muts</td>
      <td>AAAAAGCAAGGCCCAG</td>
      <td>0.125</td>
      <td>0.074556</td>
      <td></td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>959995</th>
      <td>avg2muts</td>
      <td>ATTCACCACCAGGTAG</td>
      <td>4.000</td>
      <td>0.074859</td>
      <td>Y508W</td>
    </tr>
    <tr>
      <th>959996</th>
      <td>avg2muts</td>
      <td>TAGAATGTATTATTCT</td>
      <td>4.000</td>
      <td>0.000000</td>
      <td>Y508W</td>
    </tr>
    <tr>
      <th>959997</th>
      <td>avg3muts</td>
      <td>CTTAAAATAGCTGGTC</td>
      <td>4.000</td>
      <td>0.000000</td>
      <td>Y508W</td>
    </tr>
    <tr>
      <th>959998</th>
      <td>avg4muts</td>
      <td>GGTCAATTATGTCGGG</td>
      <td>4.000</td>
      <td>0.000000</td>
      <td>Y508W</td>
    </tr>
    <tr>
      <th>959999</th>
      <td>avg1muts</td>
      <td>GGAACGACAGTGATCG</td>
      <td>4.000</td>
      <td>0.000000</td>
      <td>Y508W T531H</td>
    </tr>
  </tbody>
</table>
<p>960000 rows × 5 columns</p>
</div></div>
</div>
<p>For both the exact and noisy data frames, we will also compute the IC90 for each variant based on the simulated antibody mix:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">variants_escape</span> <span class="o">=</span> <span class="n">model_3ep</span><span class="o">.</span><span class="n">icXX</span><span class="p">(</span><span class="n">variants_escape</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="mf">0.9</span><span class="p">,</span> <span class="n">col</span><span class="o">=</span><span class="s2">&quot;IC90&quot;</span><span class="p">)</span>

<span class="n">noisy_variants_escape</span> <span class="o">=</span> <span class="n">model_3ep</span><span class="o">.</span><span class="n">icXX</span><span class="p">(</span><span class="n">noisy_variants_escape</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="mf">0.9</span><span class="p">,</span> <span class="n">col</span><span class="o">=</span><span class="s2">&quot;IC90&quot;</span><span class="p">)</span>

<span class="n">noisy_variants_escape</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>library</th>
      <th>barcode</th>
      <th>concentration</th>
      <th>prob_escape</th>
      <th>aa_substitutions</th>
      <th>IC90</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>avg1muts</td>
      <td>AAAAAATGTTCTATCC</td>
      <td>0.125</td>
      <td>0.097324</td>
      <td></td>
      <td>0.082119</td>
    </tr>
    <tr>
      <th>1</th>
      <td>avg1muts</td>
      <td>AAAAACAATCCGGACT</td>
      <td>0.125</td>
      <td>0.030241</td>
      <td></td>
      <td>0.082119</td>
    </tr>
    <tr>
      <th>2</th>
      <td>avg1muts</td>
      <td>AAAAACGCGGTCACTT</td>
      <td>0.125</td>
      <td>0.032749</td>
      <td></td>
      <td>0.082119</td>
    </tr>
    <tr>
      <th>3</th>
      <td>avg1muts</td>
      <td>AAAAACTTGGCTAGCT</td>
      <td>0.125</td>
      <td>0.016405</td>
      <td></td>
      <td>0.082119</td>
    </tr>
    <tr>
      <th>4</th>
      <td>avg1muts</td>
      <td>AAAAAGCAAGGCCCAG</td>
      <td>0.125</td>
      <td>0.074556</td>
      <td></td>
      <td>0.082119</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>959995</th>
      <td>avg1muts</td>
      <td>GGAACGACAGTGATCG</td>
      <td>0.250</td>
      <td>0.000000</td>
      <td>Y508W T531H</td>
      <td>0.082119</td>
    </tr>
    <tr>
      <th>959996</th>
      <td>avg1muts</td>
      <td>GGAACGACAGTGATCG</td>
      <td>0.500</td>
      <td>0.000000</td>
      <td>Y508W T531H</td>
      <td>0.082119</td>
    </tr>
    <tr>
      <th>959997</th>
      <td>avg1muts</td>
      <td>GGAACGACAGTGATCG</td>
      <td>1.000</td>
      <td>0.000000</td>
      <td>Y508W T531H</td>
      <td>0.082119</td>
    </tr>
    <tr>
      <th>959998</th>
      <td>avg1muts</td>
      <td>GGAACGACAGTGATCG</td>
      <td>2.000</td>
      <td>0.021968</td>
      <td>Y508W T531H</td>
      <td>0.082119</td>
    </tr>
    <tr>
      <th>959999</th>
      <td>avg1muts</td>
      <td>GGAACGACAGTGATCG</td>
      <td>4.000</td>
      <td>0.000000</td>
      <td>Y508W T531H</td>
      <td>0.082119</td>
    </tr>
  </tbody>
</table>
<p>960000 rows × 6 columns</p>
</div></div>
</div>
<p>We will write these data frames giving the variants <span class="math notranslate nohighlight">\(v\)</span> and their probabilities of escape <span class="math notranslate nohighlight">\(p_v\left(c\right)\)</span> at each serum concentration <span class="math notranslate nohighlight">\(c\)</span> to CSV files:</p>
<ul class="simple">
<li><p><a class="reference external" href="RBD_variants_escape_exact.csv">RBD_variants_escape_exact.csv</a> for the measurements without noise</p></li>
<li><p><a class="reference external" href="RBD_variants_escape_noisy.csv">RBD_variants_escape_noisy.csv</a> for the measurements with realistic experimental noise</p></li>
</ul>
<p>The goal is to infer the properties of the escape mutations and the serum, <span class="math notranslate nohighlight">\(a_{\rm{wt},e}\)</span> and <span class="math notranslate nohighlight">\(\beta_{m,e}\)</span>, from these data, which are what would be measured in a deep mutational scanning experiment.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">variants_escape</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span>
    <span class="s2">&quot;RBD_variants_escape_exact.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">float_format</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">%.4g</span><span class="s2">&quot;</span>
<span class="p">)</span>

<span class="n">noisy_variants_escape</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span>
    <span class="s2">&quot;RBD_variants_escape_noisy.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">float_format</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">%.4g</span><span class="s2">&quot;</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="Simulate-library-with-2-epitopes-and-non-one-Hill-coefficient-and-non-zero-non-neutralizable-fraction">
<h2>Simulate library with 2 epitopes and non-one Hill coefficient and non-zero non-neutralizable fraction<a class="headerlink" href="#Simulate-library-with-2-epitopes-and-non-one-Hill-coefficient-and-non-zero-non-neutralizable-fraction" title="Permalink to this heading">¶</a></h2>
<p>Create a two-epitope model with the Hill coefficent not equal to one, and the non-neutralizable fraction greater than zero. To do this, we just keep two epitopes from the above model and rename them:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">model_2ep</span> <span class="o">=</span> <span class="n">polyclonal</span><span class="o">.</span><span class="n">Polyclonal</span><span class="p">(</span>
    <span class="n">activity_wt_df</span><span class="o">=</span><span class="n">activity_wt_df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s2">&quot;epitope != &#39;class 1&#39;&quot;</span><span class="p">),</span>
    <span class="n">mut_escape_df</span><span class="o">=</span><span class="n">mut_escape_df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s2">&quot;epitope != &#39;class 1&#39;&quot;</span><span class="p">),</span>
    <span class="n">hill_coefficient_df</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
        <span class="p">{</span><span class="s2">&quot;epitope&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;class 2&quot;</span><span class="p">,</span> <span class="s2">&quot;class 3&quot;</span><span class="p">],</span> <span class="s2">&quot;hill_coefficient&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mf">1.2</span><span class="p">,</span> <span class="mf">1.6</span><span class="p">]}</span>
    <span class="p">),</span>
    <span class="n">non_neutralized_frac_df</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
        <span class="p">{</span><span class="s2">&quot;epitope&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;class 2&quot;</span><span class="p">,</span> <span class="s2">&quot;class 3&quot;</span><span class="p">],</span> <span class="s2">&quot;non_neutralized_frac&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mf">0.01</span><span class="p">,</span> <span class="mf">0.05</span><span class="p">]},</span>
    <span class="p">),</span>
<span class="p">)</span>

<span class="n">model_2ep</span><span class="o">.</span><span class="n">curve_specs_df</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>epitope</th>
      <th>activity</th>
      <th>hill_coefficient</th>
      <th>non_neutralized_frac</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>class 2</td>
      <td>3.2</td>
      <td>1.2</td>
      <td>0.01</td>
    </tr>
    <tr>
      <th>1</th>
      <td>class 3</td>
      <td>2.4</td>
      <td>1.6</td>
      <td>0.05</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Now compute probability of escape for 2-mutation variants, adding some noise. Also compute the true IC90 from the model. This is the simulated library:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">variants_2ep</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">model_2ep</span><span class="o">.</span><span class="n">prob_escape</span><span class="p">(</span>
        <span class="n">variants_df</span><span class="o">=</span><span class="n">variants_df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s2">&quot;library == &#39;avg2muts&#39;&quot;</span><span class="p">),</span>
        <span class="n">concentrations</span><span class="o">=</span><span class="p">[</span><span class="mf">0.25</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">],</span>
    <span class="p">)</span>
    <span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;predicted_prob_escape&quot;</span><span class="p">:</span> <span class="s2">&quot;prob_escape&quot;</span><span class="p">})</span>
    <span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;n_aa_substitutions&quot;</span><span class="p">,</span> <span class="s2">&quot;library&quot;</span><span class="p">])</span>
    <span class="o">.</span><span class="n">assign</span><span class="p">(</span>
        <span class="c1"># add Gaussian noise with standard deviation of 0.03</span>
        <span class="n">prob_escape</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="p">(</span>
            <span class="n">x</span><span class="p">[</span><span class="s2">&quot;prob_escape&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="n">scale</span><span class="o">=</span><span class="mf">0.03</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="p">))</span>
        <span class="p">)</span><span class="o">.</span><span class="n">clip</span><span class="p">(</span><span class="n">lower</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">upper</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span>
        <span class="c1"># add rare sequencing errors</span>
        <span class="n">aa_substitutions</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;aa_substitutions&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">add_subtract_mutation</span><span class="p">),</span>
    <span class="p">)</span>
<span class="p">)</span>

<span class="n">variants_2ep</span> <span class="o">=</span> <span class="n">model_2ep</span><span class="o">.</span><span class="n">icXX</span><span class="p">(</span><span class="n">variants_2ep</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="mf">0.9</span><span class="p">,</span> <span class="n">col</span><span class="o">=</span><span class="s2">&quot;true IC90&quot;</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Average escape at each concentration:&quot;</span><span class="p">)</span>
<span class="n">display</span><span class="p">(</span><span class="n">variants_2ep</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;concentration&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">aggregate</span><span class="p">({</span><span class="s2">&quot;prob_escape&quot;</span><span class="p">:</span> <span class="s2">&quot;mean&quot;</span><span class="p">}))</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Average escape at each concentration:
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>prob_escape</th>
    </tr>
    <tr>
      <th>concentration</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0.25</th>
      <td>0.070979</td>
    </tr>
    <tr>
      <th>1.00</th>
      <td>0.033363</td>
    </tr>
    <tr>
      <th>4.00</th>
      <td>0.026064</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>We write details on the 2-epitope model to files:</p>
<ul class="simple">
<li><p><a class="reference external" href="2epitope_model_curve_specs.csv">2epitope_model_curve_specs.csv</a> the neutralization curve parameters</p></li>
<li><p><a class="reference external" href="2epitope_escape.csv">2epitope_escape.csv</a> the variants and their measured probability of escape</p></li>
</ul>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">model_2ep</span><span class="o">.</span><span class="n">curve_specs_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;2epitope_model_curve_specs.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="n">variants_2ep</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;2epitope_escape.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">float_format</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">%.4g</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="index.html">
    <img class="logo" src="_static/BloomLogo.jpg" alt="Logo"/>
    
    <h1 class="logo logo-name">polyclonal</h1>
    
  </a>
</p>



<p class="blurb">Model mutational escape from polyclonal antibodies.</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=jbloomlab&repo=polyclonal&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





<h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="index.html">polyclonal documentation</a></li>
</ul>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="biophysical_model.html">Biophysical model</a></li>
<li class="toctree-l1"><a class="reference internal" href="visualize_RBD.html">Visualize antibody mix</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Simulate experiment</a></li>
<li class="toctree-l1"><a class="reference internal" href="fit_RBD.html">Fit model to data</a></li>
<li class="toctree-l1"><a class="reference internal" href="fit_2epitope.html">Fit more complex curves</a></li>
<li class="toctree-l1"><a class="reference internal" href="regularization.html">Regularization</a></li>
<li class="toctree-l1"><a class="reference internal" href="RBD_average.html">Averaging models</a></li>
<li class="toctree-l1"><a class="reference internal" href="real_LyCoV1404.html">Real antibody example</a></li>
<li class="toctree-l1"><a class="reference internal" href="reference_site_numbering.html">Reference site numbering</a></li>
<li class="toctree-l1"><a class="reference internal" href="real_cocktail.html">Real antibody cocktail</a></li>
<li class="toctree-l1"><a class="reference internal" href="specify_epitopes.html">Specifying epitopes</a></li>
<li class="toctree-l1"><a class="reference internal" href="expt_design.html">Experimental design</a></li>
<li class="toctree-l1"><a class="reference internal" href="optimization.html">Details on fitting</a></li>
<li class="toctree-l1"><a class="reference internal" href="partition_function.html">Distinct epitopes assumption</a></li>
<li class="toctree-l1"><a class="reference internal" href="polyclonal.html">polyclonal package</a></li>
<li class="toctree-l1"><a class="reference internal" href="package_index.html">package index</a></li>
<li class="toctree-l1"><a class="reference internal" href="acknowledgments.html">Acknowledgements</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="visualize_RBD.html" title="previous chapter">Visualize antibody mix</a></li>
      <li>Next: <a href="fit_RBD.html" title="next chapter">Fit model to data</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2019--2023.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 6.1.3</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.13</a>
      
      |
      <a href="_sources/simulate_RBD.nblink.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/jbloomlab/polyclonal" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>