
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Specifying epitopes &#8212; polyclonal 3.1 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css" />
    <link rel="stylesheet" type="text/css" href="_static/plot_directive.css" />
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Experimental design" href="expt_design.html" />
    <link rel="prev" title="Real serum example" href="real_serum.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt a.copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}
</style>
<section id="Specifying-epitopes">
<h1>Specifying epitopes<a class="headerlink" href="#Specifying-epitopes" title="Permalink to this headline">¶</a></h1>
<p>Polyclonal sera contains antibodies that can target multiple distinct epitopes on a viral protein, but we do not know the exact composition of this sera and which epitopes are targeted. At minimum, <code class="docutils literal notranslate"><span class="pre">Polyclonal</span></code> requires a user to specify the number of epitopes prior to fitting. Alternatively, a user may opt to seed initial guesses for the mutation-escape values of key epitope sites.</p>
<p>Here, we’ll use simulated data to show how the number of epitopes and seeding of key sites can affect the performance of <code class="docutils literal notranslate"><span class="pre">Polyclonal</span></code> models. We’ll also use this to provide guidance on how to systematically determine the optimal number of epitopes.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">itertools</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">pickle</span>

<span class="kn">import</span> <span class="nn">altair</span> <span class="k">as</span> <span class="nn">alt</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">polyclonal</span>

<span class="kn">from</span> <span class="nn">plotnine</span> <span class="kn">import</span> <span class="o">*</span>
</pre></div>
</div>
</div>
<p>First, we read in a simulated “noisy” dataset containing variants measured at three different sera concentrations. The variants in this library were simulated to contain a Poisson-distributed number of mutations, with an average of three mutations per gene.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">noisy_data</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;RBD_variants_escape_noisy.csv&quot;</span><span class="p">,</span> <span class="n">na_filter</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
    <span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">&#39;library == &quot;avg3muts&quot;&#39;</span><span class="p">)</span>
    <span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s2">&quot;concentration in [0.25, 1, 4]&quot;</span><span class="p">)</span>
    <span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="p">)</span>
<span class="n">noisy_data</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>library</th>
      <th>barcode</th>
      <th>concentration</th>
      <th>prob_escape</th>
      <th>aa_substitutions</th>
      <th>IC90</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>avg3muts</td>
      <td>AAAACTGCTGAGGAGA</td>
      <td>0.25</td>
      <td>0.054700</td>
      <td></td>
      <td>0.08212</td>
    </tr>
    <tr>
      <th>1</th>
      <td>avg3muts</td>
      <td>AAAAGCAGGCTACTCT</td>
      <td>0.25</td>
      <td>0.000000</td>
      <td></td>
      <td>0.08212</td>
    </tr>
    <tr>
      <th>2</th>
      <td>avg3muts</td>
      <td>AAAAGCTATAGGTGCC</td>
      <td>0.25</td>
      <td>0.007613</td>
      <td></td>
      <td>0.08212</td>
    </tr>
    <tr>
      <th>3</th>
      <td>avg3muts</td>
      <td>AAAAGGTATTAGTGGC</td>
      <td>0.25</td>
      <td>0.001363</td>
      <td></td>
      <td>0.08212</td>
    </tr>
    <tr>
      <th>4</th>
      <td>avg3muts</td>
      <td>AAAAGTGCCTTCGTTA</td>
      <td>0.25</td>
      <td>0.000000</td>
      <td></td>
      <td>0.08212</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>119995</th>
      <td>avg3muts</td>
      <td>GAGCATGATCGACGAA</td>
      <td>1.00</td>
      <td>0.000000</td>
      <td>Y508V H519I</td>
      <td>0.10830</td>
    </tr>
    <tr>
      <th>119996</th>
      <td>avg3muts</td>
      <td>GAGCATGATCGACGAA</td>
      <td>4.00</td>
      <td>0.000000</td>
      <td>Y508V H519I</td>
      <td>0.10830</td>
    </tr>
    <tr>
      <th>119997</th>
      <td>avg3muts</td>
      <td>CTTAAAATAGCTGGTC</td>
      <td>0.25</td>
      <td>0.000000</td>
      <td>Y508W</td>
      <td>0.08212</td>
    </tr>
    <tr>
      <th>119998</th>
      <td>avg3muts</td>
      <td>CTTAAAATAGCTGGTC</td>
      <td>1.00</td>
      <td>0.012260</td>
      <td>Y508W</td>
      <td>0.08212</td>
    </tr>
    <tr>
      <th>119999</th>
      <td>avg3muts</td>
      <td>CTTAAAATAGCTGGTC</td>
      <td>4.00</td>
      <td>0.000000</td>
      <td>Y508W</td>
      <td>0.08212</td>
    </tr>
  </tbody>
</table>
<p>120000 rows × 6 columns</p>
</div></div>
</div>
<p>We’ll also read in the “true” mutation-escape values, <span class="math notranslate nohighlight">\(\beta_{m,e}\)</span>, as a reference for later.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">true_mut_escape_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;RBD_mut_escape_df.csv&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span>
    <span class="s2">&quot;class &quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">regex</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>
<span class="n">true_mut_escape_df</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>epitope</th>
      <th>mutation</th>
      <th>escape</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>1</td>
      <td>N331A</td>
      <td>0.000</td>
    </tr>
    <tr>
      <th>1</th>
      <td>1</td>
      <td>N331D</td>
      <td>0.000</td>
    </tr>
    <tr>
      <th>2</th>
      <td>1</td>
      <td>N331E</td>
      <td>0.000</td>
    </tr>
    <tr>
      <th>3</th>
      <td>1</td>
      <td>N331F</td>
      <td>0.000</td>
    </tr>
    <tr>
      <th>4</th>
      <td>1</td>
      <td>N331G</td>
      <td>1.701</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>5791</th>
      <td>3</td>
      <td>T531R</td>
      <td>0.000</td>
    </tr>
    <tr>
      <th>5792</th>
      <td>3</td>
      <td>T531S</td>
      <td>0.000</td>
    </tr>
    <tr>
      <th>5793</th>
      <td>3</td>
      <td>T531V</td>
      <td>0.000</td>
    </tr>
    <tr>
      <th>5794</th>
      <td>3</td>
      <td>T531W</td>
      <td>0.000</td>
    </tr>
    <tr>
      <th>5795</th>
      <td>3</td>
      <td>T531Y</td>
      <td>0.000</td>
    </tr>
  </tbody>
</table>
<p>5796 rows × 3 columns</p>
</div></div>
</div>
<p>Additionally, we’ll make a directory for storing our fit models as <a class="reference external" href="https://docs.python.org/3/library/pickle.html#module-pickle">pickle</a> files, so that we can conveniently load them in the future without having to fit again.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="s2">&quot;fit_polyclonal_models&quot;</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<section id="Specifying-epitopes-without-prior-knowledge">
<h2>Specifying epitopes without prior knowledge<a class="headerlink" href="#Specifying-epitopes-without-prior-knowledge" title="Permalink to this headline">¶</a></h2>
<p>First, we’ll simulate a case in which we have no prior knowledge about which sites confer antibody escape. Thus, we will only supply guesses for the number of epitopes targeted by the polyclonal sera. We can guess that there are probably at least 2 epitopes targeted by polyclonal sera, so we’ll start by specifying 2 epitopes and iteratively fit <code class="docutils literal notranslate"><span class="pre">Polyclonal</span></code> models with increasing number of epitopes.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">n_epitopes</span> <span class="o">=</span> <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">]</span>

<span class="c1"># store fit models for future lookup</span>
<span class="n">fit_mut_escape</span> <span class="o">=</span> <span class="p">{}</span>

<span class="k">for</span> <span class="n">e</span> <span class="ow">in</span> <span class="n">n_epitopes</span><span class="p">:</span>
    <span class="c1"># The key for the fit model</span>
    <span class="n">model_string</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;noisy_[0.25, 1, 4]conc_3muts_</span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">epitopes_noinit&quot;</span>

    <span class="c1"># If the pickled model exists in the fit_polyclonal_models directory, load it.</span>
    <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;fit_polyclonal_models/</span><span class="si">{</span><span class="n">model_string</span><span class="si">}</span><span class="s2">.pkl&quot;</span><span class="p">)</span> <span class="ow">is</span> <span class="kc">True</span><span class="p">:</span>
        <span class="n">model</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;fit_polyclonal_models/</span><span class="si">{</span><span class="n">model_string</span><span class="si">}</span><span class="s2">.pkl&quot;</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">))</span>
        <span class="n">fit_mut_escape</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="n">model_string</span><span class="p">:</span> <span class="n">model</span><span class="p">})</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Model with </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2"> epitopes specified was already fit.&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># Else, fit a model and save it to the fit_polyclonal_models directory.</span>
        <span class="n">model</span> <span class="o">=</span> <span class="n">polyclonal</span><span class="o">.</span><span class="n">Polyclonal</span><span class="p">(</span>
            <span class="n">data_to_fit</span><span class="o">=</span><span class="n">noisy_data</span><span class="p">,</span>
            <span class="n">n_epitopes</span><span class="o">=</span><span class="n">e</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">opt_res</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span>
            <span class="n">logfreq</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span>
            <span class="n">reg_escape_weight</span><span class="o">=</span><span class="mf">0.01</span><span class="p">,</span>
            <span class="n">reg_uniqueness2_weight</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">fit_mut_escape</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="n">model_string</span><span class="p">:</span> <span class="n">model</span><span class="p">})</span>
        <span class="n">pickle</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;fit_polyclonal_models/</span><span class="si">{</span><span class="n">model_string</span><span class="si">}</span><span class="s2">.pkl&quot;</span><span class="p">,</span> <span class="s2">&quot;wb&quot;</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model with 2 epitopes specified was already fit.
Model with 3 epitopes specified was already fit.
Model with 4 epitopes specified was already fit.
</pre></div></div>
</div>
<p>We can correlate the <span class="math notranslate nohighlight">\(\beta_{m,e}\)</span> values for each of the epitopes fit by <code class="docutils literal notranslate"><span class="pre">Polyclonal</span></code> to the <span class="math notranslate nohighlight">\(\beta_{m,e}\)</span> values for each of the three “true” epitopes. These <span class="math notranslate nohighlight">\(\beta_{m,e}\)</span> values represent the extent to which mutations mediate escape from specific epitopes.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">plot_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
<span class="k">for</span> <span class="n">e</span> <span class="ow">in</span> <span class="n">n_epitopes</span><span class="p">:</span>
    <span class="n">model</span> <span class="o">=</span> <span class="n">fit_mut_escape</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;noisy_[0.25, 1, 4]conc_3muts_</span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">epitopes_noinit&quot;</span><span class="p">]</span>

    <span class="c1"># get correlation of mutation-escape values between all pairwise combinations of fit vs. true epitopes</span>
    <span class="k">for</span> <span class="n">true_e</span><span class="p">,</span> <span class="n">fit_e</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span>
        <span class="n">itertools</span><span class="o">.</span><span class="n">product</span><span class="p">(</span>
            <span class="n">true_mut_escape_df</span><span class="p">[</span><span class="s2">&quot;epitope&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">(),</span>
            <span class="n">model</span><span class="o">.</span><span class="n">mut_escape_df</span><span class="p">[</span><span class="s2">&quot;epitope&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">(),</span>
        <span class="p">)</span>
    <span class="p">):</span>

        <span class="n">df</span> <span class="o">=</span> <span class="n">true_mut_escape_df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;epitope == &quot;</span><span class="si">{</span><span class="n">true_e</span><span class="si">}</span><span class="s1">&quot;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span>
            <span class="p">(</span>
                <span class="n">model</span><span class="o">.</span><span class="n">mut_escape_df</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span>
                    <span class="n">columns</span><span class="o">=</span><span class="p">{</span>
                        <span class="s2">&quot;escape&quot;</span><span class="p">:</span> <span class="s2">&quot;predicted escape&quot;</span><span class="p">,</span>
                        <span class="s2">&quot;epitope&quot;</span><span class="p">:</span> <span class="s2">&quot;predicted epitope&quot;</span><span class="p">,</span>
                    <span class="p">}</span>
                <span class="p">)</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;`predicted epitope` == &quot;</span><span class="si">{</span><span class="n">fit_e</span><span class="si">}</span><span class="s1">&quot;&#39;</span><span class="p">)</span>
            <span class="p">),</span>
            <span class="n">on</span><span class="o">=</span><span class="s2">&quot;mutation&quot;</span><span class="p">,</span>
            <span class="n">validate</span><span class="o">=</span><span class="s2">&quot;one_to_one&quot;</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="n">corr</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;escape&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">corr</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;predicted escape&quot;</span><span class="p">]</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span>

        <span class="n">plot_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span>
            <span class="p">[</span>
                <span class="n">plot_data</span><span class="p">,</span>
                <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
                    <span class="n">data</span><span class="o">=</span><span class="p">{</span>
                        <span class="s2">&quot;Number of fit epitopes&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">e</span><span class="p">],</span>
                        <span class="s2">&quot;True epitope class&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">true_e</span><span class="p">],</span>
                        <span class="s2">&quot;Fit epitope class&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">fit_e</span><span class="p">],</span>
                        <span class="s2">&quot;Correlation (R^2)&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">corr</span><span class="p">],</span>
                    <span class="p">}</span>
                <span class="p">),</span>
            <span class="p">]</span>
        <span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">alt</span><span class="o">.</span><span class="n">Chart</span><span class="p">(</span><span class="n">plot_data</span><span class="p">)</span><span class="o">.</span><span class="n">mark_rect</span><span class="p">(</span><span class="n">opacity</span><span class="o">=</span><span class="mf">0.95</span><span class="p">,</span> <span class="n">stroke</span><span class="o">=</span><span class="s2">&quot;black&quot;</span><span class="p">,)</span><span class="o">.</span><span class="n">encode</span><span class="p">(</span>
    <span class="n">x</span><span class="o">=</span><span class="s2">&quot;Fit epitope class:O&quot;</span><span class="p">,</span>
    <span class="n">y</span><span class="o">=</span><span class="s2">&quot;True epitope class:O&quot;</span><span class="p">,</span>
    <span class="n">color</span><span class="o">=</span><span class="n">alt</span><span class="o">.</span><span class="n">Color</span><span class="p">(</span><span class="s2">&quot;Correlation (R^2):Q&quot;</span><span class="p">,</span> <span class="n">scale</span><span class="o">=</span><span class="n">alt</span><span class="o">.</span><span class="n">Scale</span><span class="p">(</span><span class="n">scheme</span><span class="o">=</span><span class="s2">&quot;viridis&quot;</span><span class="p">)),</span>
    <span class="n">tooltip</span><span class="o">=</span><span class="p">[</span>
        <span class="s2">&quot;Fit epitope class:O&quot;</span><span class="p">,</span>
        <span class="s2">&quot;True epitope class:O&quot;</span><span class="p">,</span>
        <span class="n">alt</span><span class="o">.</span><span class="n">Tooltip</span><span class="p">(</span><span class="s2">&quot;Correlation (R^2)&quot;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;.3f&quot;</span><span class="p">),</span>
    <span class="p">],</span>
<span class="p">)</span><span class="o">.</span><span class="n">properties</span><span class="p">(</span><span class="n">width</span><span class="o">=</span><span class="mi">130</span><span class="p">,</span> <span class="n">height</span><span class="o">=</span><span class="mi">100</span><span class="p">,)</span><span class="o">.</span><span class="n">facet</span><span class="p">(</span>
    <span class="s2">&quot;Number of fit epitopes&quot;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/tyu2/.local/lib/python3.8/site-packages/altair/utils/core.py:317: FutureWarning: iteritems is deprecated and will be removed in a future version. Use .items instead.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div id="altair-viz-914a441a5a754e7ea1bd9d5952d6b399"></div>
<script type="text/javascript">
  var VEGA_DEBUG = (typeof VEGA_DEBUG == "undefined") ? {} : VEGA_DEBUG;
  (function(spec, embedOpt){
    let outputDiv = document.currentScript.previousElementSibling;
    if (outputDiv.id !== "altair-viz-914a441a5a754e7ea1bd9d5952d6b399") {
      outputDiv = document.getElementById("altair-viz-914a441a5a754e7ea1bd9d5952d6b399");
    }
    const paths = {
      "vega": "https://cdn.jsdelivr.net/npm//vega@5?noext",
      "vega-lib": "https://cdn.jsdelivr.net/npm//vega-lib?noext",
      "vega-lite": "https://cdn.jsdelivr.net/npm//vega-lite@5.2.0?noext",
      "vega-embed": "https://cdn.jsdelivr.net/npm//vega-embed@6?noext",
    };

    function maybeLoadScript(lib, version) {
      var key = `${lib.replace("-", "")}_version`;
      return (VEGA_DEBUG[key] == version) ?
        Promise.resolve(paths[lib]) :
        new Promise(function(resolve, reject) {
          var s = document.createElement('script');
          document.getElementsByTagName("head")[0].appendChild(s);
          s.async = true;
          s.onload = () => {
            VEGA_DEBUG[key] = version;
            return resolve(paths[lib]);
          };
          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);
          s.src = paths[lib];
        });
    }

    function showError(err) {
      outputDiv.innerHTML = `<div class="error" style="color:red;">${err}</div>`;
      throw err;
    }

    function displayChart(vegaEmbed) {
      vegaEmbed(outputDiv, spec, embedOpt)
        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));
    }

    if(typeof define === "function" && define.amd) {
      requirejs.config({paths});
      require(["vega-embed"], displayChart, err => showError(`Error loading script: ${err.message}`));
    } else {
      maybeLoadScript("vega", "5")
        .then(() => maybeLoadScript("vega-lite", "5.2.0"))
        .then(() => maybeLoadScript("vega-embed", "6"))
        .catch(showError)
        .then(() => displayChart(vegaEmbed));
    }
  })({"config": {"view": {"continuousWidth": 400, "continuousHeight": 300}}, "data": {"name": "data-4a4d86f009e05898a8cc6bcfb030112f"}, "facet": {"field": "Number of fit epitopes", "type": "quantitative"}, "spec": {"mark": {"type": "rect", "opacity": 0.95, "stroke": "black"}, "encoding": {"color": {"field": "Correlation (R^2)", "scale": {"scheme": "viridis"}, "type": "quantitative"}, "tooltip": [{"field": "Fit epitope class", "type": "ordinal"}, {"field": "True epitope class", "type": "ordinal"}, {"field": "Correlation (R^2)", "format": ".3f", "type": "quantitative"}], "x": {"field": "Fit epitope class", "type": "ordinal"}, "y": {"field": "True epitope class", "type": "ordinal"}}, "height": 100, "width": 130}, "$schema": "https://vega.github.io/schema/vega-lite/v5.2.0.json", "datasets": {"data-4a4d86f009e05898a8cc6bcfb030112f": [{"Number of fit epitopes": 2, "True epitope class": "1", "Fit epitope class": "1", "Correlation (R^2)": -0.07933196078372097}, {"Number of fit epitopes": 2, "True epitope class": "1", "Fit epitope class": "2", "Correlation (R^2)": 0.14026078544320103}, {"Number of fit epitopes": 2, "True epitope class": "2", "Fit epitope class": "1", "Correlation (R^2)": -0.05838225069214065}, {"Number of fit epitopes": 2, "True epitope class": "2", "Fit epitope class": "2", "Correlation (R^2)": 0.8329342819237245}, {"Number of fit epitopes": 2, "True epitope class": "3", "Fit epitope class": "1", "Correlation (R^2)": 0.9310395874345984}, {"Number of fit epitopes": 2, "True epitope class": "3", "Fit epitope class": "2", "Correlation (R^2)": -0.05679500705590149}, {"Number of fit epitopes": 3, "True epitope class": "1", "Fit epitope class": "1", "Correlation (R^2)": 0.05932797106008702}, {"Number of fit epitopes": 3, "True epitope class": "1", "Fit epitope class": "2", "Correlation (R^2)": 0.8034772146578196}, {"Number of fit epitopes": 3, "True epitope class": "1", "Fit epitope class": "3", "Correlation (R^2)": -0.08978254555367132}, {"Number of fit epitopes": 3, "True epitope class": "2", "Fit epitope class": "1", "Correlation (R^2)": 0.9683604199853935}, {"Number of fit epitopes": 3, "True epitope class": "2", "Fit epitope class": "2", "Correlation (R^2)": 0.2002915394609174}, {"Number of fit epitopes": 3, "True epitope class": "2", "Fit epitope class": "3", "Correlation (R^2)": -0.057752566089737674}, {"Number of fit epitopes": 3, "True epitope class": "3", "Fit epitope class": "1", "Correlation (R^2)": -0.058457093398139995}, {"Number of fit epitopes": 3, "True epitope class": "3", "Fit epitope class": "2", "Correlation (R^2)": -0.07297049359510571}, {"Number of fit epitopes": 3, "True epitope class": "3", "Fit epitope class": "3", "Correlation (R^2)": 0.968577771568023}, {"Number of fit epitopes": 4, "True epitope class": "1", "Fit epitope class": "1", "Correlation (R^2)": 0.05844472475558725}, {"Number of fit epitopes": 4, "True epitope class": "1", "Fit epitope class": "2", "Correlation (R^2)": 0.024333631899603216}, {"Number of fit epitopes": 4, "True epitope class": "1", "Fit epitope class": "3", "Correlation (R^2)": 0.8104243392070462}, {"Number of fit epitopes": 4, "True epitope class": "1", "Fit epitope class": "4", "Correlation (R^2)": -0.09086533707859155}, {"Number of fit epitopes": 4, "True epitope class": "2", "Fit epitope class": "1", "Correlation (R^2)": 0.9482396080367836}, {"Number of fit epitopes": 4, "True epitope class": "2", "Fit epitope class": "2", "Correlation (R^2)": 0.317110482176994}, {"Number of fit epitopes": 4, "True epitope class": "2", "Fit epitope class": "3", "Correlation (R^2)": 0.20501454517936693}, {"Number of fit epitopes": 4, "True epitope class": "2", "Fit epitope class": "4", "Correlation (R^2)": -0.05800983471876021}, {"Number of fit epitopes": 4, "True epitope class": "3", "Fit epitope class": "1", "Correlation (R^2)": -0.05783996213533368}, {"Number of fit epitopes": 4, "True epitope class": "3", "Fit epitope class": "2", "Correlation (R^2)": 0.0686254822938297}, {"Number of fit epitopes": 4, "True epitope class": "3", "Fit epitope class": "3", "Correlation (R^2)": -0.07255798523416271}, {"Number of fit epitopes": 4, "True epitope class": "3", "Fit epitope class": "4", "Correlation (R^2)": 0.9659369637182172}]}}, {"mode": "vega-lite"});
</script></div>
</div>
<p>We can also visualize the wildtype activity values, <span class="math notranslate nohighlight">\(a_{wt,e}\)</span>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="k">for</span> <span class="n">e</span> <span class="ow">in</span> <span class="n">n_epitopes</span><span class="p">:</span>
    <span class="n">model</span> <span class="o">=</span> <span class="n">fit_mut_escape</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;noisy_[0.25, 1, 4]conc_3muts_</span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">epitopes_noinit&quot;</span><span class="p">]</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Number of fit epitopes: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">display</span><span class="p">(</span><span class="n">model</span><span class="o">.</span><span class="n">activity_wt_barplot</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Number of fit epitopes: 2
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div id="altair-viz-5cedfaee897e475bb40cafa278b705ac"></div>
<script type="text/javascript">
  var VEGA_DEBUG = (typeof VEGA_DEBUG == "undefined") ? {} : VEGA_DEBUG;
  (function(spec, embedOpt){
    let outputDiv = document.currentScript.previousElementSibling;
    if (outputDiv.id !== "altair-viz-5cedfaee897e475bb40cafa278b705ac") {
      outputDiv = document.getElementById("altair-viz-5cedfaee897e475bb40cafa278b705ac");
    }
    const paths = {
      "vega": "https://cdn.jsdelivr.net/npm//vega@5?noext",
      "vega-lib": "https://cdn.jsdelivr.net/npm//vega-lib?noext",
      "vega-lite": "https://cdn.jsdelivr.net/npm//vega-lite@5.2.0?noext",
      "vega-embed": "https://cdn.jsdelivr.net/npm//vega-embed@6?noext",
    };

    function maybeLoadScript(lib, version) {
      var key = `${lib.replace("-", "")}_version`;
      return (VEGA_DEBUG[key] == version) ?
        Promise.resolve(paths[lib]) :
        new Promise(function(resolve, reject) {
          var s = document.createElement('script');
          document.getElementsByTagName("head")[0].appendChild(s);
          s.async = true;
          s.onload = () => {
            VEGA_DEBUG[key] = version;
            return resolve(paths[lib]);
          };
          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);
          s.src = paths[lib];
        });
    }

    function showError(err) {
      outputDiv.innerHTML = `<div class="error" style="color:red;">${err}</div>`;
      throw err;
    }

    function displayChart(vegaEmbed) {
      vegaEmbed(outputDiv, spec, embedOpt)
        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));
    }

    if(typeof define === "function" && define.amd) {
      requirejs.config({paths});
      require(["vega-embed"], displayChart, err => showError(`Error loading script: ${err.message}`));
    } else {
      maybeLoadScript("vega", "5")
        .then(() => maybeLoadScript("vega-lite", "5.2.0"))
        .then(() => maybeLoadScript("vega-embed", "6"))
        .catch(showError)
        .then(() => displayChart(vegaEmbed));
    }
  })({"config": {"view": {"continuousWidth": 400, "continuousHeight": 300}, "axis": {"grid": false}}, "data": {"name": "data-c745bdc6811a50f90c2c6669d9a3794c"}, "mark": {"type": "bar", "size": 18.75}, "encoding": {"color": {"field": "epitope", "legend": null, "scale": {"domain": ["1", "2"], "range": ["#0072B2", "#CC79A7"]}, "type": "nominal"}, "tooltip": [{"field": "epitope", "type": "nominal"}, {"field": "activity", "format": ".2f", "type": "quantitative"}], "x": {"field": "activity", "title": "activity", "type": "quantitative"}, "y": {"field": "epitope", "sort": ["1", "2"], "type": "nominal"}}, "height": {"step": 25}, "width": 110, "$schema": "https://vega.github.io/schema/vega-lite/v5.2.0.json", "datasets": {"data-c745bdc6811a50f90c2c6669d9a3794c": [{"epitope": "1", "activity": 2.617384938766121}, {"epitope": "2", "activity": 4.073439704937735}]}}, {"mode": "vega-lite"});
</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Number of fit epitopes: 3
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div id="altair-viz-dd722b34985140d4b606eb3af564960f"></div>
<script type="text/javascript">
  var VEGA_DEBUG = (typeof VEGA_DEBUG == "undefined") ? {} : VEGA_DEBUG;
  (function(spec, embedOpt){
    let outputDiv = document.currentScript.previousElementSibling;
    if (outputDiv.id !== "altair-viz-dd722b34985140d4b606eb3af564960f") {
      outputDiv = document.getElementById("altair-viz-dd722b34985140d4b606eb3af564960f");
    }
    const paths = {
      "vega": "https://cdn.jsdelivr.net/npm//vega@5?noext",
      "vega-lib": "https://cdn.jsdelivr.net/npm//vega-lib?noext",
      "vega-lite": "https://cdn.jsdelivr.net/npm//vega-lite@5.2.0?noext",
      "vega-embed": "https://cdn.jsdelivr.net/npm//vega-embed@6?noext",
    };

    function maybeLoadScript(lib, version) {
      var key = `${lib.replace("-", "")}_version`;
      return (VEGA_DEBUG[key] == version) ?
        Promise.resolve(paths[lib]) :
        new Promise(function(resolve, reject) {
          var s = document.createElement('script');
          document.getElementsByTagName("head")[0].appendChild(s);
          s.async = true;
          s.onload = () => {
            VEGA_DEBUG[key] = version;
            return resolve(paths[lib]);
          };
          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);
          s.src = paths[lib];
        });
    }

    function showError(err) {
      outputDiv.innerHTML = `<div class="error" style="color:red;">${err}</div>`;
      throw err;
    }

    function displayChart(vegaEmbed) {
      vegaEmbed(outputDiv, spec, embedOpt)
        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));
    }

    if(typeof define === "function" && define.amd) {
      requirejs.config({paths});
      require(["vega-embed"], displayChart, err => showError(`Error loading script: ${err.message}`));
    } else {
      maybeLoadScript("vega", "5")
        .then(() => maybeLoadScript("vega-lite", "5.2.0"))
        .then(() => maybeLoadScript("vega-embed", "6"))
        .catch(showError)
        .then(() => displayChart(vegaEmbed));
    }
  })({"config": {"view": {"continuousWidth": 400, "continuousHeight": 300}, "axis": {"grid": false}}, "data": {"name": "data-99c8b3068a3909b55366c1108e404eae"}, "mark": {"type": "bar", "size": 18.75}, "encoding": {"color": {"field": "epitope", "legend": null, "scale": {"domain": ["1", "2", "3"], "range": ["#0072B2", "#CC79A7", "#4C3549"]}, "type": "nominal"}, "tooltip": [{"field": "epitope", "type": "nominal"}, {"field": "activity", "format": ".2f", "type": "quantitative"}], "x": {"field": "activity", "title": "activity", "type": "quantitative"}, "y": {"field": "epitope", "sort": ["1", "2", "3"], "type": "nominal"}}, "height": {"step": 25}, "width": 110, "$schema": "https://vega.github.io/schema/vega-lite/v5.2.0.json", "datasets": {"data-99c8b3068a3909b55366c1108e404eae": [{"epitope": "1", "activity": 3.160735996502857}, {"epitope": "2", "activity": 1.8115102597437482}, {"epitope": "3", "activity": 2.339303050117927}]}}, {"mode": "vega-lite"});
</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Number of fit epitopes: 4
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div id="altair-viz-72d270f3a2924a69b5c95db9908b0523"></div>
<script type="text/javascript">
  var VEGA_DEBUG = (typeof VEGA_DEBUG == "undefined") ? {} : VEGA_DEBUG;
  (function(spec, embedOpt){
    let outputDiv = document.currentScript.previousElementSibling;
    if (outputDiv.id !== "altair-viz-72d270f3a2924a69b5c95db9908b0523") {
      outputDiv = document.getElementById("altair-viz-72d270f3a2924a69b5c95db9908b0523");
    }
    const paths = {
      "vega": "https://cdn.jsdelivr.net/npm//vega@5?noext",
      "vega-lib": "https://cdn.jsdelivr.net/npm//vega-lib?noext",
      "vega-lite": "https://cdn.jsdelivr.net/npm//vega-lite@5.2.0?noext",
      "vega-embed": "https://cdn.jsdelivr.net/npm//vega-embed@6?noext",
    };

    function maybeLoadScript(lib, version) {
      var key = `${lib.replace("-", "")}_version`;
      return (VEGA_DEBUG[key] == version) ?
        Promise.resolve(paths[lib]) :
        new Promise(function(resolve, reject) {
          var s = document.createElement('script');
          document.getElementsByTagName("head")[0].appendChild(s);
          s.async = true;
          s.onload = () => {
            VEGA_DEBUG[key] = version;
            return resolve(paths[lib]);
          };
          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);
          s.src = paths[lib];
        });
    }

    function showError(err) {
      outputDiv.innerHTML = `<div class="error" style="color:red;">${err}</div>`;
      throw err;
    }

    function displayChart(vegaEmbed) {
      vegaEmbed(outputDiv, spec, embedOpt)
        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));
    }

    if(typeof define === "function" && define.amd) {
      requirejs.config({paths});
      require(["vega-embed"], displayChart, err => showError(`Error loading script: ${err.message}`));
    } else {
      maybeLoadScript("vega", "5")
        .then(() => maybeLoadScript("vega-lite", "5.2.0"))
        .then(() => maybeLoadScript("vega-embed", "6"))
        .catch(showError)
        .then(() => displayChart(vegaEmbed));
    }
  })({"config": {"view": {"continuousWidth": 400, "continuousHeight": 300}, "axis": {"grid": false}}, "data": {"name": "data-b55fceadbce8f72ea9cee0f2eacc13ae"}, "mark": {"type": "bar", "size": 18.75}, "encoding": {"color": {"field": "epitope", "legend": null, "scale": {"domain": ["1", "2", "3", "4"], "range": ["#0072B2", "#CC79A7", "#4C3549", "#009E73"]}, "type": "nominal"}, "tooltip": [{"field": "epitope", "type": "nominal"}, {"field": "activity", "format": ".2f", "type": "quantitative"}], "x": {"field": "activity", "title": "activity", "type": "quantitative"}, "y": {"field": "epitope", "sort": ["1", "2", "3", "4"], "type": "nominal"}}, "height": {"step": 25}, "width": 110, "$schema": "https://vega.github.io/schema/vega-lite/v5.2.0.json", "datasets": {"data-b55fceadbce8f72ea9cee0f2eacc13ae": [{"epitope": "1", "activity": 3.1551618606741103}, {"epitope": "2", "activity": -4.346848500280315}, {"epitope": "3", "activity": 1.8530799972133278}, {"epitope": "4", "activity": 2.337792844468794}]}}, {"mode": "vega-lite"});
</script></div>
</div>
<p>In the simulation ground truth, there are three epitopes (class 1, 2, and 3). The order of immunodominance from highest to lowest is class 2 &gt; class 3 &gt; class 1.</p>
<ul class="simple">
<li><p>The model that specified two epitopes (left heatmap) fit mutation-escape values that strongly correlate with that of the true class 2 and 3 epitopes. This is expected, as the model prioritizes epitopes by their immunodominance and ignores the most subdominant class 1 epitope.</p></li>
<li><p>The model that specified three epitopes (center heatmap) fit mutation-escape values that strongly correlate with that of the true class 1, 2, and 3 epitopes. This is also expected, suggesting that the class 1 epitope, though subdominant, is not redundant.</p></li>
<li><p>The model that specified four epitopes (right heatmap) fit three epitopes with mutation-escape values that strongly correlate with that of the true class 1, 2, and 3 epitopes. However, the fourth fit epitope did not correlate with any of the true epitopes. This “redundant” epitope also had a strongly negative wildtype activity value, suggesting that antibodies targeting that “epitope” don’t make an appreciable contribution to the observed data.</p></li>
</ul>
<p>Additionally, we can look at the correlation between “true” and predicted IC90’s for each of the fit models. To do this, we’ll predict the IC90’s of variants in a separate library with a different (higher) mutation rate. We therefore read in the “exact” simulated data from a library containing variants with an average of four mutations per gene.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">exact_data</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;RBD_variants_escape_exact.csv&quot;</span><span class="p">,</span> <span class="n">na_filter</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
    <span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">&#39;library == &quot;avg4muts&quot;&#39;</span><span class="p">)</span>
    <span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s2">&quot;concentration in [1]&quot;</span><span class="p">)</span>
    <span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<p>We’ll make the comparison on a log scale, and clip IC90s at values &gt;50 as that is likely to be way outside the dynamic range given the concentrations used.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ic90_corrs</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
<span class="n">ic90_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>

<span class="n">max_ic90</span> <span class="o">=</span> <span class="mi">50</span>
<span class="k">for</span> <span class="n">e</span> <span class="ow">in</span> <span class="n">n_epitopes</span><span class="p">:</span>
    <span class="n">model</span> <span class="o">=</span> <span class="n">fit_mut_escape</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;noisy_[0.25, 1, 4]conc_3muts_</span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">epitopes_noinit&quot;</span><span class="p">]</span>

    <span class="n">ic90s</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">exact_data</span><span class="p">[[</span><span class="s2">&quot;aa_substitutions&quot;</span><span class="p">,</span> <span class="s2">&quot;IC90&quot;</span><span class="p">]]</span>
        <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">IC90</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;IC90&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">clip</span><span class="p">(</span><span class="n">upper</span><span class="o">=</span><span class="n">max_ic90</span><span class="p">))</span>
        <span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
    <span class="p">)</span>
    <span class="n">ic90s</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">filter_variants_by_seen_muts</span><span class="p">(</span><span class="n">ic90s</span><span class="p">)</span>
    <span class="n">ic90s</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">icXX</span><span class="p">(</span><span class="n">ic90s</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="mf">0.9</span><span class="p">,</span> <span class="n">col</span><span class="o">=</span><span class="s2">&quot;predicted_IC90&quot;</span><span class="p">,</span> <span class="n">max_c</span><span class="o">=</span><span class="n">max_ic90</span><span class="p">)</span>

    <span class="n">ic90s</span> <span class="o">=</span> <span class="n">ic90s</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
        <span class="n">log_IC90</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s2">&quot;IC90&quot;</span><span class="p">]),</span>
        <span class="n">predicted_log_IC90</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s2">&quot;predicted_IC90&quot;</span><span class="p">]),</span>
    <span class="p">)</span>

    <span class="n">corr</span> <span class="o">=</span> <span class="n">ic90s</span><span class="p">[</span><span class="s2">&quot;log_IC90&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">corr</span><span class="p">(</span><span class="n">ic90s</span><span class="p">[</span><span class="s2">&quot;predicted_log_IC90&quot;</span><span class="p">])</span> <span class="o">**</span> <span class="mi">2</span>

    <span class="n">ic90_corrs</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span>
        <span class="p">[</span>
            <span class="n">ic90_corrs</span><span class="p">,</span>
            <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
                <span class="p">{</span>
                    <span class="s2">&quot;number_fit_epitopes&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">e</span><span class="p">],</span>
                    <span class="s2">&quot;Correlation (R^2)&quot;</span><span class="p">:</span> <span class="s2">&quot;</span><span class="si">{:.3f}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">corr</span><span class="p">),</span>
                <span class="p">}</span>
            <span class="p">),</span>
        <span class="p">]</span>
    <span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="n">ic90_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span>
        <span class="p">[</span>
            <span class="n">ic90_data</span><span class="p">,</span>
            <span class="p">(</span>
                <span class="n">ic90s</span><span class="p">[[</span><span class="s2">&quot;log_IC90&quot;</span><span class="p">,</span> <span class="s2">&quot;predicted_log_IC90&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
                    <span class="o">**</span><span class="p">{</span><span class="s2">&quot;number_fit_epitopes&quot;</span><span class="p">:</span> <span class="n">e</span><span class="p">}</span>
                <span class="p">)</span>
            <span class="p">),</span>
        <span class="p">]</span>
    <span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ic90_corrs</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>number_fit_epitopes</th>
      <th>Correlation (R^2)</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>2</td>
      <td>0.798</td>
    </tr>
    <tr>
      <th>1</th>
      <td>3</td>
      <td>0.947</td>
    </tr>
    <tr>
      <th>2</th>
      <td>4</td>
      <td>0.900</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="p">(</span>
    <span class="n">ggplot</span><span class="p">(</span><span class="n">ic90_data</span><span class="p">,</span> <span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s2">&quot;predicted_log_IC90&quot;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s2">&quot;log_IC90&quot;</span><span class="p">))</span>
    <span class="o">+</span> <span class="n">geom_point</span><span class="p">(</span><span class="n">alpha</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">facet_wrap</span><span class="p">(</span><span class="s2">&quot;number_fit_epitopes&quot;</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">theme_classic</span><span class="p">()</span>
    <span class="o">+</span> <span class="n">theme</span><span class="p">(</span><span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mf">2.5</span><span class="p">))</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/tyu2/.local/lib/python3.8/site-packages/plotnine/utils.py:371: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/specify_epitopes_23_1.png" src="_images/specify_epitopes_23_1.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;ggplot: (8752897379118)&gt;
</pre></div></div>
</div>
<p>Although the IC90 prediction appears excellent regardless of the number of fit epitopes, adding the 3rd epitope does lead to a small improvement in the prediction.</p>
</section>
<section id="Specifying-epitopes-with-prior-knowledge">
<h2>Specifying epitopes with prior knowledge<a class="headerlink" href="#Specifying-epitopes-with-prior-knowledge" title="Permalink to this headline">¶</a></h2>
<p>Next, we’ll explore the consequences of cases where we think we have prior knowledge of some sites that confer antibody escape, but do not know the exact number of epitopes that are targeted. We can start by “seeding” initial guesses for 2 epitopes and iteratively fitting <code class="docutils literal notranslate"><span class="pre">Polyclonal</span></code> models with more initial guesses.</p>
<ul class="simple">
<li><p><strong>Case 1</strong>: We guess from prior knowledge that site 484 and site 444 are key epitope sites, so we seed them with large mutation-escape values.</p></li>
<li><p><strong>Case 2</strong>: Building on Case 1, we seed a 3rd key site, 417, which is often considered to be in the most subdominant class 1 epitope.</p></li>
<li><p><strong>Case 3</strong>: Building on Case 2, we add a 4th key site, 386, that is <strong>not</strong> actually in an epitope that is targeted by polyclonal antibodies.</p></li>
<li><p><strong>Case 4</strong>: Building on Case 2, we add a 4th key site that is already located in an epitope that we’ve specified. Specifically, we’ll seed it with site 460, which is in the class 1 epitope along with site 417.</p></li>
</ul>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">model_params</span> <span class="o">=</span> <span class="p">{</span>
    <span class="c1">### Case 1</span>
    <span class="s2">&quot;noisy_[0.25, 1, 4]conc_3muts_2epitopes_seed484_444&quot;</span><span class="p">:</span> <span class="p">[</span>
        <span class="c1"># wildtype activity guesses</span>
        <span class="p">[(</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="mf">2.0</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;2&quot;</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">)],</span>
        <span class="c1"># key epitope site guesses</span>
        <span class="p">[</span>
            <span class="p">(</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="mi">484</span><span class="p">,</span> <span class="mf">10.0</span><span class="p">),</span>
            <span class="p">(</span><span class="s2">&quot;2&quot;</span><span class="p">,</span> <span class="mi">444</span><span class="p">,</span> <span class="mf">10.0</span><span class="p">),</span>
        <span class="p">],</span>
    <span class="p">],</span>
    <span class="c1">### Case 2</span>
    <span class="s2">&quot;noisy_[0.25, 1, 4]conc_3muts_3epitopes_seed417_484_444&quot;</span><span class="p">:</span> <span class="p">[</span>
        <span class="c1"># wildtype activity guesses</span>
        <span class="p">[</span>
            <span class="p">(</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">),</span>
            <span class="p">(</span><span class="s2">&quot;2&quot;</span><span class="p">,</span> <span class="mf">3.0</span><span class="p">),</span>
            <span class="p">(</span><span class="s2">&quot;3&quot;</span><span class="p">,</span> <span class="mf">2.0</span><span class="p">),</span>
        <span class="p">],</span>
        <span class="c1"># key epitope site guesses</span>
        <span class="p">[</span>
            <span class="p">(</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="mi">417</span><span class="p">,</span> <span class="mf">10.0</span><span class="p">),</span>
            <span class="p">(</span><span class="s2">&quot;2&quot;</span><span class="p">,</span> <span class="mi">484</span><span class="p">,</span> <span class="mf">10.0</span><span class="p">),</span>
            <span class="p">(</span><span class="s2">&quot;3&quot;</span><span class="p">,</span> <span class="mi">444</span><span class="p">,</span> <span class="mf">10.0</span><span class="p">),</span>
        <span class="p">],</span>
    <span class="p">],</span>
    <span class="c1">### Case 3</span>
    <span class="s2">&quot;noisy_[0.25, 1, 4]conc_3muts_4epitopes_seed417_484_444_386&quot;</span><span class="p">:</span> <span class="p">[</span>
        <span class="c1"># wildtype activity guesses</span>
        <span class="p">[</span>
            <span class="p">(</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="mf">2.0</span><span class="p">),</span>
            <span class="p">(</span><span class="s2">&quot;2&quot;</span><span class="p">,</span> <span class="mf">4.0</span><span class="p">),</span>
            <span class="p">(</span><span class="s2">&quot;3&quot;</span><span class="p">,</span> <span class="mf">3.0</span><span class="p">),</span>
            <span class="p">(</span><span class="s2">&quot;4&quot;</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">),</span>
        <span class="p">],</span>
        <span class="c1"># key epitope site guesses</span>
        <span class="p">[</span>
            <span class="p">(</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="mi">417</span><span class="p">,</span> <span class="mf">10.0</span><span class="p">),</span>
            <span class="p">(</span><span class="s2">&quot;2&quot;</span><span class="p">,</span> <span class="mi">484</span><span class="p">,</span> <span class="mf">10.0</span><span class="p">),</span>
            <span class="p">(</span><span class="s2">&quot;3&quot;</span><span class="p">,</span> <span class="mi">444</span><span class="p">,</span> <span class="mf">10.0</span><span class="p">),</span>
            <span class="p">(</span><span class="s2">&quot;4&quot;</span><span class="p">,</span> <span class="mi">386</span><span class="p">,</span> <span class="mf">10.0</span><span class="p">),</span>
        <span class="p">],</span>
    <span class="p">],</span>
    <span class="c1">### Case 4</span>
    <span class="s2">&quot;noisy_[0.25, 1, 4]conc_3muts_4epitopes_seed417_484_444_460&quot;</span><span class="p">:</span> <span class="p">[</span>
        <span class="c1"># wildtype activity guesses</span>
        <span class="p">[</span>
            <span class="p">(</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="mf">2.0</span><span class="p">),</span>
            <span class="p">(</span><span class="s2">&quot;2&quot;</span><span class="p">,</span> <span class="mf">4.0</span><span class="p">),</span>
            <span class="p">(</span><span class="s2">&quot;3&quot;</span><span class="p">,</span> <span class="mf">3.0</span><span class="p">),</span>
            <span class="p">(</span><span class="s2">&quot;4&quot;</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">),</span>
        <span class="p">],</span>
        <span class="c1"># key epitope site guesses</span>
        <span class="p">[</span>
            <span class="p">(</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="mi">417</span><span class="p">,</span> <span class="mf">10.0</span><span class="p">),</span>
            <span class="p">(</span><span class="s2">&quot;2&quot;</span><span class="p">,</span> <span class="mi">484</span><span class="p">,</span> <span class="mf">10.0</span><span class="p">),</span>
            <span class="p">(</span><span class="s2">&quot;3&quot;</span><span class="p">,</span> <span class="mi">444</span><span class="p">,</span> <span class="mf">10.0</span><span class="p">),</span>
            <span class="p">(</span><span class="s2">&quot;4&quot;</span><span class="p">,</span> <span class="mi">460</span><span class="p">,</span> <span class="mf">10.0</span><span class="p">),</span>
        <span class="p">],</span>
    <span class="p">],</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
<p>Fit <code class="docutils literal notranslate"><span class="pre">Polyclonal</span></code> models for each of the cases above.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="k">for</span> <span class="n">model_string</span> <span class="ow">in</span> <span class="n">model_params</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
    <span class="c1"># If the pickled model exists in the fit_polyclonal_models directory, load it.</span>
    <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;fit_polyclonal_models/</span><span class="si">{</span><span class="n">model_string</span><span class="si">}</span><span class="s2">.pkl&quot;</span><span class="p">)</span> <span class="ow">is</span> <span class="kc">True</span><span class="p">:</span>
        <span class="n">model</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;fit_polyclonal_models/</span><span class="si">{</span><span class="n">model_string</span><span class="si">}</span><span class="s2">.pkl&quot;</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">))</span>
        <span class="n">fit_mut_escape</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="n">model_string</span><span class="p">:</span> <span class="n">model</span><span class="p">})</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">model_string</span><span class="si">}</span><span class="s2"> was already fit.&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># Else, fit a model and save it to the fit_polyclonal_models directory.</span>
        <span class="n">model</span> <span class="o">=</span> <span class="n">polyclonal</span><span class="o">.</span><span class="n">Polyclonal</span><span class="p">(</span>
            <span class="n">data_to_fit</span><span class="o">=</span><span class="n">noisy_data</span><span class="p">,</span>
            <span class="n">activity_wt_df</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_records</span><span class="p">(</span>
                <span class="n">model_params</span><span class="p">[</span><span class="n">model_string</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span>
                <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;epitope&quot;</span><span class="p">,</span> <span class="s2">&quot;activity&quot;</span><span class="p">],</span>
            <span class="p">),</span>
            <span class="n">site_escape_df</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_records</span><span class="p">(</span>
                <span class="n">model_params</span><span class="p">[</span><span class="n">model_string</span><span class="p">][</span><span class="mi">1</span><span class="p">],</span>
                <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;epitope&quot;</span><span class="p">,</span> <span class="s2">&quot;site&quot;</span><span class="p">,</span> <span class="s2">&quot;escape&quot;</span><span class="p">],</span>
            <span class="p">),</span>
            <span class="n">data_mut_escape_overlap</span><span class="o">=</span><span class="s2">&quot;fill_to_data&quot;</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="n">opt_res</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span>
            <span class="n">logfreq</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span>
            <span class="n">reg_escape_weight</span><span class="o">=</span><span class="mf">0.01</span><span class="p">,</span>
            <span class="n">reg_uniqueness2_weight</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">fit_mut_escape</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="n">model_string</span><span class="p">:</span> <span class="n">model</span><span class="p">})</span>
        <span class="n">pickle</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;fit_polyclonal_models/</span><span class="si">{</span><span class="n">model_string</span><span class="si">}</span><span class="s2">.pkl&quot;</span><span class="p">,</span> <span class="s2">&quot;wb&quot;</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
noisy_[0.25, 1, 4]conc_3muts_2epitopes_seed484_444 was already fit.
noisy_[0.25, 1, 4]conc_3muts_3epitopes_seed417_484_444 was already fit.
noisy_[0.25, 1, 4]conc_3muts_4epitopes_seed417_484_444_386 was already fit.
noisy_[0.25, 1, 4]conc_3muts_4epitopes_seed417_484_444_460 was already fit.
</pre></div></div>
</div>
<p>Like before, we can correlate the <span class="math notranslate nohighlight">\(\beta_{m,e}\)</span> values for each of the epitopes fit by these <code class="docutils literal notranslate"><span class="pre">Polyclonal</span></code> models to the <span class="math notranslate nohighlight">\(\beta_{m,e}\)</span> values for each of the three “true” epitopes.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">plot_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
<span class="k">for</span> <span class="n">st</span> <span class="ow">in</span> <span class="p">[</span>
    <span class="s2">&quot;2epitopes_seed484_444&quot;</span><span class="p">,</span>
    <span class="s2">&quot;3epitopes_seed417_484_444&quot;</span><span class="p">,</span>
    <span class="s2">&quot;4epitopes_seed417_484_444_386&quot;</span><span class="p">,</span>
    <span class="s2">&quot;4epitopes_seed417_484_444_460&quot;</span><span class="p">,</span>
<span class="p">]:</span>
    <span class="n">model</span> <span class="o">=</span> <span class="n">fit_mut_escape</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;noisy_[0.25, 1, 4]conc_3muts_</span><span class="si">{</span><span class="n">st</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">true_e</span><span class="p">,</span> <span class="n">fit_e</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span>
        <span class="n">itertools</span><span class="o">.</span><span class="n">product</span><span class="p">(</span>
            <span class="n">true_mut_escape_df</span><span class="p">[</span><span class="s2">&quot;epitope&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">(),</span>
            <span class="n">model</span><span class="o">.</span><span class="n">mut_escape_df</span><span class="p">[</span><span class="s2">&quot;epitope&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">(),</span>
        <span class="p">)</span>
    <span class="p">):</span>

        <span class="n">df</span> <span class="o">=</span> <span class="n">true_mut_escape_df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;epitope == &quot;</span><span class="si">{</span><span class="n">true_e</span><span class="si">}</span><span class="s1">&quot;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span>
            <span class="p">(</span>
                <span class="n">model</span><span class="o">.</span><span class="n">mut_escape_df</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span>
                    <span class="n">columns</span><span class="o">=</span><span class="p">{</span>
                        <span class="s2">&quot;escape&quot;</span><span class="p">:</span> <span class="s2">&quot;predicted escape&quot;</span><span class="p">,</span>
                        <span class="s2">&quot;epitope&quot;</span><span class="p">:</span> <span class="s2">&quot;predicted epitope&quot;</span><span class="p">,</span>
                    <span class="p">}</span>
                <span class="p">)</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;`predicted epitope` == &quot;</span><span class="si">{</span><span class="n">fit_e</span><span class="si">}</span><span class="s1">&quot;&#39;</span><span class="p">)</span>
            <span class="p">),</span>
            <span class="n">on</span><span class="o">=</span><span class="s2">&quot;mutation&quot;</span><span class="p">,</span>
            <span class="n">validate</span><span class="o">=</span><span class="s2">&quot;one_to_one&quot;</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="n">corr</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;escape&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">corr</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;predicted escape&quot;</span><span class="p">]</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span>

        <span class="n">plot_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span>
            <span class="p">[</span>
                <span class="n">plot_data</span><span class="p">,</span>
                <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
                    <span class="n">data</span><span class="o">=</span><span class="p">{</span>
                        <span class="s2">&quot;Epitope seeding&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">st</span><span class="p">],</span>
                        <span class="s2">&quot;True epitope class&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">true_e</span><span class="p">],</span>
                        <span class="s2">&quot;Fit epitope class&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">fit_e</span><span class="p">],</span>
                        <span class="s2">&quot;Correlation (R^2)&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">corr</span><span class="p">],</span>
                    <span class="p">}</span>
                <span class="p">),</span>
            <span class="p">]</span>
        <span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">alt</span><span class="o">.</span><span class="n">Chart</span><span class="p">(</span><span class="n">plot_data</span><span class="p">)</span><span class="o">.</span><span class="n">mark_rect</span><span class="p">(</span><span class="n">opacity</span><span class="o">=</span><span class="mf">0.95</span><span class="p">,</span> <span class="n">stroke</span><span class="o">=</span><span class="s2">&quot;black&quot;</span><span class="p">,)</span><span class="o">.</span><span class="n">encode</span><span class="p">(</span>
    <span class="n">x</span><span class="o">=</span><span class="s2">&quot;Fit epitope class:O&quot;</span><span class="p">,</span>
    <span class="n">y</span><span class="o">=</span><span class="s2">&quot;True epitope class:O&quot;</span><span class="p">,</span>
    <span class="n">color</span><span class="o">=</span><span class="n">alt</span><span class="o">.</span><span class="n">Color</span><span class="p">(</span><span class="s2">&quot;Correlation (R^2):Q&quot;</span><span class="p">,</span> <span class="n">scale</span><span class="o">=</span><span class="n">alt</span><span class="o">.</span><span class="n">Scale</span><span class="p">(</span><span class="n">scheme</span><span class="o">=</span><span class="s2">&quot;viridis&quot;</span><span class="p">)),</span>
    <span class="n">tooltip</span><span class="o">=</span><span class="p">[</span>
        <span class="s2">&quot;Fit epitope class:O&quot;</span><span class="p">,</span>
        <span class="s2">&quot;True epitope class:O&quot;</span><span class="p">,</span>
        <span class="n">alt</span><span class="o">.</span><span class="n">Tooltip</span><span class="p">(</span><span class="s2">&quot;Correlation (R^2)&quot;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;.3f&quot;</span><span class="p">),</span>
    <span class="p">],</span>
<span class="p">)</span><span class="o">.</span><span class="n">properties</span><span class="p">(</span><span class="n">width</span><span class="o">=</span><span class="mi">130</span><span class="p">,</span> <span class="n">height</span><span class="o">=</span><span class="mi">100</span><span class="p">,)</span><span class="o">.</span><span class="n">facet</span><span class="p">(</span>
    <span class="s2">&quot;Epitope seeding&quot;</span><span class="p">,</span>
    <span class="n">columns</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/tyu2/.local/lib/python3.8/site-packages/altair/utils/core.py:317: FutureWarning: iteritems is deprecated and will be removed in a future version. Use .items instead.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div id="altair-viz-8deb89b4697a4a148d622e4c509eddbc"></div>
<script type="text/javascript">
  var VEGA_DEBUG = (typeof VEGA_DEBUG == "undefined") ? {} : VEGA_DEBUG;
  (function(spec, embedOpt){
    let outputDiv = document.currentScript.previousElementSibling;
    if (outputDiv.id !== "altair-viz-8deb89b4697a4a148d622e4c509eddbc") {
      outputDiv = document.getElementById("altair-viz-8deb89b4697a4a148d622e4c509eddbc");
    }
    const paths = {
      "vega": "https://cdn.jsdelivr.net/npm//vega@5?noext",
      "vega-lib": "https://cdn.jsdelivr.net/npm//vega-lib?noext",
      "vega-lite": "https://cdn.jsdelivr.net/npm//vega-lite@5.2.0?noext",
      "vega-embed": "https://cdn.jsdelivr.net/npm//vega-embed@6?noext",
    };

    function maybeLoadScript(lib, version) {
      var key = `${lib.replace("-", "")}_version`;
      return (VEGA_DEBUG[key] == version) ?
        Promise.resolve(paths[lib]) :
        new Promise(function(resolve, reject) {
          var s = document.createElement('script');
          document.getElementsByTagName("head")[0].appendChild(s);
          s.async = true;
          s.onload = () => {
            VEGA_DEBUG[key] = version;
            return resolve(paths[lib]);
          };
          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);
          s.src = paths[lib];
        });
    }

    function showError(err) {
      outputDiv.innerHTML = `<div class="error" style="color:red;">${err}</div>`;
      throw err;
    }

    function displayChart(vegaEmbed) {
      vegaEmbed(outputDiv, spec, embedOpt)
        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));
    }

    if(typeof define === "function" && define.amd) {
      requirejs.config({paths});
      require(["vega-embed"], displayChart, err => showError(`Error loading script: ${err.message}`));
    } else {
      maybeLoadScript("vega", "5")
        .then(() => maybeLoadScript("vega-lite", "5.2.0"))
        .then(() => maybeLoadScript("vega-embed", "6"))
        .catch(showError)
        .then(() => displayChart(vegaEmbed));
    }
  })({"config": {"view": {"continuousWidth": 400, "continuousHeight": 300}}, "data": {"name": "data-624eccb16bc5cb91d0ff3e60db505df9"}, "facet": {"field": "Epitope seeding", "type": "nominal"}, "spec": {"mark": {"type": "rect", "opacity": 0.95, "stroke": "black"}, "encoding": {"color": {"field": "Correlation (R^2)", "scale": {"scheme": "viridis"}, "type": "quantitative"}, "tooltip": [{"field": "Fit epitope class", "type": "ordinal"}, {"field": "True epitope class", "type": "ordinal"}, {"field": "Correlation (R^2)", "format": ".3f", "type": "quantitative"}], "x": {"field": "Fit epitope class", "type": "ordinal"}, "y": {"field": "True epitope class", "type": "ordinal"}}, "height": 100, "width": 130}, "columns": 2, "$schema": "https://vega.github.io/schema/vega-lite/v5.2.0.json", "datasets": {"data-624eccb16bc5cb91d0ff3e60db505df9": [{"Epitope seeding": "2epitopes_seed484_444", "True epitope class": "1", "Fit epitope class": "1", "Correlation (R^2)": 0.14033468079970324}, {"Epitope seeding": "2epitopes_seed484_444", "True epitope class": "1", "Fit epitope class": "2", "Correlation (R^2)": -0.07919895838689506}, {"Epitope seeding": "2epitopes_seed484_444", "True epitope class": "2", "Fit epitope class": "1", "Correlation (R^2)": 0.8331029273272299}, {"Epitope seeding": "2epitopes_seed484_444", "True epitope class": "2", "Fit epitope class": "2", "Correlation (R^2)": -0.05838687444866966}, {"Epitope seeding": "2epitopes_seed484_444", "True epitope class": "3", "Fit epitope class": "1", "Correlation (R^2)": -0.05680267925273883}, {"Epitope seeding": "2epitopes_seed484_444", "True epitope class": "3", "Fit epitope class": "2", "Correlation (R^2)": 0.9317040309748076}, {"Epitope seeding": "3epitopes_seed417_484_444", "True epitope class": "1", "Fit epitope class": "1", "Correlation (R^2)": 0.7903875492822057}, {"Epitope seeding": "3epitopes_seed417_484_444", "True epitope class": "1", "Fit epitope class": "2", "Correlation (R^2)": 0.06082976043744945}, {"Epitope seeding": "3epitopes_seed417_484_444", "True epitope class": "1", "Fit epitope class": "3", "Correlation (R^2)": -0.08890947070044011}, {"Epitope seeding": "3epitopes_seed417_484_444", "True epitope class": "2", "Fit epitope class": "1", "Correlation (R^2)": 0.19953126292094775}, {"Epitope seeding": "3epitopes_seed417_484_444", "True epitope class": "2", "Fit epitope class": "2", "Correlation (R^2)": 0.9745353554161248}, {"Epitope seeding": "3epitopes_seed417_484_444", "True epitope class": "2", "Fit epitope class": "3", "Correlation (R^2)": -0.05789153218602261}, {"Epitope seeding": "3epitopes_seed417_484_444", "True epitope class": "3", "Fit epitope class": "1", "Correlation (R^2)": -0.0729574780270961}, {"Epitope seeding": "3epitopes_seed417_484_444", "True epitope class": "3", "Fit epitope class": "2", "Correlation (R^2)": -0.05857458437519138}, {"Epitope seeding": "3epitopes_seed417_484_444", "True epitope class": "3", "Fit epitope class": "3", "Correlation (R^2)": 0.9686315897159}, {"Epitope seeding": "4epitopes_seed417_484_444_386", "True epitope class": "1", "Fit epitope class": "1", "Correlation (R^2)": 0.81115217422517}, {"Epitope seeding": "4epitopes_seed417_484_444_386", "True epitope class": "1", "Fit epitope class": "2", "Correlation (R^2)": 0.0605909832863074}, {"Epitope seeding": "4epitopes_seed417_484_444_386", "True epitope class": "1", "Fit epitope class": "3", "Correlation (R^2)": -0.09048622257029895}, {"Epitope seeding": "4epitopes_seed417_484_444_386", "True epitope class": "1", "Fit epitope class": "4", "Correlation (R^2)": 0.025717684402448034}, {"Epitope seeding": "4epitopes_seed417_484_444_386", "True epitope class": "2", "Fit epitope class": "1", "Correlation (R^2)": 0.20392058044236427}, {"Epitope seeding": "4epitopes_seed417_484_444_386", "True epitope class": "2", "Fit epitope class": "2", "Correlation (R^2)": 0.9735081429927286}, {"Epitope seeding": "4epitopes_seed417_484_444_386", "True epitope class": "2", "Fit epitope class": "3", "Correlation (R^2)": -0.057887712407979505}, {"Epitope seeding": "4epitopes_seed417_484_444_386", "True epitope class": "2", "Fit epitope class": "4", "Correlation (R^2)": 0.2833074530213572}, {"Epitope seeding": "4epitopes_seed417_484_444_386", "True epitope class": "3", "Fit epitope class": "1", "Correlation (R^2)": -0.0713306605662217}, {"Epitope seeding": "4epitopes_seed417_484_444_386", "True epitope class": "3", "Fit epitope class": "2", "Correlation (R^2)": -0.05883050701786486}, {"Epitope seeding": "4epitopes_seed417_484_444_386", "True epitope class": "3", "Fit epitope class": "3", "Correlation (R^2)": 0.9689377371924416}, {"Epitope seeding": "4epitopes_seed417_484_444_386", "True epitope class": "3", "Fit epitope class": "4", "Correlation (R^2)": 0.09856381126201633}, {"Epitope seeding": "4epitopes_seed417_484_444_460", "True epitope class": "1", "Fit epitope class": "1", "Correlation (R^2)": 0.8085034447857993}, {"Epitope seeding": "4epitopes_seed417_484_444_460", "True epitope class": "1", "Fit epitope class": "2", "Correlation (R^2)": 0.05828425695909513}, {"Epitope seeding": "4epitopes_seed417_484_444_460", "True epitope class": "1", "Fit epitope class": "3", "Correlation (R^2)": -0.09090722827396527}, {"Epitope seeding": "4epitopes_seed417_484_444_460", "True epitope class": "1", "Fit epitope class": "4", "Correlation (R^2)": 0.023732359089012597}, {"Epitope seeding": "4epitopes_seed417_484_444_460", "True epitope class": "2", "Fit epitope class": "1", "Correlation (R^2)": 0.20748442762134653}, {"Epitope seeding": "4epitopes_seed417_484_444_460", "True epitope class": "2", "Fit epitope class": "2", "Correlation (R^2)": 0.9471526883904942}, {"Epitope seeding": "4epitopes_seed417_484_444_460", "True epitope class": "2", "Fit epitope class": "3", "Correlation (R^2)": -0.058097471666383856}, {"Epitope seeding": "4epitopes_seed417_484_444_460", "True epitope class": "2", "Fit epitope class": "4", "Correlation (R^2)": 0.24987441650354186}, {"Epitope seeding": "4epitopes_seed417_484_444_460", "True epitope class": "3", "Fit epitope class": "1", "Correlation (R^2)": -0.0715104282971576}, {"Epitope seeding": "4epitopes_seed417_484_444_460", "True epitope class": "3", "Fit epitope class": "2", "Correlation (R^2)": -0.05803957980756185}, {"Epitope seeding": "4epitopes_seed417_484_444_460", "True epitope class": "3", "Fit epitope class": "3", "Correlation (R^2)": 0.9662508843107203}, {"Epitope seeding": "4epitopes_seed417_484_444_460", "True epitope class": "3", "Fit epitope class": "4", "Correlation (R^2)": 0.07677303236603701}]}}, {"mode": "vega-lite"});
</script></div>
</div>
<p>We can also visualize the wildtype activity values, <span class="math notranslate nohighlight">\(a_{wt,e}\)</span>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="k">for</span> <span class="n">st</span> <span class="ow">in</span> <span class="p">[</span>
    <span class="s2">&quot;2epitopes_seed484_444&quot;</span><span class="p">,</span>
    <span class="s2">&quot;3epitopes_seed417_484_444&quot;</span><span class="p">,</span>
    <span class="s2">&quot;4epitopes_seed417_484_444_386&quot;</span><span class="p">,</span>
    <span class="s2">&quot;4epitopes_seed417_484_444_460&quot;</span><span class="p">,</span>
<span class="p">]:</span>
    <span class="n">model</span> <span class="o">=</span> <span class="n">fit_mut_escape</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;noisy_[0.25, 1, 4]conc_3muts_</span><span class="si">{</span><span class="n">st</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Epitope seeding: </span><span class="si">{</span><span class="n">st</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">display</span><span class="p">(</span><span class="n">model</span><span class="o">.</span><span class="n">activity_wt_barplot</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Epitope seeding: 2epitopes_seed484_444
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div id="altair-viz-2079d7ac933e4bb497694c6617bdccd1"></div>
<script type="text/javascript">
  var VEGA_DEBUG = (typeof VEGA_DEBUG == "undefined") ? {} : VEGA_DEBUG;
  (function(spec, embedOpt){
    let outputDiv = document.currentScript.previousElementSibling;
    if (outputDiv.id !== "altair-viz-2079d7ac933e4bb497694c6617bdccd1") {
      outputDiv = document.getElementById("altair-viz-2079d7ac933e4bb497694c6617bdccd1");
    }
    const paths = {
      "vega": "https://cdn.jsdelivr.net/npm//vega@5?noext",
      "vega-lib": "https://cdn.jsdelivr.net/npm//vega-lib?noext",
      "vega-lite": "https://cdn.jsdelivr.net/npm//vega-lite@5.2.0?noext",
      "vega-embed": "https://cdn.jsdelivr.net/npm//vega-embed@6?noext",
    };

    function maybeLoadScript(lib, version) {
      var key = `${lib.replace("-", "")}_version`;
      return (VEGA_DEBUG[key] == version) ?
        Promise.resolve(paths[lib]) :
        new Promise(function(resolve, reject) {
          var s = document.createElement('script');
          document.getElementsByTagName("head")[0].appendChild(s);
          s.async = true;
          s.onload = () => {
            VEGA_DEBUG[key] = version;
            return resolve(paths[lib]);
          };
          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);
          s.src = paths[lib];
        });
    }

    function showError(err) {
      outputDiv.innerHTML = `<div class="error" style="color:red;">${err}</div>`;
      throw err;
    }

    function displayChart(vegaEmbed) {
      vegaEmbed(outputDiv, spec, embedOpt)
        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));
    }

    if(typeof define === "function" && define.amd) {
      requirejs.config({paths});
      require(["vega-embed"], displayChart, err => showError(`Error loading script: ${err.message}`));
    } else {
      maybeLoadScript("vega", "5")
        .then(() => maybeLoadScript("vega-lite", "5.2.0"))
        .then(() => maybeLoadScript("vega-embed", "6"))
        .catch(showError)
        .then(() => displayChart(vegaEmbed));
    }
  })({"config": {"view": {"continuousWidth": 400, "continuousHeight": 300}, "axis": {"grid": false}}, "data": {"name": "data-23315a8d26f786150f9671b5f1514309"}, "mark": {"type": "bar", "size": 18.75}, "encoding": {"color": {"field": "epitope", "legend": null, "scale": {"domain": ["1", "2"], "range": ["#0072B2", "#CC79A7"]}, "type": "nominal"}, "tooltip": [{"field": "epitope", "type": "nominal"}, {"field": "activity", "format": ".2f", "type": "quantitative"}], "x": {"field": "activity", "title": "activity", "type": "quantitative"}, "y": {"field": "epitope", "sort": ["1", "2"], "type": "nominal"}}, "height": {"step": 25}, "width": 110, "$schema": "https://vega.github.io/schema/vega-lite/v5.2.0.json", "datasets": {"data-23315a8d26f786150f9671b5f1514309": [{"epitope": "1", "activity": 4.073916387633656}, {"epitope": "2", "activity": 2.617279148416267}]}}, {"mode": "vega-lite"});
</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Epitope seeding: 3epitopes_seed417_484_444
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div id="altair-viz-f58abd45de9344adb0725cf9a87878b2"></div>
<script type="text/javascript">
  var VEGA_DEBUG = (typeof VEGA_DEBUG == "undefined") ? {} : VEGA_DEBUG;
  (function(spec, embedOpt){
    let outputDiv = document.currentScript.previousElementSibling;
    if (outputDiv.id !== "altair-viz-f58abd45de9344adb0725cf9a87878b2") {
      outputDiv = document.getElementById("altair-viz-f58abd45de9344adb0725cf9a87878b2");
    }
    const paths = {
      "vega": "https://cdn.jsdelivr.net/npm//vega@5?noext",
      "vega-lib": "https://cdn.jsdelivr.net/npm//vega-lib?noext",
      "vega-lite": "https://cdn.jsdelivr.net/npm//vega-lite@5.2.0?noext",
      "vega-embed": "https://cdn.jsdelivr.net/npm//vega-embed@6?noext",
    };

    function maybeLoadScript(lib, version) {
      var key = `${lib.replace("-", "")}_version`;
      return (VEGA_DEBUG[key] == version) ?
        Promise.resolve(paths[lib]) :
        new Promise(function(resolve, reject) {
          var s = document.createElement('script');
          document.getElementsByTagName("head")[0].appendChild(s);
          s.async = true;
          s.onload = () => {
            VEGA_DEBUG[key] = version;
            return resolve(paths[lib]);
          };
          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);
          s.src = paths[lib];
        });
    }

    function showError(err) {
      outputDiv.innerHTML = `<div class="error" style="color:red;">${err}</div>`;
      throw err;
    }

    function displayChart(vegaEmbed) {
      vegaEmbed(outputDiv, spec, embedOpt)
        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));
    }

    if(typeof define === "function" && define.amd) {
      requirejs.config({paths});
      require(["vega-embed"], displayChart, err => showError(`Error loading script: ${err.message}`));
    } else {
      maybeLoadScript("vega", "5")
        .then(() => maybeLoadScript("vega-lite", "5.2.0"))
        .then(() => maybeLoadScript("vega-embed", "6"))
        .catch(showError)
        .then(() => displayChart(vegaEmbed));
    }
  })({"config": {"view": {"continuousWidth": 400, "continuousHeight": 300}, "axis": {"grid": false}}, "data": {"name": "data-c17ca8edc2d6029358e09dbf87966ef4"}, "mark": {"type": "bar", "size": 18.75}, "encoding": {"color": {"field": "epitope", "legend": null, "scale": {"domain": ["1", "2", "3"], "range": ["#0072B2", "#CC79A7", "#4C3549"]}, "type": "nominal"}, "tooltip": [{"field": "epitope", "type": "nominal"}, {"field": "activity", "format": ".2f", "type": "quantitative"}], "x": {"field": "activity", "title": "activity", "type": "quantitative"}, "y": {"field": "epitope", "sort": ["1", "2", "3"], "type": "nominal"}}, "height": {"step": 25}, "width": 110, "$schema": "https://vega.github.io/schema/vega-lite/v5.2.0.json", "datasets": {"data-c17ca8edc2d6029358e09dbf87966ef4": [{"epitope": "1", "activity": 1.794815516621941}, {"epitope": "2", "activity": 3.1683056913179564}, {"epitope": "3", "activity": 2.34320341449989}]}}, {"mode": "vega-lite"});
</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Epitope seeding: 4epitopes_seed417_484_444_386
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div id="altair-viz-d88e21ac541a4057906b4f5caaa8fc98"></div>
<script type="text/javascript">
  var VEGA_DEBUG = (typeof VEGA_DEBUG == "undefined") ? {} : VEGA_DEBUG;
  (function(spec, embedOpt){
    let outputDiv = document.currentScript.previousElementSibling;
    if (outputDiv.id !== "altair-viz-d88e21ac541a4057906b4f5caaa8fc98") {
      outputDiv = document.getElementById("altair-viz-d88e21ac541a4057906b4f5caaa8fc98");
    }
    const paths = {
      "vega": "https://cdn.jsdelivr.net/npm//vega@5?noext",
      "vega-lib": "https://cdn.jsdelivr.net/npm//vega-lib?noext",
      "vega-lite": "https://cdn.jsdelivr.net/npm//vega-lite@5.2.0?noext",
      "vega-embed": "https://cdn.jsdelivr.net/npm//vega-embed@6?noext",
    };

    function maybeLoadScript(lib, version) {
      var key = `${lib.replace("-", "")}_version`;
      return (VEGA_DEBUG[key] == version) ?
        Promise.resolve(paths[lib]) :
        new Promise(function(resolve, reject) {
          var s = document.createElement('script');
          document.getElementsByTagName("head")[0].appendChild(s);
          s.async = true;
          s.onload = () => {
            VEGA_DEBUG[key] = version;
            return resolve(paths[lib]);
          };
          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);
          s.src = paths[lib];
        });
    }

    function showError(err) {
      outputDiv.innerHTML = `<div class="error" style="color:red;">${err}</div>`;
      throw err;
    }

    function displayChart(vegaEmbed) {
      vegaEmbed(outputDiv, spec, embedOpt)
        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));
    }

    if(typeof define === "function" && define.amd) {
      requirejs.config({paths});
      require(["vega-embed"], displayChart, err => showError(`Error loading script: ${err.message}`));
    } else {
      maybeLoadScript("vega", "5")
        .then(() => maybeLoadScript("vega-lite", "5.2.0"))
        .then(() => maybeLoadScript("vega-embed", "6"))
        .catch(showError)
        .then(() => displayChart(vegaEmbed));
    }
  })({"config": {"view": {"continuousWidth": 400, "continuousHeight": 300}, "axis": {"grid": false}}, "data": {"name": "data-824c783b427b1eacf88fe0fcc92a849d"}, "mark": {"type": "bar", "size": 18.75}, "encoding": {"color": {"field": "epitope", "legend": null, "scale": {"domain": ["1", "2", "3", "4"], "range": ["#0072B2", "#CC79A7", "#4C3549", "#009E73"]}, "type": "nominal"}, "tooltip": [{"field": "epitope", "type": "nominal"}, {"field": "activity", "format": ".2f", "type": "quantitative"}], "x": {"field": "activity", "title": "activity", "type": "quantitative"}, "y": {"field": "epitope", "sort": ["1", "2", "3", "4"], "type": "nominal"}}, "height": {"step": 25}, "width": 110, "$schema": "https://vega.github.io/schema/vega-lite/v5.2.0.json", "datasets": {"data-824c783b427b1eacf88fe0fcc92a849d": [{"epitope": "1", "activity": 1.84167770253583}, {"epitope": "2", "activity": 3.1522670446303374}, {"epitope": "3", "activity": 2.3412090367137375}, {"epitope": "4", "activity": -4.177652040505196}]}}, {"mode": "vega-lite"});
</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Epitope seeding: 4epitopes_seed417_484_444_460
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div id="altair-viz-542a2b7e20e44e0796ba8943668c7d33"></div>
<script type="text/javascript">
  var VEGA_DEBUG = (typeof VEGA_DEBUG == "undefined") ? {} : VEGA_DEBUG;
  (function(spec, embedOpt){
    let outputDiv = document.currentScript.previousElementSibling;
    if (outputDiv.id !== "altair-viz-542a2b7e20e44e0796ba8943668c7d33") {
      outputDiv = document.getElementById("altair-viz-542a2b7e20e44e0796ba8943668c7d33");
    }
    const paths = {
      "vega": "https://cdn.jsdelivr.net/npm//vega@5?noext",
      "vega-lib": "https://cdn.jsdelivr.net/npm//vega-lib?noext",
      "vega-lite": "https://cdn.jsdelivr.net/npm//vega-lite@5.2.0?noext",
      "vega-embed": "https://cdn.jsdelivr.net/npm//vega-embed@6?noext",
    };

    function maybeLoadScript(lib, version) {
      var key = `${lib.replace("-", "")}_version`;
      return (VEGA_DEBUG[key] == version) ?
        Promise.resolve(paths[lib]) :
        new Promise(function(resolve, reject) {
          var s = document.createElement('script');
          document.getElementsByTagName("head")[0].appendChild(s);
          s.async = true;
          s.onload = () => {
            VEGA_DEBUG[key] = version;
            return resolve(paths[lib]);
          };
          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);
          s.src = paths[lib];
        });
    }

    function showError(err) {
      outputDiv.innerHTML = `<div class="error" style="color:red;">${err}</div>`;
      throw err;
    }

    function displayChart(vegaEmbed) {
      vegaEmbed(outputDiv, spec, embedOpt)
        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));
    }

    if(typeof define === "function" && define.amd) {
      requirejs.config({paths});
      require(["vega-embed"], displayChart, err => showError(`Error loading script: ${err.message}`));
    } else {
      maybeLoadScript("vega", "5")
        .then(() => maybeLoadScript("vega-lite", "5.2.0"))
        .then(() => maybeLoadScript("vega-embed", "6"))
        .catch(showError)
        .then(() => displayChart(vegaEmbed));
    }
  })({"config": {"view": {"continuousWidth": 400, "continuousHeight": 300}, "axis": {"grid": false}}, "data": {"name": "data-beb7593959a3287244b1f74c3e6bbfac"}, "mark": {"type": "bar", "size": 18.75}, "encoding": {"color": {"field": "epitope", "legend": null, "scale": {"domain": ["1", "2", "3", "4"], "range": ["#0072B2", "#CC79A7", "#4C3549", "#009E73"]}, "type": "nominal"}, "tooltip": [{"field": "epitope", "type": "nominal"}, {"field": "activity", "format": ".2f", "type": "quantitative"}], "x": {"field": "activity", "title": "activity", "type": "quantitative"}, "y": {"field": "epitope", "sort": ["1", "2", "3", "4"], "type": "nominal"}}, "height": {"step": 25}, "width": 110, "$schema": "https://vega.github.io/schema/vega-lite/v5.2.0.json", "datasets": {"data-beb7593959a3287244b1f74c3e6bbfac": [{"epitope": "1", "activity": 1.8564744476965511}, {"epitope": "2", "activity": 3.149552645367495}, {"epitope": "3", "activity": 2.3394524556341634}, {"epitope": "4", "activity": -4.200335461430708}]}}, {"mode": "vega-lite"});
</script></div>
</div>
<p>These results look similar to the ones obtained without seeding sites above. But importantly, we can see that the models are not negatively impacted by seeding sites that are not within a targeted epitope or within the same epitope as an already seeded site. Although not seen on this simulated dataset, keep in mind that seeding sites could be helpful and lead to better results if the library mutation rate is low.</p>
<p>Additionally, we can look at the correlation between “true” and predicted IC90’s for each of these fit models.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ic90_corrs</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
<span class="n">ic90_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>

<span class="n">max_ic90</span> <span class="o">=</span> <span class="mi">50</span>
<span class="k">for</span> <span class="n">st</span> <span class="ow">in</span> <span class="p">[</span>
    <span class="s2">&quot;2epitopes_seed484_444&quot;</span><span class="p">,</span>
    <span class="s2">&quot;3epitopes_seed417_484_444&quot;</span><span class="p">,</span>
    <span class="s2">&quot;4epitopes_seed417_484_444_386&quot;</span><span class="p">,</span>
    <span class="s2">&quot;4epitopes_seed417_484_444_460&quot;</span><span class="p">,</span>
<span class="p">]:</span>
    <span class="n">model</span> <span class="o">=</span> <span class="n">fit_mut_escape</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;noisy_[0.25, 1, 4]conc_3muts_</span><span class="si">{</span><span class="n">st</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span>

    <span class="n">ic90s</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">exact_data</span><span class="p">[[</span><span class="s2">&quot;aa_substitutions&quot;</span><span class="p">,</span> <span class="s2">&quot;IC90&quot;</span><span class="p">]]</span>
        <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">IC90</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;IC90&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">clip</span><span class="p">(</span><span class="n">upper</span><span class="o">=</span><span class="n">max_ic90</span><span class="p">))</span>
        <span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
    <span class="p">)</span>
    <span class="n">ic90s</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">filter_variants_by_seen_muts</span><span class="p">(</span><span class="n">ic90s</span><span class="p">)</span>
    <span class="n">ic90s</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">icXX</span><span class="p">(</span><span class="n">ic90s</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="mf">0.9</span><span class="p">,</span> <span class="n">col</span><span class="o">=</span><span class="s2">&quot;predicted_IC90&quot;</span><span class="p">,</span> <span class="n">max_c</span><span class="o">=</span><span class="n">max_ic90</span><span class="p">)</span>

    <span class="n">ic90s</span> <span class="o">=</span> <span class="n">ic90s</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
        <span class="n">log_IC90</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s2">&quot;IC90&quot;</span><span class="p">]),</span>
        <span class="n">predicted_log_IC90</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s2">&quot;predicted_IC90&quot;</span><span class="p">]),</span>
    <span class="p">)</span>

    <span class="n">corr</span> <span class="o">=</span> <span class="n">ic90s</span><span class="p">[</span><span class="s2">&quot;log_IC90&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">corr</span><span class="p">(</span><span class="n">ic90s</span><span class="p">[</span><span class="s2">&quot;predicted_log_IC90&quot;</span><span class="p">])</span> <span class="o">**</span> <span class="mi">2</span>

    <span class="n">ic90_corrs</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span>
        <span class="p">[</span>
            <span class="n">ic90_corrs</span><span class="p">,</span>
            <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
                <span class="p">{</span>
                    <span class="s2">&quot;epitope_seeding&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">st</span><span class="p">],</span>
                    <span class="s2">&quot;Correlation (R^2)&quot;</span><span class="p">:</span> <span class="s2">&quot;</span><span class="si">{:.3f}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">corr</span><span class="p">),</span>
                <span class="p">}</span>
            <span class="p">),</span>
        <span class="p">]</span>
    <span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="n">ic90_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span>
        <span class="p">[</span>
            <span class="n">ic90_data</span><span class="p">,</span>
            <span class="p">(</span>
                <span class="n">ic90s</span><span class="p">[[</span><span class="s2">&quot;log_IC90&quot;</span><span class="p">,</span> <span class="s2">&quot;predicted_log_IC90&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
                    <span class="o">**</span><span class="p">{</span><span class="s2">&quot;epitope_seeding&quot;</span><span class="p">:</span> <span class="n">st</span><span class="p">}</span>
                <span class="p">)</span>
            <span class="p">),</span>
        <span class="p">]</span>
    <span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ic90_corrs</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>epitope_seeding</th>
      <th>Correlation (R^2)</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>2epitopes_seed484_444</td>
      <td>0.799</td>
    </tr>
    <tr>
      <th>1</th>
      <td>3epitopes_seed417_484_444</td>
      <td>0.965</td>
    </tr>
    <tr>
      <th>2</th>
      <td>4epitopes_seed417_484_444_386</td>
      <td>0.966</td>
    </tr>
    <tr>
      <th>3</th>
      <td>4epitopes_seed417_484_444_460</td>
      <td>0.899</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">_</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">ggplot</span><span class="p">(</span><span class="n">ic90_data</span><span class="p">,</span> <span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s2">&quot;predicted_log_IC90&quot;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s2">&quot;log_IC90&quot;</span><span class="p">))</span>
    <span class="o">+</span> <span class="n">geom_point</span><span class="p">(</span><span class="n">alpha</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">facet_wrap</span><span class="p">(</span><span class="s2">&quot;epitope_seeding&quot;</span><span class="p">,</span> <span class="n">ncol</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">theme_classic</span><span class="p">()</span>
    <span class="o">+</span> <span class="n">theme</span><span class="p">(</span><span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mf">5.5</span><span class="p">))</span>
<span class="p">)</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/tyu2/.local/lib/python3.8/site-packages/plotnine/utils.py:371: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/specify_epitopes_40_1.png" src="_images/specify_epitopes_40_1.png" />
</div>
</div>
</section>
<section id="Summary">
<h2>Summary<a class="headerlink" href="#Summary" title="Permalink to this headline">¶</a></h2>
<p>These simulation experiments provide a general guideline for specifying epitopes. When fitting <code class="docutils literal notranslate"><span class="pre">Polyclonal</span></code> models, one can start with 2 epitopes and iteratively fit models with increasing number of epitopes. At some point, the newly seeded <span class="math notranslate nohighlight">\(N\)</span>-th epitope will become redundant, as evidenced by a strongly negative wildtype activity value, <span class="math notranslate nohighlight">\(a_{wt,e}\)</span>. This is indication to the user that the previous fit model, containing <span class="math notranslate nohighlight">\(N - 1\)</span> epitopes, is the one that best captures the data
and polyclonal mix.</p>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="index.html">
    <img class="logo" src="_static/BloomLogo.jpg" alt="Logo"/>
    
    <h1 class="logo logo-name">polyclonal</h1>
    
  </a>
</p>



<p class="blurb">Model mutational escape from polyclonal antibodies.</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=jbloomlab&repo=polyclonal&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





<h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="index.html">polyclonal documentation</a></li>
</ul>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="biophysical_model.html">Biophysical model</a></li>
<li class="toctree-l1"><a class="reference internal" href="visualize_RBD.html">Visualize antibody mix</a></li>
<li class="toctree-l1"><a class="reference internal" href="simulate_RBD.html">Simulate experiment</a></li>
<li class="toctree-l1"><a class="reference internal" href="fit_RBD.html">Fit model to data</a></li>
<li class="toctree-l1"><a class="reference internal" href="regularization.html">Regularization</a></li>
<li class="toctree-l1"><a class="reference internal" href="RBD_bootstrap.html">Bootstrapping model fits</a></li>
<li class="toctree-l1"><a class="reference internal" href="RBD_average.html">Averaging models</a></li>
<li class="toctree-l1"><a class="reference internal" href="real_LyCoV1404.html">Real antibody data example</a></li>
<li class="toctree-l1"><a class="reference internal" href="reference_site_numbering.html">Reference site numbering</a></li>
<li class="toctree-l1"><a class="reference internal" href="real_cocktail.html">Real antibody cocktail example</a></li>
<li class="toctree-l1"><a class="reference internal" href="real_serum.html">Real serum example</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Specifying epitopes</a></li>
<li class="toctree-l1"><a class="reference internal" href="expt_design.html">Experimental design</a></li>
<li class="toctree-l1"><a class="reference internal" href="optimization.html">Details on fitting</a></li>
<li class="toctree-l1"><a class="reference internal" href="partition_function.html">Distinct epitopes assumption</a></li>
<li class="toctree-l1"><a class="reference internal" href="polyclonal.html">polyclonal package</a></li>
<li class="toctree-l1"><a class="reference internal" href="package_index.html">package index</a></li>
<li class="toctree-l1"><a class="reference internal" href="acknowledgments.html">Acknowledgements</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="real_serum.html" title="previous chapter">Real serum example</a></li>
      <li>Next: <a href="expt_design.html" title="next chapter">Experimental design</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2019--2022.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 4.2.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/specify_epitopes.nblink.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/jbloomlab/polyclonal" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>