
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>polyclonal.polyclonal &#8212; polyclonal 2.1 documentation</title>
    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/alabaster.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/plot_directive.css" />
    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
   
  <link rel="stylesheet" href="../../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <h1>Source code for polyclonal.polyclonal</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">==========</span>
<span class="sd">polyclonal</span>
<span class="sd">==========</span>

<span class="sd">Defines :class:`Polyclonal` objects for handling antibody mixtures.</span>

<span class="sd">&quot;&quot;&quot;</span>


<span class="kn">import</span> <span class="nn">collections</span>
<span class="kn">import</span> <span class="nn">copy</span>  <span class="c1"># noqa: F401</span>
<span class="kn">import</span> <span class="nn">inspect</span>
<span class="kn">import</span> <span class="nn">itertools</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">time</span>


<span class="kn">import</span> <span class="nn">binarymap</span>

<span class="kn">import</span> <span class="nn">frozendict</span>

<span class="kn">import</span> <span class="nn">natsort</span>

<span class="kn">import</span> <span class="nn">numpy</span>

<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="kn">import</span> <span class="nn">scipy.optimize</span>
<span class="kn">import</span> <span class="nn">scipy.special</span>

<span class="kn">import</span> <span class="nn">polyclonal</span>
<span class="kn">import</span> <span class="nn">polyclonal.alphabets</span>
<span class="kn">import</span> <span class="nn">polyclonal.pdb_utils</span>
<span class="kn">import</span> <span class="nn">polyclonal.plot</span>
<span class="kn">import</span> <span class="nn">polyclonal.utils</span>


<div class="viewcode-block" id="PolyclonalFitError"><a class="viewcode-back" href="../../polyclonal.polyclonal.html#polyclonal.polyclonal.PolyclonalFitError">[docs]</a><span class="k">class</span> <span class="nc">PolyclonalFitError</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Error fitting in :meth:`Polyclonal.fit`.&quot;&quot;&quot;</span>

    <span class="k">pass</span></div>


<div class="viewcode-block" id="PolyclonalHarmonizeError"><a class="viewcode-back" href="../../polyclonal.polyclonal.html#polyclonal.polyclonal.PolyclonalHarmonizeError">[docs]</a><span class="k">class</span> <span class="nc">PolyclonalHarmonizeError</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Error harmonizing epitopes in :meth:`Polyclonal.epitope_harmonized_model`.&quot;&quot;&quot;</span>

    <span class="k">pass</span></div>


<div class="viewcode-block" id="Polyclonal"><a class="viewcode-back" href="../../polyclonal.polyclonal.html#polyclonal.polyclonal.Polyclonal">[docs]</a><span class="k">class</span> <span class="nc">Polyclonal</span><span class="p">:</span>
    <span class="sa">r</span><span class="sd">&quot;&quot;&quot;Represent polyclonal antibody mixtures targeting multiple epitopes.</span>

<span class="sd">    Note</span>
<span class="sd">    ----</span>
<span class="sd">    At several concentrations :math:`c` of an antibody mixture, we measure</span>
<span class="sd">    :math:`p_v\left(c\right)`, the probability that variant :math:`v` is</span>
<span class="sd">    **not** bound (or neutralized). We assume antibodies act independently on</span>
<span class="sd">    one of :math:`E` epitopes, so the probability :math:`U_e\left(v, c\right)`</span>
<span class="sd">    that :math:`v` is unbound at concentration :math:`c` is related to the</span>
<span class="sd">    probability that epitope :math:`e` is unbound by</span>

<span class="sd">    .. math::</span>
<span class="sd">       :label: p_v</span>

<span class="sd">       p_v\left(c\right) = \prod_{e=1}^E U_e\left(v, c\right).</span>

<span class="sd">    We furthermore assume that :math:`U_e\left(v, c\right)` is related to the</span>
<span class="sd">    total binding activity :math:`\phi_e\left(v\right)` of antibodies targeting</span>
<span class="sd">    epitope :math:`e` on variant :math:`v` by</span>

<span class="sd">    .. math::</span>
<span class="sd">       :label: U_e</span>

<span class="sd">       U_e\left(v,c\right)=\frac{1}{1+c\exp\left(-\phi_e\left(v\right)\right)}</span>

<span class="sd">    where smaller (more negative) values of :math:`\phi_e\left(v\right)`</span>
<span class="sd">    correspond to higher overall binding activity against epitope :math:`e`</span>
<span class="sd">    variant :math:`v`.</span>

<span class="sd">    We define :math:`\phi_e\left(v\right)` in terms of the underlying</span>
<span class="sd">    quantities of biological interest as</span>

<span class="sd">    .. math::</span>
<span class="sd">       :label: phi_</span>

<span class="sd">       \phi_e\left(v\right) = -a_{\rm{wt}, e} +</span>
<span class="sd">                              \sum_{m=1}^M \beta_{m,e} b\left(v\right)_m,</span>

<span class="sd">    where :math:`a_{\rm{wt}, e}` is the activity of the serum against</span>
<span class="sd">    epitope :math:`e` for the &quot;wildtype&quot; (unmutated) protein (larger values</span>
<span class="sd">    indicate higher activity against this epitope), :math:`\beta_{m,e}`</span>
<span class="sd">    is the extent to which mutation :math:`m` (where :math:`1 \le m \le M`)</span>
<span class="sd">    escapes binding from antibodies targeting epitope :math:`e` (larger</span>
<span class="sd">    values indicate more escape by this mutation), and</span>
<span class="sd">    :math:`b\left(v\right)_m` is 1 if variant :math:`v` has mutation :math:`m`</span>
<span class="sd">    and 0 otherwise.</span>

<span class="sd">    Note</span>
<span class="sd">    ----</span>
<span class="sd">    You can initialize a :class:`Polyclonal` object in three ways:</span>

<span class="sd">    1. With known epitope activities :math:`a_{\rm{wt}, e}` and mutation-escape</span>
<span class="sd">       values :math:`\beta_{m,e}`, and no data. Use this approach if you</span>
<span class="sd">       already know these values and just want to visualize the polyclonal</span>
<span class="sd">       antibody mixture properties or predict escape of variants. To do this,</span>
<span class="sd">       initialize with ``activity_wt_df`` and ``mut_escape_df`` storing the</span>
<span class="sd">       known values, and ``data_to_fit=None``.</span>

<span class="sd">    2. With data to fit the epitope activities and mutation-escape values,</span>
<span class="sd">       and initial guesses for the epitope activities and mutation-escape</span>
<span class="sd">       values. To do this, initialize with ``data_to_fit`` holding the data,</span>
<span class="sd">       ``activity_wt_df`` holding initial guesses of activities, and</span>
<span class="sd">       ``mut_escape_df`` or ``site_escape_df`` holding initial guesses for</span>
<span class="sd">       mutation escapes (see also ``init_missing`` and</span>
<span class="sd">       ``data_mut_escape_overlap``). Then call :meth:`Polyclonal.fit`.</span>

<span class="sd">    3. With data to fit the epitope activities and mutation-escape values,</span>
<span class="sd">       but no initial guesses. To do this, initialize with ``data_to_fit``</span>
<span class="sd">       holding the data, ``activity_wt_df=None``, ``mut_escape_df=None``,</span>
<span class="sd">       and ``n_epitopes`` holding the number of epitopes. Then call</span>
<span class="sd">       :meth:`Polyclonal.fit`.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    data_to_fit : pandas.DataFrame or None</span>
<span class="sd">        Should have columns named &#39;aa_substitutions&#39;, &#39;concentration&#39;, and</span>
<span class="sd">        &#39;prob_escape&#39;. The &#39;aa_substitutions&#39; column defines each variant</span>
<span class="sd">        :math:`v` as a string of substitutions (e.g., &#39;M3A K5G&#39;). The</span>
<span class="sd">        &#39;prob_escape&#39; column gives the :math:`p_v\left(c\right)` value for</span>
<span class="sd">        each variant at each concentration :math:`c`.</span>
<span class="sd">    activity_wt_df : pandas.DataFrame or None</span>
<span class="sd">        Should have columns named &#39;epitope&#39; and &#39;activity&#39;, giving the names</span>
<span class="sd">        of the epitopes and the activity against epitope in the wildtype</span>
<span class="sd">        protein, :math:`a_{\rm{wt}, e}`.</span>
<span class="sd">    mut_escape_df : pandas.DataFrame or None</span>
<span class="sd">        Should have columns named &#39;mutation&#39;, &#39;epitope&#39;, and &#39;escape&#39; that</span>
<span class="sd">        give the :math:`\beta_{m,e}` values (in the &#39;escape&#39; column), with</span>
<span class="sd">        mutations written like &quot;G7M&quot;.</span>
<span class="sd">    site_escape_df : pandas.DataFrame or None</span>
<span class="sd">        Use if you want to initialize all mutations at a given site to have</span>
<span class="sd">        the same :math:`\beta_{m,e}` values. In this case, columns should be</span>
<span class="sd">        &#39;site&#39;, &#39;epitope&#39;, and &#39;escape&#39;. This option is mutually exclusive</span>
<span class="sd">        with ``mut_escape_df``.</span>
<span class="sd">    n_epitopes : int or None</span>
<span class="sd">        If initializing with ``activity_wt_df=None``, specifies number</span>
<span class="sd">        of epitopes.</span>
<span class="sd">    collapse_identical_variants : {&#39;mean&#39;, &#39;median&#39;, False}</span>
<span class="sd">        If identical variants in ``data_to_fit`` (same &#39;aa_substitutions&#39;),</span>
<span class="sd">        collapse them and make weight proportional to number of collapsed</span>
<span class="sd">        variants? Collapse by taking mean or median of &#39;prob_escape&#39;, or</span>
<span class="sd">        (if `False`) do not collapse at all. Collapsing will make fitting faster,</span>
<span class="sd">        but *not* a good idea if you are doing bootstrapping.</span>
<span class="sd">    alphabet : array-like</span>
<span class="sd">        Allowed characters in mutation strings.</span>
<span class="sd">    sites : array-like or None</span>
<span class="sd">        By default, sites are assumed to be sequential integer values are and inferred</span>
<span class="sd">        from ``data_to_fit`` or ``mut_escape_df``. However, you can also have</span>
<span class="sd">        non-sequential integer sites, or sites with lower-case letter suffixes</span>
<span class="sd">        (eg, `214a`) if your protein is numbered against a reference that it has</span>
<span class="sd">        indels relative to. In that case, provide list of all expected in order</span>
<span class="sd">        here; we require that order to be natsorted.</span>
<span class="sd">    epitope_colors : array-like or dict</span>
<span class="sd">        Maps each epitope to the color used for plotting. Either a dict keyed</span>
<span class="sd">        by each epitope, or an array of colors that are sequentially assigned</span>
<span class="sd">        to the epitopes.</span>
<span class="sd">    init_missing : &#39;zero&#39; or int</span>
<span class="sd">        How to initialize activities or mutation-escape values not specified in</span>
<span class="sd">        ``activity_wt_df`` or ``mut_escape_df`` / ``site_escape_df``. If</span>
<span class="sd">        &#39;zero&#39;, set mutation-escapes to zero and activities uniformly spaced</span>
<span class="sd">        from 1 to 0. Otherwise draw uniformly from between 0 and 1 using</span>
<span class="sd">        specified random number seed.</span>
<span class="sd">    data_mut_escape_overlap : {&#39;exact_match&#39;, &#39;fill_to_data&#39;, &#39;prune_to_data&#39;}</span>
<span class="sd">        If ``data_to_fit`` and ``mut_escape_df`` (or ``site_escape_df``) both</span>
<span class="sd">        specified, what if they don&#39;t specify same mutations.</span>
<span class="sd">        If &#39;exact_match&#39;, raise error. If &#39;fill_to_data&#39;, then take</span>
<span class="sd">        sites / wildtypes / mutations from ``data_to_fit`` and fill init</span>
<span class="sd">        values from any not specified in ``mut_escape_df`` as indicated by</span>
<span class="sd">        ``init_missing``. If &#39;prune_to_data&#39;, remove any extra mutations</span>
<span class="sd">        from ``mut_escape_df`` that are not in ``data_to_fit``.</span>

<span class="sd">    Attributes</span>
<span class="sd">    ----------</span>
<span class="sd">    epitopes : tuple</span>
<span class="sd">        Names of all epitopes.</span>
<span class="sd">    mutations : tuple</span>
<span class="sd">        All mutations for which we have escape values.</span>
<span class="sd">    alphabet : tuple</span>
<span class="sd">        Allowed characters in mutation strings.</span>
<span class="sd">    sites : tuple</span>
<span class="sd">        List of all sites. These are the sites provided via the ``sites`` parameter,</span>
<span class="sd">        or inferred from ``data_to_fit`` or ``mut_escape_df`` if that isn&#39;t provided.</span>
<span class="sd">        If `sequential_integer_sites` is `False`, these are str, otherwise int.</span>
<span class="sd">    sequential_integer_sites : bool</span>
<span class="sd">        True if sites are sequential and integer, False otherwise.</span>
<span class="sd">    wts : dict</span>
<span class="sd">        Keyed by site, value is wildtype at that site.</span>
<span class="sd">    epitope_colors : dict</span>
<span class="sd">        Maps each epitope to its color.</span>
<span class="sd">    data_to_fit : pandas.DataFrame or None</span>
<span class="sd">        Data to fit as passed when initializing this :class:`Polyclonal` object.</span>
<span class="sd">        If using ``collapse_identical_variants``, then identical variants</span>
<span class="sd">        are collapsed on columns &#39;concentration&#39;, &#39;aa_substitutions&#39;,</span>
<span class="sd">        and &#39;prob_escape&#39;, and a column &#39;weight&#39; is added to represent number</span>
<span class="sd">        of collapsed variants. Also, row-order may be changed.</span>
<span class="sd">    mutations_times_seen : frozendict.frozendict or None</span>
<span class="sd">        If `data_to_fit` is not `None`, keyed by all mutations with escape values</span>
<span class="sd">        and values are number of variants in which the mutation is seen. It is formally</span>
<span class="sd">        calculated as the number of variants with mutation across all concentrations</span>
<span class="sd">        divided by the number of concentrations, so can have non-integer values if</span>
<span class="sd">        there are variants only observed at some concentrations.</span>

<span class="sd">    Example</span>
<span class="sd">    -------</span>
<span class="sd">    Simple example with two epitopes (`e1` and `e2`) and a few mutations where</span>
<span class="sd">    we know the activities and mutation-level escape values ahead of time:</span>

<span class="sd">    &gt;&gt;&gt; activity_wt_df = pd.DataFrame({&#39;epitope&#39;:  [&#39;e1&#39;, &#39;e2&#39;],</span>
<span class="sd">    ...                                &#39;activity&#39;: [ 2.0,  1.0]})</span>
<span class="sd">    &gt;&gt;&gt; mut_escape_df = pd.DataFrame({</span>
<span class="sd">    ...   &#39;mutation&#39;: [&#39;M1C&#39;, &#39;M1C&#39;, &#39;G2A&#39;, &#39;G2A&#39;, &#39;A4K&#39;, &#39;A4K&#39;, &#39;A4L&#39;, &#39;A4L&#39;],</span>
<span class="sd">    ...   &#39;epitope&#39;:  [ &#39;e1&#39;,  &#39;e2&#39;,  &#39;e1&#39;,  &#39;e2&#39;,  &#39;e1&#39;,  &#39;e2&#39;,  &#39;e1&#39;,  &#39;e2&#39;],</span>
<span class="sd">    ...   &#39;escape&#39;:   [  2.0,   0.0,   3.0,   0.0,  0.0,    2.5,   0.0,   1.5],</span>
<span class="sd">    ...   })</span>
<span class="sd">    &gt;&gt;&gt; model = Polyclonal(</span>
<span class="sd">    ...     activity_wt_df=activity_wt_df,</span>
<span class="sd">    ...     mut_escape_df=mut_escape_df,</span>
<span class="sd">    ...     collapse_identical_variants=&quot;mean&quot;,</span>
<span class="sd">    ... )</span>
<span class="sd">    &gt;&gt;&gt; model.epitopes</span>
<span class="sd">    (&#39;e1&#39;, &#39;e2&#39;)</span>
<span class="sd">    &gt;&gt;&gt; model.mutations</span>
<span class="sd">    (&#39;M1C&#39;, &#39;G2A&#39;, &#39;A4K&#39;, &#39;A4L&#39;)</span>
<span class="sd">    &gt;&gt;&gt; model.mutations_times_seen is None</span>
<span class="sd">    True</span>
<span class="sd">    &gt;&gt;&gt; model.sites</span>
<span class="sd">    (1, 2, 4)</span>
<span class="sd">    &gt;&gt;&gt; model.wts</span>
<span class="sd">    {1: &#39;M&#39;, 2: &#39;G&#39;, 4: &#39;A&#39;}</span>
<span class="sd">    &gt;&gt;&gt; model.activity_wt_df</span>
<span class="sd">      epitope  activity</span>
<span class="sd">    0      e1       2.0</span>
<span class="sd">    1      e2       1.0</span>
<span class="sd">    &gt;&gt;&gt; model.mut_escape_df</span>
<span class="sd">      epitope  site wildtype mutant mutation  escape</span>
<span class="sd">    0      e1     1        M      C      M1C     2.0</span>
<span class="sd">    1      e1     2        G      A      G2A     3.0</span>
<span class="sd">    2      e1     4        A      K      A4K     0.0</span>
<span class="sd">    3      e1     4        A      L      A4L     0.0</span>
<span class="sd">    4      e2     1        M      C      M1C     0.0</span>
<span class="sd">    5      e2     2        G      A      G2A     0.0</span>
<span class="sd">    6      e2     4        A      K      A4K     2.5</span>
<span class="sd">    7      e2     4        A      L      A4L     1.5</span>

<span class="sd">    We can also summarize the mutation-level escape at the site level:</span>

<span class="sd">    &gt;&gt;&gt; pd.set_option(&quot;display.max_columns&quot;, None)</span>
<span class="sd">    &gt;&gt;&gt; pd.set_option(&quot;display.width&quot;, 89)</span>
<span class="sd">    &gt;&gt;&gt; model.mut_escape_site_summary_df()</span>
<span class="sd">      epitope  site wildtype  mean  total positive  max  min  total negative  n mutations</span>
<span class="sd">    0      e1     1        M   2.0             2.0  2.0  2.0             0.0            1</span>
<span class="sd">    1      e1     2        G   3.0             3.0  3.0  3.0             0.0            1</span>
<span class="sd">    2      e1     4        A   0.0             0.0  0.0  0.0             0.0            2</span>
<span class="sd">    3      e2     1        M   0.0             0.0  0.0  0.0             0.0            1</span>
<span class="sd">    4      e2     2        G   0.0             0.0  0.0  0.0             0.0            1</span>
<span class="sd">    5      e2     4        A   2.0             4.0  2.5  1.5             0.0            2</span>

<span class="sd">    Note that we can **not** initialize a :class:`Polyclonal` object if we are</span>
<span class="sd">    missing escape estimates for any mutations for any epitopes:</span>

<span class="sd">    &gt;&gt;&gt; Polyclonal(activity_wt_df=activity_wt_df,</span>
<span class="sd">    ...            mut_escape_df=mut_escape_df.head(n=5))</span>
<span class="sd">    Traceback (most recent call last):</span>
<span class="sd">      ...</span>
<span class="sd">    ValueError: invalid set of mutations for epitope=&#39;e2&#39;</span>

<span class="sd">    Now make a data frame with some variants:</span>

<span class="sd">    &gt;&gt;&gt; variants_df = pd.DataFrame.from_records(</span>
<span class="sd">    ...         [(&#39;AA&#39;, &#39;&#39;),</span>
<span class="sd">    ...          (&#39;AC&#39;, &#39;M1C&#39;),</span>
<span class="sd">    ...          (&#39;AG&#39;, &#39;G2A&#39;),</span>
<span class="sd">    ...          (&#39;AT&#39;, &#39;A4K&#39;),</span>
<span class="sd">    ...          (&#39;TA&#39;, &#39;A4L&#39;),</span>
<span class="sd">    ...          (&#39;CA&#39;, &#39;M1C G2A&#39;),</span>
<span class="sd">    ...          (&#39;CG&#39;, &#39;M1C A4K&#39;),</span>
<span class="sd">    ...          (&#39;CC&#39;, &#39;G2A A4K&#39;),</span>
<span class="sd">    ...          (&#39;TC&#39;, &#39;G2A A4L&#39;),</span>
<span class="sd">    ...          (&#39;CT&#39;, &#39;M1C G2A A4K&#39;),</span>
<span class="sd">    ...          (&#39;TG&#39;, &#39;M1C G2A A4L&#39;),</span>
<span class="sd">    ...          (&#39;GA&#39;, &#39;M1C&#39;),</span>
<span class="sd">    ...          ],</span>
<span class="sd">    ...         columns=[&#39;barcode&#39;, &#39;aa_substitutions&#39;])</span>

<span class="sd">    Get the escape probabilities predicted on these variants from</span>
<span class="sd">    the values in the :class:`Polyclonal` object:</span>

<span class="sd">    &gt;&gt;&gt; escape_probs = model.prob_escape(variants_df=variants_df,</span>
<span class="sd">    ...                                  concentrations=[1.0, 2.0, 4.0])</span>
<span class="sd">    &gt;&gt;&gt; escape_probs.round(3)</span>
<span class="sd">       barcode aa_substitutions  concentration  predicted_prob_escape</span>
<span class="sd">    0       AA                             1.0                  0.032</span>
<span class="sd">    1       AT              A4K            1.0                  0.097</span>
<span class="sd">    2       TA              A4L            1.0                  0.074</span>
<span class="sd">    3       AG              G2A            1.0                  0.197</span>
<span class="sd">    4       CC          G2A A4K            1.0                  0.598</span>
<span class="sd">    5       TC          G2A A4L            1.0                  0.455</span>
<span class="sd">    6       AC              M1C            1.0                  0.134</span>
<span class="sd">    7       GA              M1C            1.0                  0.134</span>
<span class="sd">    8       CG          M1C A4K            1.0                  0.409</span>
<span class="sd">    9       CA          M1C G2A            1.0                  0.256</span>
<span class="sd">    10      CT      M1C G2A A4K            1.0                  0.779</span>
<span class="sd">    11      TG      M1C G2A A4L            1.0                  0.593</span>
<span class="sd">    12      AA                             2.0                  0.010</span>
<span class="sd">    13      AT              A4K            2.0                  0.044</span>
<span class="sd">    14      TA              A4L            2.0                  0.029</span>
<span class="sd">    15      AG              G2A            2.0                  0.090</span>
<span class="sd">    16      CC          G2A A4K            2.0                  0.398</span>
<span class="sd">    17      TC          G2A A4L            2.0                  0.260</span>
<span class="sd">    18      AC              M1C            2.0                  0.052</span>
<span class="sd">    19      GA              M1C            2.0                  0.052</span>
<span class="sd">    20      CG          M1C A4K            2.0                  0.230</span>
<span class="sd">    21      CA          M1C G2A            2.0                  0.141</span>
<span class="sd">    22      CT      M1C G2A A4K            2.0                  0.629</span>
<span class="sd">    23      TG      M1C G2A A4L            2.0                  0.411</span>
<span class="sd">    24      AA                             4.0                  0.003</span>
<span class="sd">    25      AT              A4K            4.0                  0.017</span>
<span class="sd">    26      TA              A4L            4.0                  0.010</span>
<span class="sd">    27      AG              G2A            4.0                  0.034</span>
<span class="sd">    28      CC          G2A A4K            4.0                  0.214</span>
<span class="sd">    29      TC          G2A A4L            4.0                  0.118</span>
<span class="sd">    30      AC              M1C            4.0                  0.017</span>
<span class="sd">    31      GA              M1C            4.0                  0.017</span>
<span class="sd">    32      CG          M1C A4K            4.0                  0.106</span>
<span class="sd">    33      CA          M1C G2A            4.0                  0.070</span>
<span class="sd">    34      CT      M1C G2A A4K            4.0                  0.441</span>
<span class="sd">    35      TG      M1C G2A A4L            4.0                  0.243</span>

<span class="sd">    We can also get predicted escape probabilities by including concentrations</span>
<span class="sd">    in the data frame passed to :meth:`Polyclonal.prob_escape`:</span>

<span class="sd">    &gt;&gt;&gt; model.prob_escape(</span>
<span class="sd">    ...         variants_df=pd.concat([variants_df.assign(concentration=c)</span>
<span class="sd">    ...                                for c in [1.0, 2.0, 4.0]])</span>
<span class="sd">    ...         ).equals(escape_probs)</span>
<span class="sd">    True</span>

<span class="sd">    We can also compute the IC50s:</span>

<span class="sd">    &gt;&gt;&gt; model.icXX(variants_df).round(3)</span>
<span class="sd">       barcode aa_substitutions   IC50</span>
<span class="sd">    0       AA                   0.085</span>
<span class="sd">    1       AC              M1C  0.230</span>
<span class="sd">    2       GA              M1C  0.230</span>
<span class="sd">    3       AG              G2A  0.296</span>
<span class="sd">    4       AT              A4K  0.128</span>
<span class="sd">    5       TA              A4L  0.117</span>
<span class="sd">    6       CA          M1C G2A  0.355</span>
<span class="sd">    7       CG          M1C A4K  0.722</span>
<span class="sd">    8       CC          G2A A4K  1.414</span>
<span class="sd">    9       TC          G2A A4L  0.858</span>
<span class="sd">    10      CT      M1C G2A A4K  3.237</span>
<span class="sd">    11      TG      M1C G2A A4L  1.430</span>

<span class="sd">    Or the IC90s:</span>

<span class="sd">    &gt;&gt;&gt; model.icXX(variants_df, x=0.9, col=&#39;IC90&#39;).round(3)</span>
<span class="sd">       barcode aa_substitutions    IC90</span>
<span class="sd">    0       AA                    0.464</span>
<span class="sd">    1       AC              M1C   1.260</span>
<span class="sd">    2       GA              M1C   1.260</span>
<span class="sd">    3       AG              G2A   1.831</span>
<span class="sd">    4       AT              A4K   0.976</span>
<span class="sd">    5       TA              A4L   0.782</span>
<span class="sd">    6       CA          M1C G2A   2.853</span>
<span class="sd">    7       CG          M1C A4K   4.176</span>
<span class="sd">    8       CC          G2A A4K   7.473</span>
<span class="sd">    9       TC          G2A A4L   4.532</span>
<span class="sd">    10      CT      M1C G2A A4K  18.717</span>
<span class="sd">    11      TG      M1C G2A A4L   9.532</span>

<span class="sd">    Example</span>
<span class="sd">    -------</span>
<span class="sd">    Initialize with ``escape_probs`` created above as data to fit. In order</span>
<span class="sd">    to do this, we need to change the name of the column with the</span>
<span class="sd">    predicted escape probs to just be escape probs as we are now assuming</span>
<span class="sd">    these are the real values:</span>

<span class="sd">    &gt;&gt;&gt; data_to_fit = (</span>
<span class="sd">    ...         escape_probs</span>
<span class="sd">    ...         .rename(columns={&#39;predicted_prob_escape&#39;: &#39;prob_escape&#39;})</span>
<span class="sd">    ...         )</span>

<span class="sd">    &gt;&gt;&gt; model_data = Polyclonal(</span>
<span class="sd">    ...     data_to_fit=data_to_fit,</span>
<span class="sd">    ...     n_epitopes=2,</span>
<span class="sd">    ...     collapse_identical_variants=&quot;mean&quot;,</span>
<span class="sd">    ... )</span>

<span class="sd">    The mutations are those in ``data_to_fit``:</span>

<span class="sd">    &gt;&gt;&gt; model_data.mutations</span>
<span class="sd">    (&#39;M1C&#39;, &#39;G2A&#39;, &#39;A4K&#39;, &#39;A4L&#39;)</span>
<span class="sd">    &gt;&gt;&gt; dict(model_data.mutations_times_seen)</span>
<span class="sd">    {&#39;G2A&#39;: 6, &#39;M1C&#39;: 6, &#39;A4K&#39;: 4, &#39;A4L&#39;: 3}</span>

<span class="sd">    The activities are evenly spaced from 1 to 0, while the mutation escapes</span>
<span class="sd">    are all initialized to zero:</span>

<span class="sd">    &gt;&gt;&gt; model_data.activity_wt_df</span>
<span class="sd">      epitope  activity</span>
<span class="sd">    0       1       1.0</span>
<span class="sd">    1       2       0.0</span>
<span class="sd">    &gt;&gt;&gt; model_data.mut_escape_df</span>
<span class="sd">      epitope  site wildtype mutant mutation  escape  times_seen</span>
<span class="sd">    0       1     1        M      C      M1C     0.0           6</span>
<span class="sd">    1       1     2        G      A      G2A     0.0           6</span>
<span class="sd">    2       1     4        A      K      A4K     0.0           4</span>
<span class="sd">    3       1     4        A      L      A4L     0.0           3</span>
<span class="sd">    4       2     1        M      C      M1C     0.0           6</span>
<span class="sd">    5       2     2        G      A      G2A     0.0           6</span>
<span class="sd">    6       2     4        A      K      A4K     0.0           4</span>
<span class="sd">    7       2     4        A      L      A4L     0.0           3</span>

<span class="sd">    You can initialize to random numbers by setting ``init_missing`` to seed</span>
<span class="sd">    (in this example we also don&#39;t include all variants for one concentration):</span>

<span class="sd">    &gt;&gt;&gt; model_data2 = Polyclonal(</span>
<span class="sd">    ...     data_to_fit=data_to_fit.head(30),</span>
<span class="sd">    ...     n_epitopes=2,</span>
<span class="sd">    ...     init_missing=1,</span>
<span class="sd">    ...     collapse_identical_variants=&quot;mean&quot;,</span>
<span class="sd">    ... )</span>
<span class="sd">    &gt;&gt;&gt; model_data2.activity_wt_df.round(3)</span>
<span class="sd">      epitope  activity</span>
<span class="sd">    0       1     0.417</span>
<span class="sd">    1       2     0.720</span>

<span class="sd">    You can set some or all mutation escapes to initial values:</span>

<span class="sd">    &gt;&gt;&gt; model_data3 = Polyclonal(</span>
<span class="sd">    ...     data_to_fit=data_to_fit,</span>
<span class="sd">    ...     activity_wt_df=activity_wt_df,</span>
<span class="sd">    ...     mut_escape_df=pd.DataFrame({&#39;epitope&#39;: [&#39;e1&#39;],</span>
<span class="sd">    ...                                 &#39;mutation&#39;: [&#39;M1C&#39;],</span>
<span class="sd">    ...                                 &#39;escape&#39;: [4]}),</span>
<span class="sd">    ...     data_mut_escape_overlap=&#39;fill_to_data&#39;,</span>
<span class="sd">    ...     collapse_identical_variants=&quot;mean&quot;,</span>
<span class="sd">    ... )</span>
<span class="sd">    &gt;&gt;&gt; model_data3.mut_escape_df</span>
<span class="sd">      epitope  site wildtype mutant mutation  escape  times_seen</span>
<span class="sd">    0      e1     1        M      C      M1C     4.0           6</span>
<span class="sd">    1      e1     2        G      A      G2A     0.0           6</span>
<span class="sd">    2      e1     4        A      K      A4K     0.0           4</span>
<span class="sd">    3      e1     4        A      L      A4L     0.0           3</span>
<span class="sd">    4      e2     1        M      C      M1C     0.0           6</span>
<span class="sd">    5      e2     2        G      A      G2A     0.0           6</span>
<span class="sd">    6      e2     4        A      K      A4K     0.0           4</span>
<span class="sd">    7      e2     4        A      L      A4L     0.0           3</span>

<span class="sd">    You can initialize **sites** to escape values via ``site_activity_df``:</span>

<span class="sd">    &gt;&gt;&gt; model_data4 = Polyclonal(</span>
<span class="sd">    ...     data_to_fit=data_to_fit,</span>
<span class="sd">    ...     activity_wt_df=activity_wt_df,</span>
<span class="sd">    ...     site_escape_df=pd.DataFrame.from_records(</span>
<span class="sd">    ...         [(&#39;e1&#39;, 1, 1.0), (&#39;e1&#39;, 4, 0.0),</span>
<span class="sd">    ...          (&#39;e2&#39;, 1, 0.0), (&#39;e2&#39;, 4, 2.0)],</span>
<span class="sd">    ...         columns=[&#39;epitope&#39;, &#39;site&#39;, &#39;escape&#39;],</span>
<span class="sd">    ...     ),</span>
<span class="sd">    ...     data_mut_escape_overlap=&#39;fill_to_data&#39;,</span>
<span class="sd">    ...     collapse_identical_variants=&quot;mean&quot;,</span>
<span class="sd">    ... )</span>
<span class="sd">    &gt;&gt;&gt; model_data4.mut_escape_df</span>
<span class="sd">      epitope  site wildtype mutant mutation  escape  times_seen</span>
<span class="sd">    0      e1     1        M      C      M1C     1.0           6</span>
<span class="sd">    1      e1     2        G      A      G2A     0.0           6</span>
<span class="sd">    2      e1     4        A      K      A4K     0.0           4</span>
<span class="sd">    3      e1     4        A      L      A4L     0.0           3</span>
<span class="sd">    4      e2     1        M      C      M1C     0.0           6</span>
<span class="sd">    5      e2     2        G      A      G2A     0.0           6</span>
<span class="sd">    6      e2     4        A      K      A4K     2.0           4</span>
<span class="sd">    7      e2     4        A      L      A4L     2.0           3</span>

<span class="sd">    Fit the data using :meth:`Polyclonal.fit`, and make sure the new</span>
<span class="sd">    predicted escape probabilities are close to the real ones being fit.</span>
<span class="sd">    Reduce weight on regularization since there is so little data in this</span>
<span class="sd">    toy example:</span>

<span class="sd">    &gt;&gt;&gt; for m in [model_data, model_data2, model_data3, model_data4]:</span>
<span class="sd">    ...     opt_res = m.fit(</span>
<span class="sd">    ...         reg_escape_weight=0.001,</span>
<span class="sd">    ...         reg_spread_weight=0.001,</span>
<span class="sd">    ...         reg_activity_weight=0.0001,</span>
<span class="sd">    ...     )</span>
<span class="sd">    ...     pred_df = m.prob_escape(variants_df=data_to_fit)</span>
<span class="sd">    ...     if not numpy.allclose(pred_df[&#39;prob_escape&#39;],</span>
<span class="sd">    ...                           pred_df[&#39;predicted_prob_escape&#39;],</span>
<span class="sd">    ...                           atol=0.01):</span>
<span class="sd">    ...          raise ValueError(f&quot;wrong predictions\n{pred_df}&quot;)</span>
<span class="sd">    ...     if not numpy.allclose(</span>
<span class="sd">    ...              activity_wt_df[&#39;activity&#39;].sort_values(),</span>
<span class="sd">    ...              m.activity_wt_df[&#39;activity&#39;].sort_values(),</span>
<span class="sd">    ...              atol=0.1,</span>
<span class="sd">    ...              ):</span>
<span class="sd">    ...          raise ValueError(f&quot;wrong activities\n{m.activity_wt_df}&quot;)</span>
<span class="sd">    ...     if not numpy.allclose(</span>
<span class="sd">    ...              mut_escape_df[&#39;escape&#39;].sort_values(),</span>
<span class="sd">    ...              m.mut_escape_df[&#39;escape&#39;].sort_values(),</span>
<span class="sd">    ...              atol=0.05,</span>
<span class="sd">    ...              ):</span>
<span class="sd">    ...          raise ValueError(f&quot;wrong escapes\n{m.mut_escape_df}&quot;)</span>

<span class="sd">    &gt;&gt;&gt; model_data.mut_escape_site_summary_df().round(1)</span>
<span class="sd">      epitope  site wildtype  mean  total positive  max  min  total negative  n mutations</span>
<span class="sd">    0       1     1        M   0.0             0.0  0.0  0.0             0.0            1</span>
<span class="sd">    1       1     2        G   0.0             0.0  0.0  0.0             0.0            1</span>
<span class="sd">    2       1     4        A   2.0             4.0  2.5  1.5             0.0            2</span>
<span class="sd">    3       2     1        M   2.0             2.0  2.0  2.0             0.0            1</span>
<span class="sd">    4       2     2        G   3.0             3.0  3.0  3.0             0.0            1</span>
<span class="sd">    5       2     4        A   0.0             0.0  0.0  0.0             0.0            2</span>
<span class="sd">    &gt;&gt;&gt; model_data.mut_escape_site_summary_df(min_times_seen=4).round(1)</span>
<span class="sd">      epitope  site wildtype  mean  total positive  max  min  total negative  n mutations</span>
<span class="sd">    0       1     1        M   0.0             0.0  0.0  0.0             0.0            1</span>
<span class="sd">    1       1     2        G   0.0             0.0  0.0  0.0             0.0            1</span>
<span class="sd">    2       1     4        A   2.5             2.5  2.5  2.5             0.0            1</span>
<span class="sd">    3       2     1        M   2.0             2.0  2.0  2.0             0.0            1</span>
<span class="sd">    4       2     2        G   3.0             3.0  3.0  3.0             0.0            1</span>
<span class="sd">    5       2     4        A   0.0             0.0  0.0  0.0             0.0            1</span>

<span class="sd">    You can also exclude mutations to specific characters (typically you would want to</span>
<span class="sd">    do this for stop codons and/or gaps):</span>

<span class="sd">    &gt;&gt;&gt; model_data.mut_escape_site_summary_df(exclude_chars={&quot;C&quot;, &quot;K&quot;}).round(1)</span>
<span class="sd">      epitope  site wildtype  mean  total positive  max  min  total negative  n mutations</span>
<span class="sd">    0       1     2        G   0.0             0.0  0.0  0.0             0.0            1</span>
<span class="sd">    1       1     4        A   1.5             1.5  1.5  1.5             0.0            1</span>
<span class="sd">    2       2     2        G   3.0             3.0  3.0  3.0             0.0            1</span>
<span class="sd">    3       2     4        A   0.0             0.0  0.0  0.0             0.0            1</span>

<span class="sd">    Example</span>
<span class="sd">    -------</span>
<span class="sd">    You can convert a :class:`Polyclonal` model into a site-level model via</span>
<span class="sd">    the transformation of :func:`polyclonal.utils.site_level_variants`. The</span>
<span class="sd">    site-level model is another :class:`Polyclonal` model that just keeps</span>
<span class="sd">    track of whether or not sites are mutated using a 2-letter wildtype/mutant</span>
<span class="sd">    alphabet, and is generated using :meth:`Polyclonal.site_level_model`:</span>

<span class="sd">    &gt;&gt;&gt; model_site = model_data4.site_level_model()</span>
<span class="sd">    &gt;&gt;&gt; model_site.alphabet</span>
<span class="sd">    (&#39;w&#39;, &#39;m&#39;)</span>
<span class="sd">    &gt;&gt;&gt; (model_site.mut_escape_df</span>
<span class="sd">    ...  .assign(escape=lambda x: x[&#39;escape&#39;].abs()).round(1))</span>
<span class="sd">      epitope  site wildtype mutant mutation  escape  times_seen</span>
<span class="sd">    0      e1     1        w      m      w1m     2.0           5</span>
<span class="sd">    1      e1     2        w      m      w2m     3.0           6</span>
<span class="sd">    2      e1     4        w      m      w4m     0.0           7</span>
<span class="sd">    3      e2     1        w      m      w1m     0.0           5</span>
<span class="sd">    4      e2     2        w      m      w2m     0.0           6</span>
<span class="sd">    5      e2     4        w      m      w4m     2.0           7</span>
<span class="sd">    &gt;&gt;&gt; model_site.data_to_fit.head(n=5).round(3)</span>
<span class="sd">       concentration aa_substitutions  weight  prob_escape</span>
<span class="sd">    0            1.0                        1        0.032</span>
<span class="sd">    1            1.0              w1m       1        0.134</span>
<span class="sd">    2            1.0          w1m w2m       1        0.256</span>
<span class="sd">    3            1.0      w1m w2m w4m       2        0.686</span>
<span class="sd">    4            1.0          w1m w4m       1        0.409</span>

<span class="sd">    Example</span>
<span class="sd">    -------</span>
<span class="sd">    Epitope assignments are arbitrary, so you can calculate correlations</span>
<span class="sd">    among their mutation-escape values and create harmonized models that use</span>
<span class="sd">    the same label to refer to epitopes with similar mutation-escape values.</span>

<span class="sd">    First, correlations of a model to itself:</span>

<span class="sd">    &gt;&gt;&gt; ref_model = copy.deepcopy(model)</span>
<span class="sd">    &gt;&gt;&gt; model.mut_escape_corr(ref_model).round(3)</span>
<span class="sd">      ref_epitope self_epitope  correlation</span>
<span class="sd">    0          e1           e1        1.000</span>
<span class="sd">    1          e1           e2       -0.907</span>
<span class="sd">    2          e2           e1       -0.907</span>
<span class="sd">    3          e2           e2        1.000</span>

<span class="sd">    Make another model with epitope assignments inverted and a few mutations missing:</span>

<span class="sd">    &gt;&gt;&gt; inverted_model = Polyclonal(</span>
<span class="sd">    ...     activity_wt_df=(</span>
<span class="sd">    ...         activity_wt_df</span>
<span class="sd">    ...         .assign(epitope=lambda x: x[&quot;epitope&quot;].map({&quot;e1&quot;: &quot;e2&quot;, &quot;e2&quot;: &quot;e1&quot;}))</span>
<span class="sd">    ...     ),</span>
<span class="sd">    ...     mut_escape_df=(</span>
<span class="sd">    ...         mut_escape_df</span>
<span class="sd">    ...         .query(&quot;mutation != &#39;A4K&#39;&quot;)</span>
<span class="sd">    ...         .assign(epitope=lambda x: x[&quot;epitope&quot;].map({&quot;e1&quot;: &quot;e2&quot;, &quot;e2&quot;: &quot;e1&quot;}))</span>
<span class="sd">    ...     ),</span>
<span class="sd">    ... )</span>
<span class="sd">    &gt;&gt;&gt; inverted_model.mut_escape_corr(ref_model).round(3)</span>
<span class="sd">      ref_epitope self_epitope  correlation</span>
<span class="sd">    0          e1           e1       -0.945</span>
<span class="sd">    1          e1           e2        1.000</span>
<span class="sd">    2          e2           e1        1.000</span>
<span class="sd">    3          e2           e2       -0.945</span>

<span class="sd">    Now actually get epitope-harmonized models:</span>

<span class="sd">    &gt;&gt;&gt; model_harmonized, harmonize_df = model.epitope_harmonized_model(ref_model)</span>
<span class="sd">    &gt;&gt;&gt; harmonize_df</span>
<span class="sd">      self_initial_epitope self_harmonized_epitope ref_epitope  correlation</span>
<span class="sd">    0                   e1                      e1          e1          1.0</span>
<span class="sd">    1                   e2                      e2          e2          1.0</span>
<span class="sd">    &gt;&gt;&gt; assert model.mut_escape_df.equals(model_harmonized.mut_escape_df)</span>

<span class="sd">    &gt;&gt;&gt; inverted_harmonized, harmonize_df = inverted_model.epitope_harmonized_model(</span>
<span class="sd">    ...     ref_model</span>
<span class="sd">    ... )</span>
<span class="sd">    &gt;&gt;&gt; harmonize_df</span>
<span class="sd">      self_initial_epitope self_harmonized_epitope ref_epitope  correlation</span>
<span class="sd">    0                   e1                      e2          e2          1.0</span>
<span class="sd">    1                   e2                      e1          e1          1.0</span>
<span class="sd">    &gt;&gt;&gt; inverted_harmonized.mut_escape_df</span>
<span class="sd">      epitope  site wildtype mutant mutation  escape</span>
<span class="sd">    0      e1     1        M      C      M1C     2.0</span>
<span class="sd">    1      e1     2        G      A      G2A     3.0</span>
<span class="sd">    2      e1     4        A      L      A4L     0.0</span>
<span class="sd">    3      e2     1        M      C      M1C     0.0</span>
<span class="sd">    4      e2     2        G      A      G2A     0.0</span>
<span class="sd">    5      e2     4        A      L      A4L     1.5</span>

<span class="sd">    Example</span>
<span class="sd">    -------</span>
<span class="sd">    Filter variants by how often they are seen in data:</span>

<span class="sd">    &gt;&gt;&gt; model_data.filter_variants_by_seen_muts(variants_df)</span>
<span class="sd">    ... # doctest: +NORMALIZE_WHITESPACE</span>
<span class="sd">       barcode aa_substitutions</span>
<span class="sd">    0       AA</span>
<span class="sd">    1       AC              M1C</span>
<span class="sd">    2       AG              G2A</span>
<span class="sd">    3       AT              A4K</span>
<span class="sd">    4       TA              A4L</span>
<span class="sd">    5       CA          M1C G2A</span>
<span class="sd">    6       CG          M1C A4K</span>
<span class="sd">    7       CC          G2A A4K</span>
<span class="sd">    8       TC          G2A A4L</span>
<span class="sd">    9       CT      M1C G2A A4K</span>
<span class="sd">    10      TG      M1C G2A A4L</span>
<span class="sd">    11      GA              M1C</span>

<span class="sd">    &gt;&gt;&gt; model_data.filter_variants_by_seen_muts(variants_df, min_times_seen=5)</span>
<span class="sd">    ... # doctest: +NORMALIZE_WHITESPACE</span>
<span class="sd">      barcode aa_substitutions</span>
<span class="sd">    0      AA</span>
<span class="sd">    1      AC              M1C</span>
<span class="sd">    2      AG              G2A</span>
<span class="sd">    3      CA          M1C G2A</span>
<span class="sd">    4      GA              M1C</span>

<span class="sd">    &gt;&gt;&gt; model_data.filter_variants_by_seen_muts(variants_df, min_times_seen=4)</span>
<span class="sd">    ... # doctest: +NORMALIZE_WHITESPACE</span>
<span class="sd">      barcode aa_substitutions</span>
<span class="sd">    0      AA</span>
<span class="sd">    1      AC              M1C</span>
<span class="sd">    2      AG              G2A</span>
<span class="sd">    3      AT              A4K</span>
<span class="sd">    4      CA          M1C G2A</span>
<span class="sd">    5      CG          M1C A4K</span>
<span class="sd">    6      CC          G2A A4K</span>
<span class="sd">    7      CT      M1C G2A A4K</span>
<span class="sd">    8      GA              M1C</span>


<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="o">*</span><span class="p">,</span>
        <span class="n">activity_wt_df</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">mut_escape_df</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">data_to_fit</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">site_escape_df</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">n_epitopes</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">collapse_identical_variants</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">alphabet</span><span class="o">=</span><span class="n">polyclonal</span><span class="o">.</span><span class="n">AAS</span><span class="p">,</span>
        <span class="n">sites</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">epitope_colors</span><span class="o">=</span><span class="n">polyclonal</span><span class="o">.</span><span class="n">plot</span><span class="o">.</span><span class="n">DEFAULT_POSITIVE_COLORS</span><span class="p">,</span>
        <span class="n">init_missing</span><span class="o">=</span><span class="s2">&quot;zero&quot;</span><span class="p">,</span>
        <span class="n">data_mut_escape_overlap</span><span class="o">=</span><span class="s2">&quot;exact_match&quot;</span><span class="p">,</span>
    <span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;See main class docstring.&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">init_missing</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
            <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">init_missing</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">init_missing</span> <span class="o">!=</span> <span class="s2">&quot;zero&quot;</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;invalid </span><span class="si">{</span><span class="n">init_missing</span><span class="si">=}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">sites</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">sites</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">(</span><span class="n">sites</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">sites</span> <span class="o">!=</span> <span class="nb">tuple</span><span class="p">(</span><span class="n">natsort</span><span class="o">.</span><span class="n">natsorted</span><span class="p">(</span><span class="n">sites</span><span class="p">,</span> <span class="n">alg</span><span class="o">=</span><span class="n">natsort</span><span class="o">.</span><span class="n">ns</span><span class="o">.</span><span class="n">SIGNED</span><span class="p">)):</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;`sites` not natsorted&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="nb">any</span><span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">r</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">int</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">sites</span><span class="p">)</span> <span class="ow">or</span> <span class="n">sites</span> <span class="o">!=</span> <span class="nb">tuple</span><span class="p">(</span>
                <span class="nb">range</span><span class="p">(</span><span class="n">sites</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">sites</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
            <span class="p">):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">sequential_integer_sites</span> <span class="o">=</span> <span class="kc">False</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">sites</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">str</span><span class="p">,</span> <span class="n">sites</span><span class="p">))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">sequential_integer_sites</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">sites</span> <span class="o">=</span> <span class="n">sites</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">sites</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">sequential_integer_sites</span> <span class="o">=</span> <span class="kc">True</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">alphabet</span><span class="p">))</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">alphabet</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;duplicate letters in `alphabet`&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">alphabet</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">(</span><span class="n">alphabet</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_mutparser</span> <span class="o">=</span> <span class="n">polyclonal</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">MutationParser</span><span class="p">(</span>
            <span class="n">alphabet</span><span class="p">,</span>
            <span class="n">letter_suffixed_sites</span><span class="o">=</span><span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequential_integer_sites</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="c1"># get any epitope labels as str, not int</span>
        <span class="k">if</span> <span class="n">activity_wt_df</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">activity_wt_df</span> <span class="o">=</span> <span class="n">activity_wt_df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
                <span class="n">epitope</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;epitope&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
            <span class="p">)</span>
        <span class="k">if</span> <span class="n">site_escape_df</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">site_escape_df</span> <span class="o">=</span> <span class="n">site_escape_df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
                <span class="n">epitope</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;epitope&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
            <span class="p">)</span>
        <span class="k">if</span> <span class="n">mut_escape_df</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">mut_escape_df</span> <span class="o">=</span> <span class="n">mut_escape_df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
                <span class="n">epitope</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;epitope&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
            <span class="p">)</span>

        <span class="k">if</span> <span class="n">site_escape_df</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">mut_escape_df</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="s2">&quot;cannot set both `site_escape_df` and &quot;</span> <span class="s2">&quot;`mut_escape_df`&quot;</span>
                <span class="p">)</span>
            <span class="k">if</span> <span class="n">activity_wt_df</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="s2">&quot;cannot set `site_escape_df` without &quot;</span> <span class="s2">&quot;setting `activity_wt_df`&quot;</span>
                <span class="p">)</span>
            <span class="k">if</span> <span class="n">data_to_fit</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="s2">&quot;cannot set `site_escape_df` without &quot;</span> <span class="s2">&quot;setting `data_to_fit`&quot;</span>
                <span class="p">)</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">activity_wt_df</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span>
            <span class="p">(</span><span class="n">mut_escape_df</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="n">site_escape_df</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span>
        <span class="p">):</span>
            <span class="k">if</span> <span class="n">n_epitopes</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;specify `activity_wt_df` or `n_epitopes`&quot;</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">pd</span><span class="o">.</span><span class="n">isnull</span><span class="p">(</span><span class="n">activity_wt_df</span><span class="p">[</span><span class="s2">&quot;epitope&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">any</span><span class="p">():</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;epitope name cannot be null&quot;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">(</span><span class="n">activity_wt_df</span><span class="p">[</span><span class="s2">&quot;epitope&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">activity_wt_df</span><span class="p">):</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;duplicate epitopes in `activity_wt_df`&quot;</span><span class="p">)</span>

        <span class="k">elif</span> <span class="p">(</span><span class="n">activity_wt_df</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">mut_escape_df</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">):</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="nb">isinstance</span><span class="p">(</span><span class="n">n_epitopes</span><span class="p">,</span> <span class="nb">int</span><span class="p">)</span> <span class="ow">and</span> <span class="n">n_epitopes</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">):</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="s2">&quot;`n_epitopes` must be int &gt; 1 if no &quot;</span> <span class="s2">&quot;`activity_wt_df`&quot;</span>
                <span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_epitopes</span><span class="p">))</span>

            <span class="c1"># initialize activities</span>
            <span class="n">activity_wt_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
                <span class="p">{</span>
                    <span class="s2">&quot;epitope&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">,</span>
                    <span class="s2">&quot;activity&quot;</span><span class="p">:</span> <span class="p">(</span>
                        <span class="n">numpy</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">))</span>
                        <span class="k">if</span> <span class="n">init_missing</span> <span class="o">==</span> <span class="s2">&quot;zero&quot;</span>
                        <span class="k">else</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">))</span>
                    <span class="p">),</span>
                <span class="p">}</span>
            <span class="p">)</span>

            <span class="k">if</span> <span class="n">data_to_fit</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="s2">&quot;specify `data_to_fit` if `activity_wt_df` &quot;</span>
                    <span class="s2">&quot;and `mut_escape_df` are `None`&quot;</span>
                <span class="p">)</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;initialize both or neither `activity_wt_df` &quot;</span>
                <span class="s2">&quot;and `mut_escape_df` or `site_escape_df`&quot;</span>
            <span class="p">)</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">epitope_colors</span><span class="p">,</span> <span class="nb">dict</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">epitope_colors</span> <span class="o">=</span> <span class="p">{</span><span class="n">e</span><span class="p">:</span> <span class="n">epitope_colors</span><span class="p">[</span><span class="n">e</span><span class="p">]</span> <span class="k">for</span> <span class="n">e</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">}</span>
        <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">epitope_colors</span><span class="p">)</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;not enough `epitope_colors`&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">epitope_colors</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">,</span> <span class="n">epitope_colors</span><span class="p">))</span>

        <span class="k">def</span> <span class="nf">_init_mut_escape_df</span><span class="p">(</span><span class="n">mutations</span><span class="p">):</span>
            <span class="c1"># initialize mutation escape values</span>
            <span class="k">if</span> <span class="n">init_missing</span> <span class="o">==</span> <span class="s2">&quot;zero&quot;</span><span class="p">:</span>
                <span class="n">init</span> <span class="o">=</span> <span class="mf">0.0</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">init</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">)</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">mutations</span><span class="p">))</span>
            <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
                <span class="p">{</span>
                    <span class="s2">&quot;epitope&quot;</span><span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">)</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">mutations</span><span class="p">),</span>
                    <span class="s2">&quot;mutation&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">m</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">mutations</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">],</span>
                    <span class="s2">&quot;escape&quot;</span><span class="p">:</span> <span class="n">init</span><span class="p">,</span>
                <span class="p">}</span>
            <span class="p">)</span>

        <span class="c1"># get wildtype, sites, and mutations</span>
        <span class="k">if</span> <span class="n">data_to_fit</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">wts2</span><span class="p">,</span> <span class="n">sites2</span><span class="p">,</span> <span class="n">muts2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_muts_from_data_to_fit</span><span class="p">(</span><span class="n">data_to_fit</span><span class="p">)</span>
            <span class="k">if</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sites</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="nb">set</span><span class="p">(</span><span class="n">sites2</span><span class="p">)</span><span class="o">.</span><span class="n">issubset</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sites</span><span class="p">):</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;sites in `data_to_fit` not all in `sites`&quot;</span><span class="p">)</span>
            <span class="n">times_seen</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">data_to_fit</span><span class="p">[</span><span class="s2">&quot;aa_substitutions&quot;</span><span class="p">]</span>
                <span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
                <span class="o">.</span><span class="n">explode</span><span class="p">()</span>
                <span class="o">.</span><span class="n">dropna</span><span class="p">()</span>
                <span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
                <span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
                <span class="o">/</span> <span class="n">data_to_fit</span><span class="p">[</span><span class="s2">&quot;concentration&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">nunique</span><span class="p">()</span>
            <span class="p">)</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">times_seen</span> <span class="o">==</span> <span class="n">times_seen</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">))</span><span class="o">.</span><span class="n">all</span><span class="p">():</span>
                <span class="n">times_seen</span> <span class="o">=</span> <span class="n">times_seen</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">mutations_times_seen</span> <span class="o">=</span> <span class="n">frozendict</span><span class="o">.</span><span class="n">frozendict</span><span class="p">(</span><span class="n">times_seen</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">mutations_times_seen</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">if</span> <span class="n">site_escape_df</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="c1"># construct mut_escape_df from site_escape_df and mutations</span>
            <span class="c1"># from data_to_fit</span>
            <span class="n">req_cols</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;epitope&quot;</span><span class="p">,</span> <span class="s2">&quot;site&quot;</span><span class="p">,</span> <span class="s2">&quot;escape&quot;</span><span class="p">}</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">req_cols</span><span class="o">.</span><span class="n">issubset</span><span class="p">(</span><span class="n">site_escape_df</span><span class="o">.</span><span class="n">columns</span><span class="p">):</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;`site_escape_df` lacks columns </span><span class="si">{</span><span class="n">req_cols</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">assert</span> <span class="p">(</span><span class="n">data_to_fit</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">mut_escape_df</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequential_integer_sites</span><span class="p">:</span>
                <span class="n">site_escape_df</span> <span class="o">=</span> <span class="n">site_escape_df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
                    <span class="n">site</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;site&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
                <span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="nb">set</span><span class="p">(</span><span class="n">site_escape_df</span><span class="p">[</span><span class="s2">&quot;epitope&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">issubset</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">):</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;`site_escape_df` has unrecognized epitopes&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="nb">set</span><span class="p">(</span><span class="n">site_escape_df</span><span class="p">[</span><span class="s2">&quot;site&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">issubset</span><span class="p">(</span><span class="n">sites2</span><span class="p">):</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;site_escape_df has sites not in data_to_fit&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">site_escape_df</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span>
                <span class="n">site_escape_df</span><span class="p">[[</span><span class="s2">&quot;site&quot;</span><span class="p">,</span> <span class="s2">&quot;epitope&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
            <span class="p">):</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="s2">&quot;`site_escape_df` rows do not each represent &quot;</span>
                    <span class="sa">f</span><span class="s2">&quot;unique epitope / site:</span><span class="se">\n</span><span class="si">{</span><span class="n">site_escape_df</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="p">)</span>
            <span class="n">mut_records</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">epitope</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">:</span>
                <span class="n">site_escape</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="n">site_escape_df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s2">&quot;epitope == @epitope&quot;</span><span class="p">)</span>
                    <span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;site&quot;</span><span class="p">)[</span><span class="s2">&quot;escape&quot;</span><span class="p">]</span>
                    <span class="o">.</span><span class="n">to_dict</span><span class="p">()</span>
                <span class="p">)</span>
                <span class="k">for</span> <span class="n">mut</span> <span class="ow">in</span> <span class="n">muts2</span><span class="p">:</span>
                    <span class="p">(</span><span class="n">_</span><span class="p">,</span> <span class="n">site</span><span class="p">,</span> <span class="n">_</span><span class="p">)</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_mutparser</span><span class="o">.</span><span class="n">parse_mut</span><span class="p">(</span><span class="n">mut</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">site</span> <span class="ow">in</span> <span class="n">site_escape</span><span class="p">:</span>
                        <span class="n">mut_records</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">epitope</span><span class="p">,</span> <span class="n">mut</span><span class="p">,</span> <span class="n">site_escape</span><span class="p">[</span><span class="n">site</span><span class="p">]))</span>
            <span class="n">mut_escape_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_records</span><span class="p">(</span>
                <span class="n">mut_records</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;epitope&quot;</span><span class="p">,</span> <span class="s2">&quot;mutation&quot;</span><span class="p">,</span> <span class="s2">&quot;escape&quot;</span><span class="p">]</span>
            <span class="p">)</span>

        <span class="k">if</span> <span class="n">mut_escape_df</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">wts</span><span class="p">,</span> <span class="n">sites</span><span class="p">,</span> <span class="n">muts</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_muts_from_mut_escape_df</span><span class="p">(</span><span class="n">mut_escape_df</span><span class="p">)</span>
            <span class="k">if</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sites</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="nb">set</span><span class="p">(</span><span class="n">sites</span><span class="p">)</span><span class="o">.</span><span class="n">issubset</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sites</span><span class="p">):</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;`mut_escape_df` has sites not in `sites`&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">mut_escape_df</span> <span class="ow">is</span> <span class="n">data_to_fit</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;initialize `mut_escape_df` or `data_to_fit`&quot;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">mut_escape_df</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">wts</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutations</span> <span class="o">=</span> <span class="n">wts2</span><span class="p">,</span> <span class="n">muts2</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">sites</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">sites</span> <span class="o">=</span> <span class="n">sites2</span>
            <span class="n">mut_escape_df</span> <span class="o">=</span> <span class="n">_init_mut_escape_df</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutations</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">data_to_fit</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">wts</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutations</span> <span class="o">=</span> <span class="n">wts</span><span class="p">,</span> <span class="n">muts</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">sites</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">sites</span> <span class="o">=</span> <span class="n">sites</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">data_mut_escape_overlap</span> <span class="o">==</span> <span class="s2">&quot;exact_match&quot;</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">sites</span> <span class="o">==</span> <span class="n">sites2</span> <span class="ow">and</span> <span class="n">wts</span> <span class="o">==</span> <span class="n">wts2</span> <span class="ow">and</span> <span class="n">muts</span> <span class="o">==</span> <span class="n">muts2</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">wts</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutations</span> <span class="o">=</span> <span class="n">wts</span><span class="p">,</span> <span class="n">muts</span>
                    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">sites</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">sites</span> <span class="o">=</span> <span class="n">sites</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                        <span class="s2">&quot;`data_to_fit` and `mut_escape_df` give different mutations. &quot;</span>
                        <span class="s2">&quot;Fix or set data_mut_escape_overlap=&#39;fill_to_data&#39;&quot;</span>
                    <span class="p">)</span>
            <span class="k">elif</span> <span class="n">data_mut_escape_overlap</span> <span class="o">==</span> <span class="s2">&quot;fill_to_data&quot;</span><span class="p">:</span>
                <span class="c1"># sites are in mut_escape_df, sites2 in data_to_fit</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">sites</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="nb">set</span><span class="p">(</span><span class="n">sites</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="nb">set</span><span class="p">(</span><span class="n">sites2</span><span class="p">):</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">sites</span> <span class="o">=</span> <span class="n">sites2</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                        <span class="s2">&quot;`mut_escape_df` has more sites than `data_to_fit`&quot;</span>
                    <span class="p">)</span>
                <span class="k">if</span> <span class="n">wts</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="o">&lt;=</span> <span class="n">wts2</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">wts</span> <span class="o">=</span> <span class="n">wts2</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;`mut_escape_df` has wts not in `data_to_fit`&quot;</span><span class="p">)</span>
                <span class="k">if</span> <span class="nb">set</span><span class="p">(</span><span class="n">muts</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="nb">set</span><span class="p">(</span><span class="n">muts2</span><span class="p">):</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">mutations</span> <span class="o">=</span> <span class="n">muts2</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                        <span class="s2">&quot;`mut_escape_df` has mutations not in `data_to_fit`&quot;</span>
                    <span class="p">)</span>
                <span class="c1"># take values from `mut_escape_df` and fill missing</span>
                <span class="n">mut_escape_df</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="n">mut_escape_df</span><span class="o">.</span><span class="n">set_index</span><span class="p">([</span><span class="s2">&quot;epitope&quot;</span><span class="p">,</span> <span class="s2">&quot;mutation&quot;</span><span class="p">])[</span><span class="s2">&quot;escape&quot;</span><span class="p">]</span>
                    <span class="o">.</span><span class="n">combine_first</span><span class="p">(</span>
                        <span class="n">_init_mut_escape_df</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutations</span><span class="p">)</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span>
                            <span class="p">[</span><span class="s2">&quot;epitope&quot;</span><span class="p">,</span> <span class="s2">&quot;mutation&quot;</span><span class="p">]</span>
                        <span class="p">)[</span><span class="s2">&quot;escape&quot;</span><span class="p">]</span>
                    <span class="p">)</span>
                    <span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
                <span class="p">)</span>
            <span class="k">elif</span> <span class="n">data_mut_escape_overlap</span> <span class="o">==</span> <span class="s2">&quot;prune_to_data&quot;</span><span class="p">:</span>
                <span class="c1"># sites are in mut_escape_df, sites2 in data_to_fit</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">sites</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="nb">set</span><span class="p">(</span><span class="n">sites</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="nb">set</span><span class="p">(</span><span class="n">sites2</span><span class="p">):</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">sites</span> <span class="o">=</span> <span class="n">sites2</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                        <span class="s2">&quot;`mut_escape_df` has fewer sites than `data_to_fit`&quot;</span>
                    <span class="p">)</span>
                <span class="k">if</span> <span class="n">wts</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="o">&gt;=</span> <span class="n">wts2</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">wts</span> <span class="o">=</span> <span class="n">wts2</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;`mut_escape_df` fewer wts than `data_to_fit`&quot;</span><span class="p">)</span>
                <span class="k">if</span> <span class="nb">set</span><span class="p">(</span><span class="n">muts</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="nb">set</span><span class="p">(</span><span class="n">muts2</span><span class="p">):</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">mutations</span> <span class="o">=</span> <span class="n">muts2</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                        <span class="s2">&quot;`mut_escape_df` has fewer mutations than `data_to_fit`&quot;</span>
                    <span class="p">)</span>
                <span class="n">mut_escape_df</span> <span class="o">=</span> <span class="n">mut_escape_df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s2">&quot;mutation in @self.mutations&quot;</span><span class="p">)</span>
                <span class="k">assert</span> <span class="nb">set</span><span class="p">(</span><span class="n">mut_escape_df</span><span class="p">[</span><span class="s2">&quot;mutation&quot;</span><span class="p">])</span> <span class="o">==</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutations</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;invalid </span><span class="si">{</span><span class="n">data_mut_escape_overlap</span><span class="si">=}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="nb">set</span><span class="p">(</span><span class="n">mut_escape_df</span><span class="p">[</span><span class="s2">&quot;epitope&quot;</span><span class="p">])</span> <span class="o">!=</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;`mut_escape_df` does not have same epitopes as &quot;</span> <span class="s2">&quot;`activity_wt_df`&quot;</span>
            <span class="p">)</span>
        <span class="k">for</span> <span class="n">epitope</span><span class="p">,</span> <span class="n">df</span> <span class="ow">in</span> <span class="n">mut_escape_df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;epitope&quot;</span><span class="p">):</span>
            <span class="k">if</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;mutation&quot;</span><span class="p">])</span> <span class="o">!=</span> <span class="nb">sorted</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutations</span><span class="p">):</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;invalid set of mutations for </span><span class="si">{</span><span class="n">epitope</span><span class="si">=}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># set internal params with activities and escapes</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_params</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_params_from_dfs</span><span class="p">(</span><span class="n">activity_wt_df</span><span class="p">,</span> <span class="n">mut_escape_df</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">data_to_fit</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_one_binarymap</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_binarymaps</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_cs</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_pvs</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_weights</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">data_to_fit</span><span class="p">,</span>
            <span class="p">)</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_binarymaps_from_df</span><span class="p">(</span><span class="n">data_to_fit</span><span class="p">,</span> <span class="kc">True</span><span class="p">,</span> <span class="n">collapse_identical_variants</span><span class="p">)</span>
            <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_pvs</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data_to_fit</span><span class="p">)</span>
            <span class="c1"># for each site get mask of indices in the binary map</span>
            <span class="c1"># that correspond to that site</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_one_binarymap</span><span class="p">:</span>
                <span class="n">binary_sites</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_binarymaps</span><span class="o">.</span><span class="n">binary_sites</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">binary_sites</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_binarymaps</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">binary_sites</span>
                <span class="k">assert</span> <span class="nb">all</span><span class="p">(</span>
                    <span class="p">(</span><span class="n">binary_sites</span> <span class="o">==</span> <span class="n">bmap</span><span class="o">.</span><span class="n">binary_sites</span><span class="p">)</span><span class="o">.</span><span class="n">all</span><span class="p">()</span>
                    <span class="k">for</span> <span class="n">bmap</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_binarymaps</span>
                <span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_binary_sites</span> <span class="o">=</span> <span class="p">{</span>
                <span class="n">site</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">binary_sites</span> <span class="o">==</span> <span class="n">site</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">site</span> <span class="ow">in</span> <span class="n">numpy</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">binary_sites</span><span class="p">)</span>
            <span class="p">}</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">data_to_fit</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="k">def</span> <span class="nf">_binarymaps_from_df</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">df</span><span class="p">,</span>
        <span class="n">get_pv</span><span class="p">,</span>
        <span class="n">collapse_identical_variants</span><span class="p">,</span>
    <span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Get variants and and other information from data frame.</span>

<span class="sd">        Get `(one_binarymap, binarymaps, cs, pvs, weights, sorted_df)`. If</span>
<span class="sd">        `get_pv=False` then `pvs` is `None`. If `collapse_identical_variants`</span>
<span class="sd">        is `False` then `weights` is `None`. If same variants for all</span>
<span class="sd">        concentrations, `binarymaps` is a BinaryMap and `one_binarymap` is</span>
<span class="sd">        `True`. Otherwise, `binarymaps` lists BinaryMap for each concentration.</span>
<span class="sd">        `sorted_df` is version of `df` with variants/concentrations in same</span>
<span class="sd">        order as `binarymaps`, while `pvs` is 1D array with the prob escapes</span>
<span class="sd">        for these variants. We handle separately cases when BinaryMap same or</span>
<span class="sd">        different for concentrations as more efficient if all concentrations</span>
<span class="sd">        have same BinaryMap.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">cols</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;concentration&quot;</span><span class="p">,</span> <span class="s2">&quot;aa_substitutions&quot;</span><span class="p">]</span>
        <span class="k">if</span> <span class="s2">&quot;weight&quot;</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="n">cols</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                <span class="s2">&quot;weight&quot;</span>
            <span class="p">)</span>  <span class="c1"># will be overwritten if `collapse_identical_variants`</span>
        <span class="k">if</span> <span class="n">get_pv</span><span class="p">:</span>
            <span class="n">cols</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;prob_escape&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">df</span><span class="p">[</span><span class="n">cols</span><span class="p">]</span><span class="o">.</span><span class="n">notnull</span><span class="p">()</span><span class="o">.</span><span class="n">all</span><span class="p">()</span><span class="o">.</span><span class="n">all</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;null entries in data frame of variants:</span><span class="se">\n</span><span class="si">{</span><span class="n">df</span><span class="p">[</span><span class="n">cols</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">collapse_identical_variants</span><span class="p">:</span>
            <span class="n">agg_dict</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;weight&quot;</span><span class="p">:</span> <span class="s2">&quot;sum&quot;</span><span class="p">}</span>
            <span class="k">if</span> <span class="n">get_pv</span><span class="p">:</span>
                <span class="n">agg_dict</span><span class="p">[</span><span class="s2">&quot;prob_escape&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">collapse_identical_variants</span>
            <span class="n">df</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">df</span><span class="p">[</span><span class="n">cols</span><span class="p">]</span>
                <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">weight</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
                <span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s2">&quot;concentration&quot;</span><span class="p">,</span> <span class="s2">&quot;aa_substitutions&quot;</span><span class="p">],</span> <span class="n">as_index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
                <span class="o">.</span><span class="n">aggregate</span><span class="p">(</span><span class="n">agg_dict</span><span class="p">)</span>
            <span class="p">)</span>
        <span class="n">sorted_df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">([</span><span class="s2">&quot;concentration&quot;</span><span class="p">,</span> <span class="s2">&quot;aa_substitutions&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span>
            <span class="n">drop</span><span class="o">=</span><span class="kc">True</span>
        <span class="p">)</span>
        <span class="n">cs</span> <span class="o">=</span> <span class="n">sorted_df</span><span class="p">[</span><span class="s2">&quot;concentration&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span><span class="o">.</span><span class="n">sort_values</span><span class="p">()</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="n">cs</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">all</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;concentrations must be &gt; 0&quot;</span><span class="p">)</span>
        <span class="n">binarymaps</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">pvs</span> <span class="o">=</span> <span class="p">[]</span> <span class="k">if</span> <span class="n">get_pv</span> <span class="k">else</span> <span class="kc">None</span>
        <span class="n">weights</span> <span class="o">=</span> <span class="p">[]</span> <span class="k">if</span> <span class="n">collapse_identical_variants</span> <span class="k">else</span> <span class="kc">None</span>
        <span class="n">one_binarymap</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">c</span><span class="p">,</span> <span class="n">i_df</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">sorted_df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;concentration&quot;</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="kc">False</span><span class="p">)):</span>
            <span class="k">assert</span> <span class="n">c</span> <span class="o">==</span> <span class="n">cs</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="n">i_variants</span> <span class="o">=</span> <span class="n">i_df</span><span class="p">[</span><span class="s2">&quot;aa_substitutions&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">first_variants</span> <span class="o">=</span> <span class="n">i_variants</span>
            <span class="k">elif</span> <span class="n">one_binarymap</span><span class="p">:</span>
                <span class="n">one_binarymap</span> <span class="o">=</span> <span class="n">first_variants</span><span class="o">.</span><span class="n">equals</span><span class="p">(</span><span class="n">i_variants</span><span class="p">)</span>
            <span class="n">binarymaps</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_get_binarymap</span><span class="p">(</span><span class="n">i_df</span><span class="p">))</span>
            <span class="k">if</span> <span class="n">get_pv</span><span class="p">:</span>
                <span class="n">pvs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i_df</span><span class="p">[</span><span class="s2">&quot;prob_escape&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(</span><span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">))</span>
            <span class="k">if</span> <span class="n">collapse_identical_variants</span><span class="p">:</span>
                <span class="n">weights</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i_df</span><span class="p">[</span><span class="s2">&quot;weight&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(</span><span class="n">dtype</span><span class="o">=</span><span class="nb">int</span><span class="p">))</span>
        <span class="k">if</span> <span class="n">one_binarymap</span><span class="p">:</span>
            <span class="n">binarymaps</span> <span class="o">=</span> <span class="n">binarymaps</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">get_pv</span><span class="p">:</span>
            <span class="n">pvs</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span><span class="n">pvs</span><span class="p">)</span>
            <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">pvs</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">sorted_df</span><span class="p">)</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">pvs</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">any</span><span class="p">()</span> <span class="ow">or</span> <span class="p">(</span><span class="n">pvs</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">any</span><span class="p">():</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;`prob_escape` must be between 0 and 1&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">collapse_identical_variants</span><span class="p">:</span>
            <span class="n">weights</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span><span class="n">weights</span><span class="p">)</span>
            <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">weights</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">sorted_df</span><span class="p">)</span>
            <span class="k">assert</span> <span class="p">(</span><span class="n">weights</span> <span class="o">&gt;=</span> <span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">all</span><span class="p">()</span>

        <span class="k">return</span> <span class="p">(</span><span class="n">one_binarymap</span><span class="p">,</span> <span class="n">binarymaps</span><span class="p">,</span> <span class="n">cs</span><span class="p">,</span> <span class="n">pvs</span><span class="p">,</span> <span class="n">weights</span><span class="p">,</span> <span class="n">sorted_df</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_params_from_dfs</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">activity_wt_df</span><span class="p">,</span> <span class="n">mut_escape_df</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Params vector from data frames of activities and escapes.&quot;&quot;&quot;</span>
        <span class="c1"># first E entries are activities</span>
        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">activity_wt_df</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">)</span>
        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">)</span> <span class="o">==</span> <span class="n">activity_wt_df</span><span class="p">[</span><span class="s2">&quot;epitope&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">nunique</span><span class="p">()</span>
        <span class="k">assert</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">)</span> <span class="o">==</span> <span class="nb">set</span><span class="p">(</span><span class="n">activity_wt_df</span><span class="p">[</span><span class="s2">&quot;epitope&quot;</span><span class="p">])</span>
        <span class="n">params</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">activity_wt_df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
                <span class="n">epitope</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span>
                    <span class="n">x</span><span class="p">[</span><span class="s2">&quot;epitope&quot;</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">,</span> <span class="n">ordered</span><span class="o">=</span><span class="kc">True</span>
                <span class="p">)</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s2">&quot;epitope&quot;</span><span class="p">)[</span><span class="s2">&quot;activity&quot;</span><span class="p">]</span>
            <span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
        <span class="p">)</span>

        <span class="c1"># Remaining MxE entries are beta values</span>
        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">mut_escape_df</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">)</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutations</span><span class="p">)</span>
        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">mut_escape_df</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">mut_escape_df</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s2">&quot;mutation&quot;</span><span class="p">,</span> <span class="s2">&quot;epitope&quot;</span><span class="p">]))</span>
        <span class="k">assert</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">)</span> <span class="o">==</span> <span class="nb">set</span><span class="p">(</span><span class="n">mut_escape_df</span><span class="p">[</span><span class="s2">&quot;epitope&quot;</span><span class="p">])</span>
        <span class="k">assert</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutations</span><span class="p">)</span> <span class="o">==</span> <span class="nb">set</span><span class="p">(</span><span class="n">mut_escape_df</span><span class="p">[</span><span class="s2">&quot;mutation&quot;</span><span class="p">])</span>
        <span class="n">params</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span>
            <span class="n">mut_escape_df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
                <span class="n">epitope</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span>
                    <span class="n">x</span><span class="p">[</span><span class="s2">&quot;epitope&quot;</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">,</span> <span class="n">ordered</span><span class="o">=</span><span class="kc">True</span>
                <span class="p">),</span>
                <span class="n">mutation</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span>
                    <span class="n">x</span><span class="p">[</span><span class="s2">&quot;mutation&quot;</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutations</span><span class="p">,</span> <span class="n">ordered</span><span class="o">=</span><span class="kc">True</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">sort_values</span><span class="p">([</span><span class="s2">&quot;mutation&quot;</span><span class="p">,</span> <span class="s2">&quot;epitope&quot;</span><span class="p">])[</span><span class="s2">&quot;escape&quot;</span><span class="p">]</span>
            <span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
        <span class="p">)</span>

        <span class="n">params</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">params</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">numpy</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">params</span><span class="p">)</span><span class="o">.</span><span class="n">any</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;some parameters are NaN&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">params</span>

    <span class="k">def</span> <span class="nf">_a_beta_from_params</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">params</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Vector of activities and MxE matrix of betas from params vector.&quot;&quot;&quot;</span>
        <span class="n">params_len</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutations</span><span class="p">))</span>
        <span class="k">if</span> <span class="n">params</span><span class="o">.</span><span class="n">shape</span> <span class="o">!=</span> <span class="p">(</span><span class="n">params_len</span><span class="p">,):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;invalid </span><span class="si">{</span><span class="n">params</span><span class="o">.</span><span class="n">shape</span><span class="si">=}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">a</span> <span class="o">=</span> <span class="n">params</span><span class="p">[:</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">)]</span>
        <span class="n">beta</span> <span class="o">=</span> <span class="n">params</span><span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">)</span> <span class="p">:]</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span>
            <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutations</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">)</span>
        <span class="p">)</span>
        <span class="k">assert</span> <span class="n">a</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">),)</span>
        <span class="k">assert</span> <span class="n">beta</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutations</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">))</span>
        <span class="k">assert</span> <span class="p">(</span><span class="ow">not</span> <span class="n">numpy</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">a</span><span class="p">)</span><span class="o">.</span><span class="n">any</span><span class="p">())</span> <span class="ow">and</span> <span class="p">(</span><span class="ow">not</span> <span class="n">numpy</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">beta</span><span class="p">)</span><span class="o">.</span><span class="n">any</span><span class="p">())</span>
        <span class="k">return</span> <span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="n">beta</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_muts_from_data_to_fit</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data_to_fit</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Get wildtypes, sites, and mutations from ``data_to_fit``.&quot;&quot;&quot;</span>
        <span class="n">wts</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">mutations</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="nb">set</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">variant</span> <span class="ow">in</span> <span class="n">data_to_fit</span><span class="p">[</span><span class="s2">&quot;aa_substitutions&quot;</span><span class="p">]:</span>
            <span class="k">for</span> <span class="n">mutation</span> <span class="ow">in</span> <span class="n">variant</span><span class="o">.</span><span class="n">split</span><span class="p">():</span>
                <span class="n">wt</span><span class="p">,</span> <span class="n">site</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_mutparser</span><span class="o">.</span><span class="n">parse_mut</span><span class="p">(</span><span class="n">mutation</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">site</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">wts</span><span class="p">:</span>
                    <span class="n">wts</span><span class="p">[</span><span class="n">site</span><span class="p">]</span> <span class="o">=</span> <span class="n">wt</span>
                <span class="k">elif</span> <span class="n">wts</span><span class="p">[</span><span class="n">site</span><span class="p">]</span> <span class="o">!=</span> <span class="n">wt</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;inconsistent wildtype for site </span><span class="si">{</span><span class="n">site</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">mutations</span><span class="p">[</span><span class="n">site</span><span class="p">]</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">mutation</span><span class="p">)</span>
        <span class="n">sites</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">(</span><span class="n">natsort</span><span class="o">.</span><span class="n">natsorted</span><span class="p">(</span><span class="n">wts</span><span class="o">.</span><span class="n">keys</span><span class="p">(),</span> <span class="n">alg</span><span class="o">=</span><span class="n">natsort</span><span class="o">.</span><span class="n">ns</span><span class="o">.</span><span class="n">SIGNED</span><span class="p">))</span>
        <span class="n">wts</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">natsort</span><span class="o">.</span><span class="n">natsorted</span><span class="p">(</span><span class="n">wts</span><span class="o">.</span><span class="n">items</span><span class="p">(),</span> <span class="n">alg</span><span class="o">=</span><span class="n">natsort</span><span class="o">.</span><span class="n">ns</span><span class="o">.</span><span class="n">SIGNED</span><span class="p">))</span>
        <span class="k">assert</span> <span class="nb">set</span><span class="p">(</span><span class="n">mutations</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span> <span class="o">==</span> <span class="nb">set</span><span class="p">(</span><span class="n">sites</span><span class="p">)</span> <span class="o">==</span> <span class="nb">set</span><span class="p">(</span><span class="n">wts</span><span class="p">)</span>
        <span class="n">char_order</span> <span class="o">=</span> <span class="p">{</span><span class="n">c</span><span class="p">:</span> <span class="n">i</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">c</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alphabet</span><span class="p">)}</span>
        <span class="n">mutations</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">(</span>
            <span class="n">mut</span>
            <span class="k">for</span> <span class="n">site</span> <span class="ow">in</span> <span class="n">sites</span>
            <span class="k">for</span> <span class="n">mut</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">mutations</span><span class="p">[</span><span class="n">site</span><span class="p">],</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">m</span><span class="p">:</span> <span class="n">char_order</span><span class="p">[</span><span class="n">m</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]])</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="p">(</span><span class="n">wts</span><span class="p">,</span> <span class="n">sites</span><span class="p">,</span> <span class="n">mutations</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_muts_from_mut_escape_df</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">mut_escape_df</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Get wildtypes, sites, and mutations from ``mut_escape_df``.&quot;&quot;&quot;</span>
        <span class="n">wts</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">mutations</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="nb">set</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">mutation</span> <span class="ow">in</span> <span class="n">mut_escape_df</span><span class="p">[</span><span class="s2">&quot;mutation&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">():</span>
            <span class="n">wt</span><span class="p">,</span> <span class="n">site</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_mutparser</span><span class="o">.</span><span class="n">parse_mut</span><span class="p">(</span><span class="n">mutation</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">site</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">wts</span><span class="p">:</span>
                <span class="n">wts</span><span class="p">[</span><span class="n">site</span><span class="p">]</span> <span class="o">=</span> <span class="n">wt</span>
            <span class="k">elif</span> <span class="n">wts</span><span class="p">[</span><span class="n">site</span><span class="p">]</span> <span class="o">!=</span> <span class="n">wt</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;inconsistent wildtype for site </span><span class="si">{</span><span class="n">site</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">mutations</span><span class="p">[</span><span class="n">site</span><span class="p">]</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">mutation</span><span class="p">)</span>
        <span class="n">sites</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">(</span><span class="n">natsort</span><span class="o">.</span><span class="n">natsorted</span><span class="p">(</span><span class="n">wts</span><span class="o">.</span><span class="n">keys</span><span class="p">(),</span> <span class="n">alg</span><span class="o">=</span><span class="n">natsort</span><span class="o">.</span><span class="n">ns</span><span class="o">.</span><span class="n">SIGNED</span><span class="p">))</span>
        <span class="n">wts</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">natsort</span><span class="o">.</span><span class="n">natsorted</span><span class="p">(</span><span class="n">wts</span><span class="o">.</span><span class="n">items</span><span class="p">(),</span> <span class="n">alg</span><span class="o">=</span><span class="n">natsort</span><span class="o">.</span><span class="n">ns</span><span class="o">.</span><span class="n">SIGNED</span><span class="p">))</span>
        <span class="k">assert</span> <span class="nb">set</span><span class="p">(</span><span class="n">mutations</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span> <span class="o">==</span> <span class="nb">set</span><span class="p">(</span><span class="n">sites</span><span class="p">)</span> <span class="o">==</span> <span class="nb">set</span><span class="p">(</span><span class="n">wts</span><span class="p">)</span>
        <span class="n">char_order</span> <span class="o">=</span> <span class="p">{</span><span class="n">c</span><span class="p">:</span> <span class="n">i</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">c</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alphabet</span><span class="p">)}</span>
        <span class="n">mutations</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">(</span>
            <span class="n">mut</span>
            <span class="k">for</span> <span class="n">site</span> <span class="ow">in</span> <span class="n">sites</span>
            <span class="k">for</span> <span class="n">mut</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">mutations</span><span class="p">[</span><span class="n">site</span><span class="p">],</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">m</span><span class="p">:</span> <span class="n">char_order</span><span class="p">[</span><span class="n">m</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]])</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="p">(</span><span class="n">wts</span><span class="p">,</span> <span class="n">sites</span><span class="p">,</span> <span class="n">mutations</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">activity_wt_df</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;pandas.DataFrame: Activities :math:`a_{\rm{wt,e}}` for epitopes.&quot;&quot;&quot;</span>
        <span class="n">a</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_a_beta_from_params</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_params</span><span class="p">)</span>
        <span class="k">assert</span> <span class="n">a</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">),)</span>
        <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
            <span class="p">{</span>
                <span class="s2">&quot;epitope&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">,</span>
                <span class="s2">&quot;activity&quot;</span><span class="p">:</span> <span class="n">a</span><span class="p">,</span>
            <span class="p">}</span>
        <span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">mut_escape_df</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;pandas.DataFrame: Escape :math:`\beta_{m,e}` for each mutation.&quot;&quot;&quot;</span>
        <span class="n">_</span><span class="p">,</span> <span class="n">beta</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_a_beta_from_params</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_params</span><span class="p">)</span>
        <span class="k">assert</span> <span class="n">beta</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutations</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">))</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span>
            <span class="p">[</span>
                <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
                    <span class="p">{</span>
                        <span class="s2">&quot;mutation&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutations</span><span class="p">,</span>
                        <span class="s2">&quot;escape&quot;</span><span class="p">:</span> <span class="n">b</span><span class="p">,</span>
                        <span class="s2">&quot;epitope&quot;</span><span class="p">:</span> <span class="n">e</span><span class="p">,</span>
                    <span class="p">}</span>
                <span class="p">)</span>
                <span class="k">for</span> <span class="n">e</span><span class="p">,</span> <span class="n">b</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">,</span> <span class="n">beta</span><span class="o">.</span><span class="n">transpose</span><span class="p">())</span>
            <span class="p">],</span>
            <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="p">)</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
            <span class="n">site</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;mutation&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">m</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">_mutparser</span><span class="o">.</span><span class="n">parse_mut</span><span class="p">(</span><span class="n">m</span><span class="p">)[</span><span class="mi">1</span><span class="p">]),</span>
            <span class="n">mutant</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;mutation&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
                <span class="k">lambda</span> <span class="n">m</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">_mutparser</span><span class="o">.</span><span class="n">parse_mut</span><span class="p">(</span><span class="n">m</span><span class="p">)[</span><span class="mi">2</span><span class="p">]</span>
            <span class="p">),</span>
            <span class="n">wildtype</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;site&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">wts</span><span class="p">),</span>
        <span class="p">)[</span>
            <span class="p">[</span><span class="s2">&quot;epitope&quot;</span><span class="p">,</span> <span class="s2">&quot;site&quot;</span><span class="p">,</span> <span class="s2">&quot;wildtype&quot;</span><span class="p">,</span> <span class="s2">&quot;mutant&quot;</span><span class="p">,</span> <span class="s2">&quot;mutation&quot;</span><span class="p">,</span> <span class="s2">&quot;escape&quot;</span><span class="p">]</span>
        <span class="p">]</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutations_times_seen</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">df</span><span class="p">[</span><span class="s2">&quot;times_seen&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;mutation&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutations_times_seen</span><span class="p">)</span>
            <span class="k">assert</span> <span class="n">df</span><span class="o">.</span><span class="n">notnull</span><span class="p">()</span><span class="o">.</span><span class="n">all</span><span class="p">()</span><span class="o">.</span><span class="n">all</span><span class="p">()</span>
        <span class="k">assert</span> <span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;wildtype&quot;</span><span class="p">]</span> <span class="o">!=</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;mutant&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">all</span><span class="p">()</span>
        <span class="k">return</span> <span class="n">df</span>

<div class="viewcode-block" id="Polyclonal.mut_escape_site_summary_df"><a class="viewcode-back" href="../../polyclonal.polyclonal.html#polyclonal.polyclonal.Polyclonal.mut_escape_site_summary_df">[docs]</a>    <span class="k">def</span> <span class="nf">mut_escape_site_summary_df</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="o">*</span><span class="p">,</span>
        <span class="n">min_times_seen</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
        <span class="n">mutation_whitelist</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">exclude_chars</span><span class="o">=</span><span class="nb">frozenset</span><span class="p">([</span><span class="s2">&quot;*&quot;</span><span class="p">]),</span>
    <span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Site-level summaries of mutation escape.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        min_times_seen : int</span>
<span class="sd">            Only include in summaries mutations seen in at least this many variants.</span>
<span class="sd">        mutation_whitelist : None or set</span>
<span class="sd">            Only include in summaries these mutations.</span>
<span class="sd">        exclude_chars : set or list</span>
<span class="sd">            Exclude mutations to these characters when calculating site summaries.</span>
<span class="sd">            Useful if you want to ignore stop codons (``*``), and perhaps in some</span>
<span class="sd">            cases also gaps (``-``).</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        pandas.DataFrame</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">escape_metrics</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;mean&quot;</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">NamedAgg</span><span class="p">(</span><span class="s2">&quot;escape&quot;</span><span class="p">,</span> <span class="s2">&quot;mean&quot;</span><span class="p">),</span>
            <span class="s2">&quot;total positive&quot;</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">NamedAgg</span><span class="p">(</span><span class="s2">&quot;escape_gt_0&quot;</span><span class="p">,</span> <span class="s2">&quot;sum&quot;</span><span class="p">),</span>
            <span class="s2">&quot;max&quot;</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">NamedAgg</span><span class="p">(</span><span class="s2">&quot;escape&quot;</span><span class="p">,</span> <span class="s2">&quot;max&quot;</span><span class="p">),</span>
            <span class="s2">&quot;min&quot;</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">NamedAgg</span><span class="p">(</span><span class="s2">&quot;escape&quot;</span><span class="p">,</span> <span class="s2">&quot;min&quot;</span><span class="p">),</span>
            <span class="s2">&quot;total negative&quot;</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">NamedAgg</span><span class="p">(</span><span class="s2">&quot;escape_lt_0&quot;</span><span class="p">,</span> <span class="s2">&quot;sum&quot;</span><span class="p">),</span>
            <span class="s2">&quot;n mutations&quot;</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">NamedAgg</span><span class="p">(</span><span class="s2">&quot;mutation&quot;</span><span class="p">,</span> <span class="s2">&quot;count&quot;</span><span class="p">),</span>
        <span class="p">}</span>
        <span class="n">mut_df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mut_escape_df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s2">&quot;mutant not in @exclude_chars&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutations_times_seen</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">mut_df</span> <span class="o">=</span> <span class="n">mut_df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s2">&quot;times_seen &gt;= @min_times_seen&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">mutation_whitelist</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">mut_df</span> <span class="o">=</span> <span class="n">mut_df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s2">&quot;mutation in @mutation_whitelist&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="p">(</span>
            <span class="n">mut_df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
                <span class="n">escape_gt_0</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;escape&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">clip</span><span class="p">(</span><span class="n">lower</span><span class="o">=</span><span class="mi">0</span><span class="p">),</span>
                <span class="n">escape_lt_0</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;escape&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">clip</span><span class="p">(</span><span class="n">upper</span><span class="o">=</span><span class="mi">0</span><span class="p">),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s2">&quot;epitope&quot;</span><span class="p">,</span> <span class="s2">&quot;site&quot;</span><span class="p">,</span> <span class="s2">&quot;wildtype&quot;</span><span class="p">],</span> <span class="n">as_index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            <span class="o">.</span><span class="n">aggregate</span><span class="p">(</span><span class="o">**</span><span class="n">escape_metrics</span><span class="p">)</span>
        <span class="p">)</span></div>

<div class="viewcode-block" id="Polyclonal.prob_escape"><a class="viewcode-back" href="../../polyclonal.polyclonal.html#polyclonal.polyclonal.Polyclonal.prob_escape">[docs]</a>    <span class="k">def</span> <span class="nf">prob_escape</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="o">*</span><span class="p">,</span>
        <span class="n">variants_df</span><span class="p">,</span>
        <span class="n">concentrations</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="p">):</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;Compute predicted probability of escape :math:`p_v\left(c\right)`.</span>

<span class="sd">        Computed using current mutation-escape values :math:`\beta_{m,e}` and</span>
<span class="sd">        epitope activities :math:`a_{\rm{wt},e}` stored in this</span>
<span class="sd">        :class:`Polyclonal` object.</span>

<span class="sd">        Arguments</span>
<span class="sd">        ---------</span>
<span class="sd">        variants_df : pandas.DataFrame</span>
<span class="sd">            Input data frame defining variants. Should have a column</span>
<span class="sd">            named &#39;aa_substitutions&#39; that defines variants as space-delimited</span>
<span class="sd">            strings of substitutions (e.g., &#39;M1A K3T&#39;). Should also have a</span>
<span class="sd">            column &#39;concentration&#39; if ``concentrations=None``.</span>
<span class="sd">        concentrations : array-like or None</span>
<span class="sd">            Concentrations at which we compute probability of escape.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        pandas.DataFrame</span>
<span class="sd">            Version of ``variants_df`` with columns named &#39;concentration&#39;</span>
<span class="sd">            and &#39;predicted_prob_escape&#39; giving predicted probability of escape</span>
<span class="sd">            :math:`p_v\left(c\right)` for each variant at each concentration.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">prob_escape_col</span> <span class="o">=</span> <span class="s2">&quot;predicted_prob_escape&quot;</span>
        <span class="k">if</span> <span class="n">prob_escape_col</span> <span class="ow">in</span> <span class="n">variants_df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;`variants_df` has column </span><span class="si">{</span><span class="n">prob_escape_col</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># add concentrations column to variants_df</span>
        <span class="k">if</span> <span class="n">concentrations</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="s2">&quot;concentration&quot;</span> <span class="ow">in</span> <span class="n">variants_df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="s1">&#39;`variants_df` has &quot;concentration&quot; column &#39;</span>
                    <span class="s2">&quot;and `concentrations` not `None`&quot;</span>
                <span class="p">)</span>
            <span class="n">variants_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span>
                <span class="p">[</span><span class="n">variants_df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">concentration</span><span class="o">=</span><span class="n">c</span><span class="p">)</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">concentrations</span><span class="p">],</span>
                <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="p">)</span>

        <span class="p">(</span><span class="n">one_binarymap</span><span class="p">,</span> <span class="n">binarymaps</span><span class="p">,</span> <span class="n">cs</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">variants_df</span><span class="p">)</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_binarymaps_from_df</span><span class="p">(</span>
            <span class="n">variants_df</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">False</span>
        <span class="p">)</span>

        <span class="n">p_v_c</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_compute_1d_pvs</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_params</span><span class="p">,</span> <span class="n">one_binarymap</span><span class="p">,</span> <span class="n">binarymaps</span><span class="p">,</span> <span class="n">cs</span><span class="p">)</span>
        <span class="k">assert</span> <span class="n">p_v_c</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">variants_df</span><span class="p">),)</span>
        <span class="n">variants_df</span><span class="p">[</span><span class="n">prob_escape_col</span><span class="p">]</span> <span class="o">=</span> <span class="n">p_v_c</span>

        <span class="k">return</span> <span class="n">variants_df</span></div>

    <span class="k">def</span> <span class="nf">_check_close_activities</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Check that no two epitopes have near-identical activities.&quot;&quot;&quot;</span>
        <span class="n">a</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_a_beta_from_params</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_params</span><span class="p">)</span>
        <span class="n">a_sorted</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">a</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">a1</span><span class="p">,</span> <span class="n">a2</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">a_sorted</span><span class="p">,</span> <span class="n">a_sorted</span><span class="p">[</span><span class="mi">1</span><span class="p">:]):</span>
            <span class="k">if</span> <span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">a1</span><span class="p">,</span> <span class="n">a2</span><span class="p">):</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="s2">&quot;Near-identical activities for two epitopes, &quot;</span>
                    <span class="s2">&quot;will cause problems in fitting. Reinitialize&quot;</span>
                    <span class="sa">f</span><span class="s2">&quot; with more distinct activities:</span><span class="se">\n</span><span class="si">{</span><span class="n">a</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="p">)</span>

<div class="viewcode-block" id="Polyclonal.site_level_model"><a class="viewcode-back" href="../../polyclonal.polyclonal.html#polyclonal.polyclonal.Polyclonal.site_level_model">[docs]</a>    <span class="k">def</span> <span class="nf">site_level_model</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="o">*</span><span class="p">,</span>
        <span class="n">aggregate_mut_escapes</span><span class="o">=</span><span class="s2">&quot;mean&quot;</span><span class="p">,</span>
        <span class="n">collapse_identical_variants</span><span class="o">=</span><span class="s2">&quot;mean&quot;</span><span class="p">,</span>
    <span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Model with mutations collapsed at site level.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        aggregate_mut_escapes : {&#39;mean&#39;}</span>
<span class="sd">            How to aggregate mutation-level escape values to site-level</span>
<span class="sd">            ones in ``mut_escape_df``.</span>
<span class="sd">        collapse_identical_variants : {&quot;mean&quot;, &quot;median&quot;, False}</span>
<span class="sd">            Same meaning as for :class:`Polyclonal` initialization.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        :class:`Polyclonal`</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">data_to_fit</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">site_data_to_fit</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">site_data_to_fit</span> <span class="o">=</span> <span class="n">polyclonal</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">site_level_variants</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">data_to_fit</span><span class="p">,</span>
                <span class="n">original_alphabet</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">alphabet</span><span class="p">,</span>
                <span class="n">letter_suffixed_sites</span><span class="o">=</span><span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequential_integer_sites</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="n">site_escape_df</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">polyclonal</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">site_level_variants</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">mut_escape_df</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;mutation&quot;</span><span class="p">:</span> <span class="s2">&quot;aa_substitutions&quot;</span><span class="p">}),</span>
                <span class="n">original_alphabet</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">alphabet</span><span class="p">,</span>
                <span class="n">letter_suffixed_sites</span><span class="o">=</span><span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequential_integer_sites</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;aa_substitutions&quot;</span><span class="p">:</span> <span class="s2">&quot;mutation&quot;</span><span class="p">})</span>
            <span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s2">&quot;epitope&quot;</span><span class="p">,</span> <span class="s2">&quot;mutation&quot;</span><span class="p">],</span> <span class="n">as_index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            <span class="o">.</span><span class="n">aggregate</span><span class="p">({</span><span class="s2">&quot;escape&quot;</span><span class="p">:</span> <span class="n">aggregate_mut_escapes</span><span class="p">})</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">Polyclonal</span><span class="p">(</span>
            <span class="n">activity_wt_df</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">activity_wt_df</span><span class="p">,</span>
            <span class="n">mut_escape_df</span><span class="o">=</span><span class="n">site_escape_df</span><span class="p">,</span>
            <span class="n">data_to_fit</span><span class="o">=</span><span class="n">site_data_to_fit</span><span class="p">,</span>
            <span class="n">alphabet</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="s2">&quot;m&quot;</span><span class="p">),</span>
            <span class="n">sites</span><span class="o">=</span><span class="kc">None</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequential_integer_sites</span> <span class="k">else</span> <span class="bp">self</span><span class="o">.</span><span class="n">sites</span><span class="p">,</span>
            <span class="n">epitope_colors</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">epitope_colors</span><span class="p">,</span>
            <span class="n">collapse_identical_variants</span><span class="o">=</span><span class="n">collapse_identical_variants</span><span class="p">,</span>
        <span class="p">)</span></div>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_scaled_pseudo_huber</span><span class="p">(</span><span class="n">delta</span><span class="p">,</span> <span class="n">r</span><span class="p">,</span> <span class="n">calc_grad</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;Compute scaled Pseudo-Huber loss (and potentially its gradient).</span>

<span class="sd">        :math:`h = \delta \left(\sqrt{1+\left(r/\delta\right)^2} - 1\right)`;</span>
<span class="sd">        this is actually :math:`1/\delta` times ``scipy.special.pseudo_huber``,</span>
<span class="sd">        and so has slope of one in the linear range.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        delta : float</span>
<span class="sd">        r : numpy.ndarray</span>
<span class="sd">        calc_grad : bool</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        (h, dh)</span>
<span class="sd">            Arrays of same length as ``r``, if ``calc_grad=False`` then</span>
<span class="sd">            ``dh`` is None.</span>

<span class="sd">        &gt;&gt;&gt; h, _ = Polyclonal._scaled_pseudo_huber(2, [1, 2, 4, 8], True)</span>
<span class="sd">        &gt;&gt;&gt; h.round(2)</span>
<span class="sd">        array([0.24, 0.83, 2.47, 6.25])</span>
<span class="sd">        &gt;&gt;&gt; err = scipy.optimize.check_grad(</span>
<span class="sd">        ...       lambda r: Polyclonal._scaled_pseudo_huber(2, r, False)[0],</span>
<span class="sd">        ...       lambda r: Polyclonal._scaled_pseudo_huber(2, r, True)[1],</span>
<span class="sd">        ...       [2])</span>
<span class="sd">        &gt;&gt;&gt; err &lt; 1e-7</span>
<span class="sd">        True</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">delta</span> <span class="o">&lt;=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;PseudoHuber delta must be &gt; 0&quot;</span><span class="p">)</span>
        <span class="n">h</span> <span class="o">=</span> <span class="n">scipy</span><span class="o">.</span><span class="n">special</span><span class="o">.</span><span class="n">pseudo_huber</span><span class="p">(</span><span class="n">delta</span><span class="p">,</span> <span class="n">r</span><span class="p">)</span> <span class="o">/</span> <span class="n">delta</span>
        <span class="k">if</span> <span class="n">calc_grad</span><span class="p">:</span>
            <span class="n">dh</span> <span class="o">=</span> <span class="n">r</span> <span class="o">/</span> <span class="p">(</span><span class="n">h</span> <span class="o">+</span> <span class="n">delta</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">dh</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">return</span> <span class="n">h</span><span class="p">,</span> <span class="n">dh</span>

    <span class="k">def</span> <span class="nf">_loss_dloss</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">params</span><span class="p">,</span> <span class="n">delta</span><span class="p">):</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;Loss on :math:`p_v\left(c\right)` and derivative wrt params.&quot;&quot;&quot;</span>
        <span class="n">pred_pvs</span><span class="p">,</span> <span class="n">dpred_pvs_dparams</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_compute_1d_pvs</span><span class="p">(</span>
            <span class="n">params</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_one_binarymap</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_binarymaps</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_cs</span><span class="p">,</span> <span class="n">calc_grad</span><span class="o">=</span><span class="kc">True</span>
        <span class="p">)</span>
        <span class="k">assert</span> <span class="n">pred_pvs</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">_pvs</span><span class="o">.</span><span class="n">shape</span>
        <span class="k">assert</span> <span class="n">dpred_pvs_dparams</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">params</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_pvs</span><span class="p">))</span>
        <span class="k">assert</span> <span class="nb">type</span><span class="p">(</span><span class="n">dpred_pvs_dparams</span><span class="p">)</span> <span class="o">==</span> <span class="n">scipy</span><span class="o">.</span><span class="n">sparse</span><span class="o">.</span><span class="n">csr_matrix</span>
        <span class="n">residuals</span> <span class="o">=</span> <span class="n">pred_pvs</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">_pvs</span>
        <span class="n">loss</span><span class="p">,</span> <span class="n">dloss_dr</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_scaled_pseudo_huber</span><span class="p">(</span><span class="n">delta</span><span class="p">,</span> <span class="n">residuals</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>
        <span class="k">assert</span> <span class="n">loss</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="n">dloss_dr</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">_pvs</span><span class="o">.</span><span class="n">shape</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_weights</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">loss</span> <span class="o">=</span> <span class="n">loss</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">assert</span> <span class="n">loss</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">_weights</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="n">dloss_dr</span><span class="o">.</span><span class="n">shape</span>
            <span class="n">loss</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_weights</span> <span class="o">*</span> <span class="n">loss</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
            <span class="n">dloss_dr</span> <span class="o">=</span> <span class="n">dloss_dr</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">_weights</span>
        <span class="n">dloss_dparams</span> <span class="o">=</span> <span class="n">dpred_pvs_dparams</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">dloss_dr</span><span class="p">)</span>
        <span class="k">assert</span> <span class="n">dloss_dparams</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="n">params</span><span class="o">.</span><span class="n">shape</span>
        <span class="k">assert</span> <span class="nb">type</span><span class="p">(</span><span class="n">dloss_dparams</span><span class="p">)</span> <span class="o">==</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span>
        <span class="k">return</span> <span class="p">(</span><span class="n">loss</span><span class="p">,</span> <span class="n">dloss_dparams</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_reg_escape</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">params</span><span class="p">,</span> <span class="n">weight</span><span class="p">,</span> <span class="n">delta</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Regularization on escape and its gradient.&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">weight</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">params</span><span class="o">.</span><span class="n">shape</span><span class="p">))</span>
        <span class="k">elif</span> <span class="n">weight</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">weight</span><span class="si">=}</span><span class="s2"> for escape regularization not &gt;= 0&quot;</span><span class="p">)</span>
        <span class="n">_</span><span class="p">,</span> <span class="n">beta</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_a_beta_from_params</span><span class="p">(</span><span class="n">params</span><span class="p">)</span>
        <span class="n">h</span><span class="p">,</span> <span class="n">dh</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_scaled_pseudo_huber</span><span class="p">(</span><span class="n">delta</span><span class="p">,</span> <span class="n">beta</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>
        <span class="n">reg</span> <span class="o">=</span> <span class="n">h</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> <span class="o">*</span> <span class="n">weight</span>
        <span class="k">assert</span> <span class="n">dh</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="n">beta</span><span class="o">.</span><span class="n">shape</span>
        <span class="n">dreg</span> <span class="o">=</span> <span class="n">weight</span> <span class="o">*</span> <span class="n">numpy</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">)),</span> <span class="n">dh</span><span class="o">.</span><span class="n">ravel</span><span class="p">()])</span>
        <span class="k">assert</span> <span class="n">dreg</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="n">params</span><span class="o">.</span><span class="n">shape</span>
        <span class="k">assert</span> <span class="n">numpy</span><span class="o">.</span><span class="n">isfinite</span><span class="p">(</span><span class="n">dreg</span><span class="p">)</span><span class="o">.</span><span class="n">all</span><span class="p">()</span>
        <span class="k">assert</span> <span class="n">reg</span> <span class="o">&gt;=</span> <span class="mi">0</span>
        <span class="k">return</span> <span class="n">reg</span><span class="p">,</span> <span class="n">dreg</span>

    <span class="k">def</span> <span class="nf">_reg_activity</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">params</span><span class="p">,</span> <span class="n">weight</span><span class="p">,</span> <span class="n">delta</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Regularization on activity and its gradient.&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">weight</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">params</span><span class="o">.</span><span class="n">shape</span><span class="p">))</span>
        <span class="k">elif</span> <span class="n">weight</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">weight</span><span class="si">=}</span><span class="s2"> for activity regularization not &gt;= 0&quot;</span><span class="p">)</span>
        <span class="n">a</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_a_beta_from_params</span><span class="p">(</span><span class="n">params</span><span class="p">)</span>
        <span class="n">h</span><span class="p">,</span> <span class="n">dh</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_scaled_pseudo_huber</span><span class="p">(</span><span class="n">delta</span><span class="p">,</span> <span class="n">a</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>
        <span class="n">h</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">a</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">,</span> <span class="n">h</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">)</span>
        <span class="n">dh</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">a</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">,</span> <span class="n">dh</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">)</span>
        <span class="n">reg</span> <span class="o">=</span> <span class="n">h</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> <span class="o">*</span> <span class="n">weight</span>
        <span class="k">assert</span> <span class="n">dh</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="n">a</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">),)</span>
        <span class="n">dreg</span> <span class="o">=</span> <span class="n">weight</span> <span class="o">*</span> <span class="n">numpy</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">dh</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">params</span><span class="p">)</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="n">a</span><span class="p">))])</span>
        <span class="k">assert</span> <span class="n">dreg</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="n">params</span><span class="o">.</span><span class="n">shape</span>
        <span class="k">assert</span> <span class="n">numpy</span><span class="o">.</span><span class="n">isfinite</span><span class="p">(</span><span class="n">dreg</span><span class="p">)</span><span class="o">.</span><span class="n">all</span><span class="p">()</span>
        <span class="k">assert</span> <span class="n">reg</span> <span class="o">&gt;=</span> <span class="mi">0</span>
        <span class="k">return</span> <span class="n">reg</span><span class="p">,</span> <span class="n">dreg</span>

    <span class="k">def</span> <span class="nf">_reg_spread</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">params</span><span class="p">,</span> <span class="n">weight</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Regularization on spread of escape at each site and its gradient.&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">weight</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">params</span><span class="o">.</span><span class="n">shape</span><span class="p">))</span>
        <span class="k">elif</span> <span class="n">weight</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">weight</span><span class="si">=}</span><span class="s2"> for spread regularization not &gt;= 0&quot;</span><span class="p">)</span>
        <span class="n">_</span><span class="p">,</span> <span class="n">beta</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_a_beta_from_params</span><span class="p">(</span><span class="n">params</span><span class="p">)</span>
        <span class="k">assert</span> <span class="n">beta</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutations</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">))</span>
        <span class="n">reg</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">dreg</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">beta</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">siteindex</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_binary_sites</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="n">sitebetas</span> <span class="o">=</span> <span class="n">beta</span><span class="p">[</span><span class="n">siteindex</span><span class="p">]</span>
            <span class="n">mi</span> <span class="o">=</span> <span class="n">sitebetas</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">assert</span> <span class="n">sitebetas</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="p">(</span><span class="n">mi</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">))</span>
            <span class="n">sitemeans</span> <span class="o">=</span> <span class="n">sitebetas</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
            <span class="k">assert</span> <span class="n">sitemeans</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">),)</span>
            <span class="n">beta_minus_mean</span> <span class="o">=</span> <span class="n">sitebetas</span> <span class="o">-</span> <span class="n">sitemeans</span>
            <span class="n">reg</span> <span class="o">+=</span> <span class="n">weight</span> <span class="o">*</span> <span class="p">(</span><span class="n">beta_minus_mean</span><span class="o">**</span><span class="mi">2</span><span class="p">)</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
            <span class="n">dreg_site</span> <span class="o">=</span> <span class="mi">2</span> <span class="o">*</span> <span class="n">weight</span> <span class="o">/</span> <span class="n">mi</span> <span class="o">*</span> <span class="n">beta_minus_mean</span>
            <span class="k">assert</span> <span class="n">dreg_site</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="p">(</span><span class="n">mi</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">))</span>
            <span class="n">dreg</span><span class="p">[</span><span class="n">siteindex</span><span class="p">]</span> <span class="o">+=</span> <span class="n">dreg_site</span>
        <span class="k">assert</span> <span class="n">reg</span> <span class="o">&gt;=</span> <span class="mi">0</span>
        <span class="n">dreg</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">)),</span> <span class="n">dreg</span><span class="o">.</span><span class="n">ravel</span><span class="p">()])</span>
        <span class="k">assert</span> <span class="n">dreg</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="n">params</span><span class="o">.</span><span class="n">shape</span>
        <span class="k">assert</span> <span class="n">numpy</span><span class="o">.</span><span class="n">isfinite</span><span class="p">(</span><span class="n">dreg</span><span class="p">)</span><span class="o">.</span><span class="n">all</span><span class="p">()</span>
        <span class="k">return</span> <span class="n">reg</span><span class="p">,</span> <span class="n">dreg</span>

    <span class="n">DEFAULT_SCIPY_MINIMIZE_KWARGS</span> <span class="o">=</span> <span class="n">frozendict</span><span class="o">.</span><span class="n">frozendict</span><span class="p">(</span>
        <span class="p">{</span>
            <span class="s2">&quot;method&quot;</span><span class="p">:</span> <span class="s2">&quot;L-BFGS-B&quot;</span><span class="p">,</span>
            <span class="s2">&quot;options&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;maxfun&quot;</span><span class="p">:</span> <span class="mf">1e7</span><span class="p">,</span>
                <span class="s2">&quot;maxiter&quot;</span><span class="p">:</span> <span class="mf">1e6</span><span class="p">,</span>
                <span class="s2">&quot;ftol&quot;</span><span class="p">:</span> <span class="mf">1e-7</span><span class="p">,</span>
            <span class="p">},</span>
        <span class="p">}</span>
    <span class="p">)</span>
    <span class="sd">&quot;&quot;&quot;frozendict.frozendict: default ``scipy_minimize_kwargs`` to ``fit``.&quot;&quot;&quot;</span>

<div class="viewcode-block" id="Polyclonal.fit"><a class="viewcode-back" href="../../polyclonal.polyclonal.html#polyclonal.polyclonal.Polyclonal.fit">[docs]</a>    <span class="k">def</span> <span class="nf">fit</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="o">*</span><span class="p">,</span>
        <span class="n">loss_delta</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>
        <span class="n">reg_escape_weight</span><span class="o">=</span><span class="mf">0.02</span><span class="p">,</span>
        <span class="n">reg_escape_delta</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>
        <span class="n">reg_spread_weight</span><span class="o">=</span><span class="mf">0.25</span><span class="p">,</span>
        <span class="n">reg_activity_weight</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span>
        <span class="n">reg_activity_delta</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>
        <span class="n">fit_site_level_first</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="n">scipy_minimize_kwargs</span><span class="o">=</span><span class="n">DEFAULT_SCIPY_MINIMIZE_KWARGS</span><span class="p">,</span>
        <span class="n">log</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">logfreq</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="p">):</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;Fit parameters (activities and mutation escapes) to the data.</span>

<span class="sd">        Requires :attr:`Polyclonal.data_to_fit` be set at initialization of</span>
<span class="sd">        this :class:`Polyclonal` object. After calling this method, the</span>
<span class="sd">        :math:`a_{\rm{wt},e}` and :math:`\beta_{m,e}` have been optimized, and</span>
<span class="sd">        can be accessed using other methods of the :class:`Polyclonal` object.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        loss_delta : float</span>
<span class="sd">            Pseudo-Huber :math:`\delta` parameter for loss on</span>
<span class="sd">            :math:`p_v\left(c\right)` fitting.</span>
<span class="sd">        reg_escape_weight : float</span>
<span class="sd">            Strength of Pseudo-Huber regularization on :math:`\beta_{m,e}`.</span>
<span class="sd">        reg_escape_delta : float</span>
<span class="sd">            Pseudo-Huber :math:`\delta` for regularizing :math:`\beta_{m,e}`.</span>
<span class="sd">        reg_spread_weight : float</span>
<span class="sd">            Strength of regularization on variance of :math:`\beta_{m,e}`</span>
<span class="sd">            values at each site.</span>
<span class="sd">        reg_activity_weight : float</span>
<span class="sd">            Strength of Pseudo-Huber regularization on :math:`a_{\rm{wt},e}`.</span>
<span class="sd">            Only positive values regularized.</span>
<span class="sd">        reg_activity_delta : float</span>
<span class="sd">            Pseudo-Huber :math:`\delta` for regularizing :math:`a_{\rm{wt},e}`.</span>
<span class="sd">        fit_site_level_first : bool</span>
<span class="sd">            First fit a site-level model, then use those activities /</span>
<span class="sd">            escapes to initialize fit of this model. Generally works better.</span>
<span class="sd">        scipy_minimize_kwargs : dict</span>
<span class="sd">            Keyword arguments passed to ``scipy.optimize.minimize``.</span>
<span class="sd">        log : None or writable file-like object</span>
<span class="sd">            Where to log output. If ``None``, use ``sys.stdout``.</span>
<span class="sd">        logfreq : None or int</span>
<span class="sd">            How frequently to write updates on fitting to ``log``.</span>

<span class="sd">        Return</span>
<span class="sd">        ------</span>
<span class="sd">        scipy.optimize.OptimizeResult</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">data_to_fit</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;cannot fit if `data_to_fit` not set&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">log</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">log</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">stdout</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="n">logfreq</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="p">(</span><span class="nb">isinstance</span><span class="p">(</span><span class="n">logfreq</span><span class="p">,</span> <span class="nb">int</span><span class="p">)</span> <span class="ow">and</span> <span class="n">logfreq</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">logfreq</span><span class="si">=}</span><span class="s2"> not an integer &gt; 0&quot;</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_check_close_activities</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">fit_site_level_first</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">logfreq</span><span class="p">:</span>
                <span class="n">log</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;# First fitting site-level model.</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="c1"># get arg passed to fit: https://stackoverflow.com/a/65927265</span>
            <span class="n">myframe</span> <span class="o">=</span> <span class="n">inspect</span><span class="o">.</span><span class="n">currentframe</span><span class="p">()</span>
            <span class="n">keys</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">values</span> <span class="o">=</span> <span class="n">inspect</span><span class="o">.</span><span class="n">getargvalues</span><span class="p">(</span><span class="n">myframe</span><span class="p">)</span>
            <span class="n">fit_kwargs</span> <span class="o">=</span> <span class="p">{</span><span class="n">key</span><span class="p">:</span> <span class="n">values</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">keys</span> <span class="k">if</span> <span class="n">key</span> <span class="o">!=</span> <span class="s2">&quot;self&quot;</span><span class="p">}</span>
            <span class="n">fit_kwargs</span><span class="p">[</span><span class="s2">&quot;fit_site_level_first&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">False</span>
            <span class="n">site_model</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">site_level_model</span><span class="p">()</span>
            <span class="n">site_model</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="o">**</span><span class="n">fit_kwargs</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_params</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_params_from_dfs</span><span class="p">(</span>
                <span class="n">activity_wt_df</span><span class="o">=</span><span class="n">site_model</span><span class="o">.</span><span class="n">activity_wt_df</span><span class="p">,</span>
                <span class="n">mut_escape_df</span><span class="o">=</span><span class="p">(</span>
                    <span class="n">site_model</span><span class="o">.</span><span class="n">mut_escape_df</span><span class="p">[[</span><span class="s2">&quot;epitope&quot;</span><span class="p">,</span> <span class="s2">&quot;site&quot;</span><span class="p">,</span> <span class="s2">&quot;escape&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">mut_escape_df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="s2">&quot;escape&quot;</span><span class="p">),</span>
                        <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;epitope&quot;</span><span class="p">,</span> <span class="s2">&quot;site&quot;</span><span class="p">],</span>
                        <span class="n">how</span><span class="o">=</span><span class="s2">&quot;right&quot;</span><span class="p">,</span>
                        <span class="n">validate</span><span class="o">=</span><span class="s2">&quot;one_to_many&quot;</span><span class="p">,</span>
                    <span class="p">)</span>
                <span class="p">),</span>
            <span class="p">)</span>

        <span class="k">class</span> <span class="nc">LossReg</span><span class="p">:</span>
            <span class="c1"># compute loss in class to remember last call</span>
            <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="n">self_</span><span class="p">):</span>
                <span class="n">self_</span><span class="o">.</span><span class="n">last_loss</span> <span class="o">=</span> <span class="kc">None</span>
                <span class="n">self_</span><span class="o">.</span><span class="n">last_params</span> <span class="o">=</span> <span class="kc">None</span>

            <span class="k">def</span> <span class="nf">loss_reg</span><span class="p">(</span><span class="n">self_</span><span class="p">,</span> <span class="n">params</span><span class="p">,</span> <span class="n">breakdown</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
                <span class="k">if</span> <span class="p">(</span><span class="n">self_</span><span class="o">.</span><span class="n">last_params</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="n">params</span> <span class="o">!=</span> <span class="n">self_</span><span class="o">.</span><span class="n">last_params</span><span class="p">)</span><span class="o">.</span><span class="n">any</span><span class="p">():</span>
                    <span class="n">fitloss</span><span class="p">,</span> <span class="n">dfitloss</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_loss_dloss</span><span class="p">(</span><span class="n">params</span><span class="p">,</span> <span class="n">loss_delta</span><span class="p">)</span>
                    <span class="n">regescape</span><span class="p">,</span> <span class="n">dregescape</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_reg_escape</span><span class="p">(</span>
                        <span class="n">params</span><span class="p">,</span> <span class="n">reg_escape_weight</span><span class="p">,</span> <span class="n">reg_escape_delta</span>
                    <span class="p">)</span>
                    <span class="n">regspread</span><span class="p">,</span> <span class="n">dregspread</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_reg_spread</span><span class="p">(</span><span class="n">params</span><span class="p">,</span> <span class="n">reg_spread_weight</span><span class="p">)</span>
                    <span class="n">regactivity</span><span class="p">,</span> <span class="n">dregactivity</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_reg_activity</span><span class="p">(</span>
                        <span class="n">params</span><span class="p">,</span>
                        <span class="n">reg_activity_weight</span><span class="p">,</span>
                        <span class="n">reg_activity_delta</span><span class="p">,</span>
                    <span class="p">)</span>
                    <span class="n">loss</span> <span class="o">=</span> <span class="n">fitloss</span> <span class="o">+</span> <span class="n">regescape</span> <span class="o">+</span> <span class="n">regspread</span> <span class="o">+</span> <span class="n">regactivity</span>
                    <span class="n">dloss</span> <span class="o">=</span> <span class="n">dfitloss</span> <span class="o">+</span> <span class="n">dregescape</span> <span class="o">+</span> <span class="n">dregspread</span> <span class="o">+</span> <span class="n">dregactivity</span>
                    <span class="n">self_</span><span class="o">.</span><span class="n">last_params</span> <span class="o">=</span> <span class="n">params</span>
                    <span class="n">self_</span><span class="o">.</span><span class="n">last_loss</span> <span class="o">=</span> <span class="p">(</span>
                        <span class="n">loss</span><span class="p">,</span>
                        <span class="n">dloss</span><span class="p">,</span>
                        <span class="p">{</span>
                            <span class="s2">&quot;fit_loss&quot;</span><span class="p">:</span> <span class="n">fitloss</span><span class="p">,</span>
                            <span class="s2">&quot;reg_escape&quot;</span><span class="p">:</span> <span class="n">regescape</span><span class="p">,</span>
                            <span class="s2">&quot;reg_spread&quot;</span><span class="p">:</span> <span class="n">regspread</span><span class="p">,</span>
                            <span class="s2">&quot;reg_activity&quot;</span><span class="p">:</span> <span class="n">regactivity</span><span class="p">,</span>
                        <span class="p">},</span>
                    <span class="p">)</span>
                <span class="k">return</span> <span class="n">self_</span><span class="o">.</span><span class="n">last_loss</span> <span class="k">if</span> <span class="n">breakdown</span> <span class="k">else</span> <span class="n">self_</span><span class="o">.</span><span class="n">last_loss</span><span class="p">[:</span><span class="mi">2</span><span class="p">]</span>

        <span class="n">lossreg</span> <span class="o">=</span> <span class="n">LossReg</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">logfreq</span><span class="p">:</span>
            <span class="n">log</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;# Starting optimization of </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_params</span><span class="p">)</span><span class="si">}</span><span class="s2"> &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;parameters at </span><span class="si">{</span><span class="n">time</span><span class="o">.</span><span class="n">asctime</span><span class="p">()</span><span class="si">}</span><span class="s2">.</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="p">)</span>

            <span class="k">class</span> <span class="nc">Callback</span><span class="p">:</span>
                <span class="c1"># to log minimization</span>
                <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="n">self_</span><span class="p">,</span> <span class="n">interval</span><span class="p">,</span> <span class="n">start</span><span class="p">):</span>
                    <span class="n">self_</span><span class="o">.</span><span class="n">interval</span> <span class="o">=</span> <span class="n">interval</span>
                    <span class="n">self_</span><span class="o">.</span><span class="n">i</span> <span class="o">=</span> <span class="mi">0</span>
                    <span class="n">self_</span><span class="o">.</span><span class="n">start</span> <span class="o">=</span> <span class="n">start</span>

                <span class="k">def</span> <span class="nf">callback</span><span class="p">(</span><span class="n">self_</span><span class="p">,</span> <span class="n">params</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">force_output</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
                    <span class="k">if</span> <span class="n">force_output</span> <span class="ow">or</span> <span class="p">(</span><span class="n">self_</span><span class="o">.</span><span class="n">i</span> <span class="o">%</span> <span class="n">self_</span><span class="o">.</span><span class="n">interval</span> <span class="o">==</span> <span class="mi">0</span><span class="p">):</span>
                        <span class="n">loss</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">breakdown</span> <span class="o">=</span> <span class="n">lossreg</span><span class="o">.</span><span class="n">loss_reg</span><span class="p">(</span><span class="n">params</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>
                        <span class="k">if</span> <span class="n">header</span><span class="p">:</span>
                            <span class="n">cols</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;step&quot;</span><span class="p">,</span> <span class="s2">&quot;time_sec&quot;</span><span class="p">,</span> <span class="s2">&quot;loss&quot;</span><span class="p">,</span> <span class="o">*</span><span class="n">breakdown</span><span class="o">.</span><span class="n">keys</span><span class="p">()]</span>
                            <span class="n">log</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{:&gt;13}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">cols</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
                        <span class="n">sec</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span> <span class="o">-</span> <span class="n">self_</span><span class="o">.</span><span class="n">start</span>
                        <span class="n">log</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                            <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                                <span class="s2">&quot;</span><span class="si">{:&gt;13.5g}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
                                <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="p">[</span><span class="n">self_</span><span class="o">.</span><span class="n">i</span><span class="p">,</span> <span class="n">sec</span><span class="p">,</span> <span class="n">loss</span><span class="p">,</span> <span class="o">*</span><span class="n">breakdown</span><span class="o">.</span><span class="n">values</span><span class="p">()]</span>
                            <span class="p">)</span>
                            <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>
                        <span class="p">)</span>
                        <span class="n">log</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
                    <span class="n">self_</span><span class="o">.</span><span class="n">i</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="n">scipy_minimize_kwargs</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">scipy_minimize_kwargs</span><span class="p">)</span>
            <span class="n">callback_logger</span> <span class="o">=</span> <span class="n">Callback</span><span class="p">(</span><span class="n">logfreq</span><span class="p">,</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">())</span>
            <span class="n">callback_logger</span><span class="o">.</span><span class="n">callback</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_params</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">force_output</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">scipy_minimize_kwargs</span><span class="p">[</span><span class="s2">&quot;callback&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">callback_logger</span><span class="o">.</span><span class="n">callback</span>

        <span class="n">opt_res</span> <span class="o">=</span> <span class="n">scipy</span><span class="o">.</span><span class="n">optimize</span><span class="o">.</span><span class="n">minimize</span><span class="p">(</span>
            <span class="n">fun</span><span class="o">=</span><span class="n">lossreg</span><span class="o">.</span><span class="n">loss_reg</span><span class="p">,</span>
            <span class="n">x0</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_params</span><span class="p">,</span>
            <span class="n">jac</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="o">**</span><span class="n">scipy_minimize_kwargs</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_params</span> <span class="o">=</span> <span class="n">opt_res</span><span class="o">.</span><span class="n">x</span>
        <span class="k">if</span> <span class="n">logfreq</span><span class="p">:</span>
            <span class="n">callback_logger</span><span class="o">.</span><span class="n">callback</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_params</span><span class="p">,</span> <span class="n">force_output</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">log</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;# Successfully finished at </span><span class="si">{</span><span class="n">time</span><span class="o">.</span><span class="n">asctime</span><span class="p">()</span><span class="si">}</span><span class="s2">.</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">opt_res</span><span class="o">.</span><span class="n">success</span><span class="p">:</span>
            <span class="n">log</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;# Optimization FAILED at </span><span class="si">{</span><span class="n">time</span><span class="o">.</span><span class="n">asctime</span><span class="p">()</span><span class="si">}</span><span class="s2">.</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">raise</span> <span class="n">PolyclonalFitError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Optimization failed:</span><span class="se">\n</span><span class="si">{</span><span class="n">opt_res</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">opt_res</span></div>

<div class="viewcode-block" id="Polyclonal.activity_wt_barplot"><a class="viewcode-back" href="../../polyclonal.polyclonal.html#polyclonal.polyclonal.Polyclonal.activity_wt_barplot">[docs]</a>    <span class="k">def</span> <span class="nf">activity_wt_barplot</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;Bar plot of activity against each epitope, :math:`a_{\rm{wt},e}`.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        **kwargs</span>
<span class="sd">            Keyword args for :func:`polyclonal.plot.activity_wt_barplot`.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        altair.Chart</span>
<span class="sd">            Interactive plot.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;activity_wt_df&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">activity_wt_df</span>
        <span class="k">if</span> <span class="s2">&quot;epitope_colors&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
            <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;epitope_colors&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">epitope_colors</span>
        <span class="k">return</span> <span class="n">polyclonal</span><span class="o">.</span><span class="n">plot</span><span class="o">.</span><span class="n">activity_wt_barplot</span><span class="p">(</span><span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></div>

<div class="viewcode-block" id="Polyclonal.mut_escape_pdb_b_factor"><a class="viewcode-back" href="../../polyclonal.polyclonal.html#polyclonal.polyclonal.Polyclonal.mut_escape_pdb_b_factor">[docs]</a>    <span class="k">def</span> <span class="nf">mut_escape_pdb_b_factor</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="o">*</span><span class="p">,</span>
        <span class="n">input_pdbfile</span><span class="p">,</span>
        <span class="n">chains</span><span class="p">,</span>
        <span class="n">metric</span><span class="p">,</span>
        <span class="n">outdir</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">outfile</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">{metric}</span><span class="s2">-</span><span class="si">{epitope}</span><span class="s2">.pdb&quot;</span><span class="p">,</span>
        <span class="n">missing_metric</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
        <span class="n">min_times_seen</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
    <span class="p">):</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;Create PDB files with B factors from a site&#39;s mutation escape.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        input_pdbfile : str</span>
<span class="sd">            Path to input PDB file.</span>
<span class="sd">        chains : str or array-like</span>
<span class="sd">            Single chain or list of them to re-color.</span>
<span class="sd">        metric : str</span>
<span class="sd">            Which site-level summary metric to use. Can be any metric in</span>
<span class="sd">            :meth:`Polyclonal.mut_escape_site_summary_df`.</span>
<span class="sd">        outdir : str</span>
<span class="sd">            Output directory for created PDB files.</span>
<span class="sd">        outfile : str</span>
<span class="sd">            Output file name, with formatting used to replace metric and</span>
<span class="sd">            epitope in curly brackets.</span>
<span class="sd">        missing_metric : float or dict</span>
<span class="sd">            How do we handle sites in PDB that are missing in escape metric?</span>
<span class="sd">            If a float, reassign B factors for all missing sites to this value.</span>
<span class="sd">            If a dict, should be keyed by chain and assign all missing sites in</span>
<span class="sd">            each chain to indicated value.</span>
<span class="sd">        min_times_seen : int</span>
<span class="sd">            Value passed to :meth:`Polyclonal.mut_escape_site_summary_df`.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        pandas.DataFrame</span>
<span class="sd">            Gives name of created B-factor re-colored PDB for each epitope.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mut_escape_site_summary_df</span><span class="p">(</span><span class="n">min_times_seen</span><span class="o">=</span><span class="n">min_times_seen</span><span class="p">)</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">metric</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">metric</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">{</span><span class="s2">&quot;epitope&quot;</span><span class="p">,</span> <span class="s2">&quot;site&quot;</span><span class="p">,</span> <span class="s2">&quot;wildtype&quot;</span><span class="p">}):</span>
            <span class="n">metric_col</span> <span class="o">=</span> <span class="n">metric</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">chains</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">chains</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">chains</span> <span class="o">=</span> <span class="p">[</span><span class="n">chains</span><span class="p">]</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">chain</span><span class="o">=</span><span class="n">chain</span><span class="p">)</span> <span class="k">for</span> <span class="n">chain</span> <span class="ow">in</span> <span class="n">chains</span><span class="p">],</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">result_files</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">epitope</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">outdir</span><span class="p">:</span>
                <span class="n">output_pdbfile</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">outdir</span><span class="p">,</span> <span class="n">outfile</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">output_pdbfile</span> <span class="o">=</span> <span class="n">outfile</span>
            <span class="n">output_pdbfile</span> <span class="o">=</span> <span class="n">output_pdbfile</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">epitope</span><span class="o">=</span><span class="n">epitope</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="n">metric</span>
            <span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">output_pdbfile</span><span class="p">):</span>
                <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">output_pdbfile</span><span class="p">),</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">result_files</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">epitope</span><span class="p">,</span> <span class="n">output_pdbfile</span><span class="p">))</span>
            <span class="n">polyclonal</span><span class="o">.</span><span class="n">pdb_utils</span><span class="o">.</span><span class="n">reassign_b_factor</span><span class="p">(</span>
                <span class="n">input_pdbfile</span><span class="p">,</span>
                <span class="n">output_pdbfile</span><span class="p">,</span>
                <span class="n">df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s2">&quot;epitope == @epitope&quot;</span><span class="p">),</span>
                <span class="n">metric_col</span><span class="p">,</span>
                <span class="n">missing_metric</span><span class="o">=</span><span class="n">missing_metric</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">result_files</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;epitope&quot;</span><span class="p">,</span> <span class="s2">&quot;PDB file&quot;</span><span class="p">])</span></div>

<div class="viewcode-block" id="Polyclonal.mut_escape_plot"><a class="viewcode-back" href="../../polyclonal.polyclonal.html#polyclonal.polyclonal.Polyclonal.mut_escape_plot">[docs]</a>    <span class="k">def</span> <span class="nf">mut_escape_plot</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="o">*</span><span class="p">,</span>
        <span class="n">biochem_order_aas</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="n">prefix_epitope</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">df_to_merge</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="o">**</span><span class="n">kwargs</span><span class="p">,</span>
    <span class="p">):</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;Make plot of the mutation escape values, :math:`\beta_{m,e}`.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        biochem_order_aas : bool</span>
<span class="sd">            Biochemically order the amino-acid alphabet in :attr:`Polyclonal.alphabet`</span>
<span class="sd">            by passing it through :func:`polyclonal.alphabets.biochem_order_aas`.</span>
<span class="sd">        prefix_epitope : bool or None</span>
<span class="sd">            Do we add the prefix &quot;epitope &quot; to the epitope labels? If `None`, do</span>
<span class="sd">            only if epitope is integer.</span>
<span class="sd">        df_to_merge : None or pandas.DataFrame or list</span>
<span class="sd">            To include additional properties, specify data frame or list of them which</span>
<span class="sd">            are merged with :attr:`Polyclonal.mut_escape_df` before being passed</span>
<span class="sd">            to :func:`polyclonal.plot.lineplot_and_heatmap`. Properties will</span>
<span class="sd">            only be included in plot if relevant columns are passed to</span>
<span class="sd">            :func:`polyclonal.plot.lineplot_and_heatmap` via `addtl_slider_stats`,</span>
<span class="sd">            `addtl_tooltip_stats`, or `site_zoom_bar_color_col`.</span>
<span class="sd">        **kwargs</span>
<span class="sd">            Keyword args for :func:`polyclonal.plot.lineplot_and_heatmap`.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        altair.Chart</span>
<span class="sd">            Interactive line plot and heatmap.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;data_df&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span>
            <span class="p">[</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">mut_escape_df</span><span class="p">,</span>
                <span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">mut_escape_df</span><span class="p">[[</span><span class="s2">&quot;site&quot;</span><span class="p">,</span> <span class="s2">&quot;wildtype&quot;</span><span class="p">,</span> <span class="s2">&quot;epitope&quot;</span><span class="p">]]</span>
                    <span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
                    <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">escape</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">mutant</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;wildtype&quot;</span><span class="p">])</span>
                <span class="p">),</span>
            <span class="p">],</span>
        <span class="p">)</span>

        <span class="k">if</span> <span class="n">df_to_merge</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">df_to_merge</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
                <span class="n">df_to_merge</span> <span class="o">=</span> <span class="p">[</span><span class="n">df_to_merge</span><span class="p">]</span>
            <span class="k">elif</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">df_to_merge</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;`df_to_merge` must be pandas.DataFrame or list&quot;</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">df</span> <span class="ow">in</span> <span class="n">df_to_merge</span><span class="p">:</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequential_integer_sites</span> <span class="ow">and</span> <span class="s2">&quot;site&quot;</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
                    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">site</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;site&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">))</span>
                <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;data_df&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;data_df&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="s2">&quot;category_colors&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
            <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;category_colors&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">epitope_colors</span>

        <span class="k">if</span> <span class="n">prefix_epitope</span> <span class="ow">or</span> <span class="p">(</span>
            <span class="n">prefix_epitope</span> <span class="ow">is</span> <span class="kc">None</span>
            <span class="ow">and</span> <span class="nb">all</span><span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">e</span><span class="p">)</span> <span class="o">==</span> <span class="nb">int</span> <span class="ow">or</span> <span class="n">e</span><span class="o">.</span><span class="n">isnumeric</span><span class="p">()</span> <span class="k">for</span> <span class="n">e</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">)</span>
        <span class="p">):</span>
            <span class="n">prefixed</span> <span class="o">=</span> <span class="p">{</span><span class="n">e</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;epitope </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">e</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">}</span>
            <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;data_df&quot;</span><span class="p">][</span><span class="s2">&quot;epitope&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;data_df&quot;</span><span class="p">][</span><span class="s2">&quot;epitope&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">prefixed</span><span class="p">)</span>
            <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;category_colors&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
                <span class="n">prefixed</span><span class="p">[</span><span class="n">e</span><span class="p">]:</span> <span class="n">color</span> <span class="k">for</span> <span class="n">e</span><span class="p">,</span> <span class="n">color</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;category_colors&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
            <span class="p">}</span>

        <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;stat_col&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;escape&quot;</span>
        <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;category_col&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;epitope&quot;</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutations_times_seen</span><span class="p">:</span>
            <span class="k">if</span> <span class="s2">&quot;addtl_slider_stats&quot;</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
                <span class="k">if</span> <span class="s2">&quot;times_seen&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;addtl_slider_stats&quot;</span><span class="p">]:</span>
                    <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;addtl_slider_stats&quot;</span><span class="p">][</span><span class="s2">&quot;times_seen&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;addtl_slider_stats&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;times_seen&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">}</span>

        <span class="k">if</span> <span class="s2">&quot;sites&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
            <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;sites&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sites</span>

        <span class="k">if</span> <span class="s2">&quot;alphabet&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
            <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;alphabet&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">alphabet</span>
        <span class="k">if</span> <span class="n">biochem_order_aas</span><span class="p">:</span>
            <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;alphabet&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">polyclonal</span><span class="o">.</span><span class="n">alphabets</span><span class="o">.</span><span class="n">biochem_order_aas</span><span class="p">(</span>
                <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;alphabet&quot;</span><span class="p">]</span>
            <span class="p">)</span>

        <span class="k">return</span> <span class="n">polyclonal</span><span class="o">.</span><span class="n">plot</span><span class="o">.</span><span class="n">lineplot_and_heatmap</span><span class="p">(</span><span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></div>

<div class="viewcode-block" id="Polyclonal.filter_variants_by_seen_muts"><a class="viewcode-back" href="../../polyclonal.polyclonal.html#polyclonal.polyclonal.Polyclonal.filter_variants_by_seen_muts">[docs]</a>    <span class="k">def</span> <span class="nf">filter_variants_by_seen_muts</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">variants_df</span><span class="p">,</span>
        <span class="n">min_times_seen</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
        <span class="n">subs_col</span><span class="o">=</span><span class="s2">&quot;aa_substitutions&quot;</span><span class="p">,</span>
    <span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Remove variants that contain mutations not seen during model fitting.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        variants_df : pandas.DataFrame</span>
<span class="sd">            Contains variants as rows.</span>
<span class="sd">        min_times_seen : int</span>
<span class="sd">            Require mutations to be seen &gt;= this many times in data used to fit model.</span>
<span class="sd">        subs_col : str</span>
<span class="sd">            Column in `variants_df` with mutations in each variant.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        variants_df : pandas.DataFrame</span>
<span class="sd">            Copy of input dataframe, with rows of variants</span>
<span class="sd">            that have unseen mutations removed.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">variants_df</span> <span class="o">=</span> <span class="n">variants_df</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">subs_col</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">variants_df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;`variants_df` lacks column </span><span class="si">{</span><span class="n">subs_col</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="n">filter_col</span> <span class="o">=</span> <span class="s2">&quot;_pass_filter&quot;</span>
        <span class="k">if</span> <span class="n">filter_col</span> <span class="ow">in</span> <span class="n">variants_df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;`variants_df` cannot have column </span><span class="si">{</span><span class="n">filter_col</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">min_times_seen</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">allowed_muts</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutations</span>
        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutations_times_seen</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">allowed_muts</span> <span class="o">=</span> <span class="p">{</span>
                <span class="n">m</span> <span class="k">for</span> <span class="p">(</span><span class="n">m</span><span class="p">,</span> <span class="n">n</span><span class="p">)</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutations_times_seen</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">n</span> <span class="o">&gt;=</span> <span class="n">min_times_seen</span>
            <span class="p">}</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Cannot use </span><span class="si">{</span><span class="n">min_times_seen</span><span class="si">=}</span><span class="s2"> without data to fit&quot;</span><span class="p">)</span>

        <span class="n">variants_df</span><span class="p">[</span><span class="n">filter_col</span><span class="p">]</span> <span class="o">=</span> <span class="n">variants_df</span><span class="p">[</span><span class="n">subs_col</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">s</span><span class="p">:</span> <span class="nb">set</span><span class="p">(</span><span class="n">s</span><span class="o">.</span><span class="n">split</span><span class="p">())</span><span class="o">.</span><span class="n">issubset</span><span class="p">(</span><span class="n">allowed_muts</span><span class="p">)</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="p">(</span>
            <span class="n">variants_df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s2">&quot;_pass_filter == True&quot;</span><span class="p">)</span>
            <span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="s2">&quot;_pass_filter&quot;</span><span class="p">)</span>
            <span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="p">)</span></div>

<div class="viewcode-block" id="Polyclonal.icXX"><a class="viewcode-back" href="../../polyclonal.polyclonal.html#polyclonal.polyclonal.Polyclonal.icXX">[docs]</a>    <span class="k">def</span> <span class="nf">icXX</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">variants_df</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">col</span><span class="o">=</span><span class="s2">&quot;IC50&quot;</span><span class="p">,</span> <span class="n">min_c</span><span class="o">=</span><span class="mf">1e-5</span><span class="p">,</span> <span class="n">max_c</span><span class="o">=</span><span class="mf">1e5</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Concentration at which a given fraction is neutralized (eg, IC50).</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        variants_df : pandas.DataFrame</span>
<span class="sd">            Data frame defining variants. Should have column named</span>
<span class="sd">            &#39;aa_substitutions&#39; that defines variants as space-delimited</span>
<span class="sd">            strings of substitutions (e.g., &#39;M1A K3T&#39;).</span>
<span class="sd">        x : float</span>
<span class="sd">            Compute concentration at which this fraction is neutralized for</span>
<span class="sd">            each variant. So set to 0.5 for IC50, and 0.9 for IC90.</span>
<span class="sd">        col : str</span>
<span class="sd">            Name of column in returned data frame with the ICXX value.</span>
<span class="sd">        min_c : float</span>
<span class="sd">            Minimum allowed icXX, truncate values &lt; this at this.</span>
<span class="sd">        max_c : float</span>
<span class="sd">            Maximum allowed icXX, truncate values &gt; this at this.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        pandas.DataFrame</span>
<span class="sd">            Copy of ``variants_df`` with added column ``col`` containing icXX.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="mi">0</span> <span class="o">&lt;</span> <span class="n">x</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">x</span><span class="si">=}</span><span class="s2"> not &gt;0 and &lt;1&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">variants_df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;`variants_df` cannot have </span><span class="si">{</span><span class="n">col</span><span class="si">=}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="n">reduced_df</span> <span class="o">=</span> <span class="n">variants_df</span><span class="p">[[</span><span class="s2">&quot;aa_substitutions&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
        <span class="n">bmap</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_binarymap</span><span class="p">(</span><span class="n">reduced_df</span><span class="p">)</span>
        <span class="n">a</span><span class="p">,</span> <span class="n">beta</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_a_beta_from_params</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_params</span><span class="p">)</span>
        <span class="n">exp_phi_e_v</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="o">-</span><span class="n">bmap</span><span class="o">.</span><span class="n">binary_variants</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">beta</span><span class="p">)</span> <span class="o">+</span> <span class="n">a</span><span class="p">)</span>
        <span class="k">assert</span> <span class="n">exp_phi_e_v</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="p">(</span><span class="n">bmap</span><span class="o">.</span><span class="n">nvariants</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">))</span>
        <span class="n">variants</span> <span class="o">=</span> <span class="n">reduced_df</span><span class="p">[</span><span class="s2">&quot;aa_substitutions&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">variants</span><span class="p">)</span> <span class="o">==</span> <span class="n">exp_phi_e_v</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

        <span class="n">records</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">variant</span><span class="p">,</span> <span class="n">exp_phi_e</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">variants</span><span class="p">,</span> <span class="n">exp_phi_e_v</span><span class="p">):</span>
            <span class="k">assert</span> <span class="n">exp_phi_e</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">),)</span>

            <span class="k">def</span> <span class="nf">_func</span><span class="p">(</span><span class="n">c</span><span class="p">,</span> <span class="n">expterm</span><span class="p">):</span>
                <span class="n">pv</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">prod</span><span class="p">(</span><span class="mf">1.0</span> <span class="o">/</span> <span class="p">(</span><span class="mf">1.0</span> <span class="o">+</span> <span class="n">c</span> <span class="o">*</span> <span class="n">expterm</span><span class="p">))</span>
                <span class="k">return</span> <span class="mi">1</span> <span class="o">-</span> <span class="n">x</span> <span class="o">-</span> <span class="n">pv</span>

            <span class="k">if</span> <span class="n">_func</span><span class="p">(</span><span class="n">min_c</span><span class="p">,</span> <span class="n">exp_phi_e</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">ic</span> <span class="o">=</span> <span class="n">min_c</span>
            <span class="k">elif</span> <span class="n">_func</span><span class="p">(</span><span class="n">max_c</span><span class="p">,</span> <span class="n">exp_phi_e</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">ic</span> <span class="o">=</span> <span class="n">max_c</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">sol</span> <span class="o">=</span> <span class="n">scipy</span><span class="o">.</span><span class="n">optimize</span><span class="o">.</span><span class="n">root_scalar</span><span class="p">(</span>
                    <span class="n">_func</span><span class="p">,</span>
                    <span class="n">args</span><span class="o">=</span><span class="p">(</span><span class="n">exp_phi_e</span><span class="p">,),</span>
                    <span class="n">x0</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
                    <span class="n">bracket</span><span class="o">=</span><span class="p">(</span><span class="n">min_c</span><span class="p">,</span> <span class="n">max_c</span><span class="p">),</span>
                    <span class="n">method</span><span class="o">=</span><span class="s2">&quot;brenth&quot;</span><span class="p">,</span>
                <span class="p">)</span>
                <span class="n">ic</span> <span class="o">=</span> <span class="n">sol</span><span class="o">.</span><span class="n">root</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">sol</span><span class="o">.</span><span class="n">converged</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;root finding failed:</span><span class="se">\n</span><span class="si">{</span><span class="n">sol</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">records</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">variant</span><span class="p">,</span> <span class="n">ic</span><span class="p">))</span>

        <span class="n">ic_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_records</span><span class="p">(</span><span class="n">records</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;aa_substitutions&quot;</span><span class="p">,</span> <span class="n">col</span><span class="p">])</span>

        <span class="n">return_df</span> <span class="o">=</span> <span class="n">variants_df</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span>
            <span class="n">ic_df</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;aa_substitutions&quot;</span><span class="p">,</span> <span class="n">validate</span><span class="o">=</span><span class="s2">&quot;many_to_one&quot;</span>
        <span class="p">)</span>
        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">return_df</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">variants_df</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">return_df</span></div>

    <span class="k">def</span> <span class="nf">_compute_1d_pvs</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">params</span><span class="p">,</span> <span class="n">one_binarymap</span><span class="p">,</span> <span class="n">binarymaps</span><span class="p">,</span> <span class="n">cs</span><span class="p">,</span> <span class="n">calc_grad</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;Get 1D raveled array of :math:`p_v\left(c\right)` values.</span>

<span class="sd">        Differs from :meth:`Polyclonal._compute_pv` in that it works if just</span>
<span class="sd">        one or multiple BinaryMap objects.</span>

<span class="sd">        If `calc_grad` is `True`, also returns `scipy.sparse.csr_matrix`</span>
<span class="sd">        of gradient as described in :meth:`Polyclonal._compute_pv`.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">one_binarymap</span><span class="p">:</span>
            <span class="n">tup</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_compute_pv</span><span class="p">(</span><span class="n">params</span><span class="p">,</span> <span class="n">binarymaps</span><span class="p">,</span> <span class="n">cs</span><span class="p">,</span> <span class="n">calc_grad</span><span class="o">=</span><span class="n">calc_grad</span><span class="p">)</span>
            <span class="n">p_vc</span> <span class="o">=</span> <span class="n">tup</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">if</span> <span class="n">calc_grad</span> <span class="k">else</span> <span class="n">tup</span>
            <span class="n">n_vc</span> <span class="o">=</span> <span class="n">binarymaps</span><span class="o">.</span><span class="n">nvariants</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">cs</span><span class="p">)</span>
            <span class="k">assert</span> <span class="n">p_vc</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="p">(</span><span class="n">n_vc</span><span class="p">,),</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">p_vc</span><span class="o">.</span><span class="n">shape</span><span class="si">=}</span><span class="s2">, </span><span class="si">{</span><span class="n">n_vc</span><span class="si">=}</span><span class="s2">&quot;</span>
            <span class="k">if</span> <span class="n">calc_grad</span><span class="p">:</span>
                <span class="n">dpvc_dparams</span> <span class="o">=</span> <span class="n">tup</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
                <span class="k">assert</span> <span class="n">dpvc_dparams</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">params</span><span class="p">),</span> <span class="n">n_vc</span><span class="p">)</span>
                <span class="k">return</span> <span class="p">(</span><span class="n">p_vc</span><span class="p">,</span> <span class="n">dpvc_dparams</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">p_vc</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">cs</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">binarymaps</span><span class="p">)</span>
            <span class="n">p_vc</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="n">dpvc_dparams</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="n">n_vc</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">for</span> <span class="n">c</span><span class="p">,</span> <span class="n">bmap</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">cs</span><span class="p">,</span> <span class="n">binarymaps</span><span class="p">):</span>
                <span class="n">n_vc</span> <span class="o">+=</span> <span class="n">bmap</span><span class="o">.</span><span class="n">nvariants</span>
                <span class="n">tup</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_compute_pv</span><span class="p">(</span>
                    <span class="n">params</span><span class="p">,</span> <span class="n">bmap</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">c</span><span class="p">]),</span> <span class="n">calc_grad</span><span class="o">=</span><span class="n">calc_grad</span>
                <span class="p">)</span>
                <span class="n">p_vc</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tup</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">if</span> <span class="n">calc_grad</span> <span class="k">else</span> <span class="n">tup</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">calc_grad</span><span class="p">:</span>
                    <span class="n">dpvc_dparams</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tup</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
            <span class="n">p_vc</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span><span class="n">p_vc</span><span class="p">)</span>
            <span class="k">assert</span> <span class="n">p_vc</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="p">(</span><span class="n">n_vc</span><span class="p">,)</span>
            <span class="k">if</span> <span class="n">calc_grad</span><span class="p">:</span>
                <span class="n">dpvc_dparams</span> <span class="o">=</span> <span class="n">scipy</span><span class="o">.</span><span class="n">sparse</span><span class="o">.</span><span class="n">hstack</span><span class="p">(</span><span class="n">dpvc_dparams</span><span class="p">)</span><span class="o">.</span><span class="n">tocsr</span><span class="p">()</span>
                <span class="k">assert</span> <span class="n">dpvc_dparams</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">params</span><span class="p">),</span> <span class="n">n_vc</span><span class="p">)</span>
                <span class="k">return</span> <span class="p">(</span><span class="n">p_vc</span><span class="p">,</span> <span class="n">dpvc_dparams</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">p_vc</span>

    <span class="k">def</span> <span class="nf">_compute_pv</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">params</span><span class="p">,</span> <span class="n">bmap</span><span class="p">,</span> <span class="n">cs</span><span class="p">,</span> <span class="n">calc_grad</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;Compute :math:`p_v\left(c\right)` and its derivative.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        params : numpy.ndarray</span>
<span class="sd">        bmap : binarymap.BinaryMap</span>
<span class="sd">        cs : numpy.ndarray</span>
<span class="sd">        calc_grad : bool</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        p_vc, dpvc_dparams</span>
<span class="sd">            ``p_vc`` is 1D array ordered by concentration and then variant</span>
<span class="sd">            variant. So elements are `ivariant + iconcentration * nvariants`,</span>
<span class="sd">            and length is nconcentrations * nvariants. If ``calc_grad=True``,</span>
<span class="sd">            then ``dpvc_dparams`` is `scipy.sparse.csr_matrix` of shape</span>
<span class="sd">            (len(params), nconcentrations * nvariants). Note that</span>
<span class="sd">            len(params) is nepitopes * (1 + binarylength).</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">a</span><span class="p">,</span> <span class="n">beta</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_a_beta_from_params</span><span class="p">(</span><span class="n">params</span><span class="p">)</span>
        <span class="k">assert</span> <span class="n">a</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">),)</span>
        <span class="k">assert</span> <span class="n">beta</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="p">(</span><span class="n">bmap</span><span class="o">.</span><span class="n">binarylength</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">))</span>
        <span class="k">assert</span> <span class="n">beta</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="n">bmap</span><span class="o">.</span><span class="n">binary_variants</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
        <span class="k">assert</span> <span class="n">bmap</span><span class="o">.</span><span class="n">binary_variants</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="p">(</span><span class="n">bmap</span><span class="o">.</span><span class="n">nvariants</span><span class="p">,</span> <span class="n">bmap</span><span class="o">.</span><span class="n">binarylength</span><span class="p">)</span>
        <span class="k">assert</span> <span class="p">(</span><span class="n">cs</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">all</span><span class="p">()</span>
        <span class="k">assert</span> <span class="n">cs</span><span class="o">.</span><span class="n">ndim</span> <span class="o">==</span> <span class="mi">1</span>
        <span class="n">phi_e_v</span> <span class="o">=</span> <span class="n">bmap</span><span class="o">.</span><span class="n">binary_variants</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">beta</span><span class="p">)</span> <span class="o">-</span> <span class="n">a</span>
        <span class="k">assert</span> <span class="n">phi_e_v</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="p">(</span><span class="n">bmap</span><span class="o">.</span><span class="n">nvariants</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">))</span>
        <span class="n">exp_minus_phi_e_v</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="o">-</span><span class="n">phi_e_v</span><span class="p">)</span>
        <span class="n">U_v_e_c</span> <span class="o">=</span> <span class="mf">1.0</span> <span class="o">/</span> <span class="p">(</span><span class="mf">1.0</span> <span class="o">+</span> <span class="n">numpy</span><span class="o">.</span><span class="n">multiply</span><span class="o">.</span><span class="n">outer</span><span class="p">(</span><span class="n">exp_minus_phi_e_v</span><span class="p">,</span> <span class="n">cs</span><span class="p">))</span>
        <span class="k">assert</span> <span class="n">U_v_e_c</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="p">(</span><span class="n">bmap</span><span class="o">.</span><span class="n">nvariants</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">cs</span><span class="p">))</span>
        <span class="n">n_vc</span> <span class="o">=</span> <span class="n">bmap</span><span class="o">.</span><span class="n">nvariants</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">cs</span><span class="p">)</span>
        <span class="n">U_vc_e</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">moveaxis</span><span class="p">(</span><span class="n">U_v_e_c</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span>
            <span class="n">n_vc</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">),</span> <span class="n">order</span><span class="o">=</span><span class="s2">&quot;F&quot;</span>
        <span class="p">)</span>
        <span class="k">assert</span> <span class="n">U_vc_e</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="p">(</span><span class="n">n_vc</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">))</span>
        <span class="n">p_vc</span> <span class="o">=</span> <span class="n">U_vc_e</span><span class="o">.</span><span class="n">prod</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="k">assert</span> <span class="n">p_vc</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="p">(</span><span class="n">n_vc</span><span class="p">,)</span>
        <span class="k">if</span> <span class="n">calc_grad</span><span class="p">:</span>
            <span class="n">dpvc_da</span> <span class="o">=</span> <span class="n">p_vc</span> <span class="o">*</span> <span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">swapaxes</span><span class="p">(</span><span class="n">U_vc_e</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span>
            <span class="k">assert</span> <span class="n">dpvc_da</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">),</span> <span class="n">n_vc</span><span class="p">)</span>
            <span class="n">dpevc</span> <span class="o">=</span> <span class="o">-</span><span class="n">dpvc_da</span><span class="o">.</span><span class="n">ravel</span><span class="p">(</span><span class="n">order</span><span class="o">=</span><span class="s2">&quot;C&quot;</span><span class="p">)</span>
            <span class="n">n_vce</span> <span class="o">=</span> <span class="n">n_vc</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">)</span>
            <span class="k">assert</span> <span class="n">dpevc</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="p">(</span><span class="n">n_vce</span><span class="p">,)</span>
            <span class="c1"># Stack then transpose C X E binary_variants to multiply dpvce</span>
            <span class="c1"># Stacking should be fast: https://stackoverflow.com/a/45990096</span>
            <span class="c1"># Note after transpose this yields CSC matrix</span>
            <span class="n">stacked_binary_variants</span> <span class="o">=</span> <span class="n">scipy</span><span class="o">.</span><span class="n">sparse</span><span class="o">.</span><span class="n">vstack</span><span class="p">(</span>
                <span class="p">[</span><span class="n">bmap</span><span class="o">.</span><span class="n">binary_variants</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">cs</span><span class="p">)</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">)</span>
            <span class="p">)</span><span class="o">.</span><span class="n">transpose</span><span class="p">()</span>
            <span class="k">assert</span> <span class="n">stacked_binary_variants</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="p">(</span><span class="n">bmap</span><span class="o">.</span><span class="n">binarylength</span><span class="p">,</span> <span class="n">n_vce</span><span class="p">)</span>
            <span class="n">dpevc_dbeta</span> <span class="o">=</span> <span class="n">stacked_binary_variants</span><span class="o">.</span><span class="n">multiply</span><span class="p">(</span>
                <span class="n">numpy</span><span class="o">.</span><span class="n">broadcast_to</span><span class="p">(</span><span class="n">dpevc</span><span class="p">,</span> <span class="p">(</span><span class="n">bmap</span><span class="o">.</span><span class="n">binarylength</span><span class="p">,</span> <span class="n">n_vce</span><span class="p">))</span>
            <span class="p">)</span>
            <span class="k">assert</span> <span class="n">dpevc_dbeta</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="p">(</span><span class="n">bmap</span><span class="o">.</span><span class="n">binarylength</span><span class="p">,</span> <span class="n">n_vce</span><span class="p">)</span>
            <span class="c1"># in params, betas sorted first by mutation, then by epitope;</span>
            <span class="c1"># dpevc_dbeta sorted by concentration, then variant, then epitope</span>
            <span class="n">dpvc_dbetaparams</span> <span class="o">=</span> <span class="n">dpevc_dbeta</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span>
                <span class="n">bmap</span><span class="o">.</span><span class="n">binarylength</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">),</span> <span class="n">n_vc</span>
            <span class="p">)</span>
            <span class="k">assert</span> <span class="nb">type</span><span class="p">(</span><span class="n">dpvc_dbetaparams</span><span class="p">)</span> <span class="o">==</span> <span class="n">scipy</span><span class="o">.</span><span class="n">sparse</span><span class="o">.</span><span class="n">coo_matrix</span>
            <span class="c1"># combine to make dpvc_dparams, noting activities before betas</span>
            <span class="c1"># in params</span>
            <span class="n">dpvc_dparams</span> <span class="o">=</span> <span class="n">scipy</span><span class="o">.</span><span class="n">sparse</span><span class="o">.</span><span class="n">vstack</span><span class="p">(</span>
                <span class="p">[</span><span class="n">scipy</span><span class="o">.</span><span class="n">sparse</span><span class="o">.</span><span class="n">csr_matrix</span><span class="p">(</span><span class="n">dpvc_da</span><span class="p">),</span> <span class="n">dpvc_dbetaparams</span><span class="o">.</span><span class="n">tocsr</span><span class="p">()]</span>
            <span class="p">)</span>
            <span class="k">assert</span> <span class="nb">type</span><span class="p">(</span><span class="n">dpvc_dparams</span><span class="p">)</span> <span class="o">==</span> <span class="n">scipy</span><span class="o">.</span><span class="n">sparse</span><span class="o">.</span><span class="n">csr_matrix</span>
            <span class="k">assert</span> <span class="n">dpvc_dparams</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">params</span><span class="p">),</span> <span class="n">n_vc</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">p_vc</span><span class="p">,</span> <span class="n">dpvc_dparams</span>
        <span class="k">return</span> <span class="n">p_vc</span>

    <span class="k">def</span> <span class="nf">_get_binarymap</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">variants_df</span><span class="p">,</span>
    <span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Get ``BinaryMap`` appropriate for use.&quot;&quot;&quot;</span>
        <span class="n">bmap</span> <span class="o">=</span> <span class="n">binarymap</span><span class="o">.</span><span class="n">BinaryMap</span><span class="p">(</span>
            <span class="n">variants_df</span><span class="p">,</span>
            <span class="n">substitutions_col</span><span class="o">=</span><span class="s2">&quot;aa_substitutions&quot;</span><span class="p">,</span>
            <span class="n">allowed_subs</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">mutations</span><span class="p">,</span>
            <span class="n">alphabet</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">alphabet</span><span class="p">,</span>
            <span class="n">sites_as_str</span><span class="o">=</span><span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequential_integer_sites</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="nb">tuple</span><span class="p">(</span><span class="n">bmap</span><span class="o">.</span><span class="n">all_subs</span><span class="p">)</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutations</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;Different mutations in BinaryMap and self:&quot;</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="si">{</span><span class="n">bmap</span><span class="o">.</span><span class="n">all_subs</span><span class="si">=}</span><span class="se">\n</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">mutations</span><span class="si">=}</span><span class="s2">&quot;</span>
            <span class="p">)</span>
        <span class="k">return</span> <span class="n">bmap</span>

<div class="viewcode-block" id="Polyclonal.mut_escape_corr"><a class="viewcode-back" href="../../polyclonal.polyclonal.html#polyclonal.polyclonal.Polyclonal.mut_escape_corr">[docs]</a>    <span class="k">def</span> <span class="nf">mut_escape_corr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ref_poly</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Correlation of mutation-escape values with another model.</span>

<span class="sd">        For each epitope, how well is this model&#39;s mutation-escape values</span>
<span class="sd">        correlation with another model?</span>

<span class="sd">        Mutations present in only one model are ignored.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ------------</span>
<span class="sd">        ref_poly : :class:`Polyclonal`</span>
<span class="sd">            Other (reference) polyclonal model with which we calculate correlations.</span>

<span class="sd">        Returns</span>
<span class="sd">        ---------</span>
<span class="sd">        corr_df : pandas.DataFrame</span>
<span class="sd">            Pairwise epitope correlations for escape.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">mut_escape_df</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">ref_poly</span><span class="o">.</span><span class="n">mut_escape_df</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Both objects must have `mut_escape_df` initialized.&quot;</span><span class="p">)</span>

        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span>
            <span class="p">[</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">mut_escape_df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
                    <span class="n">epitope</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">itertools</span><span class="o">.</span><span class="n">repeat</span><span class="p">(</span><span class="s2">&quot;self&quot;</span><span class="p">),</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;epitope&quot;</span><span class="p">]))</span>
                <span class="p">),</span>
                <span class="n">ref_poly</span><span class="o">.</span><span class="n">mut_escape_df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
                    <span class="n">epitope</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">itertools</span><span class="o">.</span><span class="n">repeat</span><span class="p">(</span><span class="s2">&quot;ref&quot;</span><span class="p">),</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;epitope&quot;</span><span class="p">]))</span>
                <span class="p">),</span>
            <span class="p">]</span>
        <span class="p">)</span>

        <span class="n">corr</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">polyclonal</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">tidy_to_corr</span><span class="p">(</span>
                <span class="n">df</span><span class="p">,</span>
                <span class="n">sample_col</span><span class="o">=</span><span class="s2">&quot;epitope&quot;</span><span class="p">,</span>
                <span class="n">label_col</span><span class="o">=</span><span class="s2">&quot;mutation&quot;</span><span class="p">,</span>
                <span class="n">value_col</span><span class="o">=</span><span class="s2">&quot;escape&quot;</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">assign</span><span class="p">(</span>
                <span class="n">ref_epitope</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;epitope_2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">tup</span><span class="p">:</span> <span class="n">tup</span><span class="p">[</span><span class="mi">1</span><span class="p">]),</span>
                <span class="n">ref_model</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;epitope_2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">tup</span><span class="p">:</span> <span class="n">tup</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span>
                <span class="n">self_epitope</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;epitope_1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">tup</span><span class="p">:</span> <span class="n">tup</span><span class="p">[</span><span class="mi">1</span><span class="p">]),</span>
                <span class="n">self_model</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;epitope_1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">tup</span><span class="p">:</span> <span class="n">tup</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s2">&quot;ref_model != self_model&quot;</span><span class="p">)[</span>
                <span class="p">[</span><span class="s2">&quot;ref_epitope&quot;</span><span class="p">,</span> <span class="s2">&quot;self_epitope&quot;</span><span class="p">,</span> <span class="s2">&quot;correlation&quot;</span><span class="p">]</span>
            <span class="p">]</span>
            <span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
            <span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">corr</span></div>

<div class="viewcode-block" id="Polyclonal.epitope_harmonized_model"><a class="viewcode-back" href="../../polyclonal.polyclonal.html#polyclonal.polyclonal.Polyclonal.epitope_harmonized_model">[docs]</a>    <span class="k">def</span> <span class="nf">epitope_harmonized_model</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ref_poly</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Get a copy of model with epitopes &quot;harmonized&quot; with a reference model.</span>

<span class="sd">        Epitopes are unidentifiable, meaning there is no guarantee that we will assign</span>
<span class="sd">        the same epitope labels across multiple models fit to similar data. This</span>
<span class="sd">        function returns a copy of the current model where the epitope labels are</span>
<span class="sd">        harmonized with those of another model with the same epitope labels. The</span>
<span class="sd">        harmonization is done by putting into correspondence the epitopes with</span>
<span class="sd">        the best correlated mutation-escape values.</span>

<span class="sd">        Parameters</span>
<span class="sd">        -----------</span>
<span class="sd">        ref_poly : :class:`Polyclonal`</span>
<span class="sd">            The reference polyclonal object to harmonize epitope labels with.</span>

<span class="sd">        Returns</span>
<span class="sd">        --------</span>
<span class="sd">        harmonized_model, harmonize_df : tuple</span>
<span class="sd">            `harmonized_model` is a :class:`Polyclonal` object that is a copy of</span>
<span class="sd">            the current model but with epitopes harmonized, and</span>
<span class="sd">            :attr:`Polyclonal.epitope_colors` taken from `ref_poly`. `harmonize_df`</span>
<span class="sd">            is a `pandas.DataFrame` that shows how the epitopes were re-assigned</span>
<span class="sd">            and the correlations in their escape values.</span>

<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        :class:`PolyclonalHarmonizeError`</span>
<span class="sd">            Raise this error if epitopes cannot be harmonized one-to-one.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">set</span><span class="p">(</span><span class="n">ref_poly</span><span class="o">.</span><span class="n">epitopes</span><span class="p">):</span>
            <span class="k">raise</span> <span class="n">PolyclonalHarmonizeError</span><span class="p">(</span>
                <span class="s2">&quot;The models being harmonized have different epitope labels:</span><span class="se">\n</span><span class="s2">&quot;</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="si">=}</span><span class="se">\n</span><span class="si">{</span><span class="n">ref_poly</span><span class="o">.</span><span class="n">epitopes</span><span class="si">=}</span><span class="s2">&quot;</span>
            <span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequential_integer_sites</span> <span class="o">!=</span> <span class="n">ref_poly</span><span class="o">.</span><span class="n">sequential_integer_sites</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">PolyclonalHarmonizeError</span><span class="p">(</span>
                <span class="s2">&quot;Models being harmonized don&#39;t have same `sequential_integer_sites`&quot;</span>
            <span class="p">)</span>

        <span class="n">corr_df</span> <span class="o">=</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">mut_escape_corr</span><span class="p">(</span><span class="n">ref_poly</span><span class="p">)</span>
            <span class="o">.</span><span class="n">sort_values</span><span class="p">([</span><span class="s2">&quot;self_epitope&quot;</span><span class="p">,</span> <span class="s2">&quot;correlation&quot;</span><span class="p">],</span> <span class="n">ascending</span><span class="o">=</span><span class="p">[</span><span class="kc">True</span><span class="p">,</span> <span class="kc">False</span><span class="p">])</span>
            <span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="p">)</span>

        <span class="n">harmonize_df</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">corr_df</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;self_epitope&quot;</span><span class="p">:</span> <span class="s2">&quot;self_initial_epitope&quot;</span><span class="p">})</span>
            <span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;self_initial_epitope&quot;</span><span class="p">,</span> <span class="n">as_index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            <span class="o">.</span><span class="n">first</span><span class="p">()</span>  <span class="c1"># will be row with highest correlation after sorting above</span>
            <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">self_harmonized_epitope</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;ref_epitope&quot;</span><span class="p">])[</span>
                <span class="p">[</span>
                    <span class="s2">&quot;self_initial_epitope&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;self_harmonized_epitope&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;ref_epitope&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;correlation&quot;</span><span class="p">,</span>
                <span class="p">]</span>
            <span class="p">]</span>
        <span class="p">)</span>
        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">harmonize_df</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">ref_poly</span><span class="o">.</span><span class="n">epitopes</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span>
            <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">)</span>
            <span class="o">==</span> <span class="nb">set</span><span class="p">(</span><span class="n">harmonize_df</span><span class="p">[</span><span class="s2">&quot;self_initial_epitope&quot;</span><span class="p">])</span>
            <span class="o">==</span> <span class="nb">set</span><span class="p">(</span><span class="n">harmonize_df</span><span class="p">[</span><span class="s2">&quot;self_harmonized_epitope&quot;</span><span class="p">])</span>
        <span class="p">):</span>
            <span class="k">raise</span> <span class="n">PolyclonalHarmonizeError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;cannot harmonize 1-to-1:</span><span class="se">\n</span><span class="si">{</span><span class="n">corr_df</span><span class="si">=}</span><span class="se">\n</span><span class="si">{</span><span class="n">harmonize_df</span><span class="si">=}</span><span class="s2">&quot;</span>
            <span class="p">)</span>

        <span class="n">map_dict</span> <span class="o">=</span> <span class="n">harmonize_df</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;self_initial_epitope&quot;</span><span class="p">)[</span>
            <span class="s2">&quot;self_harmonized_epitope&quot;</span>
        <span class="p">]</span><span class="o">.</span><span class="n">to_dict</span><span class="p">()</span>
        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">map_dict</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">epitopes</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">map_dict</span><span class="o">.</span><span class="n">values</span><span class="p">()))</span>

        <span class="c1"># for sorting epitopes</span>
        <span class="n">epitope_order</span> <span class="o">=</span> <span class="p">{</span><span class="n">e</span><span class="p">:</span> <span class="n">i</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">e</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">ref_poly</span><span class="o">.</span><span class="n">epitopes</span><span class="p">)}</span>

        <span class="n">harmonized_model</span> <span class="o">=</span> <span class="n">Polyclonal</span><span class="p">(</span>
            <span class="n">activity_wt_df</span><span class="o">=</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">activity_wt_df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">epitope</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;epitope&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">map_dict</span><span class="p">))</span>
                <span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s2">&quot;epitope&quot;</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">c</span><span class="p">:</span> <span class="n">c</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">epitope_order</span><span class="p">))</span>
                <span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="p">),</span>
            <span class="n">mut_escape_df</span><span class="o">=</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">mut_escape_df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">epitope</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;epitope&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">map_dict</span><span class="p">))</span>
                <span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s2">&quot;epitope&quot;</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">c</span><span class="p">:</span> <span class="n">c</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">epitope_order</span><span class="p">))</span>
                <span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="p">),</span>
            <span class="n">data_to_fit</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">data_to_fit</span><span class="p">,</span>
            <span class="n">collapse_identical_variants</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>  <span class="c1"># already collapsed if this done for self</span>
            <span class="n">alphabet</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">alphabet</span><span class="p">,</span>
            <span class="n">epitope_colors</span><span class="o">=</span><span class="n">ref_poly</span><span class="o">.</span><span class="n">epitope_colors</span><span class="p">,</span>
            <span class="n">data_mut_escape_overlap</span><span class="o">=</span><span class="s2">&quot;exact_match&quot;</span><span class="p">,</span>  <span class="c1"># should be exact match in self</span>
            <span class="n">sites</span><span class="o">=</span><span class="kc">None</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequential_integer_sites</span> <span class="k">else</span> <span class="bp">self</span><span class="o">.</span><span class="n">sites</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="k">assert</span> <span class="n">ref_poly</span><span class="o">.</span><span class="n">epitopes</span> <span class="o">==</span> <span class="n">harmonized_model</span><span class="o">.</span><span class="n">epitopes</span>
        <span class="k">return</span> <span class="n">harmonized_model</span><span class="p">,</span> <span class="n">harmonize_df</span></div></div>


<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="kn">import</span> <span class="nn">doctest</span>

    <span class="n">doctest</span><span class="o">.</span><span class="n">testmod</span><span class="p">()</span>
</pre></div>

          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="../../index.html">
    <img class="logo" src="../../_static/BloomLogo.jpg" alt="Logo"/>
    
    <h1 class="logo logo-name">polyclonal</h1>
    
  </a>
</p>



<p class="blurb">Model mutational escape from polyclonal antibodies.</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=jbloomlab&repo=polyclonal&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





<h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../index.html">polyclonal documentation</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../biophysical_model.html">Biophysical model</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../visualize_RBD.html">Visualize antibody mix</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../simulate_RBD.html">Simulate experiment</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../fit_RBD.html">Fit model to data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../RBD_bootstrap.html">Bootstrapping model fits</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../RBD_average.html">Averaging models</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../real_LyCoV1404.html">Real antibody data example</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../reference_site_numbering.html">Reference site numbering</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../specify_epitopes.html">Specifying epitopes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../expt_design.html">Experimental design</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../optimization.html">Details on fitting</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../partition_function.html">Distinct epitopes assumption</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../polyclonal.html">polyclonal package</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../package_index.html">package index</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../acknowledgments.html">Acknowledgements</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../../index.html">Documentation overview</a><ul>
  <li><a href="../index.html">Module code</a><ul>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2019--2022.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 4.2.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
    </div>

    
    <a href="https://github.com/jbloomlab/polyclonal" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>